// @pdfme/pdf-lib/dist/node/src/api/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PDFJavaScript = exports.PDFEmbeddedPage = exports.PDFPage = exports.PDFImage = exports.PDFFont = exports.PDFDocument = void 0;
__exportStar(require("./form"), exports);
__exportStar(require("./text"), exports);
__exportStar(require("./colors"), exports);
__exportStar(require("./errors"), exports);
__exportStar(require("./image"), exports);
__exportStar(require("./objects"), exports);
__exportStar(require("./operations"), exports);
__exportStar(require("./operators"), exports);
__exportStar(require("./rotations"), exports);
__exportStar(require("./sizes"), exports);
__exportStar(require("./PDFPageOptions"), exports);
__exportStar(require("./PDFDocumentOptions"), exports);
__exportStar(require("./StandardFonts"), exports);
var PDFDocument_1 = require("./PDFDocument");
Object.defineProperty(exports, "PDFDocument", { enumerable: true, get: function () { return __importDefault(PDFDocument_1).default; } });
var PDFFont_1 = require("./PDFFont");
Object.defineProperty(exports, "PDFFont", { enumerable: true, get: function () { return __importDefault(PDFFont_1).default; } });
var PDFImage_1 = require("./PDFImage");
Object.defineProperty(exports, "PDFImage", { enumerable: true, get: function () { return __importDefault(PDFImage_1).default; } });
var PDFPage_1 = require("./PDFPage");
Object.defineProperty(exports, "PDFPage", { enumerable: true, get: function () { return __importDefault(PDFPage_1).default; } });
var PDFEmbeddedPage_1 = require("./PDFEmbeddedPage");
Object.defineProperty(exports, "PDFEmbeddedPage", { enumerable: true, get: function () { return __importDefault(PDFEmbeddedPage_1).default; } });
var PDFJavaScript_1 = require("./PDFJavaScript");
Object.defineProperty(exports, "PDFJavaScript", { enumerable: true, get: function () { return __importDefault(PDFJavaScript_1).default; } });
