// @univerjs/sheets-filter-ui/lib/index.js
var ar = Object.defineProperty;
var cr = (t, e, r) => e in t ? ar(t, e, { enumerable: !0, configurable: !0, writable: !0, value: r }) : t[e] = r;
var v = (t, e, r) => cr(t, typeof e != "symbol" ? e + "" : e, r);
import { BooleanNumber as mt, createIdentifier as Pt, Inject as T, LocaleService as G, IUniverInstanceService as De, ILogService as ur, extractPureTextFromCell as hr, numfmt as Ye, Disposable as ue, Injector as re, Quantity as $t, Tools as be, ColorKit as pt, ICommandService as j, CommandType as Be, IContextService as We, ThemeService as Lt, RxDisposable as Mt, fromCallback as dr, VerticalAlign as Re, InterceptorEffectEnum as fr, UniverInstanceType as ie, DependentOn as xt, IConfigService as Ut, Plugin as it, merge as kt, Optional as mr, registerDependencies as pr, touchDependencies as vt } from "@univerjs/core";
import { CustomFilterOperator as u, FilterBy as w, SetSheetsFilterCriteriaCommand as V, SheetsFilterService as z, SmartToggleSheetsFilterCommand as Fe, FILTER_MUTATIONS as vr, SetSheetsFilterRangeMutation as _r, SetSheetsFilterCriteriaMutation as gr, RemoveSheetsFilterMutation as Sr, ReCalcSheetsFilterMutation as Cr, UniverSheetsFilterPlugin as Ht, SheetsFilterSyncController as Tr, ReCalcSheetsFilterCommand as st, ClearSheetsFilterCriteriaCommand as nt, RemoveSheetFilterCommand as Er, SetSheetFilterRangeCommand as Fr } from "@univerjs/sheets-filter";
import { IEditorBridgeService as _t, SetCellEditVisibleOperation as Nr, SheetSkeletonManagerService as Ir, ISheetSelectionRenderService as yr, SelectionControl as Or, attachSelectionWithCoord as br, getCoordByCell as Rr, SheetsRenderService as Dt, SheetsUIPart as Ar, getObservableWithExclusiveRange$ as wr, getCurrentRangeDisable$ as Pr, whenSheetEditorFocused as $r, SheetCanvasPopManagerService as Lr } from "@univerjs/sheets-ui";
import { ILayoutService as Mr, useDependency as D, useObservable as L, IMessageService as Bt, useComponentsOfPart as xr, ComponentContainer as Ur, getMenuHiddenObservable as ot, MenuItemType as lt, RibbonDataGroup as kr, KeyCode as Hr, MetaKeys as gt, ComponentManager as Dr, IShortcutService as Br, IMenuManagerService as Wr } from "@univerjs/ui";
import { COLOR_BLACK_RGB as St, Rect as Ct, Shape as Vr, IRenderManagerService as Wt } from "@univerjs/engine-render";
import { RefRangeService as Qr, SheetPermissionCheckController as Gr, SheetsSelectionsService as jr, getSheetCommandTarget as Yr, WorksheetFilterPermission as pe, WorksheetViewPermission as ve, RangeProtectionPermissionViewPoint as _e, expandToContinuousRange as Zr, SheetInterceptorService as Kr, SetRangeValuesMutation as qr, INTERCEPTOR_POINT as Xr } from "@univerjs/sheets";
import { BehaviorSubject as Q, ReplaySubject as zr, Subject as Jr, merge as ei, combineLatest as ti, throttleTime as Vt, startWith as Qt, map as le, shareReplay as ri, of as he, switchMap as at, filter as ii, takeUntil as si, distinctUntilChanged as ni } from "rxjs";
import { IRPCChannelService as Gt, toModule as oi, fromModule as li } from "@univerjs/rpc";
import { clsx as Ae, borderClassName as ct, Select as Tt, RadioGroup as ai, Radio as Et, Input as jt, Checkbox as ci, Tree as ui, Tooltip as hi, Switch as di, MessageType as Yt, Segmented as fi, Button as Ze } from "@univerjs/design";
import { useRef as mi, createElement as de, forwardRef as Ne, useCallback as U, useMemo as ut } from "react";
import { jsx as d, jsxs as y, Fragment as Ft } from "react/jsx-runtime";
var b = /* @__PURE__ */ ((t) => (t[t.FIRST = 0] = "FIRST", t[t.SECOND = 1] = "SECOND", t))(b || {}), g = /* @__PURE__ */ ((t) => (t.NONE = "none", t.STARTS_WITH = "startsWith", t.DOES_NOT_START_WITH = "doesNotStartWith", t.ENDS_WITH = "endsWith", t.DOES_NOT_END_WITH = "doesNotEndWith", t.CONTAINS = "contains", t.DOES_NOT_CONTAIN = "doesNotContain", t.EQUALS = "equals", t.NOT_EQUALS = "notEquals", t.EMPTY = "empty", t.NOT_EMPTY = "notEmpty", t.BETWEEN = "between", t.NOT_BETWEEN = "notBetween", t.CUSTOM = "custom", t))(g || {}), m;
((t) => {
  t.NONE = {
    label: "sheets-filter.conditions.none",
    operator: g.NONE,
    order: b.SECOND,
    numOfParameters: 0,
    getDefaultFormParams: () => {
      throw new Error("[FilterConditionItems.NONE]: should not have initial form params!");
    },
    testMappingParams: (i) => i.operator1 === g.NONE,
    mapToFilterColumn: () => null,
    testMappingFilterColumn: (i) => !i.customFilters && !i.filters ? {} : !1
  }, t.EMPTY = {
    label: "sheets-filter.conditions.empty",
    operator: g.EMPTY,
    order: b.SECOND,
    numOfParameters: 0,
    getDefaultFormParams: () => {
      throw new Error("[FilterConditionItems.EMPTY]: should not have initial form params!");
    },
    testMappingParams: ({ operator1: i }) => i === g.EMPTY,
    mapToFilterColumn: () => ({ customFilters: { customFilters: [{ val: "" }] } }),
    testMappingFilterColumn: (i) => {
      var c;
      if (((c = i.customFilters) == null ? void 0 : c.customFilters.length) !== 1)
        return !1;
      const o = i.customFilters.customFilters[0];
      return o.val === "" && o.operator === void 0 ? { operator1: g.EMPTY } : !1;
    }
  }, t.NOT_EMPTY = {
    label: "sheets-filter.conditions.not-empty",
    operator: g.NOT_EMPTY,
    order: b.SECOND,
    numOfParameters: 0,
    getDefaultFormParams: () => {
      throw new Error("[FilterConditionItems.NOT_EMPTY]: should not have initial form params!");
    },
    testMappingParams: ({ operator1: i }) => i === g.NOT_EMPTY,
    mapToFilterColumn: () => ({ customFilters: { customFilters: [{ val: "", operator: u.NOT_EQUALS }] } }),
    testMappingFilterColumn: (i) => {
      var c;
      if (((c = i.customFilters) == null ? void 0 : c.customFilters.length) !== 1)
        return !1;
      const o = i.customFilters.customFilters[0];
      return o.val === " " && o.operator === u.NOT_EQUALS ? { operator1: g.NOT_EMPTY } : !1;
    }
  }, t.TEXT_CONTAINS = {
    label: "sheets-filter.conditions.text-contains",
    operator: g.CONTAINS,
    order: b.FIRST,
    numOfParameters: 1,
    getDefaultFormParams: () => ({ operator1: g.CONTAINS, val1: "" }),
    testMappingParams: (i) => {
      const [o] = x(i);
      return o === g.CONTAINS;
    },
    mapToFilterColumn: (i) => {
      const { val1: o } = i;
      return o === "" ? null : {
        customFilters: { customFilters: [{ val: `*${o}*` }] }
      };
    },
    testMappingFilterColumn: (i) => {
      var c;
      if (((c = i.customFilters) == null ? void 0 : c.customFilters.length) !== 1)
        return !1;
      const o = i.customFilters.customFilters[0], l = o.val.toString();
      return !o.operator && l.startsWith("*") && l.endsWith("*") ? { operator1: g.CONTAINS, val1: l.slice(1, -1) } : !1;
    }
  }, t.DOES_NOT_CONTAIN = {
    label: "sheets-filter.conditions.does-not-contain",
    operator: g.DOES_NOT_CONTAIN,
    order: b.FIRST,
    numOfParameters: 1,
    getDefaultFormParams: () => ({ operator1: g.DOES_NOT_CONTAIN, val1: "" }),
    mapToFilterColumn: (i) => ({
      customFilters: { customFilters: [{ val: `*${i.val1}*`, operator: u.NOT_EQUALS }] }
    }),
    testMappingParams: (i) => {
      const [o] = x(i);
      return o === g.DOES_NOT_CONTAIN;
    },
    testMappingFilterColumn: (i) => {
      var c;
      if (((c = i.customFilters) == null ? void 0 : c.customFilters.length) !== 1)
        return !1;
      const o = i.customFilters.customFilters[0], l = o.val.toString();
      return o.operator === u.NOT_EQUALS && l.startsWith("*") && l.endsWith("*") ? { operator1: g.DOES_NOT_CONTAIN, val1: l.slice(1, -1) } : !1;
    }
  }, t.STARTS_WITH = {
    label: "sheets-filter.conditions.starts-with",
    operator: g.STARTS_WITH,
    order: b.FIRST,
    numOfParameters: 1,
    getDefaultFormParams: () => ({ operator1: g.STARTS_WITH, val1: "" }),
    mapToFilterColumn: (i) => ({
      customFilters: { customFilters: [{ val: `${i.val1}*` }] }
    }),
    testMappingParams: (i) => {
      const [o] = x(i);
      return o === g.STARTS_WITH;
    },
    testMappingFilterColumn: (i) => {
      var c;
      if (((c = i.customFilters) == null ? void 0 : c.customFilters.length) !== 1)
        return !1;
      const o = i.customFilters.customFilters[0], l = o.val.toString();
      return !o.operator && l.endsWith("*") && !l.startsWith("*") ? { operator1: g.STARTS_WITH, val1: l.slice(0, -1) } : !1;
    }
  }, t.ENDS_WITH = {
    label: "sheets-filter.conditions.ends-with",
    operator: g.ENDS_WITH,
    order: b.FIRST,
    numOfParameters: 1,
    getDefaultFormParams: () => ({ operator1: g.ENDS_WITH, val1: "" }),
    mapToFilterColumn: (i) => ({
      customFilters: { customFilters: [{ val: `*${i.val1}` }] }
    }),
    testMappingParams: (i) => {
      const [o] = x(i);
      return o === g.ENDS_WITH;
    },
    testMappingFilterColumn: (i) => {
      var c;
      if (((c = i.customFilters) == null ? void 0 : c.customFilters.length) !== 1)
        return !1;
      const o = i.customFilters.customFilters[0], l = o.val.toString();
      return !o.operator && l.startsWith("*") && !l.endsWith("*") ? { operator1: g.ENDS_WITH, val1: l.slice(1) } : !1;
    }
  }, t.EQUALS = {
    label: "sheets-filter.conditions.equals",
    operator: g.EQUALS,
    order: b.FIRST,
    numOfParameters: 1,
    getDefaultFormParams: () => ({ operator1: g.EQUALS, val1: "" }),
    testMappingParams: (i) => {
      const [o] = x(i);
      return o === g.EQUALS;
    },
    mapToFilterColumn: (i) => {
      const { val1: o } = i;
      return o === "" ? null : {
        customFilters: { customFilters: [{ val: o }] }
      };
    },
    testMappingFilterColumn: (i) => {
      var o, l, c;
      return ((l = (o = i.filters) == null ? void 0 : o.filters) == null ? void 0 : l.length) === 1 ? { operator1: g.EQUALS, val1: "" } : ((c = i.customFilters) == null ? void 0 : c.customFilters.length) === 1 && !i.customFilters.customFilters[0].operator ? { operator1: g.EQUALS, val1: i.customFilters.customFilters[0].val.toString() } : !1;
    }
  }, t.GREATER_THAN = {
    label: "sheets-filter.conditions.greater-than",
    operator: u.GREATER_THAN,
    numOfParameters: 1,
    order: b.FIRST,
    getDefaultFormParams: () => ({ operator1: u.GREATER_THAN, val1: "" }),
    mapToFilterColumn: (i) => ({
      customFilters: { customFilters: [{ val: i.val1, operator: u.GREATER_THAN }] }
    }),
    testMappingParams: (i) => {
      const [o] = x(i);
      return o === u.GREATER_THAN;
    },
    testMappingFilterColumn: (i) => {
      var l;
      if (((l = i.customFilters) == null ? void 0 : l.customFilters.length) !== 1)
        return !1;
      const o = i.customFilters.customFilters[0];
      return o.operator !== u.GREATER_THAN ? !1 : { operator1: u.GREATER_THAN, val1: o.val.toString() };
    }
  }, t.GREATER_THAN_OR_EQUAL = {
    label: "sheets-filter.conditions.greater-than-or-equal",
    operator: u.GREATER_THAN_OR_EQUAL,
    numOfParameters: 1,
    order: b.FIRST,
    getDefaultFormParams: () => ({ operator1: u.GREATER_THAN_OR_EQUAL, val1: "" }),
    testMappingParams: (i) => {
      const [o] = x(i);
      return o === u.GREATER_THAN_OR_EQUAL;
    },
    mapToFilterColumn: (i) => ({
      customFilters: { customFilters: [{ val: i.val1, operator: u.GREATER_THAN_OR_EQUAL }] }
    }),
    testMappingFilterColumn: (i) => {
      var l;
      if (((l = i.customFilters) == null ? void 0 : l.customFilters.length) !== 1)
        return !1;
      const o = i.customFilters.customFilters[0];
      return o.operator !== u.GREATER_THAN_OR_EQUAL ? !1 : { operator1: u.GREATER_THAN_OR_EQUAL, val1: o.val.toString() };
    }
  }, t.LESS_THAN = {
    label: "sheets-filter.conditions.less-than",
    operator: u.LESS_THAN,
    numOfParameters: 1,
    order: b.FIRST,
    getDefaultFormParams: () => ({ operator1: u.LESS_THAN, val1: "" }),
    testMappingParams: (i) => {
      const [o] = x(i);
      return o === u.LESS_THAN;
    },
    mapToFilterColumn: (i) => ({
      customFilters: { customFilters: [{ val: i.val1, operator: u.LESS_THAN }] }
    }),
    testMappingFilterColumn: (i) => {
      var l;
      if (((l = i.customFilters) == null ? void 0 : l.customFilters.length) !== 1)
        return !1;
      const o = i.customFilters.customFilters[0];
      return o.operator !== u.LESS_THAN ? !1 : { operator1: u.LESS_THAN, val1: o.val.toString() };
    }
  }, t.LESS_THAN_OR_EQUAL = {
    label: "sheets-filter.conditions.less-than-or-equal",
    operator: u.LESS_THAN_OR_EQUAL,
    numOfParameters: 1,
    order: b.FIRST,
    getDefaultFormParams: () => ({ operator1: u.LESS_THAN_OR_EQUAL, val1: "" }),
    testMappingParams: (i) => {
      const [o] = x(i);
      return o === u.LESS_THAN_OR_EQUAL;
    },
    mapToFilterColumn: (i) => ({
      customFilters: { customFilters: [{ val: i.val1, operator: u.LESS_THAN_OR_EQUAL }] }
    }),
    testMappingFilterColumn: (i) => {
      var l;
      if (((l = i.customFilters) == null ? void 0 : l.customFilters.length) !== 1)
        return !1;
      const o = i.customFilters.customFilters[0];
      return o.operator !== u.LESS_THAN_OR_EQUAL ? !1 : { operator1: u.LESS_THAN_OR_EQUAL, val1: o.val.toString() };
    }
  }, t.EQUAL = {
    label: "sheets-filter.conditions.equal",
    operator: u.EQUAL,
    numOfParameters: 1,
    order: b.FIRST,
    getDefaultFormParams: () => ({ operator1: u.EQUAL, val1: "" }),
    testMappingParams: (i) => {
      const [o] = x(i);
      return o === u.EQUAL;
    },
    mapToFilterColumn: (i) => ({
      customFilters: { customFilters: [{ val: i.val1, operator: u.EQUAL }] }
    }),
    testMappingFilterColumn: (i) => {
      var l;
      if (((l = i.customFilters) == null ? void 0 : l.customFilters.length) !== 1)
        return !1;
      const o = i.customFilters.customFilters[0];
      return o.operator !== u.EQUAL ? !1 : { operator1: u.EQUAL, val1: o.val.toString() };
    }
  }, t.NOT_EQUAL = {
    label: "sheets-filter.conditions.not-equal",
    operator: u.NOT_EQUALS,
    numOfParameters: 1,
    order: b.FIRST,
    getDefaultFormParams: () => ({ operator1: u.NOT_EQUALS, val1: "" }),
    testMappingParams: (i) => {
      const [o] = x(i);
      return o === u.NOT_EQUALS;
    },
    mapToFilterColumn: (i) => ({
      customFilters: { customFilters: [{ val: i.val1, operator: u.NOT_EQUALS }] }
    }),
    testMappingFilterColumn: (i) => {
      var l;
      if (((l = i.customFilters) == null ? void 0 : l.customFilters.length) !== 1)
        return !1;
      const o = i.customFilters.customFilters[0];
      return o.operator !== u.NOT_EQUALS ? !1 : { operator1: u.NOT_EQUALS, val1: o.val.toString() };
    }
  }, t.BETWEEN = {
    label: "sheets-filter.conditions.between",
    operator: g.BETWEEN,
    order: b.SECOND,
    numOfParameters: 2,
    getDefaultFormParams: () => ({
      and: !0,
      operator1: u.GREATER_THAN_OR_EQUAL,
      val1: "",
      operator2: u.LESS_THAN_OR_EQUAL,
      val2: ""
    }),
    testMappingParams: (i) => {
      const { and: o, operator1: l, operator2: c } = i;
      if (!o) return !1;
      const h = [l, c];
      return h.includes(u.GREATER_THAN_OR_EQUAL) && h.includes(u.LESS_THAN_OR_EQUAL);
    },
    mapToFilterColumn: (i) => {
      const { val1: o, val2: l, operator1: c } = i, h = c === u.GREATER_THAN_OR_EQUAL;
      return {
        customFilters: {
          and: mt.TRUE,
          customFilters: [
            { val: h ? o : l, operator: u.GREATER_THAN_OR_EQUAL },
            { val: h ? l : o, operator: u.LESS_THAN_OR_EQUAL }
          ]
        }
      };
    },
    testMappingFilterColumn: (i) => {
      var c;
      if (((c = i.customFilters) == null ? void 0 : c.customFilters.length) !== 2)
        return !1;
      const [o, l] = i.customFilters.customFilters;
      return o.operator === u.GREATER_THAN_OR_EQUAL && l.operator === u.LESS_THAN_OR_EQUAL && i.customFilters.and ? {
        and: !0,
        operator1: u.GREATER_THAN_OR_EQUAL,
        val1: o.val.toString(),
        operator2: u.LESS_THAN_OR_EQUAL,
        val2: l.val.toString()
      } : l.operator === u.GREATER_THAN_OR_EQUAL && o.operator === u.LESS_THAN_OR_EQUAL && i.customFilters.and ? {
        and: !0,
        operator1: u.GREATER_THAN_OR_EQUAL,
        val1: l.val.toString(),
        operator2: u.LESS_THAN_OR_EQUAL,
        val2: o.val.toLocaleString()
      } : !1;
    }
  }, t.NOT_BETWEEN = {
    label: "sheets-filter.conditions.not-between",
    operator: g.NOT_BETWEEN,
    order: b.SECOND,
    numOfParameters: 2,
    getDefaultFormParams: () => ({
      operator1: u.LESS_THAN,
      val1: "",
      operator2: u.GREATER_THAN,
      val2: ""
    }),
    testMappingParams: (i) => {
      const { and: o, operator1: l, operator2: c } = i;
      if (o) return !1;
      const h = [l, c];
      return h.includes(u.GREATER_THAN) && h.includes(u.LESS_THAN);
    },
    mapToFilterColumn: (i) => {
      const { val1: o, val2: l, operator1: c } = i, h = c === u.GREATER_THAN;
      return {
        customFilters: {
          customFilters: [
            { val: h ? o : l, operator: u.GREATER_THAN },
            { val: h ? l : o, operator: u.LESS_THAN }
          ]
        }
      };
    },
    testMappingFilterColumn: (i) => {
      var c;
      if (((c = i.customFilters) == null ? void 0 : c.customFilters.length) !== 2)
        return !1;
      const [o, l] = i.customFilters.customFilters;
      return o.operator === u.LESS_THAN && l.operator === u.GREATER_THAN && !i.customFilters.and ? {
        operator1: u.LESS_THAN,
        val1: o.val.toString(),
        operator2: u.GREATER_THAN,
        val2: l.val.toString()
      } : l.operator === u.LESS_THAN && o.operator === u.GREATER_THAN && !i.customFilters.and ? {
        operator1: u.GREATER_THAN,
        val1: l.val.toString(),
        operator2: u.LESS_THAN,
        val2: o.val.toLocaleString()
      } : !1;
    }
  }, t.CUSTOM = {
    label: "sheets-filter.conditions.custom",
    operator: g.CUSTOM,
    order: b.SECOND,
    numOfParameters: 2,
    getDefaultFormParams: () => ({
      operator1: g.NONE,
      val1: "",
      operator2: g.NONE,
      val2: ""
    }),
    testMappingParams: () => !0,
    mapToFilterColumn: (i) => {
      const { and: o, val1: l, val2: c, operator1: h, operator2: f } = i;
      function S(R, O) {
        for (const A of t.ALL_CONDITIONS)
          if (A.operator === R)
            return A.mapToFilterColumn({ val1: O, operator1: R });
      }
      const p = !h || h === t.NONE.operator, _ = !f || f === t.NONE.operator;
      if (p && _)
        return t.NONE.mapToFilterColumn({});
      if (p)
        return S(f, c);
      if (_)
        return S(h, l);
      const C = S(h, l), F = S(f, c), N = {
        customFilters: [
          C.customFilters.customFilters[0],
          F.customFilters.customFilters[0]
        ]
      };
      return o && (N.and = mt.TRUE), { customFilters: N };
    },
    testMappingFilterColumn: (i) => {
      var c;
      if (((c = i.customFilters) == null ? void 0 : c.customFilters.length) !== 2)
        return !1;
      const o = i.customFilters.customFilters.map((h) => a({ customFilters: { customFilters: [h] } })), l = {
        operator1: o[0][0].operator,
        val1: o[0][1].val1,
        operator2: o[1][0].operator,
        val2: o[1][1].val1
      };
      return i.customFilters.and && (l.and = !0), l;
    }
  }, t.ALL_CONDITIONS = [
    // ------------------------------
    t.NONE,
    // ------------------------------
    t.EMPTY,
    t.NOT_EMPTY,
    // ------------------------------
    t.TEXT_CONTAINS,
    t.DOES_NOT_CONTAIN,
    t.STARTS_WITH,
    t.ENDS_WITH,
    t.EQUALS,
    // ------------------------------
    t.GREATER_THAN,
    t.GREATER_THAN_OR_EQUAL,
    t.LESS_THAN,
    t.LESS_THAN_OR_EQUAL,
    t.EQUAL,
    t.NOT_EQUAL,
    t.BETWEEN,
    t.NOT_BETWEEN,
    // ------------------------------
    t.CUSTOM
  ];
  function e(i) {
    const o = t.ALL_CONDITIONS.find((l) => l.operator === i);
    if (!o)
      throw new Error(`[SheetsFilter]: no condition item found for operator: ${i}`);
    return o;
  }
  t.getItemByOperator = e;
  function r(i, o) {
    for (const l of t.ALL_CONDITIONS.filter((c) => c.numOfParameters === o))
      if (l.numOfParameters !== 0 && l.testMappingParams(i))
        return l;
    for (const l of t.ALL_CONDITIONS)
      if (l.testMappingParams(i))
        return l;
    throw new Error("[SheetsFilter]: no condition item can be mapped from the filter map params!");
  }
  t.testMappingParams = r;
  function s(i) {
    const o = t.ALL_CONDITIONS.find((l) => l.operator === i);
    return (o == null ? void 0 : o.numOfParameters) === 0 ? { operator1: o.operator } : o.getDefaultFormParams();
  }
  t.getInitialFormParams = s;
  function n(i, o) {
    return i.mapToFilterColumn(o);
  }
  t.mapToFilterColumn = n;
  function a(i) {
    if (!i)
      return [t.NONE, {}];
    for (const o of t.ALL_CONDITIONS) {
      const l = o.testMappingFilterColumn(i);
      if (l)
        return [o, l];
    }
    return [t.NONE, {}];
  }
  t.testMappingFilterColumn = a;
})(m || (m = {}));
function x(t) {
  const { operator1: e, operator2: r, val1: s, val2: n } = t;
  if (e && r)
    throw new Error("Both operator1 and operator2 are set!");
  if (!e && !r)
    throw new Error("Neither operator1 and operator2 and both not set!");
  return e ? [e, s] : [r, n];
}
function ze(t) {
  const e = [], r = [];
  let s = 0, n = 0;
  function a(i) {
    i.leaf && (i.checked ? (e.push(i), s += i.count) : (r.push(i), n += i.count)), i.children && i.children.forEach(a);
  }
  return t.forEach(a), {
    checkedItems: e,
    uncheckedItems: r,
    checked: s,
    unchecked: n
  };
}
var pi = Object.getOwnPropertyDescriptor, vi = (t, e, r, s) => {
  for (var n = s > 1 ? void 0 : s ? pi(e, r) : e, a = t.length - 1, i; a >= 0; a--)
    (i = t[a]) && (n = i(n) || n);
  return n;
}, Ke = (t, e) => (r, s) => e(r, s, t);
const ht = "sheets-filter.generate-filter-values.service", we = Pt(ht);
let Je = class extends ue {
  constructor(t, e, r) {
    super(), this._localeService = t, this._univerInstanceService = e, this._logService = r;
  }
  async getFilterValues(t) {
    var f;
    const { unitId: e, subUnitId: r, filteredOutRowsByOtherColumns: s, filterColumn: n, filters: a, blankChecked: i, iterateRange: o, alreadyChecked: l } = t, c = this._univerInstanceService.getUnit(e), h = (f = this._univerInstanceService.getUnit(e)) == null ? void 0 : f.getSheetBySheetId(r);
    return !c || !h ? [] : (this._logService.debug("[SheetsGenerateFilterValuesService]", "getFilterValues for", { unitId: e, subUnitId: r }), Zt(
      a,
      this._localeService,
      o,
      h,
      new Set(s),
      n,
      new Set(l.map(String)),
      i,
      c.getStyles()
    ));
  }
};
Je = vi([
  Ke(0, T(G)),
  Ke(1, De),
  Ke(2, ur)
], Je);
function Zt(t, e, r, s, n, a, i, o, l) {
  var F, N, R, O, A, P, Y, Z, se, M;
  const c = /* @__PURE__ */ new Map(), h = /* @__PURE__ */ new Map(), f = "yyyy-mm-dd", S = "empty", p = !t && ((a == null ? void 0 : a.filterBy) === w.COLORS || (a == null ? void 0 : a.filterBy) === w.CONDITIONS) && ((F = a.filteredOutRows) == null ? void 0 : F.size);
  let _ = 0;
  for (const E of s.iterateByColumn(r, !1, !1)) {
    const { row: Qe, rowSpan: Ie = 1 } = E;
    let ne = 0;
    for (; ne < Ie; ) {
      const lr = Qe + ne;
      if (n.has(lr)) {
        ne++;
        continue;
      }
      const J = E != null && E.value ? hr(E.value) : "";
      if (!J) {
        _ += 1, ne += Ie;
        continue;
      }
      const ye = (N = E.value) != null && N.v && !E.value.p ? (A = (O = l.get((R = E.value) == null ? void 0 : R.s)) == null ? void 0 : O.n) == null ? void 0 : A.pattern : "", dt = ye && Ye.getFormatInfo(ye).isDate;
      let ft = !1;
      if (dt) {
        const { year: k, month: K, day: I } = Ye.getFormatDateInfo(ye);
        ft = k || K || I;
      }
      if (ye && dt && ft) {
        const k = (P = s.getCellRaw(E.row, E.col)) == null ? void 0 : P.v;
        if (!k) {
          ne++;
          continue;
        }
        const K = Ye.format(f, Number(k)), [I, H, me] = K.split("-").map(Number);
        let ee = c.get(`${I}`);
        ee || (ee = {
          title: `${I}`,
          key: `${I}`,
          children: [],
          count: 0,
          leaf: !1,
          checked: !1
        }, c.set(`${I}`, ee), h.set(`${I}`, [`${I}`]));
        let W = (Y = ee.children) == null ? void 0 : Y.find((je) => je.key === `${I}-${H}`);
        W || (W = {
          title: e.t(`sheets-filter.date.${H}`),
          key: `${I}-${H}`,
          children: [],
          count: 0,
          leaf: !1,
          checked: !1
        }, (Z = ee.children) == null || Z.push(W), h.set(`${I}-${H}`, [`${I}`, `${I}-${H}`]));
        const Ge = (se = W == null ? void 0 : W.children) == null ? void 0 : se.find((je) => je.key === `${I}-${H}-${me}`);
        Ge ? (Ge.originValues.add(J), Ge.count++, W.count++, ee.count++) : ((M = W.children) == null || M.push({
          title: `${me}`,
          key: `${I}-${H}-${me}`,
          count: 1,
          originValues: /* @__PURE__ */ new Set([J]),
          leaf: !0,
          checked: p ? !1 : i.size ? i.has(J) : !o
        }), W.count++, ee.count++, h.set(`${I}-${H}-${me}`, [`${I}`, `${I}-${H}`, `${I}-${H}-${me}`]));
      } else {
        const k = J;
        let K = c.get(k);
        K ? K.count++ : (K = {
          title: J,
          leaf: !0,
          checked: p ? !1 : i.size ? i.has(J) : !o,
          key: k,
          count: 1
        }, c.set(k, K), h.set(k, [k]));
      }
      ne++;
    }
  }
  const C = p ? !1 : t ? o : !0;
  if (_ > 0) {
    const E = {
      title: e.t("sheets-filter.panel.empty"),
      count: _,
      leaf: !0,
      checked: C,
      key: S
    };
    c.set("empty", E), h.set("empty", [S]);
  }
  return {
    filterTreeItems: _i(Array.from(c.values())),
    filterTreeMapCache: h
  };
}
function _i(t) {
  return Array.from(t).sort((e, r) => e.children && !r.children ? -1 : !e.children && r.children ? 1 : gi(e.title, r.title)).map((e) => (e.children && e.children.sort((r, s) => {
    const n = Number.parseInt(r.key.split("-")[1], 10), a = Number.parseInt(s.key.split("-")[1], 10);
    return n - a;
  }).forEach((r) => {
    r.children && r.children.sort((s, n) => {
      const a = Number.parseInt(s.key.split("-")[2], 10), i = Number.parseInt(n.key.split("-")[2], 10);
      return a - i;
    });
  }), e));
}
const Nt = (t) => !Number.isNaN(Number(t)) && !Number.isNaN(Number.parseFloat(t));
function gi(t, e) {
  const r = Nt(t), s = Nt(e);
  return r && s ? Number.parseFloat(t) - Number.parseFloat(e) : r && !s ? -1 : !r && s ? 1 : t.localeCompare(e);
}
function et(t, e) {
  for (const r of t) {
    if (r.key === e)
      return r;
    if (r.children) {
      const s = et(r.children, e);
      if (s)
        return s;
    }
  }
  return null;
}
function Kt(t) {
  return t.leaf ? t.checked : t.children ? t.children.every((e) => Kt(e)) : !0;
}
function ge(t, e) {
  t.leaf && (e !== void 0 ? t.checked = e : t.checked = !t.checked), t.children && t.children.forEach((r) => ge(r, e));
}
function qt(t, e) {
  const r = [];
  return t.forEach((s) => {
    const n = s.originValues ? e.some(
      (o) => Array.from(s.originValues).some(
        (l) => l.toLowerCase().includes(o.toLowerCase())
      )
    ) : !1, a = !n && e.some(
      (o) => s.title.toLowerCase().includes(o.toLowerCase())
    );
    if (n || a)
      r.push({ ...s });
    else if (s.children) {
      const o = qt(s.children, e);
      if (o.length > 0) {
        const l = o.reduce((c, h) => c + h.count, 0);
        r.push({ ...s, count: l, children: o });
      }
    }
  }), r;
}
var Si = Object.getOwnPropertyDescriptor, Ve = (t, e, r, s) => {
  for (var n = s > 1 ? void 0 : s ? Si(e, r) : e, a = t.length - 1, i; a >= 0; a--)
    (i = t[a]) && (n = i(n) || n);
  return n;
}, Te = (t, e) => (r, s) => e(r, s, t);
Pt("sheets-filter-ui.sheets-filter-panel.service");
let X = class extends ue {
  constructor(e, r) {
    super();
    v(this, "_filterBy$", new Q(w.VALUES));
    v(this, "filterBy$", this._filterBy$.asObservable());
    v(this, "_filterByModel$", new zr(1));
    v(this, "filterByModel$", this._filterByModel$.asObservable());
    v(this, "_filterByModel", null);
    v(this, "_hasCriteria$", new Q(!1));
    v(this, "hasCriteria$", this._hasCriteria$.asObservable());
    v(this, "_filterModel", null);
    v(this, "_col$", new Q(-1));
    v(this, "col$", this._col$.asObservable());
    v(this, "_filterHeaderListener", null);
    this._injector = e, this._refRangeService = r;
  }
  get filterBy() {
    return this._filterBy$.getValue();
  }
  get filterByModel() {
    return this._filterByModel;
  }
  set filterByModel(e) {
    this._filterByModel = e, this._filterByModel$.next(e);
  }
  get filterModel() {
    return this._filterModel;
  }
  get col() {
    return this._col$.getValue();
  }
  dispose() {
    this._filterBy$.complete(), this._filterByModel$.complete(), this._hasCriteria$.complete();
  }
  setupCol(e, r) {
    this.terminate(), this._filterModel = e, this._col$.next(r);
    const s = e.getFilterColumn(r);
    if (s) {
      const n = s.getColumnData();
      if (n.customFilters) {
        this._hasCriteria$.next(!0), this._setupByConditions(e, r);
        return;
      }
      if (n.colorFilters) {
        this._hasCriteria$.next(!0), this._setupByColors(e, r);
        return;
      }
      if (n.filters) {
        this._hasCriteria$.next(!0), this._setupByValues(e, r);
        return;
      }
      this._hasCriteria$.next(!1), this._setupByValues(e, r);
      return;
    }
    this._hasCriteria$.next(!1), this._setupByValues(e, r);
  }
  changeFilterBy(e) {
    if (!this._filterModel || this.col === -1)
      return !1;
    switch (e) {
      case w.VALUES:
        this._setupByValues(this._filterModel, this.col);
        break;
      case w.COLORS:
        this._setupByColors(this._filterModel, this.col);
        break;
      case w.CONDITIONS:
        this._setupByConditions(this._filterModel, this.col);
        break;
    }
    return !0;
  }
  terminate() {
    return this._filterModel = null, this._col$.next(-1), this._disposeFilterHeaderChangeListener(), !0;
  }
  _disposeFilterHeaderChangeListener() {
    var e;
    (e = this._filterHeaderListener) == null || e.dispose(), this._filterHeaderListener = null;
  }
  _listenToFilterHeaderChange(e, r) {
    this._disposeFilterHeaderChangeListener();
    const s = e.unitId, n = e.subUnitId, a = e.getRange(), i = {
      startColumn: r,
      startRow: a.startRow,
      endRow: a.startRow,
      endColumn: r
    };
    this._filterHeaderListener = this._refRangeService.watchRange(s, n, i, (o, l) => {
      if (!l)
        this.terminate();
      else {
        const c = l.startColumn - o.startColumn;
        c !== 0 && this._filterByModel.deltaCol(c);
      }
    });
  }
  async _setupByValues(e, r) {
    this._disposePreviousModel();
    const s = e.getRange();
    if (s.startRow === s.endRow) return !1;
    const n = await $e.fromFilterColumn(
      this._injector,
      e,
      r
    );
    return this.filterByModel = n, this._filterBy$.next(w.VALUES), this._listenToFilterHeaderChange(e, r), !0;
  }
  async _setupByColors(e, r) {
    this._disposePreviousModel();
    const s = e.getRange();
    if (s.startRow === s.endRow) return !1;
    const n = await Le.fromFilterColumn(
      this._injector,
      e,
      r
    );
    return this.filterByModel = n, this._filterBy$.next(w.COLORS), this._listenToFilterHeaderChange(e, r), !0;
  }
  _setupByConditions(e, r) {
    this._disposePreviousModel();
    const s = e.getRange();
    if (s.startRow === s.endRow) return !1;
    const n = Pe.fromFilterColumn(
      this._injector,
      e,
      r,
      e.getFilterColumn(r)
    );
    return this.filterByModel = n, this._filterBy$.next(w.CONDITIONS), this._listenToFilterHeaderChange(e, r), !0;
  }
  _disposePreviousModel() {
    var e;
    (e = this._filterByModel) == null || e.dispose(), this.filterByModel = null;
  }
};
X = Ve([
  Te(0, T(re)),
  Te(1, T(Qr))
], X);
let Pe = class extends ue {
  constructor(e, r, s, n, a) {
    super();
    v(this, "canApply$", he(!0));
    v(this, "_conditionItem$");
    v(this, "conditionItem$");
    v(this, "_filterConditionFormParams$");
    v(this, "filterConditionFormParams$");
    this._filterModel = e, this.col = r, this._commandService = a, this._conditionItem$ = new Q(s), this.conditionItem$ = this._conditionItem$.asObservable(), this._filterConditionFormParams$ = new Q(n), this.filterConditionFormParams$ = this._filterConditionFormParams$.asObservable();
  }
  /**
   * Create a model with targeting filter column. If there is not a filter column, the model would be created with
   * default values.
   *
   * @param injector
   * @param filterModel
   * @param col
   * @param filterColumn
   *
   * @returns the model to control the panel's state
   */
  static fromFilterColumn(e, r, s, n) {
    const [a, i] = m.testMappingFilterColumn(n == null ? void 0 : n.getColumnData());
    return e.createInstance(Pe, r, s, a, i);
  }
  get conditionItem() {
    return this._conditionItem$.getValue();
  }
  get filterConditionFormParams() {
    return this._filterConditionFormParams$.getValue();
  }
  dispose() {
    super.dispose(), this._conditionItem$.complete(), this._filterConditionFormParams$.complete();
  }
  deltaCol(e) {
    this.col += e;
  }
  clear() {
    return this._disposed ? Promise.resolve(!1) : this._commandService.executeCommand(V.id, {
      unitId: this._filterModel.unitId,
      subUnitId: this._filterModel.subUnitId,
      col: this.col,
      criteria: null
    });
  }
  /**
   * Apply the filter condition to the target filter column.
   */
  async apply() {
    if (this._disposed) return !1;
    const e = m.mapToFilterColumn(this.conditionItem, this.filterConditionFormParams);
    return this._commandService.executeCommand(V.id, {
      unitId: this._filterModel.unitId,
      subUnitId: this._filterModel.subUnitId,
      col: this.col,
      criteria: e
    });
  }
  /**
   * This method would be called when user changes the primary condition. The model would load the corresponding
   * `IFilterConditionFormParams` and load default condition form params.
   */
  onPrimaryConditionChange(e) {
    const r = m.ALL_CONDITIONS.find((s) => s.operator === e);
    if (!r)
      throw new Error(`[ByConditionsModel]: condition item not found for operator: ${e}!`);
    this._conditionItem$.next(r), this._filterConditionFormParams$.next(m.getInitialFormParams(e));
  }
  /**
   * This method would be called when user changes the primary conditions, the input values or "AND" "OR" ratio.
   * If the primary conditions or the ratio is changed, the method would load the corresponding `IFilterCondition`.
   *
   * When the panel call this method, it only has to pass the changed keys.
   *
   * @param params
   */
  onConditionFormChange(e) {
    const r = { ...this.filterConditionFormParams, ...e };
    if (r.and !== !0 && delete r.and, typeof e.and < "u" || typeof e.operator1 < "u" || typeof e.operator2 < "u") {
      const s = m.testMappingParams(r, this.conditionItem.numOfParameters);
      this._conditionItem$.next(s);
    }
    this._filterConditionFormParams$.next(r);
  }
};
Pe = Ve([
  Te(4, j)
], Pe);
let $e = class extends ue {
  constructor(e, r, s, n, a) {
    super();
    v(this, "_rawFilterItems$");
    v(this, "rawFilterItems$");
    v(this, "filterItems$");
    v(this, "_filterItems", []);
    v(this, "_treeMapCache");
    v(this, "canApply$");
    v(this, "_manuallyUpdateFilterItems$");
    v(this, "_searchString$");
    v(this, "searchString$");
    this._filterModel = e, this.col = r, this._commandService = a, this._treeMapCache = n, this._searchString$ = new Q(""), this.searchString$ = this._searchString$.asObservable(), this._rawFilterItems$ = new Q(s), this.rawFilterItems$ = this._rawFilterItems$.asObservable(), this._manuallyUpdateFilterItems$ = new Jr(), this.filterItems$ = ei(
      ti([
        this._searchString$.pipe(
          Vt(500, void 0, { leading: !0, trailing: !0 }),
          Qt(void 0)
        ),
        this._rawFilterItems$
      ]).pipe(
        le(([i, o]) => {
          if (!i) return o;
          const c = i.toLowerCase().split(/\s+/).filter((h) => !!h);
          return qt(o, c);
        })
      ),
      this._manuallyUpdateFilterItems$
    ).pipe(ri(1)), this.canApply$ = this.filterItems$.pipe(le((i) => ze(i).checked > 0)), this.disposeWithMe(this.filterItems$.subscribe((i) => this._filterItems = i));
  }
  /**
   * Create a model with targeting filter column. If there is not a filter column, the model would be created with
   * default values.
   *
   * @param injector
   * @param filterModel
   * @param col
   *
   * @returns the model to control the panel's state
   */
  static async fromFilterColumn(e, r, s) {
    const n = e.get(De), a = e.get(G), i = e.get(we, $t.OPTIONAL), { unitId: o, subUnitId: l } = r, c = n.getUniverSheetInstance(o);
    if (!c) throw new Error(`[ByValuesModel]: Workbook not found for filter model with unitId: ${o}!`);
    const h = c == null ? void 0 : c.getSheetBySheetId(l);
    if (!h) throw new Error(`[ByValuesModel]: Worksheet not found for filter model with unitId: ${o} and subUnitId: ${l}!`);
    const f = r.getRange(), S = s, p = r.getFilterColumn(s), _ = p == null ? void 0 : p.getColumnData().filters, C = new Set(_ == null ? void 0 : _.filters), F = !!(_ && _.blank), N = r.getFilteredOutRowsExceptCol(s), R = { ...f, startRow: f.startRow + 1, startColumn: S, endColumn: S };
    let O, A;
    if (i) {
      const P = await i.getFilterValues({
        unitId: o,
        subUnitId: l,
        filteredOutRowsByOtherColumns: Array.from(N),
        filterColumn: p,
        filters: !!_,
        blankChecked: F,
        iterateRange: R,
        alreadyChecked: Array.from(C)
      });
      O = P.filterTreeItems, A = P.filterTreeMapCache;
    } else {
      const P = Zt(
        !!_,
        a,
        R,
        h,
        N,
        p,
        C,
        F,
        c.getStyles()
      );
      O = P.filterTreeItems, A = P.filterTreeMapCache;
    }
    return e.createInstance($e, r, s, O, A);
  }
  get rawFilterItems() {
    return this._rawFilterItems$.getValue();
  }
  get filterItems() {
    return this._filterItems;
  }
  get treeMapCache() {
    return this._treeMapCache;
  }
  dispose() {
    this._rawFilterItems$.complete(), this._searchString$.complete();
  }
  deltaCol(e) {
    this.col += e;
  }
  setSearchString(e) {
    this._searchString$.next(e);
  }
  onCheckAllToggled(e) {
    const r = be.deepClone(this._filterItems);
    r.forEach((s) => ge(s, e)), this._manuallyUpdateFilterItems(r);
  }
  /**
   * Toggle a filter item.
   */
  onFilterCheckToggled(e) {
    const r = be.deepClone(this._filterItems), s = et(r, e.key);
    if (!s)
      return;
    const n = Kt(s);
    ge(s, !n), this._manuallyUpdateFilterItems(r);
  }
  onFilterOnly(e) {
    const r = be.deepClone(this._filterItems);
    r.forEach((s) => ge(s, !1)), e.forEach((s) => {
      const n = et(r, s);
      n && ge(n, !0);
    }), this._manuallyUpdateFilterItems(r);
  }
  _manuallyUpdateFilterItems(e) {
    this._manuallyUpdateFilterItems$.next(e);
  }
  // expose method here to let the panel change filter items
  // #region ByValuesModel apply methods
  clear() {
    return this._disposed ? Promise.resolve(!1) : this._commandService.executeCommand(V.id, {
      unitId: this._filterModel.unitId,
      subUnitId: this._filterModel.subUnitId,
      col: this.col,
      criteria: null
    });
  }
  /**
   * Apply the filter condition to the target filter column.
   */
  async apply() {
    if (this._disposed)
      return !1;
    const e = ze(this._filterItems), { checked: r, checkedItems: s } = e, n = this.rawFilterItems;
    let a = 0;
    for (const c of n)
      a += c.count;
    const i = r === 0, o = e.checked === a, l = { colId: this.col };
    if (i)
      throw new Error("[ByValuesModel]: no checked items!");
    if (o)
      return this._commandService.executeCommand(V.id, {
        unitId: this._filterModel.unitId,
        subUnitId: this._filterModel.subUnitId,
        col: this.col,
        criteria: null
      });
    {
      l.filters = {};
      const c = s.filter((f) => f.key !== "empty");
      c.length > 0 && (l.filters = {
        filters: c.flatMap((f) => f.originValues ? Array.from(f.originValues) : [f.title])
      }), c.length !== s.length && (l.filters.blank = !0);
    }
    return this._commandService.executeCommand(V.id, {
      unitId: this._filterModel.unitId,
      subUnitId: this._filterModel.subUnitId,
      col: this.col,
      criteria: l
    });
  }
  // #endregion
};
$e = Ve([
  Te(4, j)
], $e);
let Le = class extends ue {
  constructor(e, r, s, n, a) {
    super();
    v(this, "canApply$", he(!0));
    v(this, "_cellFillColors$");
    v(this, "cellFillColors$");
    v(this, "_cellTextColors$");
    v(this, "cellTextColors$");
    this._filterModel = e, this.col = r, this._commandService = a, this._cellFillColors$ = new Q(Array.from(s.values())), this.cellFillColors$ = this._cellFillColors$.asObservable(), this._cellTextColors$ = new Q(Array.from(n.values())), this.cellTextColors$ = this._cellTextColors$.asObservable();
  }
  /**
   * Create a model with targeting filter column. If there is not a filter column, the model would be created with
   * default values.
   *
   * @param injector
   * @param filterModel
   * @param col
   *
   * @returns the model to control the panel's state
   */
  static async fromFilterColumn(e, r, s) {
    var R, O, A;
    const n = e.get(De), { unitId: a, subUnitId: i } = r, o = n.getUniverSheetInstance(a);
    if (!o) throw new Error(`[ByColorsModel]: Workbook not found for filter model with unitId: ${a}!`);
    const l = o == null ? void 0 : o.getSheetBySheetId(i);
    if (!l) throw new Error(`[ByColorsModel]: Worksheet not found for filter model with unitId: ${a} and subUnitId: ${i}!`);
    const c = r.getRange(), h = s, f = (R = r.getFilterColumn(s)) == null ? void 0 : R.getColumnData().colorFilters, S = r.getFilteredOutRowsExceptCol(s), p = { ...c, startRow: c.startRow + 1, startColumn: h, endColumn: h }, _ = /* @__PURE__ */ new Map(), C = new Set((O = f == null ? void 0 : f.cellFillColors) != null ? O : []), F = /* @__PURE__ */ new Map(), N = new Set((A = f == null ? void 0 : f.cellTextColors) != null ? A : []);
    for (const P of l.iterateByColumn(p, !1, !0)) {
      const { row: Y, col: Z, value: se } = P;
      if (S.has(Y))
        continue;
      const M = l.getComposedCellStyleByCellData(Y, Z, se);
      if (M.bg && M.bg.rgb) {
        const E = new pt(M.bg.rgb).toRgbString();
        _.has(E) || _.set(E, { color: E, checked: C.has(E) });
      } else
        _.set("default-fill-color", { color: null, checked: C.has(null) });
      if (M.cl && M.cl.rgb) {
        const E = new pt(M.cl.rgb).toRgbString();
        F.has(E) || F.set(E, { color: E, checked: N.has(E) });
      } else
        F.set("default-font-color", { color: St, checked: N.has(St) });
    }
    return e.createInstance(Le, r, s, _, F);
  }
  get cellFillColors() {
    return this._cellFillColors$.getValue();
  }
  get cellTextColors() {
    return this._cellTextColors$.getValue();
  }
  dispose() {
    super.dispose(), this._cellFillColors$.complete();
  }
  deltaCol(e) {
    this.col += e;
  }
  // expose method here to let the panel change filter items
  // #region ByColorsModel apply methods
  clear() {
    return this._disposed ? Promise.resolve(!1) : this._commandService.executeCommand(V.id, {
      unitId: this._filterModel.unitId,
      subUnitId: this._filterModel.subUnitId,
      col: this.col,
      criteria: null
    });
  }
  onFilterCheckToggled(e, r = !0) {
    const s = r ? this.cellFillColors : this.cellTextColors, n = [];
    let a = !1;
    for (let i = 0; i < s.length; i++) {
      const o = s[i];
      if (o.color === e.color) {
        a = !0, n.push({
          color: o.color,
          checked: !o.checked
        });
        continue;
      }
      n.push({
        color: o.color,
        checked: o.checked
      });
    }
    a && (this._resetColorsCheckedStatus(!r), r ? this._cellFillColors$.next([...n]) : this._cellTextColors$.next([...n]));
  }
  _resetColorsCheckedStatus(e = !0) {
    const r = e ? this.cellFillColors : this.cellTextColors, s = [];
    for (let n = 0; n < r.length; n++)
      s.push({
        color: r[n].color,
        checked: !1
      });
    e ? this._cellFillColors$.next([...s]) : this._cellTextColors$.next([...s]);
  }
  /**
   * Apply the filter condition to the target filter column.
   */
  async apply() {
    if (this._disposed)
      return !1;
    const e = this.cellFillColors.filter((n) => n.checked).map((n) => n.color), r = this.cellTextColors.filter((n) => n.checked).map((n) => n.color);
    if (e.length === 0 && r.length === 0)
      return this._commandService.executeCommand(V.id, {
        unitId: this._filterModel.unitId,
        subUnitId: this._filterModel.subUnitId,
        col: this.col,
        criteria: null
      });
    const s = { colId: this.col };
    return e.length > 0 ? s.colorFilters = {
      cellFillColors: e
    } : r.length > 0 && (s.colorFilters = {
      cellTextColors: r
    }), this._commandService.executeCommand(V.id, {
      unitId: this._filterModel.unitId,
      subUnitId: this._filterModel.subUnitId,
      col: this.col,
      criteria: s
    });
  }
  // #endregion
};
Le = Ve([
  Te(4, j)
], Le);
const ae = "FILTER_PANEL_OPENED", Me = {
  id: "sheet.operation.open-filter-panel",
  type: Be.OPERATION,
  handler: (t, e) => {
    const r = t.get(We), s = t.get(z), n = t.get(X), a = t.get(j), i = t.has(_t) ? t.get(_t) : null;
    i != null && i.isVisible().visible && a.syncExecuteCommand(Nr.id, { visible: !1 });
    const { unitId: o, subUnitId: l, col: c } = e, h = s.getFilterModel(o, l);
    return h ? (n.setupCol(h, c), r.getContextValue(ae) || r.setContextValue(ae, !0), !0) : !1;
  }
}, Se = {
  id: "sheet.operation.close-filter-panel",
  type: Be.OPERATION,
  handler: (t) => {
    const e = t.get(We), r = t.get(X), s = t.get(Mr, $t.OPTIONAL);
    return e.getContextValue(ae) ? (e.setContextValue(ae, !1), s == null || s.focus(), r.terminate()) : !1;
  }
}, Xt = {
  id: "sheet.operation.apply-filter",
  type: Be.OPERATION,
  handler: (t, e) => {
    const { filterBy: r } = e;
    return t.get(X).changeFilterBy(r);
  }
}, zt = "sheets-filter-ui.config", xe = {};
var Ci = Object.getOwnPropertyDescriptor, Ti = (t, e, r, s) => {
  for (var n = s > 1 ? void 0 : s ? Ci(e, r) : e, a = t.length - 1, i; a >= 0; a--)
    (i = t[a]) && (n = i(n) || n);
  return n;
}, oe = (t, e) => (r, s) => e(r, s, t);
let ce = class extends ue {
  constructor(t, e, r, s, n, a) {
    super(), this._sheetsFilterService = t, this._localeService = e, this._commandService = r, this._sheetPermissionCheckPermission = s, this._injector = n, this._sheetsSelectionService = a, this._commandExecutedListener();
  }
  _commandExecutedListener() {
    this.disposeWithMe(
      this._commandService.beforeCommandExecuted((t) => {
        var e, r, s;
        if (t.id === Fe.id) {
          const n = this._injector.get(De), a = Yr(n);
          if (!a) return;
          const { unitId: i, subUnitId: o, worksheet: l } = a, c = (e = this._sheetsFilterService.getFilterModel(i, o)) == null ? void 0 : e.getRange();
          let h;
          if (c)
            h = this._sheetPermissionCheckPermission.permissionCheckWithRanges({
              rangeTypes: [_e],
              worksheetTypes: [pe, ve]
            }, [c], i, o);
          else {
            const f = (r = this._sheetsSelectionService.getCurrentLastSelection()) == null ? void 0 : r.range;
            if (f) {
              let S = { ...f };
              S = f.startColumn === f.endColumn && f.startRow === f.endRow ? Zr(S, { left: !0, right: !0, up: !0, down: !0 }, l) : S, h = this._sheetPermissionCheckPermission.permissionCheckWithRanges({
                rangeTypes: [_e],
                worksheetTypes: [ve, pe]
              }, [S], i, o);
            } else
              h = this._sheetPermissionCheckPermission.permissionCheckWithoutRange({
                rangeTypes: [_e],
                worksheetTypes: [ve, pe]
              });
          }
          h || this._sheetPermissionCheckPermission.blockExecuteWithoutPermission(this._localeService.t("permission.dialog.filterErr"));
        }
        if (t.id === Me.id) {
          const n = t.params, { unitId: a, subUnitId: i } = n, o = (s = this._sheetsFilterService.getFilterModel(a, i)) == null ? void 0 : s.getRange(), l = be.deepClone(o);
          l && (l.startColumn = n.col, l.endColumn = n.col, this._sheetPermissionCheckPermission.permissionCheckWithRanges({
            rangeTypes: [_e],
            worksheetTypes: [pe, ve]
          }, [l], a, i) || this._sheetPermissionCheckPermission.blockExecuteWithoutPermission(this._localeService.t("permission.dialog.filterErr")));
        }
      })
    );
  }
};
ce = Ti([
  oe(0, T(z)),
  oe(1, T(G)),
  oe(2, j),
  oe(3, T(Gr)),
  oe(4, T(re)),
  oe(5, T(jr))
], ce);
const q = 16, Ei = new Path2D("M3.30363 3C2.79117 3 2.51457 3.60097 2.84788 3.99024L6.8 8.60593V12.5662C6.8 12.7184 6.8864 12.8575 7.02289 12.9249L8.76717 13.7863C8.96655 13.8847 9.2 13.7396 9.2 13.5173V8.60593L13.1521 3.99024C13.4854 3.60097 13.2088 3 12.6964 3H3.30363Z");
class It {
  static drawNoCriteria(e, r, s, n) {
    e.save(), Ct.drawWith(e, {
      radius: 2,
      width: q,
      height: q,
      fill: n
    }), e.lineCap = "square", e.strokeStyle = s, e.scale(r / q, r / q), e.beginPath(), e.lineWidth = 1, e.lineCap = "round", e.moveTo(3, 4), e.lineTo(13, 4), e.moveTo(4.5, 8), e.lineTo(11.5, 8), e.moveTo(6, 12), e.lineTo(10, 12), e.stroke(), e.restore();
  }
  static drawHasCriteria(e, r, s, n) {
    e.save(), Ct.drawWith(e, {
      radius: 2,
      width: q,
      height: q,
      fill: n
    }), e.scale(r / q, r / q), e.fillStyle = s, e.fill(Ei), e.restore();
  }
}
var Fi = Object.getOwnPropertyDescriptor, Ni = (t, e, r, s) => {
  for (var n = s > 1 ? void 0 : s ? Fi(e, r) : e, a = t.length - 1, i; a >= 0; a--)
    (i = t[a]) && (n = i(n) || n);
  return n;
}, qe = (t, e) => (r, s) => e(r, s, t);
const B = 16, Ce = 1;
let tt = class extends Vr {
  constructor(e, r, s, n, a) {
    super(e, r);
    v(this, "_cellWidth", 0);
    v(this, "_cellHeight", 0);
    v(this, "_filterParams");
    v(this, "_hovered", !1);
    this._contextService = s, this._commandService = n, this._themeService = a, this.setShapeProps(r), this.onPointerDown$.subscribeEvent((i) => this.onPointerDown(i)), this.onPointerEnter$.subscribeEvent(() => this.onPointerEnter()), this.onPointerLeave$.subscribeEvent(() => this.onPointerLeave());
  }
  setShapeProps(e) {
    typeof e.cellHeight < "u" && (this._cellHeight = e.cellHeight), typeof e.cellWidth < "u" && (this._cellWidth = e.cellWidth), typeof e.filterParams < "u" && (this._filterParams = e.filterParams), this.transformByState({
      width: e.width,
      height: e.height
    });
  }
  _draw(e) {
    const r = this._cellHeight, s = this._cellWidth, n = B - s, a = B - r;
    e.save();
    const i = new Path2D();
    i.rect(n, a, s, r), e.clip(i);
    const { hasCriteria: o } = this._filterParams, l = this._themeService.getColorFromTheme("primary.600"), c = this._hovered ? this._themeService.getColorFromTheme("gray.50") : "rgba(255, 255, 255, 1.0)";
    o ? It.drawHasCriteria(e, B, l, c) : It.drawNoCriteria(e, B, l, c), e.restore();
  }
  onPointerDown(e) {
    if (e.button === 2)
      return;
    const { col: r, unitId: s, subUnitId: n } = this._filterParams;
    this._contextService.getContextValue(ae) || !this._commandService.hasCommand(Me.id) || setTimeout(() => {
      this._commandService.executeCommand(Me.id, {
        unitId: s,
        subUnitId: n,
        col: r
      });
    }, 200);
  }
  onPointerEnter() {
    this._hovered = !0, this.makeDirty(!0);
  }
  onPointerLeave() {
    this._hovered = !1, this.makeDirty(!0);
  }
};
tt = Ni([
  qe(2, We),
  qe(3, j),
  qe(4, T(Lt))
], tt);
var Ii = Object.getOwnPropertyDescriptor, yi = (t, e, r, s) => {
  for (var n = s > 1 ? void 0 : s ? Ii(e, r) : e, a = t.length - 1, i; a >= 0; a--)
    (i = t[a]) && (n = i(n) || n);
  return n;
}, te = (t, e) => (r, s) => e(r, s, t);
const Oi = 1e3, bi = 5e3;
function Ri(t, e, r, s) {
  switch (s) {
    case Re.TOP:
      return t + Ce;
    case Re.MIDDLE:
      return t + Math.max(0, (r - B) / 2);
    case Re.BOTTOM:
    default:
      return e - B - Ce;
  }
}
let rt = class extends Mt {
  constructor(e, r, s, n, a, i, o, l) {
    super();
    v(this, "_filterRangeShape", null);
    v(this, "_buttonRenderDisposable", null);
    v(this, "_filterButtonShapes", []);
    this._context = e, this._injector = r, this._sheetSkeletonManagerService = s, this._sheetsFilterService = n, this._themeService = a, this._sheetInterceptorService = i, this._commandService = o, this._selectionRenderService = l, this._initRenderer();
  }
  dispose() {
    super.dispose(), this._disposeRendering();
  }
  _initRenderer() {
    this._sheetSkeletonManagerService.currentSkeleton$.pipe(
      at((e) => {
        var o, l;
        if (!e) return he(null);
        const { unit: r, unitId: s } = this._context, n = ((o = r.getActiveSheet()) == null ? void 0 : o.getSheetId()) || "", a = (l = this._sheetsFilterService.getFilterModel(s, n)) != null ? l : void 0, i = () => ({
          unitId: s,
          worksheetId: n,
          filterModel: a,
          range: a == null ? void 0 : a.getRange(),
          skeleton: e.skeleton
        });
        return dr(this._commandService.onCommandExecuted.bind(this._commandService)).pipe(
          ii(
            ([c]) => {
              var h;
              return c.type === Be.MUTATION && ((h = c.params) == null ? void 0 : h.unitId) === r.getUnitId() && (vr.has(c.id) || c.id === qr.id);
            }
          ),
          Vt(20, void 0, { leading: !1, trailing: !0 }),
          le(i),
          Qt(i())
          // must trigger once
        );
      }),
      si(this.dispose$)
    ).subscribe((e) => {
      this._disposeRendering(), !(!e || !e.range) && (this._renderRange(e.range, e.skeleton), this._renderButtons(e));
    });
  }
  _renderRange(e, r) {
    const { scene: s } = this._context, { rowHeaderWidth: n, columnHeaderHeight: a } = r, i = this._filterRangeShape = new Or(
      s,
      Oi,
      this._themeService,
      {
        rowHeaderWidth: n,
        columnHeaderHeight: a,
        enableAutoFill: !1,
        highlightHeader: !1
      }
    ), l = br({
      range: e,
      primary: null,
      style: { fill: "rgba(0, 0, 0, 0.0)" }
    }, r);
    i.updateRangeBySelectionWithCoord(l), i.setEvent(!1), s.makeDirty(!0);
  }
  _renderButtons(e) {
    const { range: r, filterModel: s, unitId: n, skeleton: a, worksheetId: i } = e, { unit: o, scene: l } = this._context, c = o.getSheetBySheetId(i);
    if (!c)
      return;
    this._interceptCellContent(n, i, e.range);
    const { startColumn: h, endColumn: f, startRow: S } = r;
    for (let p = h; p <= f; p++) {
      const _ = `sheets-filter-button-${p}`, C = Rr(S, p, l, a), F = c.getComposedCellStyle(S, p), N = (F == null ? void 0 : F.vt) || Re.BOTTOM, { startX: R, startY: O, endX: A, endY: P } = C, Y = A - R, Z = P - O;
      if (Z <= Ce || Y <= Ce)
        continue;
      const se = !!s.getFilterColumn(p), M = A - B - Ce, E = Ri(O, P, Z, N), Qe = {
        left: M,
        top: E,
        height: B,
        width: B,
        zIndex: bi,
        cellHeight: Z,
        cellWidth: Y,
        filterParams: { unitId: n, subUnitId: i, col: p, hasCriteria: se }
      }, Ie = this._injector.createInstance(tt, _, Qe);
      this._filterButtonShapes.push(Ie);
    }
    l.addObjects(this._filterButtonShapes), l.makeDirty();
  }
  _interceptCellContent(e, r, s) {
    const { startRow: n, startColumn: a, endColumn: i } = s;
    this._buttonRenderDisposable = this._sheetInterceptorService.intercept(Xr.CELL_CONTENT, {
      effect: fr.Style,
      handler: (o, l, c) => {
        const { row: h, col: f, unitId: S, subUnitId: p } = l;
        return S !== e || p !== r || h !== n || f < a || f > i || ((!o || o === l.rawData) && (o = { ...l.rawData }), o.fontRenderExtension = {
          ...o == null ? void 0 : o.fontRenderExtension,
          rightOffset: B
        }), c(o);
      },
      priority: 10
    });
  }
  _disposeRendering() {
    var e, r;
    (e = this._filterRangeShape) == null || e.dispose(), this._filterButtonShapes.forEach((s) => s.dispose()), (r = this._buttonRenderDisposable) == null || r.dispose(), this._filterRangeShape = null, this._buttonRenderDisposable = null, this._filterButtonShapes = [];
  }
};
rt = yi([
  te(1, T(re)),
  te(2, T(Ir)),
  te(3, T(z)),
  te(4, T(Lt)),
  te(5, T(Kr)),
  te(6, j),
  te(7, yr)
], rt);
var Ai = Object.getOwnPropertyDescriptor, wi = (t, e, r, s) => {
  for (var n = s > 1 ? void 0 : s ? Ai(e, r) : e, a = t.length - 1, i; a >= 0; a--)
    (i = t[a]) && (n = i(n) || n);
  return n;
}, yt = (t, e) => (r, s) => e(r, s, t);
let Ee = class extends Mt {
  constructor(t, e) {
    super(), this._renderManagerService = t, this._sheetsRenderService = e, [
      _r,
      gr,
      Sr,
      Cr
    ].forEach((r) => this.disposeWithMe(this._sheetsRenderService.registerSkeletonChangingMutations(r.id))), this.disposeWithMe(this._renderManagerService.registerRenderModule(
      ie.UNIVER_SHEET,
      [rt]
    ));
  }
};
Ee = wi([
  yt(0, Wt),
  yt(1, T(Dt))
], Ee);
var Pi = Object.defineProperty, $i = Object.getOwnPropertyDescriptor, Li = (t, e, r) => e in t ? Pi(t, e, { enumerable: !0, configurable: !0, writable: !0, value: r }) : t[e] = r, Mi = (t, e, r, s) => {
  for (var n = s > 1 ? void 0 : s ? $i(e, r) : e, a = t.length - 1, i; a >= 0; a--)
    (i = t[a]) && (n = i(n) || n);
  return n;
}, Ot = (t, e) => (r, s) => e(r, s, t), Jt = (t, e, r) => Li(t, typeof e != "symbol" ? e + "" : e, r);
const xi = "SHEET_FILTER_UI_PLUGIN";
let Ue = class extends it {
  constructor(t = xe, e, r) {
    super(), this._config = t, this._injector = e, this._configService = r;
    const { menu: s, ...n } = kt(
      {},
      xe,
      this._config
    );
    s && this._configService.setConfig("menu", s, { merge: !0 }), this._configService.setConfig(zt, n);
  }
  onStarting() {
    [
      [ce],
      [Ee]
    ].forEach((t) => this._injector.add(t));
  }
  onReady() {
    this._injector.get(ce);
  }
  onRendered() {
    this._injector.get(Ee);
  }
};
Jt(Ue, "type", ie.UNIVER_SHEET);
Jt(Ue, "pluginName", xi);
Ue = Mi([
  xt(Ht),
  Ot(1, T(re)),
  Ot(2, Ut)
], Ue);
function fe({ ref: t, ...e }) {
  const { icon: r, id: s, className: n, extend: a, ...i } = e, o = `univerjs-icon univerjs-icon-${s} ${n || ""}`.trim(), l = mi(`_${Hi()}`);
  return er(r, `${s}`, {
    defIds: r.defIds,
    idSuffix: l.current
  }, {
    ref: t,
    className: o,
    ...i
  }, a);
}
function er(t, e, r, s, n) {
  return de(t.tag, {
    key: e,
    ...Ui(t, r, n),
    ...s
  }, (ki(t, r).children || []).map((a, i) => er(a, `${e}-${t.tag}-${i}`, r, void 0, n)));
}
function Ui(t, e, r) {
  const s = { ...t.attrs };
  r != null && r.colorChannel1 && s.fill === "colorChannel1" && (s.fill = r.colorChannel1), t.tag === "mask" && s.id && (s.id = s.id + e.idSuffix), Object.entries(s).forEach(([a, i]) => {
    a === "mask" && typeof i == "string" && (s[a] = i.replace(/url\(#(.*)\)/, `url(#$1${e.idSuffix})`));
  });
  const { defIds: n } = e;
  return !n || n.length === 0 || (t.tag === "use" && s["xlink:href"] && (s["xlink:href"] = s["xlink:href"] + e.idSuffix), Object.entries(s).forEach(([a, i]) => {
    typeof i == "string" && (s[a] = i.replace(/url\(#(.*)\)/, `url(#$1${e.idSuffix})`));
  })), s;
}
function ki(t, e) {
  var s;
  const { defIds: r } = e;
  return !r || r.length === 0 ? t : t.tag === "defs" && ((s = t.children) != null && s.length) ? {
    ...t,
    children: t.children.map((n) => typeof n.attrs.id == "string" && r && r.includes(n.attrs.id) ? {
      ...n,
      attrs: {
        ...n.attrs,
        id: n.attrs.id + e.idSuffix
      }
    } : n)
  } : t;
}
function Hi() {
  return Math.random().toString(36).substring(2, 8);
}
fe.displayName = "UniverIcon";
const Di = {
  tag: "svg",
  attrs: {
    xmlns: "http://www.w3.org/2000/svg",
    fill: "none",
    viewBox: "0 0 20 20",
    width: "1em",
    height: "1em"
  },
  children: [{
    tag: "path",
    attrs: {
      fill: "currentColor",
      d: "M10 1.05957C10.356 1.05957 10.6816 1.26162 10.8408 1.58008L18.8408 17.5801L18.8799 17.668C19.0486 18.1134 18.8551 18.6232 18.4199 18.8408C17.9557 19.0727 17.3913 18.8841 17.1592 18.4199L10 4.10156L2.84082 18.4199C2.60871 18.8841 2.04434 19.0727 1.58008 18.8408C1.11587 18.6087 0.92731 18.0443 1.15918 17.5801L9.15918 1.58008C9.31841 1.26162 9.64395 1.05957 10 1.05957Z"
    }
  }, {
    tag: "path",
    attrs: {
      fill: "currentColor",
      d: "M15.3337 11.7261L15.4294 11.731C15.9035 11.779 16.2732 12.1798 16.2732 12.6665C16.2732 13.1532 15.9035 13.554 15.4294 13.602L15.3337 13.6069H4.66675C4.1476 13.6069 3.72632 13.1856 3.72632 12.6665C3.72632 12.1474 4.1476 11.7261 4.66675 11.7261H15.3337Z"
    }
  }]
}, tr = Ne(function(e, r) {
  return de(fe, Object.assign({}, e, {
    id: "a-icon",
    ref: r,
    icon: Di
  }));
});
tr.displayName = "AIcon";
const Bi = {
  tag: "svg",
  attrs: {
    xmlns: "http://www.w3.org/2000/svg",
    fill: "none",
    viewBox: "0 0 20 20",
    width: "1em",
    height: "1em"
  },
  children: [{
    tag: "path",
    attrs: {
      fill: "currentColor",
      d: "M17.0596 10C17.0596 6.10087 13.8992 2.94043 10 2.94043C6.10087 2.94043 2.94043 6.10087 2.94043 10C2.94043 13.8992 6.10087 17.0596 10 17.0596C13.8992 17.0596 17.0596 13.8992 17.0596 10ZM18.9404 10C18.9404 14.9374 14.9374 18.9404 10 18.9404C5.06257 18.9404 1.05957 14.9374 1.05957 10C1.05957 5.06257 5.06257 1.05957 10 1.05957C14.9374 1.05957 18.9404 5.06257 18.9404 10Z"
    }
  }, {
    tag: "path",
    attrs: {
      fill: "currentColor",
      d: "M4.29492 4.13476C4.63911 3.79057 5.1845 3.76906 5.55371 4.07031L5.625 4.13476L16.0244 14.5352L16.0889 14.6064C16.3902 14.9757 16.3686 15.52 16.0244 15.8643C15.6573 16.2313 15.0624 16.2313 14.6953 15.8643L4.29492 5.46484L4.23047 5.39355C3.92922 5.02434 3.95073 4.47895 4.29492 4.13476Z"
    }
  }]
}, rr = Ne(function(e, r) {
  return de(fe, Object.assign({}, e, {
    id: "ban-icon",
    ref: r,
    icon: Bi
  }));
});
rr.displayName = "BanIcon";
const Wi = {
  tag: "svg",
  attrs: {
    xmlns: "http://www.w3.org/2000/svg",
    fill: "none",
    viewBox: "0 0 17 16",
    width: "1em",
    height: "1em"
  },
  children: [{
    tag: "path",
    attrs: {
      fill: "currentColor",
      d: "M3.32182 2.60967C2.98161 2.60967 2.79671 3.0074 3.01601 3.2675L6.85819 7.8246C6.94943 7.93282 6.99947 8.06981 6.99947 8.21136V12.7338C6.99947 12.898 7.0998 13.0455 7.2525 13.1058L8.73833 13.6928C9.00085 13.7965 9.28531 13.6031 9.28531 13.3208V8.21136C9.28531 8.06981 9.33535 7.93282 9.42659 7.8246L13.2688 3.2675C13.4881 3.0074 13.3032 2.60967 12.963 2.60967H3.32182ZM2.09858 4.04101C1.22139 3.0006 1.96097 1.40967 3.32182 1.40967H12.963C14.3238 1.40967 15.0634 3.0006 14.1862 4.04101L10.4853 8.43054V13.3208C10.4853 14.4498 9.34747 15.2237 8.29742 14.8089L6.81158 14.2219C6.20078 13.9806 5.79947 13.3905 5.79947 12.7338V8.43054L2.09858 4.04101Z",
      fillRule: "evenodd",
      clipRule: "evenodd"
    }
  }]
}, ir = Ne(function(e, r) {
  return de(fe, Object.assign({}, e, {
    id: "filter-icon",
    ref: r,
    icon: Wi
  }));
});
ir.displayName = "FilterIcon";
const Vi = {
  tag: "svg",
  attrs: {
    xmlns: "http://www.w3.org/2000/svg",
    fill: "none",
    viewBox: "0 0 16 16",
    width: "1em",
    height: "1em"
  },
  children: [{
    tag: "path",
    attrs: {
      fill: "currentColor",
      d: "M8.00016 1.33203C6.68162 1.33203 5.39269 1.72302 4.29636 2.45557C3.20004 3.18811 2.34555 4.2293 1.84097 5.44747C1.33638 6.66565 1.20436 8.00609 1.4616 9.2993C1.71883 10.5925 2.35377 11.7804 3.28612 12.7127C4.21847 13.6451 5.40636 14.28 6.69956 14.5373C7.99277 14.7945 9.33321 14.6625 10.5514 14.1579C11.7696 13.6533 12.8108 12.7988 13.5433 11.7025C14.2758 10.6062 14.6668 9.31724 14.6668 7.9987C14.6649 6.23118 13.9619 4.53662 12.7121 3.2868C11.4622 2.03697 9.76768 1.33397 8.00016 1.33203ZM7.66683 3.9987C7.86461 3.9987 8.05795 4.05735 8.2224 4.16723C8.38685 4.27711 8.51502 4.43329 8.59071 4.61601C8.6664 4.79874 8.6862 4.99981 8.64762 5.19379C8.60903 5.38777 8.51379 5.56595 8.37394 5.7058C8.23409 5.84566 8.0559 5.9409 7.86192 5.97948C7.66794 6.01807 7.46687 5.99826 7.28415 5.92258C7.10142 5.84689 6.94524 5.71872 6.83536 5.55427C6.72548 5.38982 6.66683 5.19648 6.66683 4.9987C6.66683 4.73348 6.77219 4.47913 6.95972 4.29159C7.14726 4.10405 7.40162 3.9987 7.66683 3.9987ZM9.3335 11.332H6.66683C6.49002 11.332 6.32045 11.2618 6.19543 11.1368C6.0704 11.0117 6.00016 10.8422 6.00016 10.6654C6.00016 10.4886 6.0704 10.319 6.19543 10.194C6.32045 10.0689 6.49002 9.9987 6.66683 9.9987H7.3335V7.9987H6.66683C6.49002 7.9987 6.32045 7.92846 6.19543 7.80343C6.0704 7.67841 6.00016 7.50884 6.00016 7.33203C6.00016 7.15522 6.0704 6.98565 6.19543 6.86063C6.32045 6.7356 6.49002 6.66536 6.66683 6.66536H8.00016C8.17698 6.66536 8.34655 6.7356 8.47157 6.86063C8.59659 6.98565 8.66683 7.15522 8.66683 7.33203V9.9987H9.3335C9.51031 9.9987 9.67988 10.0689 9.8049 10.194C9.92993 10.319 10.0002 10.4886 10.0002 10.6654C10.0002 10.8422 9.92993 11.0117 9.8049 11.1368C9.67988 11.2618 9.51031 11.332 9.3335 11.332Z"
    }
  }]
}, sr = Ne(function(e, r) {
  return de(fe, Object.assign({}, e, {
    id: "info-icon",
    ref: r,
    icon: Vi
  }));
});
sr.displayName = "InfoIcon";
const Qi = {
  tag: "svg",
  attrs: {
    xmlns: "http://www.w3.org/2000/svg",
    fill: "none",
    viewBox: "0 0 16 16",
    width: "1em",
    height: "1em"
  },
  children: [{
    tag: "path",
    attrs: {
      fill: "currentColor",
      d: "M8 15C11.866 15 15 11.866 15 8C15 4.13401 11.866 1 8 1C4.13401 1 1 4.13401 1 8C1 11.866 4.13401 15 8 15ZM11.7245 6.42417C11.9588 6.18985 11.9588 5.80995 11.7245 5.57564C11.4901 5.34132 11.1102 5.34132 10.8759 5.57564L7.3002 9.15137L5.72446 7.57564C5.49014 7.34132 5.11025 7.34132 4.87593 7.57564C4.64162 7.80995 4.64162 8.18985 4.87593 8.42417L6.87593 10.4242C7.11025 10.6585 7.49014 10.6585 7.72446 10.4242L11.7245 6.42417Z",
      fillRule: "evenodd",
      clipRule: "evenodd"
    }
  }]
}, nr = Ne(function(e, r) {
  return de(fe, Object.assign({}, e, {
    id: "success-icon",
    ref: r,
    icon: Qi
  }));
});
nr.displayName = "SuccessIcon";
function Gi(t) {
  const { model: e } = t, r = D(G), s = L(e.cellFillColors$, [], !0), n = L(e.cellTextColors$, [], !0), a = U((o) => {
    e.onFilterCheckToggled(o);
  }, [e]), i = U((o) => {
    e.onFilterCheckToggled(o, !1);
  }, [e]);
  return /* @__PURE__ */ d(
    "div",
    {
      "data-u-comp": "sheets-filter-panel-colors-container",
      className: "univer-flex univer-h-full univer-min-h-[300px] univer-flex-col",
      children: /* @__PURE__ */ y(
        "div",
        {
          "data-u-comp": "sheets-filter-panel",
          className: Ae("univer-mt-2 univer-box-border univer-flex univer-h-[300px] univer-flex-grow univer-flex-col univer-gap-4 univer-overflow-auto univer-rounded-md univer-px-2 univer-py-2.5", ct),
          children: [
            s.length > 1 && /* @__PURE__ */ y("div", { children: [
              /* @__PURE__ */ d(
                "div",
                {
                  className: "univer-mb-2 univer-text-sm univer-text-gray-900 dark:!univer-text-white",
                  children: r.t("sheets-filter.panel.filter-by-cell-fill-color")
                }
              ),
              /* @__PURE__ */ d(
                "div",
                {
                  className: "univer-grid univer-grid-cols-8 univer-items-center univer-justify-start univer-gap-2",
                  children: s.map((o, l) => /* @__PURE__ */ y(
                    "div",
                    {
                      className: "univer-relative univer-h-6 univer-w-6",
                      onClick: () => a(o),
                      children: [
                        o.color ? /* @__PURE__ */ d(
                          "button",
                          {
                            type: "button",
                            className: Ae("univer-box-border univer-h-6 univer-w-6 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow hover:univer-ring-2 hover:univer-ring-offset-2 hover:univer-ring-offset-white"),
                            style: { backgroundColor: o.color }
                          }
                        ) : /* @__PURE__ */ d(
                          rr,
                          {
                            className: "univer-h-6 univer-w-6 univer-cursor-pointer univer-rounded-full hover:univer-ring-2 hover:univer-ring-offset-2 hover:univer-ring-offset-white"
                          }
                        ),
                        o.checked && /* @__PURE__ */ d(bt, {})
                      ]
                    },
                    `sheets-filter-cell-fill-color-${l}`
                  ))
                }
              )
            ] }),
            n.length > 1 && /* @__PURE__ */ y("div", { children: [
              /* @__PURE__ */ d(
                "div",
                {
                  className: "univer-mb-2 univer-text-sm univer-text-gray-900 dark:!univer-text-white",
                  children: r.t("sheets-filter.panel.filter-by-cell-text-color")
                }
              ),
              /* @__PURE__ */ d(
                "div",
                {
                  className: "univer-grid univer-grid-cols-8 univer-items-center univer-justify-start univer-gap-2",
                  children: n.map((o, l) => /* @__PURE__ */ y(
                    "div",
                    {
                      className: "univer-relative univer-h-6 univer-w-6",
                      onClick: () => i(o),
                      children: [
                        /* @__PURE__ */ d(
                          "div",
                          {
                            className: "univer-box-border univer-flex univer-h-full univer-w-full univer-cursor-pointer univer-items-center univer-justify-center univer-rounded-full univer-border univer-border-solid univer-border-[rgba(13,13,13,0.06)] univer-p-0.5 hover:univer-ring-2 hover:univer-ring-offset-2 hover:univer-ring-offset-white dark:!univer-border-[rgba(255,255,255,0.06)]",
                            children: /* @__PURE__ */ d(tr, { style: { color: o.color } })
                          }
                        ),
                        o.checked && /* @__PURE__ */ d(bt, {})
                      ]
                    },
                    `sheets-filter-cell-text-color-${l}`
                  ))
                }
              )
            ] }),
            s.length <= 1 && n.length <= 1 && /* @__PURE__ */ d(
              "div",
              {
                className: "univer-flex univer-h-full univer-w-full univer-items-center univer-justify-center univer-text-sm univer-text-gray-900 dark:!univer-text-gray-200",
                children: r.t("sheets-filter.panel.filter-by-color-none")
              }
            )
          ]
        }
      )
    }
  );
}
function bt() {
  return /* @__PURE__ */ d(
    "div",
    {
      className: "univer-absolute -univer-bottom-0.5 -univer-right-0.5 univer-flex univer-h-3 univer-w-3 univer-cursor-pointer univer-items-center univer-justify-center univer-rounded-full univer-bg-white",
      children: /* @__PURE__ */ d(
        nr,
        {
          className: "univer-h-full univer-w-full univer-font-bold univer-text-[#418F1F]"
        }
      )
    }
  );
}
function ji(t) {
  var p, _;
  const { model: e } = t, r = D(G), s = L(e.conditionItem$, void 0), n = L(e.filterConditionFormParams$, void 0), a = n != null && n.and ? "AND" : "OR", i = U((C) => {
    e.onConditionFormChange({ and: C === "AND" });
  }, [e]), o = Yi(r), l = U((C) => {
    e.onPrimaryConditionChange(C);
  }, [e]), c = Zi(r), h = U((C) => {
    e.onConditionFormChange(C);
  }, [e]), f = r.t("sheets-filter.panel.input-values-placeholder");
  function S(C, F, N) {
    const R = m.getItemByOperator(C).numOfParameters === 1;
    return /* @__PURE__ */ y(Ft, { children: [
      N === "operator2" && /* @__PURE__ */ y(ai, { value: a, onChange: i, children: [
        /* @__PURE__ */ d(Et, { value: "AND", children: r.t("sheets-filter.panel.and") }),
        /* @__PURE__ */ d(Et, { value: "OR", children: r.t("sheets-filter.panel.or") })
      ] }),
      /* @__PURE__ */ d(
        Tt,
        {
          value: C,
          options: c,
          onChange: (O) => h({ [N]: O })
        }
      ),
      R && /* @__PURE__ */ d("div", { children: /* @__PURE__ */ d(
        jt,
        {
          className: "univer-mt-2",
          value: F,
          placeholder: f,
          onChange: (O) => h({ [N === "operator1" ? "val1" : "val2"]: O })
        }
      ) })
    ] });
  }
  return /* @__PURE__ */ d(
    "div",
    {
      "data-u-comp": "sheets-filter-panel-conditions-container",
      className: "univer-flex univer-h-full univer-min-h-[300px] univer-flex-col",
      children: s && n && /* @__PURE__ */ y(Ft, { children: [
        /* @__PURE__ */ d(Tt, { value: s.operator, options: o, onChange: l }),
        m.getItemByOperator(s.operator).numOfParameters !== 0 ? /* @__PURE__ */ y(
          "div",
          {
            "data-u-comp": "sheets-filter-panel-conditions-container-inner",
            className: Ae("univer-mt-2 univer-flex-grow univer-overflow-hidden univer-rounded-md univer-p-2", ct),
            children: [
              s.numOfParameters >= 1 && S(n.operator1, (p = n.val1) != null ? p : "", "operator1"),
              s.numOfParameters >= 2 && S(n.operator2, (_ = n.val2) != null ? _ : "", "operator2"),
              /* @__PURE__ */ y(
                "div",
                {
                  "data-u-comp": "sheets-filter-panel-conditions-desc",
                  className: "univer-mt-2 univer-text-xs univer-text-gray-500",
                  children: [
                    r.t("sheets-filter.panel.?"),
                    /* @__PURE__ */ d("br", {}),
                    r.t("sheets-filter.panel.*")
                  ]
                }
              )
            ]
          }
        ) : null
      ] })
    }
  );
}
function Yi(t) {
  const e = t.getCurrentLocale();
  return ut(() => [
    {
      options: [
        { label: t.t(m.NONE.label), value: m.NONE.operator }
      ]
    },
    {
      options: [
        { label: t.t(m.EMPTY.label), value: m.EMPTY.operator },
        { label: t.t(m.NOT_EMPTY.label), value: m.NOT_EMPTY.operator }
      ]
    },
    {
      options: [
        { label: t.t(m.TEXT_CONTAINS.label), value: m.TEXT_CONTAINS.operator },
        { label: t.t(m.DOES_NOT_CONTAIN.label), value: m.DOES_NOT_CONTAIN.operator },
        { label: t.t(m.STARTS_WITH.label), value: m.STARTS_WITH.operator },
        { label: t.t(m.ENDS_WITH.label), value: m.ENDS_WITH.operator },
        { label: t.t(m.EQUALS.label), value: m.EQUALS.operator }
      ]
    },
    {
      options: [
        { label: t.t(m.GREATER_THAN.label), value: m.GREATER_THAN.operator },
        { label: t.t(m.GREATER_THAN_OR_EQUAL.label), value: m.GREATER_THAN_OR_EQUAL.operator },
        { label: t.t(m.LESS_THAN.label), value: m.LESS_THAN.operator },
        { label: t.t(m.LESS_THAN_OR_EQUAL.label), value: m.LESS_THAN_OR_EQUAL.operator },
        { label: t.t(m.EQUAL.label), value: m.EQUAL.operator },
        { label: t.t(m.NOT_EQUAL.label), value: m.NOT_EQUAL.operator },
        { label: t.t(m.BETWEEN.label), value: m.BETWEEN.operator },
        { label: t.t(m.NOT_BETWEEN.label), value: m.NOT_BETWEEN.operator }
      ]
    },
    {
      options: [
        { label: t.t(m.CUSTOM.label), value: m.CUSTOM.operator }
      ]
    }
  ], [e, t]);
}
function Zi(t) {
  const e = t.getCurrentLocale();
  return ut(() => m.ALL_CONDITIONS.filter((r) => r.numOfParameters !== 2).map((r) => ({ label: t.t(r.label), value: r.operator })), [e, t]);
}
function Ki(t) {
  const { model: e } = t, r = D(G), s = L(e.searchString$, "", !0), n = L(e.filterItems$, void 0, !0), a = r.t("sheets-filter.panel.filter-only"), i = ze(n), o = i.checked > 0 && i.unchecked === 0, l = i.checked > 0 && i.unchecked > 0, c = e.treeMapCache, h = U(() => {
    e.onCheckAllToggled(!o);
  }, [e, o]), f = U((p) => {
    e.setSearchString(p);
  }, [e]);
  function S(p) {
    let _ = [];
    return p.forEach((C) => {
      C.checked && _.push(C.key), C.children && (_ = _.concat(S(C.children)));
    }), _;
  }
  return /* @__PURE__ */ y(
    "div",
    {
      "data-u-comp": "sheets-filter-panel-values-container",
      className: "univer-flex univer-h-full univer-min-h-[300px] univer-flex-col",
      children: [
        /* @__PURE__ */ d(
          jt,
          {
            autoFocus: !0,
            value: s,
            placeholder: r.t("sheets-filter.panel.search-placeholder"),
            onChange: f
          }
        ),
        /* @__PURE__ */ y(
          "div",
          {
            "data-u-comp": "sheets-filter-panel",
            className: Ae("univer-mt-2 univer-box-border univer-flex univer-flex-grow univer-flex-col univer-overflow-hidden univer-rounded-md univer-px-2 univer-py-2.5", ct),
            children: [
              /* @__PURE__ */ d(
                "div",
                {
                  "data-u-comp": "sheets-filter-panel-values-item",
                  className: "univer-box-border univer-h-8 univer-w-full univer-py-0.5",
                  children: /* @__PURE__ */ y(
                    "div",
                    {
                      "data-u-comp": "sheets-filter-panel-values-item-inner",
                      className: "univer-box-border univer-flex univer-h-7 univer-items-center univer-rounded-md univer-pb-0 univer-pl-5 univer-pr-0.5 univer-pt-0 univer-text-sm",
                      children: [
                        /* @__PURE__ */ d(
                          ci,
                          {
                            indeterminate: l,
                            disabled: n.length === 0,
                            checked: o,
                            onChange: h
                          }
                        ),
                        /* @__PURE__ */ d(
                          "span",
                          {
                            "data-u-comp": "sheets-filter-panel-values-item-text",
                            className: "univer-mx-1 univer-inline-block univer-flex-shrink univer-truncate univer-text-gray-900 dark:!univer-text-white",
                            children: `${r.t("sheets-filter.panel.select-all")}`
                          }
                        ),
                        /* @__PURE__ */ d(
                          "span",
                          {
                            "data-u-comp": "sheets-filter-panel-values-item-count",
                            className: "univer-text-gray-400 dark:!univer-text-gray-500",
                            children: `(${i.checked}/${i.checked + i.unchecked})`
                          }
                        )
                      ]
                    }
                  )
                }
              ),
              /* @__PURE__ */ d("div", { "data-u-comp": "sheets-filter-panel-values-virtual", className: "univer-flex-grow", children: /* @__PURE__ */ d(
                ui,
                {
                  data: n,
                  defaultExpandAll: !1,
                  valueGroup: S(n),
                  onChange: (p) => {
                    e.onFilterCheckToggled(p);
                  },
                  defaultCache: c,
                  itemHeight: 28,
                  treeNodeClassName: `
                          univer-pr-2 univer-border-box univer-rounded-md
                          [&:hover_a]:univer-inline-block
                          hover:univer-bg-gray-50 univer-h-full
                          univer-text-gray-900 dark:hover:!univer-bg-gray-900
                          dark:!univer-text-white
                        `,
                  attachRender: (p) => /* @__PURE__ */ y(
                    "div",
                    {
                      className: "univer-ml-1 univer-flex univer-h-5 univer-flex-1 univer-cursor-pointer univer-items-center univer-justify-between univer-text-sm univer-text-primary-500",
                      children: [
                        /* @__PURE__ */ d(
                          "span",
                          {
                            "data-u-comp": "sheets-filter-panel-values-item-count",
                            className: "univer-text-gray-400 dark:!univer-text-gray-500",
                            children: `(${p.count})`
                          }
                        ),
                        /* @__PURE__ */ d(
                          "a",
                          {
                            className: "univer-box-border univer-hidden univer-h-4 univer-whitespace-nowrap univer-px-1.5",
                            onClick: () => {
                              const _ = [];
                              p.children ? p.children.forEach((C) => {
                                C.children ? C.children.forEach((F) => {
                                  _.push(F.key);
                                }) : _.push(C.key);
                              }) : _.push(p.key), e.onFilterOnly(_);
                            },
                            children: a
                          }
                        )
                      ]
                    }
                  )
                }
              ) })
            ]
          }
        )
      ]
    }
  );
}
function qi() {
  const t = D(Tr);
  if (!L(t.visible$, void 0, !0)) return null;
  const r = D(G), s = D(Bt), n = L(t.enabled$, void 0, !0);
  return /* @__PURE__ */ y(
    "div",
    {
      className: "univer-mt-2 univer-flex univer-items-center univer-justify-between univer-text-sm univer-text-gray-900 dark:!univer-text-gray-200",
      children: [
        /* @__PURE__ */ y("div", { className: "univer-flex univer-items-center univer-gap-1", children: [
          /* @__PURE__ */ d("span", { children: r.t("sheets-filter.sync.title") }),
          /* @__PURE__ */ d(
            hi,
            {
              title: n ? r.t("sheets-filter.sync.statusTips.off") : r.t("sheets-filter.sync.statusTips.on"),
              asChild: !0,
              children: /* @__PURE__ */ d(sr, { className: "univer-block" })
            }
          )
        ] }),
        /* @__PURE__ */ d(
          di,
          {
            defaultChecked: n,
            onChange: (a) => {
              const i = a ? r.t("sheets-filter.sync.switchTips.on") : r.t("sheets-filter.sync.switchTips.off");
              t.setEnabled(a), s.show({
                content: i,
                type: Yt.Success,
                duration: 2e3
              });
            }
          }
        )
      ]
    }
  );
}
function Xi() {
  var F;
  const t = D(X), e = D(G), r = D(j), s = L(t.filterBy$, void 0, !0), n = L(t.filterByModel$, void 0, !1), a = L(() => (n == null ? void 0 : n.canApply$) || he(!1), void 0, !1, [n]), i = zi(e), o = !L(t.hasCriteria$), l = U((N) => {
    r.executeCommand(Xt.id, { filterBy: N });
  }, [r]), c = U(async () => {
    await (n == null ? void 0 : n.clear()), r.executeCommand(Se.id);
  }, [n, r]), h = U(() => {
    r.executeCommand(Se.id);
  }, [r]), f = U(async () => {
    await (n == null ? void 0 : n.apply()), r.executeCommand(Se.id);
  }, [n, r]), p = (F = D(z).activeFilterModel) == null ? void 0 : F.getRange(), _ = t.col, C = xr(Ar.FILTER_PANEL_EMBED_POINT);
  return /* @__PURE__ */ y(
    "div",
    {
      "data-u-comp": "sheets-filter-panel",
      className: "univer-box-border univer-flex univer-max-h-[500px] univer-w-[400px] univer-flex-col univer-rounded-lg univer-bg-white univer-p-4 univer-shadow-lg dark:!univer-border-gray-600 dark:!univer-bg-gray-700",
      children: [
        /* @__PURE__ */ d(
          Ur,
          {
            components: C,
            sharedProps: { range: p, colIndex: _, onClose: h }
          }
        ),
        /* @__PURE__ */ d("div", { className: "univer-mb-1 univer-flex-shrink-0 univer-flex-grow-0", children: /* @__PURE__ */ d(
          fi,
          {
            value: s,
            items: i,
            onChange: (N) => l(N)
          }
        ) }),
        n ? /* @__PURE__ */ d(
          "div",
          {
            "data-u-comp": "sheets-filter-panel-content",
            className: "univer-flex-shrink univer-flex-grow univer-pt-2",
            children: s === w.VALUES ? /* @__PURE__ */ d(Ki, { model: n }) : s === w.COLORS ? /* @__PURE__ */ d(Gi, { model: n }) : /* @__PURE__ */ d(ji, { model: n })
          }
        ) : /* @__PURE__ */ d("div", { className: "univer-flex-1" }),
        /* @__PURE__ */ d(qi, {}),
        /* @__PURE__ */ y(
          "div",
          {
            "data-u-comp": "sheets-filter-panel-footer",
            className: "univer-mt-4 univer-inline-flex univer-flex-shrink-0 univer-flex-grow-0 univer-flex-nowrap univer-justify-between univer-overflow-hidden",
            children: [
              /* @__PURE__ */ d(Ze, { variant: "link", onClick: c, disabled: o, children: e.t("sheets-filter.panel.clear-filter") }),
              /* @__PURE__ */ y("span", { className: "univer-flex univer-gap-2", children: [
                /* @__PURE__ */ d(Ze, { variant: "default", onClick: h, children: e.t("sheets-filter.panel.cancel") }),
                /* @__PURE__ */ d(Ze, { disabled: !a, variant: "primary", onClick: f, children: e.t("sheets-filter.panel.confirm") })
              ] })
            ]
          }
        )
      ]
    }
  );
}
function zi(t) {
  const e = t.getCurrentLocale();
  return ut(() => [
    { label: t.t("sheets-filter.panel.by-values"), value: w.VALUES },
    { label: t.t("sheets-filter.panel.by-colors"), value: w.COLORS },
    { label: t.t("sheets-filter.panel.by-conditions"), value: w.CONDITIONS }
  ], [e, t]);
}
function Ji(t) {
  const e = t.get(z);
  return {
    id: Fe.id,
    type: lt.BUTTON_SELECTOR,
    icon: "FilterIcon",
    tooltip: "sheets-filter.toolbar.smart-toggle-filter-tooltip",
    hidden$: ot(t, ie.UNIVER_SHEET),
    activated$: e.activeFilterModel$.pipe(le((r) => !!r)),
    disabled$: wr(
      t,
      Pr(
        t,
        {
          worksheetTypes: [pe, ve],
          rangeTypes: [_e]
        }
      )
    )
  };
}
function es(t) {
  const e = t.get(z);
  return {
    id: nt.id,
    type: lt.BUTTON,
    title: "sheets-filter.toolbar.clear-filter-criteria",
    hidden$: ot(t, ie.UNIVER_SHEET),
    disabled$: e.activeFilterModel$.pipe(at((r) => {
      var s;
      return (s = r == null ? void 0 : r.hasCriteria$.pipe(le((n) => !n))) != null ? s : he(!0);
    }))
  };
}
function ts(t) {
  const e = t.get(z);
  return {
    id: st.id,
    type: lt.BUTTON,
    title: "sheets-filter.toolbar.re-calc-filter-conditions",
    hidden$: ot(t, ie.UNIVER_SHEET),
    disabled$: e.activeFilterModel$.pipe(at((r) => {
      var s;
      return (s = r == null ? void 0 : r.hasCriteria$.pipe(le((n) => !n))) != null ? s : he(!0);
    }))
  };
}
const rs = {
  [kr.ORGANIZATION]: {
    [Fe.id]: {
      order: 2,
      menuItemFactory: Ji,
      [nt.id]: {
        order: 0,
        menuItemFactory: es
      },
      [st.id]: {
        order: 1,
        menuItemFactory: ts
      }
    }
  }
}, is = {
  id: Fe.id,
  binding: Hr.L | gt.CTRL_COMMAND | gt.SHIFT,
  description: "sheets-filter.shortcut.smart-toggle-filter",
  preconditions: $r,
  group: "4_sheet-edit"
};
var ss = Object.getOwnPropertyDescriptor, ns = (t, e, r, s) => {
  for (var n = s > 1 ? void 0 : s ? ss(e, r) : e, a = t.length - 1, i; a >= 0; a--)
    (i = t[a]) && (n = i(n) || n);
  return n;
}, $ = (t, e) => (r, s) => e(r, s, t);
const Rt = "FILTER_PANEL_POPUP";
let ke = class extends Ee {
  constructor(e, r, s, n, a, i, o, l, c, h, f, S, p) {
    super(p, S);
    v(this, "_popupDisposable");
    this._injector = e, this._componentManager = r, this._sheetsFilterPanelService = s, this._sheetCanvasPopupService = n, this._sheetsFilterService = a, this._localeService = i, this._shortcutService = o, this._commandService = l, this._menuManagerService = c, this._contextService = h, this._messageService = f, this._initCommands(), this._initShortcuts(), this._initMenuItems(), this._initUI();
  }
  dispose() {
    super.dispose(), this._closeFilterPopup();
  }
  _initShortcuts() {
    [
      is
    ].forEach((e) => {
      this.disposeWithMe(this._shortcutService.registerShortcut(e));
    });
  }
  _initCommands() {
    [
      Fe,
      Er,
      Fr,
      V,
      nt,
      st,
      Xt,
      Me,
      Se
    ].forEach((e) => {
      this.disposeWithMe(this._commandService.registerCommand(e));
    });
  }
  _initMenuItems() {
    this._menuManagerService.mergeMenu(rs);
  }
  _initUI() {
    [
      [Rt, Xi],
      ["FilterIcon", ir]
    ].forEach(([e, r]) => {
      this.disposeWithMe(
        this._componentManager.register(e, r)
      );
    }), this.disposeWithMe(this._contextService.subscribeContextValue$(ae).pipe(ni()).subscribe((e) => {
      e ? this._openFilterPopup() : this._closeFilterPopup();
    })), this.disposeWithMe(this._sheetsFilterService.errorMsg$.subscribe((e) => {
      e && this._messageService.show({
        type: Yt.Error,
        content: this._localeService.t(e)
      });
    }));
  }
  _openFilterPopup() {
    const e = this._sheetsFilterPanelService.filterModel;
    if (!e)
      throw new Error("[SheetsFilterUIController]: no filter model when opening filter popup!");
    const r = e.getRange(), s = this._sheetsFilterPanelService.col, { startRow: n } = r;
    this._popupDisposable = this._sheetCanvasPopupService.attachPopupToCell(n, s, {
      componentKey: Rt,
      direction: "horizontal",
      onClickOutside: () => this._commandService.syncExecuteCommand(Se.id),
      offset: [5, 0]
    });
  }
  _closeFilterPopup() {
    var e;
    (e = this._popupDisposable) == null || e.dispose(), this._popupDisposable = null;
  }
};
ke = ns([
  $(0, T(re)),
  $(1, T(Dr)),
  $(2, T(X)),
  $(3, T(Lr)),
  $(4, T(z)),
  $(5, T(G)),
  $(6, Br),
  $(7, j),
  $(8, Wr),
  $(9, We),
  $(10, Bt),
  $(11, T(Dt)),
  $(12, Wt)
], ke);
var os = Object.defineProperty, ls = Object.getOwnPropertyDescriptor, as = (t, e, r) => e in t ? os(t, e, { enumerable: !0, configurable: !0, writable: !0, value: r }) : t[e] = r, cs = (t, e, r, s) => {
  for (var n = s > 1 ? void 0 : s ? ls(e, r) : e, a = t.length - 1, i; a >= 0; a--)
    (i = t[a]) && (n = i(n) || n);
  return n;
}, Xe = (t, e) => (r, s) => e(r, s, t), or = (t, e, r) => as(t, typeof e != "symbol" ? e + "" : e, r);
const us = "SHEET_FILTER_UI_PLUGIN";
let He = class extends it {
  constructor(t = xe, e, r, s) {
    super(), this._config = t, this._injector = e, this._configService = r, this._rpcChannelService = s;
    const { menu: n, ...a } = kt(
      {},
      xe,
      this._config
    );
    n && this._configService.setConfig("menu", n, { merge: !0 }), this._configService.setConfig(zt, a);
  }
  onStarting() {
    pr(this._injector, [
      [X],
      [ce],
      [ke]
    ]), this._config.useRemoteFilterValuesGenerator && this._rpcChannelService && this._injector.add([we, {
      useFactory: () => oi(
        this._rpcChannelService.requestChannel(ht)
      )
    }]);
  }
  onReady() {
    vt(this._injector, [
      [ce]
    ]);
  }
  onRendered() {
    vt(this._injector, [
      [ke]
    ]);
  }
};
or(He, "type", ie.UNIVER_SHEET);
or(He, "pluginName", us);
He = cs([
  xt(Ht),
  Xe(1, T(re)),
  Xe(2, Ut),
  Xe(3, mr(Gt))
], He);
var hs = Object.getOwnPropertyDescriptor, ds = (t, e, r, s) => {
  for (var n = s > 1 ? void 0 : s ? hs(e, r) : e, a = t.length - 1, i; a >= 0; a--)
    (i = t[a]) && (n = i(n) || n);
  return n;
}, At = (t, e) => (r, s) => e(r, s, t), Oe;
let wt = (Oe = class extends it {
  constructor(t, e, r) {
    super(), this._config = t, this._injector = e, this._rpcChannelService = r;
  }
  onStarting() {
    [
      [we, { useClass: Je }]
    ].forEach((t) => this._injector.add(t));
  }
  onReady() {
    this._rpcChannelService.registerChannel(
      ht,
      li(this._injector.get(we))
    );
  }
}, v(Oe, "type", ie.UNIVER_SHEET), v(Oe, "pluginName", "SHEET_FILTER_UI_WORKER_PLUGIN"), Oe);
wt = ds([
  At(1, T(re)),
  At(2, Gt)
], wt);
export {
  Xt as ChangeFilterByOperation,
  Se as CloseFilterPanelOperation,
  Me as OpenFilterPanelOperation,
  Ue as UniverSheetsFilterMobileUIPlugin,
  He as UniverSheetsFilterUIPlugin,
  wt as UniverSheetsFilterUIWorkerPlugin
};
