// @ant-design/pro-flow/es/Background/components/SwimBg.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _taggedTemplateLiteral from "@babel/runtime/helpers/esm/taggedTemplateLiteral";
var _templateObject, _templateObject2;
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
import { createStyles } from 'antd-style';
import { jsx as _jsx } from "react/jsx-runtime";
var useStyles = createStyles(function (_ref) {
  var css = _ref.css;
  return {
    container: css(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n    width: 100%;\n    height: 100%;\n    display: flex;\n    /* flex-direction: column; */\n  "]))),
    label: css(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["\n    text-align: center;\n    line-height: 40px;\n    height: 40px;\n    border-bottom: 1px solid #cbcdce;\n    color: '#A8AAAE';\n  "])))
  };
});
var SwimlaneBackground = function SwimlaneBackground(props) {
  var lanes = props.lanes,
    clm = props.className,
    style = props.style;
  var _useStyles = useStyles(),
    styles = _useStyles.styles;
  return /*#__PURE__*/_jsx("div", {
    className: "".concat(styles.container, " ").concat(clm),
    style: _objectSpread({}, style),
    children: lanes.map(function (lane) {
      return /*#__PURE__*/_jsx("div", {
        style: _objectSpread({
          top: 0,
          height: '100%',
          width: lane.width || '100%',
          backgroundColor: lane.backgroundColor || '#FAFAFA',
          border: '1px solid #cbcdce'
        }, lane.style),
        children: /*#__PURE__*/_jsx("div", {
          className: styles.label,
          style: {
            backgroundColor: lane.labelColor || '#f6f8fc'
          },
          children: lane.label
        })
      }, lane.id);
    })
  });
};
export default SwimlaneBackground;
