// @pdfme/pdf-lib/dist/node/src/utils/elements/Point.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const maths_1 = require("../maths");
const GraphElement_1 = __importDefault(require("./GraphElement"));
class Point extends GraphElement_1.default {
    constructor(coords = { x: 0, y: 0 }) {
        super();
        Object.defineProperty(this, "x", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "y", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.x = coords.x;
        this.y = coords.y;
    }
    toCoords() {
        return { x: this.x, y: this.y };
    }
    isEqual(element) {
        if (!(element instanceof Point))
            return false;
        const A = this.toCoords();
        const B = element.toCoords();
        return (0, maths_1.isEqual)(A.x, B.x) && (0, maths_1.isEqual)(A.y, B.y);
    }
    orthoProjection() {
        return new Point(this.toCoords());
    }
    plus(vect) {
        const P = new Point((0, maths_1.plus)(this.toCoords(), vect));
        return P;
    }
}
Object.defineProperty(Point, "type", {
    enumerable: true,
    configurable: true,
    writable: true,
    value: 'PointFixed'
});
exports.default = Point;
