// @svar-ui/react-core/dist/index.es.js
import { jsx as l, jsxs as $, Fragment as J } from "react/jsx-runtime";
import { createContext as ae, useContext as Q, useState as P, useCallback as y, useRef as H, useEffect as B, useMemo as E, useId as Fe, forwardRef as Le, useImperativeHandle as Ce } from "react";
import { useWritableProp as W, snippet as ne, useHotkeys as Pe } from "@svar-ui/lib-react";
import { uid as le, env as oe, clickOutside as de, locale as De, locateID as He, getDuodecade as Ve, dateToString as X, delegateClick as $e, calculatePosition as Be, setEnv as Ee } from "@svar-ui/lib-dom";
import { locale as tn } from "@svar-ui/lib-dom";
import { en as Te } from "@svar-ui/core-locales";
import { en as ln } from "@svar-ui/core-locales";
import { createPortal as Ae } from "react-dom";
const re = ae("willow"), Me = ae({}), G = ae(null), se = ae(null), Dt = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  fieldId: se,
  helpers: Me,
  i18n: G,
  theme: re
}, Symbol.toStringTag, { value: "Module" }));
function z(e) {
  const n = Q(se), [t] = P(() => e || n && n() || le());
  return t;
}
function Vt({
  value: e = "",
  id: n,
  placeholder: t = "",
  title: c = "",
  disabled: r = !1,
  error: i = !1,
  readonly: o = !1,
  onChange: s
}) {
  const a = z(n), [u, w] = W(e), f = y(
    (p) => {
      const d = p.target.value;
      w(d), s && s({ value: d, input: !0 });
    },
    [s]
  ), x = y(
    (p) => {
      const d = p.target.value;
      w(d), s && s({ value: d });
    },
    [s]
  ), m = H(null);
  return B(() => {
    const p = x, d = m.current;
    return d.addEventListener("change", p), () => {
      d && d.removeEventListener("change", p);
    };
  }, [x]), /* @__PURE__ */ l(
    "textarea",
    {
      className: `wx-3yFVAC wx-textarea ${i ? "wx-error" : ""}`,
      id: a,
      disabled: r,
      placeholder: t,
      readOnly: o,
      title: c,
      value: u,
      onInput: f,
      ref: m
    }
  );
}
function ee({
  type: e = "",
  css: n = "",
  icon: t = "",
  disabled: c = !1,
  title: r = "",
  text: i = "",
  children: o,
  onClick: s
}) {
  const a = E(() => {
    let w = e ? e.split(" ").filter((f) => f !== "").map((f) => "wx-" + f).join(" ") : "";
    return n + (n ? " " : "") + w;
  }, [e, n]), u = (w) => {
    s && s(w);
  };
  return /* @__PURE__ */ $(
    "button",
    {
      title: r,
      className: `wx-2ZWgb4 wx-button ${a} ${t && !o ? "wx-icon" : ""}`,
      disabled: c,
      onClick: u,
      children: [
        t && /* @__PURE__ */ l("i", { className: "wx-2ZWgb4 " + t }),
        o || i || " "
      ]
    }
  );
}
function Ie({
  id: e,
  label: n = "",
  inputValue: t = "",
  value: c = !1,
  onChange: r,
  disabled: i = !1
}) {
  const o = z(e), [s, a] = W(c);
  return /* @__PURE__ */ $("div", { className: "wx-2IvefP wx-checkbox", children: [
    /* @__PURE__ */ l(
      "input",
      {
        type: "checkbox",
        id: o,
        disabled: i,
        className: "wx-2IvefP wx-check",
        checked: s,
        value: t,
        onChange: ({ target: w }) => {
          const f = w.checked;
          a(f), r && r({ value: f, inputValue: t });
        }
      }
    ),
    /* @__PURE__ */ $("label", { htmlFor: o, className: "wx-2IvefP wx-label", children: [
      /* @__PURE__ */ l("span", { className: "wx-2IvefP wx-before" }),
      n && /* @__PURE__ */ l("span", { className: "wx-2IvefP wx-after", children: n })
    ] })
  ] });
}
function $t({
  options: e = [],
  value: n = [],
  type: t = "",
  onChange: c
}) {
  const r = y(
    (i) => {
      let o;
      i.value ? o = [...n, i.inputValue] : o = n.filter((s) => s !== i.inputValue), c && c({ value: o });
    },
    [n, c]
  );
  return /* @__PURE__ */ l(se.Provider, { value: null, children: /* @__PURE__ */ l("div", { className: `wx-q8xwRD wx-checkboxgroup ${t && `wx-${t}`}`, children: e.map((i) => /* @__PURE__ */ l("div", { className: "wx-q8xwRD wx-item", children: /* @__PURE__ */ l(
    Ie,
    {
      label: i.label,
      inputValue: i.id,
      value: n.includes(i.id),
      onChange: r
    }
  ) }, i.id)) }) });
}
function te({
  position: e = "bottom",
  align: n = "start",
  autoFit: t = !0,
  onCancel: c,
  width: r = "100%",
  children: i
}) {
  const o = H(null), [s, a] = W(e), [u, w] = W(n);
  return B(() => {
    if (t) {
      const f = o.current;
      if (f) {
        const x = f.getBoundingClientRect(), m = oe.getTopNode(f).getBoundingClientRect();
        x.right >= m.right && w("end"), x.bottom >= m.bottom && a("top");
      }
    }
  }, [t]), B(() => {
    if (o.current) {
      const f = (x) => {
        c && c(x);
      };
      return de(o.current, f).destroy;
    }
  }, [c]), /* @__PURE__ */ l(
    "div",
    {
      ref: o,
      className: `wx-32GZ52 wx-dropdown wx-${s}-${u}`,
      style: { width: r },
      children: i
    }
  );
}
const Oe = [
  "#00a037",
  "#37a9ef",
  "#f5a623",
  "#ff4c3b",
  "#a0a0a0",
  "#000000",
  "#ffffff"
];
function Tt({
  colors: e = Oe,
  value: n = "",
  onChange: t,
  id: c,
  clear: r = !1,
  placeholder: i = "",
  title: o = "",
  disabled: s = !1,
  error: a = !1
}) {
  const u = z(c), [w, f] = P(n), [x, m] = P(!1), p = y(
    (v, N) => {
      v.stopPropagation(), f(N), m(!1), t && t({ value: N });
    },
    [t]
  ), d = y(
    (v) => {
      v.stopPropagation(), f(""), t && t({ value: "" });
    },
    [t]
  ), h = y(() => {
    if (s) return !1;
    m(!0);
  }, [s]);
  return /* @__PURE__ */ $("div", { className: "wx-121TgJ wx-colorselect", onClick: h, children: [
    /* @__PURE__ */ l(
      "input",
      {
        className: "wx-121TgJ wx-input" + (a ? " wx-error" : "") + (x ? " wx-focus" : ""),
        title: o,
        value: w,
        readOnly: !0,
        id: u,
        placeholder: i,
        disabled: s
      }
    ),
    r && w && !s && /* @__PURE__ */ l("i", { className: "wx-121TgJ wx-clear wxi-close", onClick: d }),
    w ? /* @__PURE__ */ l(
      "div",
      {
        className: "wx-121TgJ wx-color wx-selected",
        style: { backgroundColor: w || "#00a037" }
      }
    ) : /* @__PURE__ */ l("div", { className: "wx-121TgJ wx-empty wx-selected" }),
    x && /* @__PURE__ */ l(te, { onCancel: () => m(!1), children: /* @__PURE__ */ $("div", { className: "wx-121TgJ wx-colors", children: [
      /* @__PURE__ */ l(
        "div",
        {
          className: "wx-121TgJ wx-empty",
          onClick: (v) => p(v, "")
        }
      ),
      e.map((v, N) => /* @__PURE__ */ l(
        "div",
        {
          className: "wx-121TgJ wx-color",
          style: { backgroundColor: v },
          onClick: (R) => p(R, v)
        },
        N
      ))
    ] }) })
  ] });
}
function me(e, n) {
  const t = e.firstChild, { moveBlockSlider: c, moveLineSlider: r } = n || {};
  function i(a) {
    o(a), t.focus(), window.addEventListener("mousemove", o), window.addEventListener("mouseup", s);
  }
  function o(a) {
    let u, w, f, x;
    const m = e.getBoundingClientRect();
    u = m.left, w = m.top, f = a.clientX - u, x = a.clientY - w, c && n.moveBlockSlider(f, x), r && n.moveLineSlider(f);
  }
  function s() {
    t.blur(), window.removeEventListener("mousemove", o), window.removeEventListener("mouseup", s);
  }
  return e.addEventListener("mousedown", i), {
    destroy() {
      e.removeEventListener("mousedown", i);
    }
  };
}
const q = {
  _toHex: [
    "0",
    "1",
    "2",
    "3",
    "4",
    "5",
    "6",
    "7",
    "8",
    "9",
    "A",
    "B",
    "C",
    "D",
    "E",
    "F"
  ],
  toHex(e, n) {
    e = parseInt(e, 10);
    let t = "";
    for (; e > 0; )
      t = this._toHex[e % 16] + t, e = Math.floor(e / 16);
    for (; t.length < n; ) t = `0${t}`;
    return t;
  },
  rgbToHex(e) {
    let n = [];
    return typeof e == "string" ? e.replace(/[\d+.]+/g, (t) => n.push(parseFloat(t))) : Array.isArray(e) && (n = e), n[3] === 0 ? "" : n.slice(0, 3).map((t) => this.toHex(Math.floor(t), 2)).join("");
  },
  hexToDec(e) {
    return parseInt(e, 16);
  },
  // only handles the six-digit hex color format
  hexToRgb(e) {
    e = e.charAt(0) === "#" ? e.slice(1) : e;
    const n = e.match(/.{1,2}/g);
    let [t, c, r] = n;
    return t = this.hexToDec(t), c = this.hexToDec(c), r = this.hexToDec(r), [t, c, r];
  },
  hsvToRgb(e, n, t) {
    let c, r, i, o, s, a, u, w;
    switch (c = Math.floor(e / 60) % 6, r = e / 60 - c, i = t * (1 - n), o = t * (1 - r * n), s = t * (1 - (1 - r) * n), a = 0, u = 0, w = 0, c) {
      case 0:
        a = t, u = s, w = i;
        break;
      case 1:
        a = o, u = t, w = i;
        break;
      case 2:
        a = i, u = t, w = s;
        break;
      case 3:
        a = i, u = o, w = t;
        break;
      case 4:
        a = s, u = i, w = t;
        break;
      case 5:
        a = t, u = i, w = o;
        break;
    }
    return a = Math.floor(a * 255), u = Math.floor(u * 255), w = Math.floor(w * 255), [a, u, w];
  },
  rgbToHsv(e, n, t) {
    let c, r, i;
    e /= 255, n /= 255, t /= 255;
    let o = Math.min(e, n, t), s = Math.max(e, n, t);
    return r = 0, c = s === 0 ? 0 : 1 - o / s, i = s, s === o ? r = 0 : s === e && n >= t ? r = 60 * (n - t) / (s - o) + 0 : s === e && n < t ? r = 60 * (n - t) / (s - o) + 360 : s === n ? r = 60 * (t - e) / (s - o) + 120 : s === t && (r = 60 * (e - n) / (s - o) + 240), [r, c, i];
  },
  hexToHvs(e) {
    const n = this.hexToRgb(e);
    return this.rgbToHsv(...n);
  },
  hvsToHex(e, n, t) {
    const c = this.hsvToRgb(e, n, t);
    return "#" + this.rgbToHex(c).toUpperCase();
  }
}, We = (e) => /^([A-Fa-f0-9]{6}|[A-Fa-f0-9]{3})$/.test(e);
function he(e) {
  return e ? (e = e.toString(16), e.charAt(0) === "#" && (e = e.slice(1)), We(e) ? (e.length === 3 && (e = e.slice(0, 3).split("").reduce((n, t) => n + t + t, "")), `#${e.toUpperCase()}`) : null) : null;
}
function Z() {
  return De(Te);
}
const pe = "Block", ve = "Line";
function Ye({
  value: e = "#65D3B3",
  button: n = !1,
  onChange: t
}) {
  const c = (Q(G) || Z()).getGroup("core"), r = H(null), i = H(null), [o, s] = P(), [a, u] = P(), [w, f] = P(), [x, m] = P(), [p, d] = W(e), h = E(() => he(p) || "#65D3B3", [p]), v = E(
    () => q.hvsToHex(w || 0, 1, 1),
    [w]
  ), N = y(
    (b, k) => {
      const { width: F, height: A } = r.current.getBoundingClientRect();
      k < 0 ? k = 0 : k > A && (k = A), b < 0 ? b = 0 : b > F && (b = F), s(k), u(b);
      let L, I;
      q.hexToHvs(h);
      const O = F / 100, Y = A / 100;
      L = Math.ceil(b / O) / 100, I = Math.ceil(Math.abs(k / Y - 100)) / 100;
      const _ = q.hvsToHex(
        w || 0,
        L,
        I
      );
      d(_), t && t({ value: _, input: !0 });
    },
    [h, w, t]
  ), R = y(
    (b) => {
      const k = i.current.getBoundingClientRect().width;
      b < 0 ? b = 0 : b > k && (b = k);
      const F = Math.round(b * 359 / k), A = Math.max(Math.min(F, 359), 0);
      f(A), m(b);
      let L, I;
      [, L, I] = q.hexToHvs(h);
      const O = q.hvsToHex(A, L, I);
      d(O), t && t({ value: O, input: !0 });
    },
    [h, t]
  ), g = y(() => {
    const [b, k, F] = q.hexToHvs(h);
    if (f(b), r.current) {
      const { width: A, height: L } = r.current.getBoundingClientRect();
      i.current && m(
        b * i.current.getBoundingClientRect().width / 359
      ), u(k * A), s(Math.abs(L * (F - 1)));
    }
  }, [h]), T = y(
    ({ target: b }) => {
      const k = he(b.value);
      d(k), t && t({ value: k, input: !0 }), k && g();
    },
    [t, g]
  ), M = y(
    (b) => {
      b.stopPropagation(), t && t({ value: h });
    },
    [h, t]
  ), S = y(
    (b) => {
      const k = b.target, F = k.getAttribute("data-slider") === pe, A = k.getAttribute("data-slider") === ve;
      let L = window.getComputedStyle(k), I = parseFloat(L.left), O = parseFloat(L.top);
      const Y = b.code;
      if (F) {
        switch (Y) {
          case "ArrowLeft": {
            I--;
            break;
          }
          case "ArrowRight": {
            I++;
            break;
          }
          case "ArrowDown": {
            O++;
            break;
          }
          case "ArrowUp": {
            O--;
            break;
          }
          default:
            return;
        }
        N(I, O);
      }
      if (A) {
        if (Y === "ArrowLeft" || Y === "ArrowDown") I--;
        else if (Y === "ArrowRight" || Y === "ArrowUp") I++;
        else return;
        R(I);
      }
      b.preventDefault();
    },
    [N, R]
  );
  B(() => {
    g();
  }, [g]), B(() => {
    p !== p && (d(p), g());
  }, [p, p, g]), B(() => {
    r.current && me(r.current, { moveBlockSlider: N }), i.current && me(i.current, { moveLineSlider: R });
  }, [N, R]), B(() => {
    p !== p && t && t({ value: p, input: !0 });
  }, [p, t]), B(() => {
    h && r.current && g();
  }, [h, g]);
  const D = "wx-1yoKzq";
  return /* @__PURE__ */ $("div", { className: D + " wx-colorboard", children: [
    /* @__PURE__ */ l(
      "div",
      {
        className: D + " wx-color-block",
        style: { background: v },
        ref: r,
        children: /* @__PURE__ */ l(
          "div",
          {
            className: D + " wx-color-block-slider wx-slider",
            style: {
              background: h,
              top: o + "px",
              left: a + "px"
            },
            tabIndex: 0,
            "data-slider": pe,
            onKeyDown: S
          }
        )
      }
    ),
    /* @__PURE__ */ l("div", { className: D + " wx-color-line", ref: i, children: /* @__PURE__ */ l(
      "div",
      {
        className: D + " wx-color-line-slider wx-slider",
        style: { background: v, left: x + "px" },
        tabIndex: 0,
        "data-slider": ve,
        onKeyDown: S
      }
    ) }),
    /* @__PURE__ */ $("div", { className: D + " wx-color-controls", children: [
      /* @__PURE__ */ l(
        "div",
        {
          className: D + " wx-color",
          style: { background: h }
        }
      ),
      /* @__PURE__ */ l(
        "input",
        {
          type: "text",
          className: D + " wx-text",
          value: p,
          onChange: T
        }
      )
    ] }),
    n && /* @__PURE__ */ l(ee, { onClick: M, type: "secondary", children: c("select") })
  ] });
}
function Mt({
  value: e = "",
  id: n,
  placeholder: t = "",
  title: c = "",
  disabled: r = !1,
  error: i = !1,
  clear: o = !1,
  onChange: s
}) {
  const a = z(n);
  let [u, w] = W(e);
  const [f, x] = P(!1);
  function m() {
    if (r) return !1;
    x(!0);
  }
  function p(h) {
    if (h.input) return;
    x(!1);
    const v = h.value;
    w(v), s && s({ value: v });
  }
  function d(h) {
    h.stopPropagation(), w(""), s && s({ value: "" });
  }
  return /* @__PURE__ */ $("div", { className: "wx-mmT8WH wx-colorpicker", onClick: m, children: [
    /* @__PURE__ */ l(
      "input",
      {
        className: "wx-mmT8WH wx-input" + (i ? " wx-error" : "") + (f ? " wx-focus" : ""),
        title: c,
        value: u,
        readOnly: !0,
        id: a,
        placeholder: t,
        disabled: r
      }
    ),
    /* @__PURE__ */ l("div", { className: "wx-mmT8WH wx-color", style: { background: u } }),
    o && !r && u && /* @__PURE__ */ l("i", { className: "wx-mmT8WH wxi-close", onClick: d }),
    f && /* @__PURE__ */ l(te, { onCancel: () => x(!1), children: /* @__PURE__ */ l(Ye, { value: u, button: "true", onChange: p }) })
  ] });
}
function je() {
  let e = null, n = !1, t, c, r, i;
  const o = (x, m, p, d) => {
    t = x, c = m, r = p, i = d;
  }, s = (x) => {
    e = x, n = e !== null, r(e);
  }, a = (x, m) => {
    if (x !== null && t) {
      const p = t.querySelectorAll(".wx-list > .wx-item")[x];
      p && (p.scrollIntoView({ block: "nearest" }), m && m.preventDefault());
    }
  }, u = (x, m) => {
    const p = x === null ? null : Math.max(0, Math.min(e + x, c.length - 1));
    p !== e && (s(p), t ? a(p, m) : requestAnimationFrame(() => a(p, m)));
  };
  return { move: (x) => {
    const m = He(x), p = c.findIndex((d) => d.id == m);
    p !== e && s(p);
  }, keydown: (x, m) => {
    switch (x.code) {
      case "Enter":
        n ? i() : s(0);
        break;
      case "Space":
        n || s(0);
        break;
      case "Escape":
        r(e = null);
        break;
      case "Tab":
        r(e = null);
        break;
      case "ArrowDown":
        u(n ? 1 : m || 0, x);
        break;
      case "ArrowUp":
        u(n ? -1 : m || 0, x);
        break;
    }
  }, init: o, navigate: u };
}
function we({
  items: e = [],
  children: n,
  onSelect: t,
  onReady: c
}) {
  const r = H(), i = H(je()), [o, s] = P(null), a = H(o), u = (Q(G) || Z()).getGroup("core"), w = (x) => {
    x && x.stopPropagation(), t && t({ id: e[a.current]?.id });
  };
  B(() => {
    i.current.init(
      r.current,
      e,
      (x) => {
        s(x), a.current = x;
      },
      w
    );
  }, [e, r.current]), B(() => {
    c && c(i.current);
  }, []);
  const f = y(() => {
    i.current.navigate(null);
  }, [i]);
  return o === null ? null : /* @__PURE__ */ l(te, { onCancel: f, children: /* @__PURE__ */ l(
    "div",
    {
      className: "wx-233fr7 wx-list",
      ref: r,
      onClick: w,
      onMouseMove: i.current.move,
      children: e.length ? e.map((x, m) => /* @__PURE__ */ l(
        "div",
        {
          className: `wx-233fr7 wx-item ${m === o ? "wx-focus" : ""}`,
          "data-id": x.id,
          children: n ? ne(n, { option: x }) : x.label
        },
        x.id
      )) : /* @__PURE__ */ l("div", { className: "wx-233fr7 wx-no-data", children: u("No data") })
    }
  ) });
}
function It({
  value: e = "",
  id: n,
  options: t = [],
  textOptions: c = null,
  textField: r = "label",
  placeholder: i = "",
  title: o = "",
  disabled: s = !1,
  error: a = !1,
  clear: u = !1,
  children: w,
  onChange: f
}) {
  const x = z(n), m = H(null), p = H(null), [d, h] = W(e), [v, N] = P(!1), [R, g] = P(""), T = H(null), M = H(!1), S = E(() => {
    if (v) return R;
    if (d || d === 0) {
      const C = (c || t).find((U) => U.id === d);
      if (C) return C[r];
    }
    return "";
  }, [v, R, d, c, t, r]), D = E(() => !S || !v ? t : t.filter(
    (C) => C[r].toLowerCase().includes(S.toLowerCase())
  ), [S, v, t, r]), b = y(
    () => D.findIndex((C) => C.id === d),
    [D, d]
  ), k = y((C) => {
    m.current = C.navigate, p.current = C.keydown;
  }, []), F = y(
    (C, U) => {
      if (C || C === 0) {
        let K = t.find((V) => V.id === C);
        if (N(!1), U && m.current(null), K && d !== K.id) {
          const V = K.id;
          h(V), f && f({ value: V });
        }
      }
      !M.current && U && T.current.focus();
    },
    [t, d, f]
  ), A = y(
    ({ id: C }) => {
      F(C, !0);
    },
    [F]
  ), L = y(
    (C) => {
      C && C.stopPropagation(), h(""), N(!1), f && f({ value: "" });
    },
    [f]
  ), I = y(
    (C) => {
      if (!t.length) return;
      if (C === "" && u) {
        L();
        return;
      }
      let U = t.find((V) => V[r] === C);
      U || (U = t.find(
        (V) => V[r].toLowerCase().includes(C.toLowerCase())
      ));
      const K = U ? U.id : d || t[0].id;
      F(K, !1);
    },
    [t, r, u, d, F, L]
  ), O = y(() => {
    g(T.current.value), N(!0), D.length ? m.current(0) : m.current(null);
  }, [D.length, m]), Y = y(() => {
    M.current = !0;
  }, []), _ = y(() => {
    M.current = !1, setTimeout(() => {
      M.current || I(S);
    }, 200);
  }, [I, S]);
  return /* @__PURE__ */ $(
    "div",
    {
      className: "wx-1j11Jk wx-combo",
      onClick: () => m.current(b()),
      onKeyDown: (C) => p.current(C, b()),
      title: o,
      children: [
        /* @__PURE__ */ l(
          "input",
          {
            className: "wx-1j11Jk wx-input " + (a ? "wx-error" : ""),
            id: x,
            ref: T,
            value: S,
            disabled: s,
            placeholder: i,
            onFocus: Y,
            onBlur: _,
            onInput: O
          }
        ),
        u && !s && d ? /* @__PURE__ */ l("i", { className: "wx-1j11Jk wx-icon wxi-close", onClick: L }) : /* @__PURE__ */ l("i", { className: "wx-1j11Jk wx-icon wxi-angle-down" }),
        !s && /* @__PURE__ */ l(we, { items: D, onReady: k, onSelect: A, children: ({ option: C }) => /* @__PURE__ */ l(J, { children: w ? w({ option: C }) : C[r] }) })
      ]
    }
  );
}
function fe({
  value: e = "",
  id: n,
  readonly: t = !1,
  focus: c = !1,
  select: r = !1,
  type: i = "text",
  placeholder: o = "",
  disabled: s = !1,
  error: a = !1,
  inputStyle: u = {},
  title: w = "",
  css: f = "",
  icon: x = "",
  clear: m = !1,
  onChange: p
}) {
  const d = z(n), [h, v] = W(e), N = H(null), R = E(
    () => x && f.indexOf("wx-icon-left") === -1 ? "wx-icon-right " + f : f,
    [x, f]
  ), g = E(
    () => x && f.indexOf("wx-icon-left") !== -1,
    [x, f]
  );
  B(() => {
    const b = setTimeout(() => {
      c && N.current && N.current.focus(), r && N.current && N.current.select();
    }, 1);
    return () => clearTimeout(b);
  }, [c, r]);
  const T = y(
    (b) => {
      const k = b.target.value;
      v(k), p && p({ value: k, input: !0 });
    },
    [p]
  ), M = y(
    (b) => p && p({ value: b.target.value }),
    [p]
  );
  function S(b) {
    b.stopPropagation(), v(""), p && p({ value: "" });
  }
  let D = i;
  return i !== "password" && i !== "number" && (D = "text"), B(() => {
    const b = M, k = N.current;
    return k.addEventListener("change", b), () => {
      k && k.removeEventListener("change", b);
    };
  }, [M]), /* @__PURE__ */ $(
    "div",
    {
      className: `wx-hQ64J4 wx-text ${R} ${a ? "wx-error" : ""} ${s ? "wx-disabled" : ""} ${m ? "wx-clear" : ""}`,
      children: [
        /* @__PURE__ */ l(
          "input",
          {
            className: "wx-hQ64J4 wx-input",
            ref: N,
            id: d,
            readOnly: t,
            disabled: s,
            placeholder: o,
            type: D,
            style: u,
            title: w,
            value: h,
            onInput: T
          }
        ),
        m && !s && h ? /* @__PURE__ */ $(J, { children: [
          /* @__PURE__ */ l("i", { className: "wx-hQ64J4 wx-icon wxi-close", onClick: S }),
          g && /* @__PURE__ */ l("i", { className: `wx-hQ64J4 wx-icon ${x}` })
        ] }) : x ? /* @__PURE__ */ l("i", { className: `wx-hQ64J4 wx-icon ${x}` }) : null
      ]
    }
  );
}
function Je({ date: e, type: n, part: t, onShift: c }) {
  const { calendar: r, formats: i } = Q(G).getRaw(), o = e.getFullYear(), s = E(() => {
    switch (n) {
      case "month":
        return X(i.monthYearFormat, r)(e);
      case "year":
        return X(i.yearFormat, r)(e);
      case "duodecade": {
        const { start: u, end: w } = Ve(o), f = X(i.yearFormat, r);
        return `${f(new Date(u, 0, 1))} - ${f(new Date(w, 11, 31))}`;
      }
      default:
        return "";
    }
  }, [e, n, o, r, i]);
  function a() {
    c && c({ diff: 0, type: n });
  }
  return /* @__PURE__ */ $("div", { className: "wx-8HQVQV wx-header", children: [
    t !== "right" ? /* @__PURE__ */ l(
      "i",
      {
        className: "wx-8HQVQV wx-pager wxi-angle-left",
        onClick: () => c && c({ diff: -1, type: n })
      }
    ) : /* @__PURE__ */ l("span", { className: "wx-8HQVQV wx-spacer" }),
    /* @__PURE__ */ l("span", { className: "wx-8HQVQV wx-label", onClick: a, children: s }),
    t !== "left" ? /* @__PURE__ */ l(
      "i",
      {
        className: "wx-8HQVQV wx-pager wxi-angle-right",
        onClick: () => c && c({ diff: 1, type: n })
      }
    ) : /* @__PURE__ */ l("span", { className: "wx-8HQVQV wx-spacer" })
  ] });
}
function xe({ onClick: e, children: n }) {
  return /* @__PURE__ */ l("button", { className: "wx-3s8W4d wx-button", onClick: e, children: n });
}
function Ge({
  value: e,
  current: n,
  part: t = "",
  markers: c = null,
  onCancel: r,
  onChange: i
}) {
  const o = (Q(G) || Z()).getRaw().calendar, s = (o.weekStart || 7) % 7, a = o.dayShort.slice(s).concat(o.dayShort.slice(0, s)), u = (g, T, M) => new Date(
    g.getFullYear(),
    g.getMonth() + (T || 0),
    g.getDate() + (M || 0)
  );
  let w = t !== "normal";
  function f(g) {
    const T = g.getDay();
    return T === 0 || T === 6;
  }
  function x() {
    const g = u(n, 0, 1 - n.getDate());
    return g.setDate(g.getDate() - (g.getDay() - (s - 7)) % 7), g;
  }
  function m() {
    const g = u(n, 1, -n.getDate());
    return g.setDate(g.getDate() + (6 - g.getDay() + s) % 7), g;
  }
  const p = H(0);
  function d(g, T) {
    T.timeStamp !== p.current && (p.current = T.timeStamp, T.stopPropagation(), i && i(new Date(new Date(g))), r && r());
  }
  const h = E(() => t == "normal" ? [e ? u(e).valueOf() : 0] : e ? [
    e.start ? u(e.start).valueOf() : 0,
    e.end ? u(e.end).valueOf() : 0
  ] : [0, 0], [t, e]), v = E(() => {
    const g = x(), T = m(), M = n.getMonth();
    let S = [];
    for (let D = g; D <= T; D.setDate(D.getDate() + 1)) {
      const b = {
        day: D.getDate(),
        in: D.getMonth() === M,
        date: D.valueOf()
      };
      let k = "";
      if (k += b.in ? "" : " wx-inactive", k += h.indexOf(b.date) > -1 ? " wx-selected" : "", w) {
        const F = b.date == h[0], A = b.date == h[1];
        F && !A ? k += " wx-left" : A && !F && (k += " wx-right"), b.date > h[0] && b.date < h[1] && (k += " wx-inrange");
      }
      if (k += f(D) ? " wx-weekend" : "", c) {
        const F = c(D);
        F && (k += " " + F);
      }
      S.push({ ...b, css: k });
    }
    return S;
  }, [n, h, w, c]), N = H(null);
  let R = H({});
  return R.current.click = d, B(() => {
    $e(N.current, R.current);
  }, []), /* @__PURE__ */ $("div", { children: [
    /* @__PURE__ */ l("div", { className: "wx-398RBS wx-weekdays", children: a.map((g) => /* @__PURE__ */ l("div", { className: "wx-398RBS wx-weekday", children: g }, g)) }),
    /* @__PURE__ */ l("div", { className: "wx-398RBS wx-days", ref: N, children: v.map((g) => /* @__PURE__ */ l(
      "div",
      {
        className: `wx-398RBS wx-day ${g.css} ${g.in ? "" : "wx-out"}`,
        "data-id": g.date,
        children: g.day
      },
      g.date
    )) })
  ] });
}
function Qe({
  value: e,
  current: n,
  part: t,
  onCancel: c,
  onChange: r,
  onShift: i
}) {
  const [o, s] = W(e || /* @__PURE__ */ new Date()), [a, u] = W(n || /* @__PURE__ */ new Date()), w = Q(G).getRaw().calendar, f = w.monthShort || [], x = E(() => a.getMonth(), [a]), m = y(
    (h, v) => {
      if (h != null) {
        v.stopPropagation();
        const N = new Date(a);
        N.setMonth(h), u(N), i && i({ current: N });
      }
      t === "normal" && s(new Date(a)), c && c();
    },
    [a, t, i, c]
  ), p = y(() => {
    const h = new Date(Re(o, t) || a);
    h.setMonth(a.getMonth()), h.setFullYear(a.getFullYear()), r && r(h);
  }, [o, a, t, r]), d = y(
    (h) => {
      const v = h.target.closest("[data-id]");
      if (v) {
        const N = parseInt(v.getAttribute("data-id"), 10);
        m(N, h);
      }
    },
    [m]
  );
  return /* @__PURE__ */ $(J, { children: [
    /* @__PURE__ */ l("div", { className: "wx-34U8T8 wx-months", onClick: d, children: f.map((h, v) => /* @__PURE__ */ l(
      "div",
      {
        className: "wx-34U8T8 wx-month" + (x === v ? " wx-current" : ""),
        "data-id": v,
        children: h
      },
      v
    )) }),
    /* @__PURE__ */ l("div", { className: "wx-34U8T8 wx-buttons", children: /* @__PURE__ */ l(xe, { onClick: p, children: w.done }) })
  ] });
}
const ie = "wx-1XEF33", _e = ({ value: e, current: n, onCancel: t, onChange: c, onShift: r, part: i }) => {
  const o = Q(G).getRaw().calendar, [s, a] = W(n), [u, w] = W(e), f = E(() => s.getFullYear(), [s]), x = E(() => {
    const { start: v, end: N } = Ve(f), R = [];
    for (let g = v; g <= N; ++g)
      R.push(g);
    return R;
  }, [f]), m = {
    click: p
  };
  function p(v, N) {
    if (v) {
      N.stopPropagation();
      const R = new Date(s);
      R.setFullYear(v), a(R), r && r({ current: R });
    }
    i === "normal" && w(new Date(s)), t && t();
  }
  function d() {
    const v = new Date(Re(u, i) || s);
    v.setFullYear(s.getFullYear()), c && c(v);
  }
  const h = H(null);
  return B(() => {
    h.current && $e(h.current, m);
  }, []), /* @__PURE__ */ $(J, { children: [
    /* @__PURE__ */ l("div", { className: ie + " wx-years", ref: h, children: x.map((v, N) => /* @__PURE__ */ l(
      "div",
      {
        className: ie + ` wx-year ${f == v ? "wx-current" : ""} ${N === 0 ? "wx-prev-decade" : ""} ${N === 11 ? "wx-next-decade" : ""}`,
        "data-id": v,
        children: v
      },
      N
    )) }),
    /* @__PURE__ */ l("div", { className: ie + " wx-buttons", children: /* @__PURE__ */ l(xe, { onClick: d, children: o.done }) })
  ] });
}, ge = {
  month: {
    component: Ge,
    next: ze,
    prev: Ue
  },
  year: {
    component: Qe,
    next: Ze,
    prev: Ke
  },
  duodecade: {
    component: _e,
    next: Xe,
    prev: qe
  }
};
function Ue(e) {
  return e = new Date(e), e.setMonth(e.getMonth() - 1), e;
}
function ze(e) {
  return e = new Date(e), e.setMonth(e.getMonth() + 1), e;
}
function Ke(e) {
  return e = new Date(e), e.setFullYear(e.getFullYear() - 1), e;
}
function Ze(e) {
  return e = new Date(e), e.setFullYear(e.getFullYear() + 1), e;
}
function qe(e) {
  return e = new Date(e), e.setFullYear(e.getFullYear() - 10), e;
}
function Xe(e) {
  return e = new Date(e), e.setFullYear(e.getFullYear() + 10), e;
}
function Re(e, n) {
  let t;
  if (n === "normal") t = e;
  else {
    const { start: c, end: r } = e;
    n === "left" ? t = c : n == "right" ? t = r : t = c && r;
  }
  return t;
}
const et = ["clear", "today"];
function tt(e) {
  if (e === "done") return -1;
  if (e === "clear") return null;
  if (e === "today") return /* @__PURE__ */ new Date();
}
function ce({
  value: e,
  current: n,
  onCurrentChange: t,
  part: c = "normal",
  markers: r = null,
  buttons: i,
  onShift: o,
  onChange: s
}) {
  const a = Q(G).getGroup("calendar"), [u, w] = P("month"), f = Array.isArray(i) ? i : i ? et : [], x = (v, N) => {
    v.preventDefault(), s && s({ value: N });
  }, m = () => {
    u === "duodecade" ? w("year") : u === "year" && w("month");
  }, p = (v) => {
    const { diff: N, current: R } = v;
    if (N === 0) {
      u === "month" ? w("year") : u === "year" && w("duodecade");
      return;
    }
    if (N) {
      const g = ge[u];
      t(N > 0 ? g.next(n) : g.prev(n));
    } else R && t(R);
    o && o();
  }, d = (v) => {
    w("month"), s && s({ select: !0, value: v });
  }, h = E(() => ge[u].component, [u]);
  return /* @__PURE__ */ l(
    "div",
    {
      className: `wx-2Gr4AS wx-calendar ${c !== "normal" && c !== "both" ? "wx-part" : ""}`,
      children: /* @__PURE__ */ $("div", { className: "wx-2Gr4AS wx-wrap", children: [
        /* @__PURE__ */ l(Je, { date: n, part: c, type: u, onShift: p }),
        /* @__PURE__ */ $("div", { children: [
          /* @__PURE__ */ l(
            h,
            {
              value: e,
              current: n,
              onCurrentChange: t,
              part: c,
              markers: r,
              onCancel: m,
              onChange: d,
              onShift: p
            }
          ),
          u === "month" && f.length > 0 && /* @__PURE__ */ l("div", { className: "wx-2Gr4AS wx-buttons", children: f.map((v) => /* @__PURE__ */ l("div", { className: "wx-2Gr4AS wx-button-item", children: /* @__PURE__ */ l(
            xe,
            {
              onClick: (N) => x(N, tt(v)),
              children: a(v)
            }
          ) }, v)) })
        ] })
      ] })
    }
  );
}
function ue(e) {
  let { words: n = null, optional: t = !1, children: c } = e, r = Q(G);
  const i = E(() => {
    let o = r;
    return (!o || !o.extend) && (o = De(Te)), n !== null && (o = o.extend(n, t)), o;
  }, [n, t, r]);
  return /* @__PURE__ */ l(G.Provider, { value: i, children: c });
}
function Ne(e, n, t, c) {
  if (!e || c) {
    const r = n ? new Date(n) : /* @__PURE__ */ new Date();
    r.setDate(1), t(r);
  } else if (e.getDate() !== 1) {
    const r = new Date(e);
    r.setDate(1), t(r);
  }
}
const nt = ["clear", "today"];
function lt({
  value: e,
  current: n,
  markers: t = null,
  buttons: c = nt,
  onChange: r
}) {
  const [i, o] = W(e), [s, a] = W(n);
  B(() => {
    Ne(s, i, a, !1);
  }, [i, s]);
  const u = y(
    (f) => {
      const x = f.value;
      x ? (o(new Date(x)), Ne(s, new Date(x), a, !0)) : o(null), r && r({ value: x ? new Date(x) : null });
    },
    [r, s]
  ), w = y(
    (f) => {
      a(f);
    },
    [a]
  );
  return s ? /* @__PURE__ */ l(ue, { children: /* @__PURE__ */ l(
    ce,
    {
      value: i,
      current: s,
      markers: t,
      buttons: c,
      onChange: u,
      onCurrentChange: w
    }
  ) }) : null;
}
const rt = ["clear", "today"];
function Rt({
  value: e,
  id: n,
  disabled: t = !1,
  error: c = !1,
  width: r = "unset",
  align: i = "start",
  placeholder: o = "",
  format: s = "",
  buttons: a = rt,
  css: u = "",
  title: w = "",
  editable: f = !1,
  clear: x = !1,
  onChange: m
}) {
  const { calendar: p, formats: d } = (Q(G) || Z()).getRaw(), h = s || d?.dateFormat;
  let v = typeof h == "function" ? h : X(h, p);
  const [N, R] = P(e), [g, T] = P(!1);
  B(() => {
    R(e);
  }, [e]);
  function M() {
    T(!1);
  }
  function S(k) {
    const F = k === N || k && N && k.valueOf() === N.valueOf() || !k && !N;
    R(k), F || m && m({ value: k }), setTimeout(M, 1);
  }
  const D = E(
    () => N ? v(N) : "",
    [N, v]
  );
  function b({ value: k, input: F }) {
    if (!f && !x || F) return;
    let A = typeof f == "function" ? f(k) : k ? new Date(k) : null;
    A = isNaN(A) ? N || null : A || null, S(A);
  }
  return B(() => {
    const k = M;
    return window.addEventListener("scroll", k), () => window.removeEventListener("scroll", k);
  }, []), /* @__PURE__ */ $("div", { className: "wx-1lKOFG wx-datepicker", onClick: () => T(!0), children: [
    /* @__PURE__ */ l(
      fe,
      {
        css: u,
        title: w,
        value: D,
        id: n,
        readonly: !f,
        disabled: t,
        error: c,
        placeholder: o,
        onInput: M,
        onChange: b,
        icon: "wxi-calendar",
        inputStyle: {
          cursor: "pointer",
          width: "100%",
          paddingRight: "calc(var(--wx-input-icon-size) + var(--wx-input-icon-indent) * 2)"
        },
        clear: x
      }
    ),
    g && !t && /* @__PURE__ */ l(
      te,
      {
        onCancel: M,
        width: r,
        align: i,
        autoFit: !!i,
        children: /* @__PURE__ */ l(
          lt,
          {
            buttons: a,
            value: N,
            onChange: (k) => S(k.value)
          }
        )
      }
    )
  ] });
}
const st = ["clear", "today"], ct = ({
  start: e,
  end: n,
  current: t,
  months: c = 2,
  markers: r = null,
  buttons: i = st,
  onChange: o
}) => {
  const [s, a] = W(e), [u, w] = W(n), [f, x] = P(), [m, p] = P(), d = y((T, M, S) => {
    const D = new Date(T);
    return D.setMonth(D.getMonth() + M), S && D.valueOf() === S.valueOf() ? S : D;
  }, []), h = y(
    (T) => {
      x(() => {
        const M = new Date(T);
        return M.setDate(1), p(d(M, 1)), M;
      });
    },
    [d]
  ), v = y(
    (T) => {
      p(() => {
        const M = new Date(T);
        return M.setDate(1), x(d(M, -1)), M;
      });
    },
    [d]
  );
  B(() => {
    const T = s;
    f || h(T ? new Date(T) : t || /* @__PURE__ */ new Date());
  }, [s, t, f, h]);
  const N = y(
    (T) => {
      const M = T.value;
      let S = s, D = u;
      const b = M === -1;
      return b || (T.select ? !s || u ? (a(S = M), w(D = null)) : s > M ? (w(D = s), a(S = M)) : w(D = M) : M ? (a(S = new Date(M)), w(D = new Date(M))) : (a(S = null), w(D = null))), (b || !i.includes("done")) && o && o({ start: S, end: D }), [S, D];
    },
    [s, u, i, o]
  ), R = y(
    (T) => {
      const [M] = N(T);
      M && h(new Date(M));
    },
    [h, s, u]
  ), g = y(
    (T) => {
      const [, M] = N(T);
      M && v(new Date(M));
    },
    [v, s, u]
  );
  return !f || !m ? null : c === 1 ? /* @__PURE__ */ l(ue, { children: /* @__PURE__ */ l(
    ce,
    {
      value: { start: s, end: u },
      current: f,
      setCurrent: h,
      markers: r,
      buttons: i,
      part: "both",
      onChange: R
    }
  ) }) : /* @__PURE__ */ l(ue, { children: /* @__PURE__ */ $("div", { className: "wx-p2jCaW wx-rangecalendar", children: [
    /* @__PURE__ */ l("div", { className: "wx-p2jCaW wx-half", children: /* @__PURE__ */ l(
      ce,
      {
        value: { start: s, end: u },
        current: f,
        setCurrent: h,
        markers: r,
        buttons: !1,
        part: "left",
        onChange: R,
        onCurrentChange: (T) => h(T)
      }
    ) }),
    /* @__PURE__ */ l("div", { className: "wx-p2jCaW wx-half", children: /* @__PURE__ */ l(
      ce,
      {
        value: { start: s, end: u },
        current: m,
        setCurrent: v,
        markers: r,
        buttons: i,
        part: "right",
        onChange: g,
        onCurrentChange: (T) => v(T)
      }
    ) })
  ] }) });
}, at = ["clear", "today"], St = ({
  value: e,
  onChange: n,
  id: t,
  disabled: c = !1,
  error: r = !1,
  width: i = "unset",
  align: o = "start",
  placeholder: s = "",
  css: a = "",
  title: u = "",
  format: w = "",
  months: f = 2,
  buttons: x = at,
  editable: m = !1,
  clear: p = !1
}) => {
  const [d, h] = W(e), { calendar: v, formats: N } = (Q(G) || Z()).getRaw(), R = w || N?.dateFormat, g = typeof R == "function" ? R : X(R, v), [T, M] = P(!1), S = y(() => {
    M(!1);
  }, []), D = E(() => d ? d.start ? g(d.start) + (d.end ? ` - ${g(d.end)}` : "") : g(d) : "", [d, g]), b = E(() => d && d.start || null, [d]), k = E(() => d && d.end || null, [d]), F = y(
    (L) => {
      const I = L.start || L.end ? { start: L.start, end: L.end } : null;
      h && h(I), (L.start && L.end || !L.start && !L.end) && (n && n({ value: I }), setTimeout(S, 1));
    },
    [n, S, h]
  ), A = y(
    (L) => {
      if (!m && !p) return;
      const { value: I, input: O } = L;
      if (O) return;
      const [Y, _] = I.split(" -").map((C, U) => {
        const K = C.trim();
        let V = typeof m == "function" ? m(K) : K ? new Date(K) : null, j = U === 0 ? b : k;
        return isNaN(V) ? j || null : V || null;
      });
      F({ start: Y, end: _ });
    },
    [m, p, b, k, F]
  );
  return B(() => {
    const L = () => {
      S();
    };
    return window.addEventListener("scroll", L), () => {
      window.removeEventListener("scroll", L);
    };
  }, [S]), /* @__PURE__ */ $(
    "div",
    {
      className: `wx-1b0WDQ wx-daterangepicker ${c ? "wx-disabled" : ""} ${r ? "wx-error" : ""}`,
      onClick: () => M(!0),
      children: [
        /* @__PURE__ */ l(
          fe,
          {
            css: a,
            title: u,
            value: D,
            id: t,
            readonly: !m,
            disabled: c,
            placeholder: s,
            error: r,
            onChange: A,
            icon: "wxi-calendar",
            inputStyle: {
              cursor: "pointer",
              width: "100%",
              paddingRight: "calc(var(--wx-input-icon-size) + var(--wx-input-icon-indent) * 2)"
            },
            clear: p
          }
        ),
        T && !c && /* @__PURE__ */ l(
          te,
          {
            onCancel: S,
            width: i,
            align: o,
            autoFit: !!o,
            children: /* @__PURE__ */ l(
              ct,
              {
                onCancel: S,
                buttons: x,
                start: b,
                end: k,
                months: f,
                onChange: F
              }
            )
          }
        )
      ]
    }
  );
};
function Ft({
  hotkey: e = null,
  toggleButton: n,
  children: t
}) {
  const c = H(null), [r, i] = P(!1), o = E(
    () => `wxi-${r ? "collapse" : "expand"}`,
    [r]
  ), s = H(null);
  return s.current = () => {
    const a = c.current;
    !r && a ? a.requestFullscreen() : r && document.exitFullscreen(), i((u) => !u);
  }, Pe({ [e]: () => s.current() }, c), B(() => {
    const a = () => {
      i(document.fullscreenElement === c.current);
    };
    return document.addEventListener("fullscreenchange", a), () => {
      document.removeEventListener("fullscreenchange", a);
    };
  }, []), /* @__PURE__ */ $("div", { className: "wx-13aa5W wx-fullscreen", ref: c, tabIndex: -1, children: [
    t,
    n ? n(s.current, r) : /* @__PURE__ */ l(ee, { css: "wx-13aa5W wx-fullscreen-button", onClick: s.current, children: /* @__PURE__ */ l("i", { className: `wx-13aa5W ${o} wx-fullscreen-icon`, children: " " }) })
  ] });
}
function Lt({ css: e = "", title: n = "", children: t, onClick: c }) {
  return t ? /* @__PURE__ */ l(
    "i",
    {
      title: n,
      role: "img",
      className: `wx-mdnST1 wx-icon ${e}`,
      onClick: c,
      children: t
    }
  ) : /* @__PURE__ */ l(
    "i",
    {
      title: n,
      className: `wx-mdnST1 wx-icon ${e}`,
      onClick: c
    }
  );
}
const ot = [];
function Ct({
  id: e,
  value: n = ot,
  options: t = [],
  textOptions: c = null,
  textField: r = "label",
  placeholder: i = "",
  title: o = "",
  disabled: s = !1,
  error: a = !1,
  checkboxes: u = !1,
  onChange: w,
  children: f
}) {
  const x = z(e), m = H(null), p = H(null), [d, h] = W(n), [v, N] = P(""), [R, g] = P(!1), T = H(null), M = E(() => d ? (c || t).filter((I) => d.includes(I.id)) : [], [d, c, t]), S = E(() => {
    const I = t;
    return v ? I.filter((O) => O[r].toLowerCase().includes(v.toLowerCase())) : I;
  }, [v, t, r]), D = y((I) => {
    m.current = I.navigate, p.current = I.keydown;
  }, []), b = y(() => {
    S.length ? m.current(0) : m.current(null);
  }, [S]), k = y(
    (I) => {
      const { id: O } = I;
      if (O) {
        let Y;
        d ? d.includes(O) ? Y = d.filter((_) => _ !== O) : Y = [...d, O] : Y = [O], h(Y), w && w({ value: Y }), T.current.focus();
      }
    },
    [d, w]
  ), F = y(
    (I, O) => {
      O && O.stopPropagation();
      const Y = d.filter((_) => _ !== I);
      h(Y), w && w({ value: Y });
    },
    [d, w]
  ), A = y(
    () => d && d.length ? S.findIndex((I) => I.id === d[0]) : 0,
    [d, S]
  ), L = y(() => {
    s || (T.current.focus(), m.current(A()));
  }, [s, A]);
  return /* @__PURE__ */ $(
    "div",
    {
      title: o,
      className: `wx-12Wj21 wx-multicombo ${a ? "wx-error" : ""} ${s ? "wx-disabled" : ""} ${M.length ? "wx-not-empty" : ""} ${R && !s ? "wx-focus" : ""}`,
      onClick: L,
      onKeyDown: (I) => p.current(I, A()),
      children: [
        /* @__PURE__ */ $("div", { className: "wx-12Wj21 wx-wrapper", children: [
          /* @__PURE__ */ l("div", { className: "wx-12Wj21 wx-tags", children: M.map((I) => /* @__PURE__ */ $("div", { className: "wx-12Wj21 wx-tag", children: [
            f ? ne(f, { option: I }) : I[r],
            !s && /* @__PURE__ */ l(
              "i",
              {
                className: "wx-12Wj21 wx-icon wxi-close",
                onClick: (O) => F(I.id, O)
              }
            )
          ] }, I.id)) }),
          /* @__PURE__ */ $("div", { className: "wx-12Wj21 wx-select", children: [
            /* @__PURE__ */ l(
              "input",
              {
                className: "wx-12Wj21 wx-input",
                id: x,
                type: "text",
                value: v,
                onChange: (I) => N(I.target.value),
                onInput: b,
                placeholder: i,
                disabled: s,
                onFocus: () => g(!0),
                onBlur: () => g(!1),
                ref: T
              }
            ),
            /* @__PURE__ */ l("i", { className: "wx-12Wj21 wx-icon wxi-angle-down" })
          ] })
        ] }),
        !s && /* @__PURE__ */ l(we, { items: S, onReady: D, onSelect: k, children: ({ option: I }) => /* @__PURE__ */ $(J, { children: [
          u && /* @__PURE__ */ l(
            Ie,
            {
              style: { marginRight: "8px", pointerEvents: "none" },
              name: I.id,
              value: d && d.includes(I.id)
            }
          ),
          f ? ne(f, { option: I }) : I[r]
        ] }) })
      ]
    }
  );
}
function Pt({
  left: e = 0,
  top: n = 0,
  at: t = "bottom",
  parent: c = null,
  onCancel: r,
  children: i
}) {
  const [o, s] = P(0), [a, u] = P(0), [w, f] = P("auto"), [x, m] = P(!1), p = H(null), d = y(() => {
    if (!p.current) return;
    const h = Be(p.current, c, t, e, n);
    h && (s(h.x), u(h.y), f(h.width), m(!0));
  }, [p, c, t, e, n]);
  return B(() => {
    requestAnimationFrame(d);
  }, []), B(() => {
    d(c);
  }, [c, t, e, n]), B(() => {
    const h = (v) => {
      r && r(v);
    };
    if (p.current)
      return de(p.current, h).destroy;
  }, [r]), /* @__PURE__ */ l(
    "div",
    {
      ref: p,
      className: "wx-37M6Fj wx-popup",
      style: {
        position: "absolute",
        visibility: x ? "visible" : "hidden",
        top: a + "px",
        left: o + "px",
        width: w
      },
      children: i
    }
  );
}
const Ht = ({
  total: e = 0,
  pageSize: n = 20,
  value: t = 1,
  onChange: c
}) => {
  const [r, i] = W(n), [o, s] = W(t), a = (Q(G) || Z()).getGroup("core"), u = Math.ceil(e / r), w = (o - 1) * r, f = Math.min(o * r, e), x = (d, h) => {
    Number.isNaN(d) || (h || s(d), d = Math.max(1, Math.min(d, u)), h && s(d), d !== o && setTimeout(() => {
      c && c({
        value: d,
        from: (d - 1) * r,
        to: Math.min(d * r, e),
        pageSize: r
      });
    }));
  };
  function m(d) {
    switch (d) {
      case "first":
        x(1);
        break;
      case "prev":
        x(o - 1);
        break;
      case "next":
        x(o + 1);
        break;
      case "last":
        x(u);
        break;
    }
  }
  const p = (d) => {
    const h = d.target.value * 1;
    Number.isNaN(h) || (i(h), c && c({ value: o, from: w, to: f, pageSize: h }));
  };
  return /* @__PURE__ */ $("div", { className: "wx-35Np0p wx-pager", children: [
    /* @__PURE__ */ $("div", { className: "wx-35Np0p wx-left", children: [
      /* @__PURE__ */ $("span", { children: [
        a("Rows per page"),
        ": "
      ] }),
      /* @__PURE__ */ l(
        "input",
        {
          type: "number",
          min: 1,
          className: "wx-35Np0p wx-input",
          value: r,
          onChange: p
        }
      )
    ] }),
    /* @__PURE__ */ $("div", { className: "wx-35Np0p wx-center", children: [
      /* @__PURE__ */ l(
        "i",
        {
          className: "wx-35Np0p wx-icon wxi-angle-dbl-left" + (o === 1 ? " wx-disabled" : ""),
          onClick: () => m("first")
        }
      ),
      /* @__PURE__ */ l(
        "i",
        {
          className: "wx-35Np0p wx-icon wxi-angle-left" + (o === 1 ? " wx-disabled" : ""),
          onClick: () => m("prev")
        }
      ),
      /* @__PURE__ */ l(
        "input",
        {
          type: "text",
          value: o,
          className: "wx-35Np0p wx-input",
          onChange: (d) => x(d.target.value * 1),
          onBlur: (d) => x(d.target.value * 1, !0)
        }
      ),
      /* @__PURE__ */ l(
        "i",
        {
          className: "wx-35Np0p wx-icon wxi-angle-right" + (o === u ? " wx-disabled" : ""),
          onClick: () => m("next")
        }
      ),
      /* @__PURE__ */ l(
        "i",
        {
          className: "wx-35Np0p wx-icon wxi-angle-dbl-right" + (o === u ? " wx-disabled" : ""),
          onClick: () => m("last")
        }
      )
    ] }),
    /* @__PURE__ */ $("div", { className: "wx-35Np0p wx-right", children: [
      a("Total pages"),
      ": ",
      u
    ] })
  ] });
};
function it({
  id: e,
  label: n = "",
  value: t,
  name: c = "",
  inputValue: r = "",
  disabled: i = !1,
  onChange: o
}) {
  const s = Fe(), a = z(e) || s, [u, w] = P(!1), f = t !== void 0, x = f ? t : u;
  function m(p) {
    const d = p.target.checked;
    f || w(d), d && o && o({ value: !0, inputValue: r });
  }
  return /* @__PURE__ */ $("div", { className: "wx-3lDF1s wx-radio", children: [
    /* @__PURE__ */ l(
      "input",
      {
        type: "radio",
        className: "wx-3lDF1s wx-input",
        id: a,
        disabled: i,
        name: c,
        value: r,
        checked: x,
        onChange: m
      }
    ),
    /* @__PURE__ */ $("label", { htmlFor: a, className: "wx-3lDF1s wx-label", children: [
      /* @__PURE__ */ l("span", { className: "wx-3lDF1s wx-circle" }),
      n && /* @__PURE__ */ l("span", { className: "wx-3lDF1s wx-radio-text", children: n })
    ] })
  ] });
}
function Bt({
  options: e = [{}],
  value: n = "",
  onChange: t,
  type: c = ""
}) {
  const r = le(), [i, o] = P(n);
  B(() => {
    o(n);
  }, [n]);
  function s(a) {
    const u = a.inputValue;
    o(u), t && t({ value: u });
  }
  return /* @__PURE__ */ l(se.Provider, { value: null, children: /* @__PURE__ */ l("div", { className: `wx-38w70j wx-radiogroup ${c && `wx-${c}`}`, children: e.map((a) => /* @__PURE__ */ l("div", { className: "wx-38w70j wx-item", children: /* @__PURE__ */ l(
    it,
    {
      label: a.label,
      inputValue: a.id,
      value: i === a.id,
      name: r,
      onChange: s
    }
  ) }, a.id)) }) });
}
function Et({
  value: e = "",
  options: n = [],
  textOptions: t = null,
  placeholder: c = "",
  disabled: r = !1,
  error: i = !1,
  title: o = "",
  textField: s = "label",
  clear: a = !1,
  children: u,
  onChange: w
}) {
  const f = H(null), x = H(null);
  let [m, p] = W(e);
  function d(g) {
    f.current = g.navigate, x.current = g.keydown;
  }
  const h = E(() => m || m === 0 ? (t || n).find((g) => g.id === m) : null, [m, t, n]), v = y(
    ({ id: g }) => {
      (g || g === 0) && (p(g), f.current(null), w && w({ value: g }));
    },
    [p, w]
  ), N = y(
    (g) => {
      g.stopPropagation(), p(""), w && w({ value: "" });
    },
    [p, w]
  ), R = y(() => n.findIndex((g) => g.id === m), [n, m]);
  return /* @__PURE__ */ $(
    "div",
    {
      className: `wx-2YgblL wx-richselect ${i ? "wx-2YgblL wx-error" : ""} ${r ? "wx-2YgblL wx-disabled" : ""} ${u ? "" : "wx-2YgblL wx-nowrap"}`,
      title: o,
      onClick: () => f.current(R()),
      onKeyDown: (g) => x.current(g, R()),
      tabIndex: 0,
      children: [
        /* @__PURE__ */ l("div", { className: "wx-2YgblL wx-label", children: h ? u ? u(h) : h[s] : c ? /* @__PURE__ */ l("span", { className: "wx-2YgblL wx-placeholder", children: c }) : "" }),
        a && !r && m ? /* @__PURE__ */ l("i", { className: "wx-2YgblL wx-icon wxi-close", onClick: N }) : /* @__PURE__ */ l("i", { className: "wx-2YgblL wx-icon wxi-angle-down" }),
        !r && /* @__PURE__ */ l(we, { items: n, onReady: d, onSelect: v, children: ({ option: g }) => u ? u(g) : g[s] })
      ]
    }
  );
}
const At = ({
  options: e = [],
  value: n = "",
  css: t = "",
  children: c,
  onChange: r
}) => {
  const [i, o] = P(n), s = (a) => {
    o(a), r && r({ value: a });
  };
  return B(() => {
    o(n);
  }, [n]), /* @__PURE__ */ l("div", { className: `wx-0hMO7P wx-segmented ${t}`, children: e.map((a) => /* @__PURE__ */ l(
    "button",
    {
      className: `wx-0hMO7P wx-segment ${a.css} ${a.id === i ? "wx-selected" : ""}`,
      title: a.title,
      onClick: () => s(a.id),
      children: c ? ne(c, { option: a }) : /* @__PURE__ */ $(J, { children: [
        a.icon && /* @__PURE__ */ l(
          "i",
          {
            className: `wx-0hMO7P wx-icon ${a.icon} ${a.label ? "" : "wx-only"}`
          }
        ),
        a.label && /* @__PURE__ */ l("span", { className: "wx-0hMO7P wx-label", children: a.label })
      ] })
    },
    a.id
  )) });
};
function Ot({
  value: e,
  options: n = [],
  placeholder: t = "",
  title: c = "",
  disabled: r = !1,
  error: i = !1,
  textField: o = "label",
  clear: s = !1,
  id: a,
  onChange: u
}) {
  const w = z(a), [f, x] = P(e || ""), m = y(
    (d) => {
      const h = d.target.value;
      x(h), u && u({ value: h });
    },
    [u]
  ), p = y(() => {
    x(""), u && u({ value: "" });
  }, [u]);
  return /* @__PURE__ */ $("div", { className: "wx-2yx1Fo wx-select-box", children: [
    /* @__PURE__ */ $(
      "select",
      {
        id: w,
        value: f,
        disabled: r,
        className: `wx-2yx1Fo wx-select ${i ? "wx-2yx1Fo wx-error" : ""}`,
        title: c,
        onChange: m,
        children: [
          /* @__PURE__ */ l("option", { value: "", children: t }),
          n.map((d) => /* @__PURE__ */ l("option", { value: d.id, children: d[o] }, d.id))
        ]
      }
    ),
    s && !r && f ? /* @__PURE__ */ l("i", { className: "wx-2yx1Fo wx-icon wxi-close", onClick: p }) : /* @__PURE__ */ l("i", { className: "wx-2yx1Fo wx-icon wxi-angle-down" })
  ] });
}
function ke({
  id: e,
  label: n = "",
  css: t = "",
  min: c = 0,
  max: r = 100,
  value: i = 0,
  step: o = 1,
  title: s = "",
  disabled: a = !1,
  onChange: u
}) {
  const w = z(e), [f, x] = W(i), m = H({ value: f, input: f }), p = E(
    () => (f - c) / (r - c) * 100 + "%",
    [f, c, r]
  ), d = E(() => a ? "" : `linear-gradient(90deg, var(--wx-slider-primary) 0% ${p}, var(--wx-slider-background) ${p} 100%)`, [a, p]);
  function h({ target: R }) {
    const g = R.value * 1;
    x(g), u && u({
      value: g,
      previous: m.current.input,
      input: !0
    }), m.current.input = g;
  }
  function v({ target: R }) {
    const g = R.value * 1;
    x(g), u && u({ value: g, previous: m.current.value }), m.current.value = g;
  }
  B(() => {
    x(i);
  }, [i]);
  const N = H(null);
  return B(() => {
    if (N.current)
      return N.current.addEventListener("change", v), () => {
        N.current && N.current.removeEventListener("change", v);
      };
  }, [N, v]), /* @__PURE__ */ $("div", { className: `wx-2EDJ8G wx-slider ${t}`, title: s, children: [
    n && /* @__PURE__ */ l("label", { className: "wx-2EDJ8G wx-label", htmlFor: w, children: n }),
    /* @__PURE__ */ l("div", { className: "wx-2EDJ8G wx-inner", children: /* @__PURE__ */ l(
      "input",
      {
        id: w,
        className: "wx-2EDJ8G wx-input",
        type: "range",
        min: c,
        max: r,
        step: o,
        disabled: a,
        value: f,
        onInput: h,
        style: { background: d },
        ref: N
      }
    ) })
  ] });
}
function Wt({
  id: e,
  value: n = !1,
  disabled: t = !1,
  onChange: c
}) {
  const r = z(e), [i, o] = W(n), s = y(
    (a) => {
      const u = a.target.checked;
      o(u), c && c({ value: u });
    },
    [c]
  );
  return /* @__PURE__ */ $("label", { className: "wx-2dAR5c wx-switch", children: [
    /* @__PURE__ */ l(
      "input",
      {
        className: "wx-2dAR5c wx-input",
        type: "checkbox",
        checked: i,
        onChange: s,
        disabled: t,
        id: r
      }
    ),
    /* @__PURE__ */ l("span", { className: "wx-2dAR5c wx-box" })
  ] });
}
const Yt = ({
  options: e = [],
  value: n = "",
  type: t = "top",
  css: c = "",
  onChange: r
}) => {
  const [i, o] = W(n), s = (a) => {
    o(a), r && r({ value: a });
  };
  return /* @__PURE__ */ l("div", { className: `wx-138fWJ wx-tabs wx-${t} ${c}`, children: e.map((a) => /* @__PURE__ */ $(
    "button",
    {
      className: `wx-138fWJ wx-tab ${a.css} ${a.id === i ? "wx-active" : ""}`,
      title: a.title,
      onClick: () => s(a.id),
      children: [
        a.icon && /* @__PURE__ */ l(
          "i",
          {
            className: `wx-138fWJ wx-icon ${a.icon} ${a.label ? "" : "wx-only"}`
          }
        ),
        a.label && /* @__PURE__ */ l("span", { className: "wx-138fWJ wx-label", children: a.label })
      ]
    },
    a.id
  )) });
}, jt = ({
  id: e,
  value: n = 0,
  step: t = 1,
  min: c = 0,
  max: r = 1 / 0,
  error: i = !1,
  disabled: o = !1,
  readonly: s = !1,
  onChange: a
}) => {
  const u = z(e), [w, f] = W(n), x = y(() => {
    if (s || w <= c) return;
    const h = w - t;
    f(h), a && a({ value: h });
  }, [w, s, c, t, a]), m = y(() => {
    if (s || w >= r) return;
    const h = w + t;
    f(h), a && a({ value: h });
  }, [w, s, r, t, a]), p = y(() => {
    if (!s) {
      const h = Math.round(Math.min(r, Math.max(w, c)) / t) * t, v = isNaN(h) ? Math.max(c, 0) : h;
      f(v), a && a({ value: v });
    }
  }, [s, w, r, c, t, a]), d = y(
    (h) => {
      const v = h.target.value * 1;
      f(v), a && a({ value: v, input: !0 });
    },
    [a]
  );
  return /* @__PURE__ */ $(
    "div",
    {
      className: `wx-22t21n wx-counter ${o ? "wx-disabled" : ""} ${s ? "wx-readonly" : ""} ${i ? "wx-error" : ""}`,
      children: [
        /* @__PURE__ */ l(
          "button",
          {
            "aria-label": "-",
            className: "wx-22t21n wx-btn wx-btn-dec",
            disabled: o,
            onClick: x,
            children: /* @__PURE__ */ l(
              "svg",
              {
                className: "wx-22t21n wx-dec",
                width: "12",
                height: "2",
                viewBox: "0 0 12 2",
                fill: "none",
                xmlns: "http://www.w3.org/2000/svg",
                children: /* @__PURE__ */ l("path", { d: "M11.2501 1.74994H0.750092V0.249939H11.2501V1.74994Z" })
              }
            )
          }
        ),
        /* @__PURE__ */ l(
          "input",
          {
            id: u,
            type: "text",
            className: "wx-22t21n wx-input",
            disabled: o,
            readOnly: s,
            required: !0,
            value: w,
            onBlur: p,
            onInput: d
          }
        ),
        /* @__PURE__ */ l(
          "button",
          {
            "aria-label": "-",
            className: "wx-22t21n wx-btn wx-btn-inc",
            disabled: o,
            onClick: m,
            children: /* @__PURE__ */ l(
              "svg",
              {
                className: "wx-22t21n wx-inc",
                width: "12",
                height: "12",
                viewBox: "0 0 12 12",
                fill: "none",
                xmlns: "http://www.w3.org/2000/svg",
                children: /* @__PURE__ */ l(
                  "path",
                  {
                    d: `M11.2501
								6.74994H6.75009V11.2499H5.25009V6.74994H0.750092V5.24994H5.25009V0.749939H6.75009V5.24994H11.2501V6.74994Z`
                  }
                )
              }
            )
          }
        )
      ]
    }
  );
};
function ut({ notice: e = {} }) {
  function n() {
    e.remove && e.remove();
  }
  return /* @__PURE__ */ $(
    "div",
    {
      className: `wx-11sNg5 wx-notice wx-${e.type ? e.type : ""}`,
      role: "status",
      "aria-live": "polite",
      children: [
        /* @__PURE__ */ l("div", { className: "wx-11sNg5 wx-text", children: e.text }),
        /* @__PURE__ */ l("div", { className: "wx-11sNg5 wx-button", children: /* @__PURE__ */ l("i", { className: "wx-11sNg5 wxi-close", onClick: n }) })
      ]
    }
  );
}
function dt({ data: e = [] }) {
  return /* @__PURE__ */ l("div", { className: "wx-3nwoO9 wx-notices", children: e.map((n) => /* @__PURE__ */ l(ut, { notice: n }, n.id)) });
}
function wt({
  title: e = "",
  buttons: n = ["cancel", "ok"],
  header: t,
  children: c,
  footer: r,
  onConfirm: i,
  onCancel: o
}) {
  const s = (Q(G) || Z()).getGroup("core"), a = H(null);
  B(() => {
    a.current?.focus();
  }, []);
  function u(f) {
    switch (f.code) {
      case "Enter": {
        const x = f.target.tagName;
        if (x === "TEXTAREA" || x === "BUTTON") return;
        i && i({ event: f });
        break;
      }
      case "Escape":
        o && o({ event: f });
        break;
    }
  }
  function w(f, x) {
    const m = { event: f, button: x };
    x === "cancel" ? o && o(m) : i && i(m);
  }
  return /* @__PURE__ */ l(
    "div",
    {
      className: "wx-1FxkZa wx-modal",
      ref: a,
      tabIndex: 0,
      onKeyDown: u,
      children: /* @__PURE__ */ $("div", { className: "wx-1FxkZa wx-window", children: [
        t || (e ? /* @__PURE__ */ l("div", { className: "wx-1FxkZa wx-header", children: e }) : null),
        /* @__PURE__ */ l("div", { children: c }),
        r || n && /* @__PURE__ */ l("div", { className: "wx-1FxkZa wx-buttons", children: n.map((f) => /* @__PURE__ */ l("div", { className: "wx-1FxkZa wx-button", children: /* @__PURE__ */ l(
          ee,
          {
            type: `block ${f === "ok" ? "primary" : "secondary"}`,
            onClick: (x) => w(x, f),
            children: s(f)
          }
        ) }, f)) })
      ] })
    }
  );
}
function ft({ children: e }, n) {
  const [t, c] = P(null), [r, i] = P([]);
  return Ce(
    n,
    () => ({
      showModal: (o) => {
        const s = { ...o };
        return c(s), new Promise((a, u) => {
          s.resolve = (w) => {
            c(null), a(w);
          }, s.reject = (w) => {
            c(null), u(w);
          };
        });
      },
      showNotice: (o) => {
        o = { ...o }, o.id = o.id || le(), o.remove = () => i((s) => s.filter((a) => a.id !== o.id)), o.expire != -1 && setTimeout(o.remove, o.expire || 5100), i((s) => [...s, o]);
      }
    }),
    []
  ), /* @__PURE__ */ $(J, { children: [
    e,
    t && /* @__PURE__ */ l(
      wt,
      {
        title: t.title,
        buttons: t.buttons,
        onConfirm: t.resolve,
        onCancel: t.reject,
        children: t.message
      }
    ),
    /* @__PURE__ */ l(dt, { data: r })
  ] });
}
const xt = Le(ft);
function Jt(e) {
  let { children: n } = e, t = H();
  const c = E(() => ({
    showNotice: (r) => t.current.showNotice(r),
    showModal: (r) => t.current.showModal(r)
  }), []);
  return /* @__PURE__ */ l(Me.Provider, { value: c, children: /* @__PURE__ */ l(xt, { ref: t, children: n }) });
}
function ye({
  label: e = "",
  position: n = "",
  css: t = "",
  error: c = !1,
  type: r = "",
  required: i = !1,
  children: o
}) {
  const s = H(null), a = y(() => {
    if (s.current) return s.current;
    const u = le();
    return s.current = u, u;
  }, []);
  return /* @__PURE__ */ l(se.Provider, { value: a, children: /* @__PURE__ */ $(
    "div",
    {
      className: `wx-2oVUvC wx-field wx-${n} ${t} ${c ? "wx-error" : ""} ${i ? "wx-required" : ""}`.trim(),
      children: [
        e && /* @__PURE__ */ l("label", { className: "wx-2oVUvC wx-label", htmlFor: s.current, children: e }),
        /* @__PURE__ */ l("div", { className: `wx-2oVUvC wx-field-control wx-${r}`, children: o })
      ]
    }
  ) });
}
const mt = ({
  value: e = !1,
  type: n = "",
  icon: t = "",
  disabled: c = !1,
  iconActive: r = "",
  onClick: i,
  title: o = "",
  css: s = "",
  text: a = "",
  textActive: u = "",
  children: w,
  active: f,
  onChange: x
}) => {
  const [m, p] = W(e), d = E(() => (m ? "pressed" : "") + (n ? " " + n : ""), [m, n]), h = y(
    (v) => {
      let N = !m;
      i && i(v), v.defaultPrevented || (p(N), x && x({ value: N }));
    },
    [m, i, x]
  );
  return m && f ? /* @__PURE__ */ l(
    ee,
    {
      title: o,
      text: m && u || a,
      css: s,
      type: d,
      icon: m && r || t,
      onClick: h,
      disabled: c,
      children: ne(f, { value: m })
    }
  ) : w ? /* @__PURE__ */ l(
    ee,
    {
      title: o,
      text: m && u || a,
      css: s,
      type: d,
      icon: m && r || t,
      onClick: h,
      disabled: c,
      children: w
    }
  ) : /* @__PURE__ */ l(
    ee,
    {
      title: o,
      text: m && u || a,
      css: s,
      type: d,
      icon: m && r || t,
      onClick: h,
      disabled: c
    }
  );
}, be = new Date(0, 0, 0, 0, 0);
function Gt({
  value: e = be,
  id: n,
  title: t = "",
  css: c = "",
  disabled: r = !1,
  error: i = !1,
  format: o = "",
  onChange: s
}) {
  let [a, u] = W(e);
  const { calendar: w, formats: f } = (Q(G) || Z()).getRaw(), x = w.clockFormat == 12, m = 23, p = 59, d = E(() => {
    const V = o || f?.timeFormat;
    return typeof V == "function" ? V : X(V, w);
  }, [o, f, w]), h = E(() => d(new Date(0, 0, 0, 1)).indexOf("01") != -1, [d]), v = (V, j) => (V < 10 && j ? `0${V}` : `${V}`).slice(-2), N = (V) => v(V, !0), R = (V) => `${V}`.replace(/[^\d]/g, "") || 0, g = (V) => x && (V = V % 12, V === 0) ? "12" : v(V, h), T = y((V, j) => (V = R(V), Math.min(V, j)), []), [M, S] = P(null), D = a || be, b = T(D.getHours(), m), k = T(D.getMinutes(), p), F = b > 12, A = g(b), L = N(k), I = E(
    () => d(new Date(0, 0, 0, b, k)),
    [b, k, d]
  ), O = y(() => {
    S(!0);
  }, []), Y = y(() => {
    const V = new Date(D);
    V.setHours(V.getHours() + (F ? -12 : 12)), u(V), s && s({ value: V });
  }, [D, F, s]), _ = y(
    ({ value: V }) => {
      if (D.getHours() === V) return;
      const j = new Date(D);
      j.setHours(V), u(j), s && s({ value: j });
    },
    [D, s]
  ), C = y(
    ({ value: V }) => {
      if (D.getMinutes() === V) return;
      const j = new Date(D);
      j.setMinutes(V), u(j), s && s({ value: j });
    },
    [D, s]
  ), U = y(
    (V) => (V = T(V, m), x && (V = V * 1, V === 12 && (V = 0), F && (V += 12)), V),
    [T, x, F]
  ), K = y(() => {
    S(null);
  }, []);
  return /* @__PURE__ */ $(
    "div",
    {
      className: `wx-7f497i wx-timepicker ${i ? "wx-7f497i wx-error" : ""} ${r ? "wx-7f497i wx-disabled" : ""}`,
      onClick: r ? void 0 : O,
      style: { cursor: r ? "default" : "pointer" },
      children: [
        /* @__PURE__ */ l(
          fe,
          {
            id: n,
            css: c,
            title: t,
            value: I,
            readonly: !0,
            disabled: r,
            error: i,
            icon: "wxi-clock",
            inputStyle: {
              cursor: "pointer",
              width: "100%",
              paddingRight: "calc(var(--wx-input-icon-size) + var(--wx-input-icon-indent) * 2)"
            }
          }
        ),
        M && !r && /* @__PURE__ */ l(te, { onCancel: K, width: "unset", children: /* @__PURE__ */ $("div", { className: "wx-7f497i wx-wrapper", children: [
          /* @__PURE__ */ $("div", { className: "wx-7f497i wx-timer", children: [
            /* @__PURE__ */ l(
              "input",
              {
                className: "wx-7f497i wx-digit",
                value: A,
                onChange: (V) => {
                  const j = U(V.target.value);
                  _({ value: j });
                }
              }
            ),
            /* @__PURE__ */ l("div", { className: "wx-7f497i wx-separator", children: ":" }),
            /* @__PURE__ */ l(
              "input",
              {
                className: "wx-7f497i wx-digit",
                value: L,
                onChange: (V) => {
                  const j = T(V.target.value, p);
                  C({ value: j });
                }
              }
            ),
            x && /* @__PURE__ */ l(
              mt,
              {
                value: F,
                onClick: Y,
                active: () => /* @__PURE__ */ l("span", { children: "pm" }),
                children: /* @__PURE__ */ l("span", { children: "am" })
              }
            )
          ] }),
          /* @__PURE__ */ l(ye, { width: "unset", children: /* @__PURE__ */ l(
            ke,
            {
              label: w.hours,
              value: b,
              onChange: _,
              max: m
            }
          ) }),
          /* @__PURE__ */ l(ye, { width: "unset", children: /* @__PURE__ */ l(
            ke,
            {
              label: w.minutes,
              value: k,
              onChange: C,
              max: p
            }
          ) })
        ] }) })
      ]
    }
  );
}
function Qt({ children: e }) {
  return /* @__PURE__ */ l("div", { className: "wx-KgpO9N wx-modal", children: /* @__PURE__ */ l("div", { className: "wx-KgpO9N wx-window", children: e }) });
}
function _t({ position: e = "right", children: n, onCancel: t }) {
  const c = H(null);
  return B(() => de(c.current, t).destroy, []), /* @__PURE__ */ l("div", { ref: c, className: `wx-2L733M wx-sidearea wx-pos-${e}`, children: n });
}
function Ut({ theme: e = "", target: n, children: t }) {
  const c = H(null), r = H(null), [i, o] = P(null);
  c.current || (c.current = document.createElement("div"));
  const s = Q(re);
  return B(() => {
    o(
      n || ht(r.current) || oe.getTopNode(r.current)
    );
  }, [r.current]), /* @__PURE__ */ $(J, { children: [
    /* @__PURE__ */ l("span", { ref: r, style: { display: "none" } }),
    r.current && i ? Ae(
      /* @__PURE__ */ l(
        "div",
        {
          className: `wx-3ZWsT0 wx-${e || s}-theme`,
          children: t
        }
      ),
      i
    ) : null
  ] });
}
function ht(e) {
  const n = oe.getTopNode(e);
  for (; e && e !== n && !e.getAttribute("data-wx-portal-root"); )
    e = e.parentNode;
  return e;
}
function pt() {
  return /* @__PURE__ */ l(J, {});
}
function zt(e) {
  const { fonts: n = !0, children: t } = e;
  return /* @__PURE__ */ l(re.Provider, { value: "material", children: /* @__PURE__ */ $(J, { children: [
    t && /* @__PURE__ */ l("div", { className: "wx-theme wx-material-theme", children: t }),
    n && /* @__PURE__ */ $(J, { children: [
      /* @__PURE__ */ l(
        "link",
        {
          rel: "preconnect",
          href: "https://cdn.svar.dev",
          crossOrigin: "true"
        }
      ),
      /* @__PURE__ */ l(pt, {}),
      /* @__PURE__ */ l(
        "link",
        {
          rel: "stylesheet",
          href: "https://cdn.svar.dev/fonts/wxi/wx-icons.css"
        }
      )
    ] })
  ] }) });
}
function Se() {
  return /* @__PURE__ */ l(J, {});
}
function Kt(e) {
  const { fonts: n = !0, children: t } = e;
  return /* @__PURE__ */ l(re.Provider, { value: "willow", children: /* @__PURE__ */ $(J, { children: [
    t && t && /* @__PURE__ */ l("div", { className: "wx-theme wx-willow-theme", children: t }),
    n && /* @__PURE__ */ $(J, { children: [
      /* @__PURE__ */ l(
        "link",
        {
          rel: "preconnect",
          href: "https://cdn.svar.dev",
          crossOrigin: "true"
        }
      ),
      /* @__PURE__ */ l(Se, {}),
      /* @__PURE__ */ l(
        "link",
        {
          rel: "stylesheet",
          href: "https://cdn.svar.dev/fonts/wxi/wx-icons.css"
        }
      )
    ] })
  ] }) });
}
function Zt(e) {
  const { fonts: n = !0, children: t } = e;
  return /* @__PURE__ */ l(re.Provider, { value: "willow-dark", children: /* @__PURE__ */ $(J, { children: [
    t && t && /* @__PURE__ */ l("div", { className: "wx-theme wx-willow-dark-theme", children: t }),
    n && /* @__PURE__ */ $(J, { children: [
      /* @__PURE__ */ l(
        "link",
        {
          rel: "preconnect",
          href: "https://cdn.svar.dev",
          crossOrigin: "true"
        }
      ),
      /* @__PURE__ */ l(Se, {}),
      /* @__PURE__ */ l(
        "link",
        {
          rel: "stylesheet",
          href: "https://cdn.svar.dev/fonts/wxi/wx-icons.css"
        }
      )
    ] })
  ] }) });
}
function qt(e) {
  const n = le();
  e.setAttribute("data-wx-portal-root", n);
}
Ee(oe);
export {
  ee as Button,
  lt as Calendar,
  Ie as Checkbox,
  $t as CheckboxGroup,
  Ye as ColorBoard,
  Mt as ColorPicker,
  Tt as ColorSelect,
  It as Combo,
  jt as Counter,
  Rt as DatePicker,
  St as DateRangePicker,
  te as Dropdown,
  ye as Field,
  Ft as Fullscreen,
  Jt as Globals,
  Lt as Icon,
  ue as Locale,
  zt as Material,
  wt as Modal,
  Qt as ModalArea,
  Ge as Month,
  Ct as MultiCombo,
  Ht as Pager,
  Pt as Popup,
  Ut as Portal,
  it as RadioButton,
  Bt as RadioButtonGroup,
  ct as RangeCalendar,
  Et as RichSelect,
  At as Segmented,
  Ot as Select,
  _t as SideArea,
  ke as Slider,
  we as SuggestDropdown,
  Wt as Switch,
  Yt as Tabs,
  fe as Text,
  Vt as TextArea,
  Gt as TimePicker,
  mt as TwoState,
  Kt as Willow,
  Zt as WillowDark,
  Dt as context,
  ln as en,
  tn as locale,
  qt as popupContainer
};
