// @univerjs/preset-sheets-advanced/lib/es/index.js
import { UniverProFormulaEnginePlugin as f } from "@univerjs-pro/engine-formula";
export * from "@univerjs-pro/engine-formula";
import { UniverExchangeClientPlugin as s } from "@univerjs-pro/exchange-client";
export * from "@univerjs-pro/exchange-client";
import { UniverLicensePlugin as u } from "@univerjs-pro/license";
export * from "@univerjs-pro/license";
import { UniverSheetsChartPlugin as v } from "@univerjs-pro/sheets-chart";
export * from "@univerjs-pro/sheets-chart";
import { UniverSheetsChartUIPlugin as x } from "@univerjs-pro/sheets-chart-ui";
export * from "@univerjs-pro/sheets-chart-ui";
import { UniverSheetsExchangeClientPlugin as c } from "@univerjs-pro/sheets-exchange-client";
export * from "@univerjs-pro/sheets-exchange-client";
import { UniverSheetsPivotTablePlugin as n } from "@univerjs-pro/sheets-pivot";
export * from "@univerjs-pro/sheets-pivot";
import { UniverSheetsPivotTableUIPlugin as g } from "@univerjs-pro/sheets-pivot-ui";
export * from "@univerjs-pro/sheets-pivot-ui";
import { UniverSheetsPrintPlugin as U } from "@univerjs-pro/sheets-print";
export * from "@univerjs-pro/sheets-print";
import { UniverSheetSparklinePlugin as h } from "@univerjs-pro/sheets-sparkline";
export * from "@univerjs-pro/sheets-sparkline";
import { UniverSheetSparklineUIPlugin as P } from "@univerjs-pro/sheets-sparkline-ui";
export * from "@univerjs-pro/sheets-sparkline-ui";
import "@univerjs-pro/exchange-client/lib/facade";
import "@univerjs-pro/sheets-pivot/lib/facade";
import "@univerjs-pro/engine-formula/lib/facade";
import "@univerjs-pro/sheets-print/lib/facade";
import "@univerjs-pro/sheets-chart-ui/lib/facade";
import "@univerjs-pro/sheets-sparkline/lib/facade";
export * from "@univerjs-pro/engine-chart";
export * from "@univerjs/sheets-graphics";
function j(p = {
  license: "",
  universerEndpoint: ""
}) {
  const {
    license: m,
    universerEndpoint: e,
    useWorker: t,
    formula: o,
    exchangeClientOptions: l,
    print: i
  } = p, r = e != null ? e : `${window.location.protocol}//${window.location.host}`;
  return {
    plugins: [
      [u, { license: m }],
      // TODO: @wzhudev: if we use worker, we need to add different configurations to SheetsPivotTable
      t ? [n, { notExecuteFormula: !0 }] : [n],
      g,
      [f, {
        notExecuteFormula: t,
        function: o == null ? void 0 : o.function
      }],
      [U, {
        enforceWatermark: i == null ? void 0 : i.enforceWatermark
      }],
      v,
      x,
      h,
      P,
      [s, {
        uploadFileServerUrl: `${r}/universer-api/stream/file/upload`,
        getTaskServerUrl: `${r}/universer-api/exchange/task/{taskID}`,
        signUrlServerUrl: `${r}/universer-api/file/{fileID}/sign-url`,
        importServerUrl: `${r}/universer-api/exchange/{type}/import`,
        exportServerUrl: `${r}/universer-api/exchange/{type}/export`,
        downloadEndpointUrl: `${r}/`,
        options: l
      }],
      c
    ].filter((a) => !!a)
  };
}
export {
  j as UniverSheetsAdvancedPreset
};
