// @svar-ui/react-grid/dist/index.es.js
import { jsxs as Q, jsx as c, Fragment as xe } from "react/jsx-runtime";
import Xt, { createContext as Yt, useContext as Z, useMemo as S, useRef as I, useEffect as K, useCallback as W, useState as O, forwardRef as jt, useImperativeHandle as Bt } from "react";
import { Text as Gt, RichSelect as Ut, SuggestDropdown as yt, Dropdown as qt, Calendar as Qt, context as Je, Locale as Zt, Material as dt, Willow as ut, WillowDark as ft } from "@svar-ui/react-core";
import { en as et } from "@svar-ui/grid-locales";
import { EventBusRouter as Jt } from "@svar-ui/lib-state";
import { getRenderValue as Ke, getValue as en, isCommunity as tn, getPrintFilterValue as nn, getPrintCellStyle as bt, getHeaderFooterPrintColumns as ht, getPrintColumns as on, hotkeys as rn, defaultHotkeys as ln, scrollTo as sn, DataStore as cn, assignChecks as vt, defaultMenuOptions as an, handleAction as Ct, defaultToolbarButtons as dn } from "@svar-ui/grid-store";
import { defaultMenuOptions as wo, defaultToolbarButtons as xo, getEditorConfig as mo } from "@svar-ui/grid-store";
import { useWritableProp as un, useStore as F, styleObject as St, useStoreWithCounter as fn, delegateClick as hn, writable as gn, useStoreLater as he } from "@svar-ui/lib-react";
import { locate as Qe, id as Rt, clickOutside as Nt, locateAttr as gt, locale as Dt, setEnv as pn, env as wn } from "@svar-ui/lib-dom";
import { registerMenuItem as xn, ContextMenu as Tt } from "@svar-ui/react-menu";
import { Toolbar as mn } from "@svar-ui/react-toolbar";
const oe = Yt(null);
function yn(n, e) {
  const t = new ResizeObserver((r) => {
    requestAnimationFrame(() => e(r[0].contentRect));
  });
  return t.observe(n.parentNode), {
    destroy() {
      t.disconnect();
    }
  };
}
const pt = 5, bn = 700;
function vn(n) {
  return Rt(n.getAttribute("data-id"));
}
function Pe(n) {
  const e = n.getBoundingClientRect(), t = document.body, r = e.top + t.scrollTop - t.clientTop || 0, u = e.left + t.scrollLeft - t.clientLeft || 0;
  return {
    y: Math.round(r),
    x: Math.round(u),
    width: n.offsetWidth,
    height: n.offsetHeight
  };
}
function Ze(n, e) {
  const t = Pe(e);
  return { x: n.clientX - t.x, y: n.clientY - t.y };
}
function Cn(n, e) {
  const t = e.current;
  let r = null, u, x, a = !1, b = !1;
  const d = document.createElement("DIV");
  d.className = "wx-drag-zone", d.setAttribute("tabindex", -1);
  function f() {
    clearTimeout(u), u = null;
  }
  function h(i) {
    const k = Qe(i);
    k && (r = {
      container: d,
      sourceNode: i.target,
      from: vn(k),
      pos: Ze(i, n)
    }, x = r.pos, s(i));
  }
  function s(i) {
    if (!r) return;
    const k = r.pos = Ze(i, n);
    if (!a) {
      if (!b && !i?.target?.getAttribute("draggable-data") && Math.abs(x.x - k.x) < pt && Math.abs(x.y - k.y) < pt)
        return;
      if (D(i) === !1) return R();
    }
    if (b) {
      const U = window.scrollX || document.documentElement.scrollLeft || document.body.scrollLeft, J = window.scrollY || document.documentElement.scrollTop || document.body.scrollTop;
      r.targetNode = document.elementFromPoint(
        i.pageX - U,
        i.pageY - J
      );
    } else r.targetNode = i.target;
    t.move && t.move(i, r), d.style.left = -(r.offset ? r.offset.x : 0) + "px", d.style.top = r.pos.y + (r.offset ? r.offset.y : 0) + "px";
  }
  function v(i) {
    d.parentNode && d.parentNode.removeChild(d), d.innerHTML = "", a && t.end && t.end(i, r), r = x = null, R();
  }
  function E(i) {
    t.getReorder && !t.getReorder() || i.button === 0 && (T(i), window.addEventListener("mousemove", C), window.addEventListener("mouseup", l), h(i));
  }
  function C(i) {
    s(i);
  }
  function l(i) {
    v(i);
  }
  function m(i) {
    if (t.getReorder && !t.getReorder()) return;
    u = setTimeout(() => {
      b = !0, h(i.touches[0]);
    }, bn), T(i);
    function k() {
      u && f(), i.target.removeEventListener("touchmove", g), i.target.removeEventListener("touchend", k), v(i);
    }
    i.target.addEventListener("touchmove", g), i.target.addEventListener("touchend", k), n.addEventListener("contextmenu", w);
  }
  function g(i) {
    a ? (i.preventDefault(), s(i.touches[0])) : u && f();
  }
  function w(i) {
    if (a || u)
      return i.preventDefault(), !1;
  }
  function y(i) {
    i.preventDefault();
  }
  function T(i) {
    if (!t.getDraggableInfo) return;
    const { hasDraggable: k } = t.getDraggableInfo();
    (!k || i.target.getAttribute("draggable-data")) && (document.body.style.userSelect = "none", document.body.style.webkitUserSelect = "none");
  }
  function D(i) {
    if (a = !0, t.start) {
      if (t.start(i, r) === !1) return !1;
      n.appendChild(d), document.body.style.cursor = "move";
    }
  }
  function R(i) {
    a = b = !1, document.body.style.cursor = "", document.body.style.userSelect = "", document.body.style.webkitUserSelect = "", window.removeEventListener("mousemove", C), window.removeEventListener("mouseup", l), i && (n.removeEventListener("mousedown", E), n.removeEventListener("touchstart", m), n.removeEventListener("dragstart", y));
  }
  return n.addEventListener("mousedown", E), n.addEventListener("touchstart", m), n.addEventListener("dragstart", y), {
    destroy() {
      R(!0);
    }
  };
}
const Sn = 4e-3;
function Rn() {
  return {
    dirX: 0,
    dirY: 0,
    scrollSpeedFactor: 1
  };
}
function Nn(n, e, t, r) {
  const { node: u, left: x, top: a, bottom: b, sense: d, xScroll: f, yScroll: h } = r, s = Ze(n, u);
  t.scrollState || (t.scrollState = Rn());
  let v = 0, E = 0;
  s.x < x + d ? v = -1 : s.x > e.width - d && (v = 1), s.y < a + Math.round(d / 2) ? E = -1 : s.y > e.height - b - Math.round(d / 2) && (E = 1), (t.scrollState.dirX !== v || t.scrollState.dirY !== E) && (Et(t), t.scrollState.dirX = v, t.scrollState.dirY = E), (f && t.scrollState.dirX !== 0 || h && t.scrollState.dirY !== 0) && Dn(t, r, {
    x: t.scrollState.dirX,
    y: t.scrollState.dirY
  });
}
function Dn(n, e, t) {
  n.autoScrollTimer || (n.autoScrollTimer = setTimeout(() => {
    n.activeAutoScroll = setInterval(
      Tn,
      15,
      n,
      e,
      t
    );
  }, 250));
}
function Et(n) {
  n.scrollSpeedFactor = 1, n.autoScrollTimer && (n.autoScrollTimer = clearTimeout(n.autoScrollTimer), n.activeAutoScroll = clearInterval(n.activeAutoScroll));
}
function Tn(n, e, t) {
  const { x: r, y: u } = t;
  n.scrollSpeedFactor += Sn, r !== 0 && Ln(n, e, r), u !== 0 && En(n, e, u);
}
function En(n, e, t) {
  const r = e.node.scrollTop;
  Lt(
    r + Math.round(e.sense / 3) * n.scrollSpeedFactor * t,
    "scrollTop",
    e
  );
}
function Ln(n, e, t) {
  const r = e.node.scrollLeft;
  Lt(
    r + Math.round(e.sense / 3) * n.scrollSpeedFactor * t,
    "scrollLeft",
    e
  );
}
function Lt(n, e, t) {
  t.node[e] = n;
}
function We(n, e, t, r, u, x) {
  const a = {};
  return n && (a.width = `${n}px`, a.minWidth = `${n}px`), e && (a.flexGrow = e), x && (a.height = `${x}px`), t && (a.position = "sticky", t.left && (a.left = `${r}px`), t.right && (a.right = `${u}px`)), a;
}
function Ht(n, e, t) {
  let r = "";
  if (n.fixed)
    for (const u in n.fixed)
      r += n.fixed[u] === -1 ? "wx-shadow " : "wx-fixed ";
  return r += e.rowspan > 1 ? "wx-rowspan " : "", r += e.colspan > 1 ? "wx-colspan " : "", r += e.vertical ? "wx-vertical " : "", r += t ? t(n) + " " : "", r;
}
function Hn(n) {
  const {
    row: e,
    column: t,
    cellStyle: r = null,
    columnStyle: u = null,
    children: x
  } = n, [a, b] = un(n.focusable), d = Z(oe), f = F(d, "focusCell"), h = F(d, "search"), s = F(d, "reorder"), v = S(
    () => h?.rows[e.id] && h.rows[e.id][t.id],
    [h, e.id, t.id]
  ), E = S(
    () => We(
      t.width,
      t.flexgrow,
      t.fixed,
      t.left,
      t.right
    ),
    [t.width, t.flexgrow, t.fixed, t.left, t.right]
  );
  function C(R, i) {
    let k = "wx-cell";
    return k += t.fixed ? " " + (t.fixed === -1 ? "wx-shadow" : "wx-fixed") : "", k += R ? " " + R(t) : "", k += i ? " " + i(e, t) : "", k += t.treetoggle ? " wx-tree-cell" : "", k;
  }
  const l = S(
    () => C(u, r),
    [u, r, t, e]
  ), m = S(() => typeof t.draggable == "function" ? t.draggable(e, t) !== !1 : t.draggable, [t, e]), g = I(null);
  K(() => {
    g.current && a && f?.row === e.id && f?.column === t.id && g.current.focus();
  }, [f, a, e.id, t.id]);
  const w = W(() => {
    a && !f && d.exec("focus-cell", {
      row: e.id,
      column: t.id,
      eventSource: "focus"
    });
  }, [d, a, f, e.id, t.id]);
  K(() => () => {
    a && f && (d.exec("focus-cell", { eventSource: "destroy" }), b(!1));
  }, [d, b]);
  function y(R) {
    const i = new RegExp(`(${h.value.trim()})`, "gi");
    return String(R).split(i).map((U) => ({ text: U, highlight: i.test(U) }));
  }
  const T = S(() => {
    const R = t.fixed && t.fixed.left === -1 || t.fixed.right === -1, i = t.fixed && t.fixed.right;
    return [
      l,
      R ? "wx-shadow" : "",
      i ? "wx-fixed-right" : ""
    ].filter(Boolean).join(" ");
  }, [l, t]), D = t.cell;
  return /* @__PURE__ */ Q(
    "div",
    {
      className: "wx-TSCaXsGV " + T,
      ref: g,
      onFocus: w,
      style: E,
      "data-row-id": e.id,
      "data-col-id": t.id,
      tabIndex: a ? "0" : "-1",
      role: "gridcell",
      "aria-colindex": t._colindex,
      "aria-readonly": t.editor ? void 0 : !0,
      children: [
        s && t.draggable ? m ? /* @__PURE__ */ c(
          "i",
          {
            "draggable-data": "true",
            className: "wx-TSCaXsGV wx-draggable wxi-drag"
          }
        ) : /* @__PURE__ */ c("i", { className: "wx-TSCaXsGV wx-draggable-stub" }) : null,
        t.treetoggle ? /* @__PURE__ */ Q(xe, { children: [
          /* @__PURE__ */ c("span", { style: { marginLeft: `${e.$level * 28}px` } }),
          e.$count ? /* @__PURE__ */ c(
            "i",
            {
              "data-action": "toggle-row",
              className: `wx-TSCaXsGV wx-table-tree-toggle wxi-menu-${e.open !== !1 ? "down" : "right"}`
            }
          ) : null
        ] }) : null,
        D ? /* @__PURE__ */ c(
          D,
          {
            api: d,
            row: e,
            column: t,
            onAction: ({ action: R, data: i }) => d.exec(R, i)
          }
        ) : x ? x() : v ? /* @__PURE__ */ c("span", { children: y(Ke(e, t)).map(
          ({ highlight: R, text: i }, k) => R ? /* @__PURE__ */ c("mark", { className: "wx-TSCaXsGV wx-search", children: i }, k) : /* @__PURE__ */ c("span", { children: i }, k)
        ) }) : Ke(e, t)
      ]
    }
  );
}
function wt(n, e) {
  let t, r;
  function u(b) {
    t = b.clientX, n.style.opacity = 1, document.body.style.cursor = "ew-resize", document.body.style.userSelect = "none", window.addEventListener("mousemove", x), window.addEventListener("mouseup", a), e && e.down && e.down(n);
  }
  function x(b) {
    r = b.clientX - t, e && e.move && e.move(r);
  }
  function a() {
    n.style.opacity = "", document.body.style.cursor = "", document.body.style.userSelect = "", e && e.up && e.up(r), window.removeEventListener("mousemove", x), window.removeEventListener("mouseup", a);
  }
  return n.addEventListener("mousedown", u), {
    destroy() {
      n.removeEventListener("mousedown", u);
    }
  };
}
function kn({ filter: n, column: e, action: t, filterValue: r }) {
  function u({ value: x }) {
    t({ value: x, key: e.id });
  }
  return /* @__PURE__ */ c(
    Gt,
    {
      ...n.config ?? {},
      value: r,
      onChange: u
    }
  );
}
function $n({ filter: n, column: e, action: t, filterValue: r }) {
  const u = Z(oe), x = F(u, "flatData"), a = S(
    () => n?.config?.options || e?.options || d(),
    [n, e, x]
  ), b = S(() => n?.config?.template, [n]);
  function d() {
    const s = [];
    return x.forEach((v) => {
      const E = en(v, e);
      s.includes(E) || s.push(E);
    }), s.map((v) => ({ id: v, label: v }));
  }
  function f({ value: s }) {
    t({ value: s, key: e.id });
  }
  function h(s) {
    s.key !== "Tab" && s.preventDefault();
  }
  return /* @__PURE__ */ c("div", { style: { width: "100%" }, onKeyDown: h, children: /* @__PURE__ */ c(
    Ut,
    {
      placeholder: "",
      clear: !0,
      ...n?.config ?? {},
      options: a,
      value: r,
      onChange: f,
      children: (s) => b ? b(s) : s.label
    }
  ) });
}
const Mn = {
  text: kn,
  richselect: $n
};
function An({ filter: n, column: e }) {
  const t = Z(oe), r = F(t, "filterValues");
  function u(a) {
    t.exec("filter-rows", a);
  }
  const x = S(() => Mn[n.type], [n.type]);
  return /* @__PURE__ */ c(
    x,
    {
      filter: n,
      column: e,
      action: u,
      filterValue: r[e.id]
    }
  );
}
function In(n) {
  const {
    cell: e,
    column: t,
    row: r,
    lastRow: u,
    sortRow: x,
    columnStyle: a,
    bodyHeight: b,
    hasSplit: d
  } = n, f = Z(oe), h = F(f, "sortMarks"), s = S(() => h ? h[t.id] : void 0, [h, t.id]), v = I(), E = W(
    (H) => {
      v.current = e.flexgrow ? H.parentNode.clientWidth : e.width;
    },
    [e.flexgrow, e.width]
  ), C = W(
    (H, $) => {
      f.exec("resize-column", {
        id: e.id,
        width: Math.max(1, (v.current || 0) + H),
        inProgress: $
      });
    },
    [f, e.id]
  ), l = W((H) => C(H, !0), [C]), m = W((H) => C(H, !1), [C]), g = W(
    (H) => {
      if (!t.sort || e.filter) return;
      let $ = s?.order;
      $ && ($ = $ === "asc" ? "desc" : "asc"), f.exec("sort-rows", { key: e.id, add: H.ctrlKey, order: $ });
    },
    [f, e.id, e.filter, t.sort, s?.order]
  ), w = W(
    (H) => {
      H && H.stopPropagation(), f.exec("collapse-column", { id: e.id, row: r });
    },
    [f, e.id, r]
  ), y = W(
    (H) => {
      H.key === "Enter" && w();
    },
    [w]
  ), T = W(
    (H) => {
      H.key === "Enter" && !e.filter && g(H);
    },
    [g, e.filter]
  ), D = S(
    () => e.collapsed && t.collapsed,
    [e.collapsed, t.collapsed]
  ), R = S(
    () => D && !d && e.collapsible !== "header",
    [D, d, e.collapsible]
  ), i = S(
    () => R ? { top: -b / 2, position: "absolute" } : {},
    [R, b]
  ), k = S(
    () => We(
      e.width,
      e.flexgrow,
      t.fixed,
      t.left,
      e.right ?? t.right,
      e.height + (D && R ? b : 0)
    ),
    [
      e.width,
      e.flexgrow,
      t.fixed,
      t.left,
      e.right,
      t.right,
      e.height,
      D,
      R,
      b
    ]
  ), U = S(
    () => Ht(t, e, a),
    [t, e, a]
  ), J = W(() => Object.fromEntries(
    Object.entries(e).filter(([H]) => H !== "cell")
  ), [e]), te = `wx-cell ${U} ${e.css || ""} wx-collapsed`, V = [
    "wx-cell",
    U,
    e.css || "",
    e.filter ? "wx-filter" : "",
    t.fixed && t.fixed.right ? "wx-fixed-right" : ""
  ].filter(Boolean).join(" "), j = I(null);
  return K(() => {
    const H = j.current;
    if (!H) return;
    const $ = wt(H, { down: E, move: l, up: m });
    return () => {
      typeof $ == "function" && $();
    };
  }, [E, l, m, wt]), D ? /* @__PURE__ */ c(
    "div",
    {
      className: "wx-RsQD74qC " + te,
      style: k,
      role: "button",
      "aria-label": `Expand column ${e.text || ""}`,
      "aria-expanded": !e.collapsed,
      tabIndex: 0,
      onKeyDown: y,
      onClick: w,
      "data-header-id": t.id,
      children: /* @__PURE__ */ c("div", { className: "wx-RsQD74qC wx-text", style: i, children: e.text || "" })
    }
  ) : /* @__PURE__ */ Q(
    "div",
    {
      className: "wx-RsQD74qC " + V,
      style: k,
      onClick: g,
      "data-header-id": t.id,
      tabIndex: !e._hidden && t.sort && !e.filter ? 0 : void 0,
      role: "columnheader",
      "aria-colindex": e._colindex,
      "aria-colspan": e.colspan > 1 ? e.colspan : void 0,
      "aria-rowspan": e.rowspan > 1 ? e.rowspan : void 0,
      "aria-sort": !s?.order || e.filter ? "none" : s?.order === "asc" ? "ascending" : "descending",
      onKeyDown: T,
      children: [
        e.collapsible ? /* @__PURE__ */ c(
          "div",
          {
            className: "wx-RsQD74qC wx-collapse",
            role: "button",
            "aria-label": e.collapsed ? "Expand column" : "Collapse column",
            "aria-expanded": !e.collapsed,
            tabIndex: 0,
            onKeyDown: y,
            onClick: w,
            children: /* @__PURE__ */ c(
              "i",
              {
                className: `wx-RsQD74qC wxi-angle-${e.collapsed ? "down" : "right"}`
              }
            )
          }
        ) : null,
        e.cell ? (() => {
          const H = e.cell;
          return /* @__PURE__ */ c(
            H,
            {
              api: f,
              cell: J(),
              column: t,
              row: r,
              onAction: ({ action: $, data: Ee }) => f.exec($, Ee)
            }
          );
        })() : e.filter ? /* @__PURE__ */ c(An, { filter: e.filter, column: t }) : /* @__PURE__ */ c("div", { className: "wx-RsQD74qC wx-text", children: e.text || "" }),
        t.resize && u && !e._hidden ? /* @__PURE__ */ c(
          "div",
          {
            className: "wx-RsQD74qC wx-grip",
            role: "presentation",
            "aria-label": "Resize column",
            ref: j,
            onClick: (H) => H.stopPropagation(),
            children: /* @__PURE__ */ c("div", {})
          }
        ) : null,
        x ? /* @__PURE__ */ c("div", { className: "wx-RsQD74qC wx-sort", children: s ? /* @__PURE__ */ Q(xe, { children: [
          typeof s.index < "u" ? /* @__PURE__ */ c("div", { className: "wx-RsQD74qC wx-order", children: s.index + 1 }) : null,
          /* @__PURE__ */ c(
            "i",
            {
              className: `wx-RsQD74qC wxi-arrow-${s.order === "asc" ? "up" : "down"}`
            }
          )
        ] }) : null }) : null
      ]
    }
  );
}
function Vn({ cell: n, column: e, row: t, columnStyle: r }) {
  const u = Z(oe), x = S(
    () => We(
      n?.width,
      n?.flexgrow,
      e?.fixed,
      e?.left,
      n?.right ?? e?.right,
      n?.height
    ),
    [
      n?.width,
      n?.flexgrow,
      e?.fixed,
      e?.left,
      n?.right,
      e?.right,
      n?.height
    ]
  ), a = S(
    () => Ht(e, n, r),
    [e, n, r]
  ), b = W(() => Object.fromEntries(
    Object.entries(n || {}).filter(([f]) => f !== "cell")
  ), [n]), d = `wx-6Sdi3Dfd wx-cell ${a || ""} ${n?.css || ""}` + (e?.fixed && e?.fixed.right ? " wx-fixed-right" : "");
  return /* @__PURE__ */ c("div", { className: d, style: x, children: !e?.collapsed && !n?.collapsed ? n?.cell ? Xt.createElement(n.cell, {
    api: u,
    cell: b(),
    column: e,
    row: t,
    onAction: ({ action: f, data: h }) => u.exec(f, h)
  }) : /* @__PURE__ */ c("div", { className: "wx-6Sdi3Dfd wx-text", children: n?.text || "" }) : null });
}
function xt({
  deltaLeft: n,
  contentWidth: e,
  columns: t,
  type: r = "header",
  columnStyle: u,
  bodyHeight: x
}) {
  const a = Z(oe), b = F(a, "_sizes"), d = F(a, "split"), f = S(() => b?.[`${r}RowHeights`], [b, r]), h = S(() => {
    let l = [];
    if (t && t.length) {
      const m = t[0][r].length;
      for (let g = 0; g < m; g++) {
        let w = 0;
        l.push([]), t.forEach((y, T) => {
          const D = { ...y[r][g] };
          if (w || l[g].push(D), D.colspan > 1) {
            if (w = D.colspan - 1, !tn() && y.right) {
              let R = y.right;
              for (let i = 1; i < D.colspan; i++)
                R -= t[T + i].width;
              D.right = R;
            }
          } else w && w--;
        });
      }
    }
    return l;
  }, [t, r]), s = S(() => d?.left || d?.right, [d]);
  function v(l) {
    return t.find((m) => m.id === l);
  }
  function E(l, m) {
    let g = m;
    return l.rowspan && (g += l.rowspan - 1), g === h.length - 1;
  }
  function C(l, m, g) {
    if (!g.sort) return !1;
    for (let w = h.length - 1; w >= 0; w--) {
      const y = g.header[w];
      if (!y.filter && !y._hidden) return m === w;
    }
    return E(l, m);
  }
  return /* @__PURE__ */ c(
    "div",
    {
      className: `wx-sAsPVaUK wx-${r}`,
      style: { paddingLeft: `${n}px`, width: `${e}px` },
      role: "rowgroup",
      children: h.map((l, m) => /* @__PURE__ */ c(
        "div",
        {
          className: r === "header" ? "wx-sAsPVaUK wx-h-row" : "wx-sAsPVaUK wx-f-row",
          style: { height: `${f?.[m]}px`, display: "flex" },
          role: "row",
          children: l.map((g) => {
            const w = v(g.id);
            return r === "header" ? /* @__PURE__ */ c(
              In,
              {
                cell: g,
                columnStyle: u,
                column: w,
                row: m,
                lastRow: E(g, m),
                bodyHeight: x,
                sortRow: C(g, m, w),
                hasSplit: s
              },
              g.id
            ) : /* @__PURE__ */ c(
              Vn,
              {
                cell: g,
                columnStyle: u,
                column: v(g.id),
                row: m
              },
              g.id
            );
          })
        },
        m
      ))
    }
  );
}
function zn({ overlay: n }) {
  const e = Z(oe);
  function t(u) {
    return typeof u == "function";
  }
  const r = n;
  return /* @__PURE__ */ c("div", { className: "wx-1ty666CQ wx-overlay", children: t(n) ? /* @__PURE__ */ c(r, { onAction: ({ action: u, data: x }) => e.exec(u, x) }) : n });
}
function On(n) {
  const { actions: e, editor: t } = n, [r, u] = O(t?.value || ""), x = I(null);
  K(() => {
    x.current && x.current.focus();
  }, []);
  function a() {
    x.current && (u(x.current.value), e.updateValue(x.current.value));
  }
  function b({ key: d }) {
    d === "Enter" && e.save();
  }
  return /* @__PURE__ */ c(
    "input",
    {
      className: "wx-e7Ao5ejY wx-text",
      onInput: a,
      onKeyDown: b,
      ref: x,
      type: "text",
      value: r
    }
  );
}
function Pn({ actions: n, editor: e, onAction: t }) {
  const [r, u] = O(e?.value), [x, a] = O(e?.renderedValue), [b, d] = O(e?.options || []), f = S(() => e?.config?.template, [e]), h = S(() => e?.config?.cell, [e]), s = S(() => (b || []).findIndex((w) => w.id === r), [b, r]), v = I(null), E = I(null), C = W(
    (w) => {
      v.current = w.navigate, E.current = w.keydown, v.current(s);
    },
    [s, v]
  ), l = W(
    (w) => {
      const y = w?.target?.value ?? "";
      a(y);
      const T = y ? (e?.options || []).filter(
        (D) => (D.label || "").toLowerCase().includes(y.toLowerCase())
      ) : e?.options || [];
      d(T), T.length ? v.current(-1 / 0) : v.current(null);
    },
    [e]
  ), m = I(null);
  K(() => {
    m.current && m.current.focus();
  }, []), K(() => {
    u(e?.value), a(e?.renderedValue), d(e?.options || []);
  }, [e]);
  const g = W(
    ({ id: w }) => {
      n.updateValue(w), n.save();
    },
    [n]
  );
  return /* @__PURE__ */ Q(xe, { children: [
    /* @__PURE__ */ c(
      "input",
      {
        className: "wx-0UYfSd1x wx-input",
        ref: m,
        value: x ?? "",
        onChange: l,
        onKeyDown: (w) => E.current ? E.current(w, s) : void 0
      }
    ),
    /* @__PURE__ */ c(
      yt,
      {
        items: b,
        onReady: C,
        onSelect: g,
        children: ({ option: w }) => f ? f(w) : h ? /* @__PURE__ */ c(h, { data: w, onAction: t }) : w.label
      }
    )
  ] });
}
function Kn({ actions: n, editor: e, onAction: t }) {
  const [r] = O(() => e.value || /* @__PURE__ */ new Date()), [u] = O(() => e.config?.template), [x] = O(() => e.config?.cell);
  function a({ value: d }) {
    n.updateValue(d), n.save();
  }
  const b = I(null);
  return K(() => {
    b.current && b.current.focus(), typeof window < "u" && window.getSelection && window.getSelection().removeAllRanges();
  }, []), /* @__PURE__ */ Q(xe, { children: [
    /* @__PURE__ */ c(
      "div",
      {
        className: "wx-lNWNYUb6 wx-value",
        ref: b,
        tabIndex: 0,
        onClick: () => n.cancel(),
        onKeyDown: (d) => d.preventDefault(),
        children: u ? u(r) : x ? /* @__PURE__ */ c(x, { data: e.value, onAction: t }) : /* @__PURE__ */ c("span", { className: "wx-lNWNYUb6 wx-text", children: e.renderedValue })
      }
    ),
    /* @__PURE__ */ c(qt, { width: "auto", children: /* @__PURE__ */ c(
      Qt,
      {
        value: r,
        onChange: a,
        buttons: e.config?.buttons
      }
    ) })
  ] });
}
function Wn(n) {
  const { actions: e, editor: t } = n, r = n.onAction ?? n.onaction, u = t.config || {}, [x] = O(
    t.options.find((l) => l.id === t.value)
  ), [a] = O(t.value), [b] = O(t.options), d = S(
    () => b.findIndex((l) => l.id === a),
    [b, a]
  );
  function f({ id: l }) {
    e.updateValue(l), e.save();
  }
  let h;
  const [s, v] = O();
  function E(l) {
    h = l.navigate, v(() => l.keydown), h(d);
  }
  const C = I(null);
  return K(() => {
    C.current && C.current.focus(), typeof window < "u" && window.getSelection && window.getSelection().removeAllRanges();
  }, []), /* @__PURE__ */ Q(xe, { children: [
    /* @__PURE__ */ c(
      "div",
      {
        ref: C,
        className: "wx-ywGRk611 wx-value",
        tabIndex: 0,
        onClick: () => e.cancel(),
        onKeyDown: (l) => {
          s(l, d), l.preventDefault();
        },
        children: u.template ? u.template(x) : u.cell ? (() => {
          const l = u.cell;
          return /* @__PURE__ */ c(l, { data: x, onAction: r });
        })() : /* @__PURE__ */ c("span", { className: "wx-ywGRk611 wx-text", children: t.renderedValue })
      }
    ),
    /* @__PURE__ */ c(yt, { items: b, onReady: E, onSelect: f, children: ({ option: l }) => u.template ? u.template(l) : u.cell ? (() => {
      const m = u.cell;
      return /* @__PURE__ */ c(m, { data: l, onAction: r });
    })() : l.label })
  ] });
}
const _n = {
  text: On,
  combo: Pn,
  datepicker: Kn,
  richselect: Wn
};
function Fn({ column: n, row: e }) {
  const t = Z(oe), r = F(t, "editor"), u = W(
    (C, l) => {
      t.exec("close-editor", { ignore: C }), l && t.exec("focus-cell", {
        ...l,
        eventSource: "click"
      });
    },
    [t]
  ), x = W(
    (C) => {
      const l = C ? null : { row: r?.id, column: r?.column };
      u(!1, l);
    },
    [r, u]
  ), a = W(() => {
    u(!0, { row: r?.id, column: r?.column });
  }, [r, u]), b = W(
    (C) => {
      t.exec("editor", { value: C });
    },
    [t]
  ), d = W(
    (C) => {
      C.key === "Enter" && r && a();
    },
    [r, a]
  ), f = S(
    () => We(
      n.width,
      n.flexgrow,
      n.fixed,
      n.left,
      n.right
    ),
    [n.width, n.flexgrow, n.fixed, n.left, n.right]
  ), h = S(() => {
    let C = n.editor;
    typeof C == "function" && (C = C(e, n));
    let l = typeof C == "string" ? C : C.type;
    return _n[l];
  }, [n, e]), s = I(null);
  K(() => {
    if (!s.current) return;
    const C = Nt(s.current, () => x(!0));
    return () => {
      typeof C == "function" && C();
    };
  }, [x]), K(() => {
    s.current && typeof f == "string" && s.current.setAttribute("style", f);
  }, [f]);
  const v = typeof e.$parent < "u" ? "gridcell" : "cell", E = typeof e.$parent < "u" ? !n.editor : void 0;
  return /* @__PURE__ */ c(
    "div",
    {
      className: "wx-8l724t2g wx-cell wx-editor",
      ref: s,
      style: typeof f == "object" && f !== null ? f : void 0,
      role: v,
      "aria-readonly": E,
      tabIndex: -1,
      onClick: (C) => C.stopPropagation(),
      onDoubleClick: (C) => C.stopPropagation(),
      onKeyDown: d,
      children: h ? /* @__PURE__ */ c(
        h,
        {
          editor: r,
          actions: { save: x, cancel: a, updateValue: b },
          onAction: ({ action: C, data: l }) => t.exec(C, l)
        }
      ) : null
    }
  );
}
function mt(n) {
  const { columns: e, type: t, columnStyle: r } = n, u = Z(oe), { filterValues: x, _columns: a, _sizes: b } = u.getState();
  function d(f) {
    return r ? " " + r(f) : "";
  }
  return /* @__PURE__ */ c(xe, { children: e.map((f, h) => /* @__PURE__ */ c("tr", { children: f.map((s) => {
    const v = a.find((l) => l.id == s.id), E = `wx-print-cell-${t}${d(v)}${s.filter ? " wx-print-cell-filter" : ""}${s.vertical ? " wx-vertical" : ""}`, C = s.cell;
    return /* @__PURE__ */ c(
      "th",
      {
        style: St(bt(s, b.columnWidth)),
        className: "wx-Gy81xq2u " + E,
        rowSpan: s.rowspan,
        colSpan: s.colspan,
        children: C ? /* @__PURE__ */ c(
          C,
          {
            api: u,
            cell: Object.fromEntries(
              Object.entries(s).filter(([l]) => l !== "cell")
            ),
            column: v,
            row: h
          }
        ) : s.filter ? /* @__PURE__ */ c("div", { className: "wx-Gy81xq2u wx-print-filter", children: nn(x, a, s) }) : /* @__PURE__ */ c("div", { className: "wx-Gy81xq2u wx-text", children: s.text ?? "" })
      },
      s.id
    );
  }) }, h)) });
}
function Xn(n) {
  const { columns: e, rowStyle: t, columnStyle: r, cellStyle: u, header: x, footer: a, reorder: b } = n, d = Z(oe), { flatData: f, _sizes: h } = d.getState(), s = x && ht(e, "header", h.headerRowHeights), v = a && ht(e, "footer", h.footerRowHeights);
  function E(l, m) {
    let g = "";
    return g += r ? " " + r(m) : "", g += u ? " " + u(l, m) : "", g;
  }
  function C(l, m) {
    return typeof m.draggable == "function" ? m.draggable(l, m) !== !1 : m.draggable;
  }
  return /* @__PURE__ */ Q(
    "table",
    {
      className: `wx-8NTMLH0z wx-print-grid ${e.some((l) => l.flexgrow) ? "wx-flex-columns" : ""}`,
      children: [
        x ? /* @__PURE__ */ c("thead", { children: /* @__PURE__ */ c(
          mt,
          {
            columns: s,
            type: "header",
            columnStyle: r
          }
        ) }) : null,
        /* @__PURE__ */ c("tbody", { children: f.map((l, m) => /* @__PURE__ */ c(
          "tr",
          {
            className: "wx-8NTMLH0z wx-row" + (t ? " " + t(l) : ""),
            style: { height: `${l.rowHeight || h.rowHeight}px` },
            children: e.map(
              (g) => g.collapsed ? null : /* @__PURE__ */ Q(
                "td",
                {
                  className: `wx-8NTMLH0z wx-print-cell wx-cell ${E(l, g)}`,
                  style: St(
                    bt(g, h.columnWidth)
                  ),
                  children: [
                    b && g.draggable ? /* @__PURE__ */ c("span", { className: "wx-8NTMLH0z wx-print-draggable", children: C(l, g) ? /* @__PURE__ */ c("i", { className: "wx-8NTMLH0z wxi-drag" }) : null }) : null,
                    g.treetoggle ? /* @__PURE__ */ Q(xe, { children: [
                      /* @__PURE__ */ c(
                        "span",
                        {
                          style: { marginLeft: l.$level * 28 + "px" }
                        }
                      ),
                      l.$count ? /* @__PURE__ */ c(
                        "i",
                        {
                          className: `wx-8NTMLH0z wx-print-grid-tree-toggle wxi-menu-${l.open !== !1 ? "down" : "right"}`
                        }
                      ) : null
                    ] }) : null,
                    g.cell ? (() => {
                      const w = g.cell;
                      return /* @__PURE__ */ c(w, { api: d, row: l, column: g });
                    })() : /* @__PURE__ */ c("span", { children: Ke(l, g) })
                  ]
                },
                g.id
              )
            )
          },
          m
        )) }),
        a ? /* @__PURE__ */ c("tfoot", { children: /* @__PURE__ */ c(
          mt,
          {
            columns: v,
            type: "footer",
            columnStyle: r
          }
        ) }) : null
      ]
    }
  );
}
function Yn(n) {
  const { config: e, ...t } = n, r = Z(oe), { _skin: u, _columns: x } = r.getState(), a = S(() => on(x, e), []), b = I(null);
  return K(() => {
    const d = document.body;
    d.classList.add("wx-print");
    const f = b.current;
    if (!f) return;
    const h = f.cloneNode(!0);
    d.appendChild(h);
    const s = `@media print { @page { size: ${e.paper} ${e.mode}; }`, v = document.createElement("style");
    v.setAttribute("type", "text/css"), v.setAttribute("media", "print"), document.getElementsByTagName("head")[0].appendChild(v), v.appendChild(document.createTextNode(s)), window.print(), v.remove(), d.classList.remove("wx-print"), h.remove();
  }, []), /* @__PURE__ */ c(
    "div",
    {
      className: `wx-4zwCKA7C wx-${u}-theme wx-print-container`,
      ref: b,
      children: a.map((d, f) => /* @__PURE__ */ c("div", { className: "wx-4zwCKA7C wx-print-grid-wrapper", children: /* @__PURE__ */ c(Xn, { columns: d, ...t }) }, f))
    }
  );
}
function jn(n) {
  const {
    header: e,
    footer: t,
    overlay: r,
    multiselect: u,
    onreorder: x,
    rowStyle: a,
    columnStyle: b,
    cellStyle: d,
    autoRowHeight: f,
    resize: h,
    clientWidth: s,
    clientHeight: v,
    responsiveLevel: E,
    hotkeys: C
  } = n, l = Z(oe), m = F(l, "dynamic"), g = F(l, "_columns"), w = F(l, "flatData"), y = F(l, "split"), T = F(l, "_sizes"), [D, R] = fn(l, "selectedRows"), i = F(l, "select"), k = F(l, "editor"), U = F(l, "tree"), J = F(l, "focusCell"), te = F(l, "_print"), V = F(l, "undo"), j = F(l, "reorder"), H = F(l, "_rowHeightFromData"), [$, Ee] = O(0);
  K(() => {
    Ee(Pt());
  }, []);
  const [ve, _e] = O(0), [ge, Le] = O(0), ce = S(() => (g || []).some((o) => !o.hidden && o.flexgrow), [g]), P = S(() => T?.rowHeight || 0, [T]), ee = I(null), [He, Ce] = O(null), [pe, Se] = O(null), ne = S(() => {
    let o = [], p = 0;
    return y && y.left && (o = (g || []).slice(0, y.left).filter((N) => !N.hidden).map((N) => ({ ...N })), o.forEach((N) => {
      N.fixed = { left: 1 }, N.left = p, p += N.width;
    }), o.length && (o[o.length - 1].fixed = { left: -1 })), { columns: o, width: p };
  }, [y, g]), _ = S(() => {
    let o = [], p = 0;
    if (y && y.right) {
      o = (g || []).slice(y.right * -1).filter((N) => !N.hidden).map((N) => ({ ...N }));
      for (let N = o.length - 1; N >= 0; N--) {
        const M = o[N];
        M.fixed = { right: 1 }, M.right = p, p += M.width;
      }
      o.length && (o[0].fixed = { right: -1 });
    }
    return { columns: o, width: p };
  }, [y, g]), B = S(() => {
    const o = (g || []).slice(y?.left || 0, (g || []).length - (y?.right ?? 0)).filter((p) => !p.hidden);
    return o.forEach((p) => {
      p.fixed = 0;
    }), o;
  }, [g, y]), z = S(() => (g || []).reduce((o, p) => (p.hidden || (o += p.width), o), 0), [g]), ae = 1;
  function Me(o, p, N) {
    let M = p, X = o;
    if (B.length) {
      let A = B.length;
      for (let L = o; L >= 0; L--)
        B[L][N].forEach((q) => {
          q.colspan > 1 && L > o - q.colspan && L < A && (A = L);
        });
      if (A !== B.length && A < o) {
        for (let L = A; L < o; L++)
          M -= B[L].width;
        X = A;
      }
    }
    return { index: X, delta: M };
  }
  const re = S(() => {
    let o, p, N;
    const M = ve, X = ve + (s || 0);
    let A = 0, L = 0, Y = 0, q = 0;
    B.forEach((fe, Te) => {
      M > Y && (A = Te, q = Y), Y = Y + fe.width, X > Y && (L = Te + ae);
    });
    const ie = { header: 0, footer: 0 };
    for (let fe = L; fe >= A; fe--)
      ["header", "footer"].forEach((Te) => {
        B[fe] && B[fe][Te].forEach((Ft) => {
          const qe = Ft.colspan;
          if (qe && qe > 1) {
            const at = qe - (L - fe + 1);
            at > 0 && (ie[Te] = Math.max(ie[Te], at));
          }
        });
      });
    const G = Me(A, q, "header"), se = Me(A, q, "footer"), Ue = G.delta, $e = G.index, ze = se.delta, Oe = se.index;
    return ce && z > (s || 0) ? o = p = N = [...ne.columns, ...B, ..._.columns] : (o = [
      ...ne.columns,
      ...B.slice(A, L + 1),
      ..._.columns
    ], p = [
      ...ne.columns,
      ...B.slice($e, L + ie.header + 1),
      ..._.columns
    ], N = [
      ...ne.columns,
      ...B.slice(Oe, L + ie.footer + 1),
      ..._.columns
    ]), {
      data: o || [],
      header: p || [],
      footer: N || [],
      d: q,
      df: ze,
      dh: Ue
    };
  }, [
    B,
    ne,
    _,
    ve,
    s,
    ce,
    z
  ]), me = S(
    () => e && T?.headerHeight || 0,
    [e, T]
  ), ye = S(
    () => t && T?.footerHeight || 0,
    [t, T]
  ), we = S(() => s && v ? z >= s : !1, [s, v, z]), Re = S(() => (v || 0) - me - ye - (we ? $ : 0), [v, me, ye, we, $]), tt = S(() => Math.ceil((Re || 0) / (P || 1)) + 1, [Re, P]), Fe = I([]), [nt, kt] = O(0), [ot, $t] = O(void 0), le = S(() => {
    let o = 0, p = 0;
    const N = 2;
    if (f) {
      let A = ge;
      for (; A > 0; )
        A -= Fe.current[o] || P, o++;
      p = ge - A;
      for (let L = Math.max(0, o - N - 1); L < o; L++)
        p -= Fe.current[o - L] || P;
      o = Math.max(0, o - N);
    } else {
      if (H) {
        let A = 0, L = 0;
        for (let G = 0; G < (w || []).length; G++) {
          const se = w[G].rowHeight || P;
          if (L + se > ge) {
            A = G;
            break;
          }
          L += se;
        }
        o = Math.max(0, A - N);
        for (let G = 0; G < o; G++)
          p += w[G].rowHeight || P;
        let Y = 0, q = 0;
        for (let G = A + 1; G < (w || []).length; G++) {
          const se = w[G].rowHeight || P;
          if (Y++, q + se > Re)
            break;
          q += se;
        }
        const ie = Math.min(
          m ? m.rowCount : (w || []).length,
          A + Y + N
        );
        return { d: p, start: o, end: ie };
      }
      o = Math.floor(ge / (P || 1)), o = Math.max(0, o - N), p = o * (P || 0);
    }
    const M = m ? m.rowCount : (w || []).length, X = Math.min(M, o + (tt || 0) + N);
    return { d: p, start: o, end: X };
  }, [f, H, ge, P, m, w, tt, Re]), Xe = S(() => {
    const o = m ? m.rowCount : (w || []).length;
    if (f)
      return nt + le.d + (o - (ot || 0)) * (P || 0);
    if (!H)
      return o * (P || 0);
    let p = 0;
    for (let N = 0; N < o; N++)
      p += w[N]?.rowHeight || P;
    return p;
  }, [
    m,
    w,
    P,
    f,
    H,
    nt,
    le.d,
    ot
  ]), de = S(() => s && v ? Xe + me + ye >= v - (z >= (s || 0) ? $ : 0) : !1, [
    s,
    v,
    Xe,
    me,
    ye,
    z,
    $
  ]), be = S(() => ce && z <= (s || 0) ? (s || 0) - 0 - (de ? $ : 0) : z, [ce, z, s, de, $, we]), Ye = S(() => ce && z <= (s || 0) ? s || 0 : be < (s || 0) ? z + (de ? $ : 0) : -1, [ce, z, s, be, de, $]), je = I({});
  K(() => {
    if (m && (je.current.start !== le.start || je.current.end !== le.end)) {
      const { start: o, end: p } = le;
      je.current = { start: o, end: p }, l && l.exec && l.exec("request-data", { row: { start: o, end: p } });
    }
  }, [m, le, l]);
  const ue = S(() => m ? w || [] : (w || []).slice(le.start, le.end), [m, w, le]), ke = S(() => (D || []).filter(
    (o) => (ue || []).some((p) => p.id === o)
  ), [R, ue]), Be = S(() => le.start, [le.start]), Mt = W((o) => {
    Le(o.target.scrollTop), _e(o.target.scrollLeft);
  }, []), At = W((o) => {
    o.shiftKey && o.preventDefault(), ee.current && ee.current.focus && ee.current.focus();
  }, []), rt = W(() => !!(g || []).find((o) => !!o.draggable), [g]), Ge = I(null), Ne = I(null), It = I({
    dblclick: (o, p) => {
      const N = { id: o, column: gt(p, "data-col-id") };
      l.exec("open-editor", N);
    },
    click: (o, p) => {
      if (Ge.current) return;
      const N = gt(p, "data-col-id");
      if (J?.id !== o && l.exec("focus-cell", {
        row: o,
        column: N,
        eventSource: "click"
      }), i === !1) return;
      const M = u && p.ctrlKey, X = u && p.shiftKey;
      (M || D.length > 1 || !D.includes(o)) && l.exec("select-row", { id: o, toggle: M, range: X });
    },
    "toggle-row": (o) => {
      const p = l.getRow(o);
      l.exec(p.open !== !1 ? "close-row" : "open-row", { id: o });
    },
    "ignore-click": () => !1
  }), Ae = S(() => ({
    top: me,
    bottom: ye,
    left: ne.width,
    xScroll: we,
    yScroll: de,
    sense: f && pe ? pe.offsetHeight : Math.max(T?.rowHeight || 0, 40),
    node: ee.current && ee.current.firstElementChild
  }), [
    me,
    ye,
    ne.width,
    we,
    de,
    f,
    pe,
    T
  ]);
  function Vt(o, p) {
    const { container: N, sourceNode: M, from: X } = p;
    if (rt() && !M.getAttribute("draggable-data"))
      return !1;
    Ce(X), l.getRow(X).open && l.exec("close-row", { id: X, nested: !0 });
    const L = Qe(M, "data-id"), Y = L.cloneNode(!0);
    Y.classList.remove("wx-selected"), Y.querySelectorAll("[tabindex]").forEach((se) => se.setAttribute("tabindex", "-1")), N.appendChild(Y), Se(Y);
    const q = ve - re.d, ie = de ? $ : 0;
    N.style.width = Math.min(
      (s || 0) - ie,
      ce && z <= (s || 0) ? be : be - ie
    ) + q + "px";
    const G = Pe(L);
    p.offset = {
      x: q,
      y: -Math.round(G.height / 2)
    }, Ne.current || (Ne.current = o.clientY);
  }
  function zt(o, p) {
    const { from: N } = p, M = p.pos, X = Pe(ee.current);
    M.x = X.x;
    const A = Ae.top;
    if (M.y < A) M.y = A;
    else {
      const L = X.height - (we && $ > 0 ? $ : Math.round(Ae.sense / 2)) - Ae.bottom;
      M.y > L && (M.y = L);
    }
    if (ee.current.contains(p.targetNode)) {
      const L = Qe(p.targetNode, "data-id"), Y = Rt(L?.getAttribute("data-id"));
      if (Y && Y !== N) {
        p.to = Y;
        const q = f ? pe?.offsetHeight : T?.rowHeight;
        if (pe && (ge === 0 || M.y > A + q - 1)) {
          const ie = L.getBoundingClientRect(), se = Pe(pe).y, Ue = ie.y, $e = se > Ue ? -1 : 1, ze = $e === 1 ? "after" : "before", Oe = Math.abs(l.getRowIndex(N) - l.getRowIndex(Y)), fe = Oe !== 1 ? ze === "before" ? "after" : "before" : ze;
          if (Oe === 1 && ($e === -1 && o.clientY > Ne.current || $e === 1 && o.clientY < Ne.current))
            return;
          Ne.current = o.clientY, l.exec("move-item", {
            id: N,
            target: Y,
            mode: fe,
            inProgress: !0
          });
        }
      }
      x && x({ event: o, context: p });
    }
    Nn(o, X, p, Ae);
  }
  function Ot(o, p) {
    const { from: N, to: M } = p;
    l.exec("move-item", {
      id: N,
      target: M,
      inProgress: !1
    }), Ge.current = setTimeout(() => {
      Ge.current = 0;
    }, 1), Ce(null), Se(null), Ne.current = null, Et(p);
  }
  function Pt() {
    const o = document.createElement("div");
    o.style.cssText = "position:absolute;left:-1000px;width:100px;padding:0px;margin:0px;min-height:100px;overflow-y:scroll;", document.body.appendChild(o);
    const p = o.offsetWidth - o.clientWidth;
    return document.body.removeChild(o), p;
  }
  const Kt = S(() => Ye > 0 ? { width: `${Ye}px` } : void 0, [Ye]), lt = I(null);
  function Wt() {
    Promise.resolve().then(() => {
      let o = 0, p = Be;
      const N = lt.current;
      N && (Array.from(N.children).forEach((M, X) => {
        Fe.current[Be + X] = M.offsetHeight, o += M.offsetHeight, p++;
      }), kt(o), $t(p));
    });
  }
  K(() => {
    ue && f && Wt();
  }, [ue, f, Be]);
  let [De, Ie] = O();
  K(() => {
    if (J && (!i || !ke.length || ke.includes(J.row)))
      Ie({ ...J });
    else if (ue.length && re.data.length) {
      if (!De || ke.length && !ke.includes(De.row) || ue.findIndex((o) => o.id == De.row) === -1 || re.data.findIndex(
        (o) => o.id == De.column && !o.collapsed
      ) === -1) {
        const o = ke[0] || ue[0].id, p = re.data.findIndex((N) => !N.collapsed);
        Ie(p !== -1 ? { row: o, column: re.data[p].id } : null);
      }
    } else Ie(null);
  }, [J]);
  const st = I(null);
  K(() => {
    const o = ee.current;
    if (!o) return;
    const p = yn(o, h);
    return () => {
      typeof p == "function" && p();
    };
  }, [h]);
  const it = I({});
  Object.assign(it.current, {
    start: Vt,
    move: zt,
    end: Ot,
    getReorder: () => j,
    getDraggableInfo: () => ({ hasDraggable: rt() })
  }), K(() => {
    const o = ee.current;
    return o ? Cn(o, it).destroy : void 0;
  }, [j, ee.current]), K(() => {
    const o = ee.current;
    return o ? rn(o, {
      keys: C !== !1 && {
        ...ln,
        "ctrl+z": V,
        "ctrl+y": V,
        ...C
      },
      exec: (N) => l.exec("hotkey", N)
    }).destroy : void 0;
  }, [l, V, C]);
  const Ve = I({
    scroll: l.getReactiveState().scroll
  });
  Ve.current.getWidth = () => (s || 0) - (de ? $ : 0), Ve.current.getHeight = () => Re, Ve.current.getScrollMargin = () => ne.width + _.width, K(() => {
    sn(st.current, Ve.current);
  }, []);
  const ct = I(null);
  K(() => {
    const o = ct.current;
    if (!o) return;
    const p = [];
    return p.push(
      Nt(o, () => l.exec("focus-cell", { eventSource: "click" })).destroy
    ), p.push(hn(o, It.current)), () => p.forEach((N) => N());
  }, []);
  const _t = `wx-grid ${E ? `wx-responsive-${E}` : ""}`;
  return /* @__PURE__ */ Q(xe, { children: [
    /* @__PURE__ */ c(
      "div",
      {
        className: "wx-4VuBwK2D " + _t,
        style: {
          "--header-height": `${me}px`,
          "--footer-height": `${ye}px`,
          "--split-left-width": `${ne.width}px`,
          "--split-right-width": `${_.width}px`
        },
        children: /* @__PURE__ */ c(
          "div",
          {
            ref: ee,
            className: "wx-4VuBwK2D wx-table-box",
            style: Kt,
            role: U ? "treegrid" : "grid",
            "aria-colcount": re.data.length,
            "aria-rowcount": ue.length,
            "aria-multiselectable": U && u ? !0 : void 0,
            tabIndex: -1,
            children: /* @__PURE__ */ Q(
              "div",
              {
                ref: st,
                className: "wx-4VuBwK2D wx-scroll",
                style: {
                  overflowX: we ? "scroll" : "hidden",
                  overflowY: de ? "scroll" : "hidden"
                },
                onScroll: Mt,
                children: [
                  e ? /* @__PURE__ */ c("div", { className: "wx-4VuBwK2D wx-header-wrapper", children: /* @__PURE__ */ c(
                    xt,
                    {
                      contentWidth: be,
                      deltaLeft: re.dh,
                      columns: re.header,
                      columnStyle: b,
                      bodyHeight: Re - +t
                    }
                  ) }) : null,
                  /* @__PURE__ */ Q(
                    "div",
                    {
                      ref: ct,
                      className: "wx-4VuBwK2D wx-body",
                      style: { width: `${be}px`, height: `${Xe}px` },
                      onMouseDown: (o) => At(o),
                      children: [
                        r ? /* @__PURE__ */ c(zn, { overlay: r }) : null,
                        /* @__PURE__ */ c(
                          "div",
                          {
                            ref: lt,
                            className: "wx-4VuBwK2D wx-data",
                            style: {
                              paddingTop: `${le.d}px`,
                              paddingLeft: `${re.d}px`
                            },
                            children: ue.map((o, p) => {
                              const N = D.indexOf(o.id) !== -1, M = He === o.id, X = "wx-row" + (f ? " wx-autoheight" : "") + (a ? " " + a(o) : "") + (N ? " wx-selected" : "") + (M ? " wx-inactive" : ""), A = f ? { minHeight: `${o.rowHeight || P}px` } : { height: `${o.rowHeight || P}px` };
                              return /* @__PURE__ */ c(
                                "div",
                                {
                                  className: "wx-4VuBwK2D " + X,
                                  "data-id": o.id,
                                  "data-context-id": o.id,
                                  style: A,
                                  role: "row",
                                  "aria-rowindex": p,
                                  "aria-expanded": o.open,
                                  "aria-level": U ? o.$level + 1 : void 0,
                                  "aria-selected": U ? N : void 0,
                                  tabIndex: -1,
                                  children: re.data.map((L) => L.collapsed ? /* @__PURE__ */ c(
                                    "div",
                                    {
                                      className: "wx-4VuBwK2D wx-cell wx-collapsed"
                                    },
                                    L.id
                                  ) : k?.id === o.id && k.column == L.id ? /* @__PURE__ */ c(Fn, { row: o, column: L }, L.id) : /* @__PURE__ */ c(
                                    Hn,
                                    {
                                      row: o,
                                      column: L,
                                      columnStyle: b,
                                      cellStyle: d,
                                      reorder: j,
                                      focusable: De?.row === o.id && De?.column == L.id
                                    },
                                    L.id
                                  ))
                                },
                                o.id
                              );
                            })
                          }
                        )
                      ]
                    }
                  ),
                  t && (w || []).length ? /* @__PURE__ */ c(
                    xt,
                    {
                      type: "footer",
                      contentWidth: be,
                      deltaLeft: re.df,
                      columns: re.footer,
                      columnStyle: b
                    }
                  ) : null
                ]
              }
            )
          }
        )
      }
    ),
    te ? /* @__PURE__ */ c(
      Yn,
      {
        config: te,
        rowStyle: a,
        columnStyle: b,
        cellStyle: d,
        header: e,
        footer: t,
        reorder: j
      }
    ) : null
  ] });
}
const Bn = (n) => n.split("-").map((e) => e ? e.charAt(0).toUpperCase() + e.slice(1) : "").join(""), lo = jt(function({
  data: e = [],
  columns: t = [],
  rowStyle: r = null,
  columnStyle: u = null,
  cellStyle: x = null,
  selectedRows: a,
  select: b = !0,
  multiselect: d = !1,
  header: f = !0,
  footer: h = !1,
  dynamic: s = null,
  overlay: v = null,
  reorder: E = !1,
  onReorder: C = null,
  autoRowHeight: l = !1,
  sizes: m,
  split: g,
  tree: w = !1,
  autoConfig: y = !1,
  init: T = null,
  responsive: D = null,
  sortMarks: R,
  undo: i = !1,
  hotkeys: k = null,
  ...U
}, J) {
  const te = I();
  te.current = U;
  const V = S(() => new cn(gn), []), j = S(() => V.in, [V]), H = I(null);
  H.current === null && (H.current = new Jt((_, B) => {
    const z = "on" + Bn(_);
    te.current && te.current[z] && te.current[z](B);
  }), j.setNext(H.current));
  const $ = S(
    () => ({
      getState: V.getState.bind(V),
      getReactiveState: V.getReactive.bind(V),
      getStores: () => ({ data: V }),
      exec: j.exec,
      setNext: (_) => (H.current = H.current.setNext(_), H.current),
      intercept: j.intercept.bind(j),
      on: j.on.bind(j),
      detach: j.detach.bind(j),
      getRow: V.getRow.bind(V),
      getRowIndex: V.getRowIndex.bind(V),
      getColumn: V.getColumn.bind(V)
    }),
    [V, j]
  ), [Ee, ve] = O(0), [_e, ge] = O(0), [Le, ce] = O(null), [P, ee] = O(null), He = S(() => {
    if (y && !t.length && e.length) {
      const _ = e[0], B = [];
      for (let z in _)
        if (z !== "id" && z[0] !== "$") {
          let ae = {
            id: z,
            header: z[0].toUpperCase() + z.slice(1)
          };
          typeof y == "object" && (ae = { ...ae, ...y }), B.push(ae);
        }
      return B;
    }
    return (P && P.columns) ?? t;
  }, [y, t, e, P]), Ce = S(
    () => (P && P.sizes) ?? m,
    [P, m]
  ), pe = W(
    (_) => {
      if (ve(_.width), ge(_.height), D) {
        const z = Object.keys(D).map(Number).sort((ae, Me) => ae - Me).find((ae) => _.width <= ae) ?? null;
        z !== Le && (ee(D[z]), ce(z));
      }
    },
    [D, Le]
  ), Se = Z(Je.theme), ne = I(0);
  return K(() => {
    if (!ne.current)
      T && T($);
    else {
      const _ = V.getState();
      V.init({
        data: e,
        columns: He,
        split: g || _.split,
        sizes: Ce || _.sizes,
        selectedRows: a || _.selectedRows,
        dynamic: s,
        tree: w,
        sortMarks: R || _.sortMarks,
        undo: i,
        reorder: E,
        _skin: Se,
        _select: b
      });
    }
    ne.current++;
  }, [
    V,
    e,
    He,
    g,
    Ce,
    a,
    s,
    w,
    R,
    i,
    E,
    Se,
    b,
    T,
    $
  ]), ne.current === 0 && V.init({
    data: e,
    columns: He,
    split: g || { left: 0 },
    sizes: Ce || {},
    selectedRows: a || [],
    dynamic: s,
    tree: w,
    sortMarks: R || {},
    undo: i,
    reorder: E,
    _skin: Se,
    select: b
  }), Bt(
    J,
    () => ({
      ...$
    }),
    [$]
  ), /* @__PURE__ */ c(oe.Provider, { value: $, children: /* @__PURE__ */ c(Zt, { words: et, optional: !0, children: /* @__PURE__ */ c(
    jn,
    {
      header: f,
      footer: h,
      overlay: v,
      rowStyle: r,
      columnStyle: u,
      cellStyle: x,
      onReorder: C,
      multiselect: d,
      autoRowHeight: l,
      clientWidth: Ee,
      clientHeight: _e,
      responsiveLevel: Le,
      resize: pe,
      hotkeys: k
    }
  ) }) });
});
function Gn({ item: n }) {
  return /* @__PURE__ */ Q(
    "div",
    {
      tabIndex: -1,
      role: "menuitem",
      "aria-label": n.hidden ? `Show ${n.text} column` : `Hide ${n.text} column`,
      children: [
        /* @__PURE__ */ c(
          "div",
          {
            className: "wx-v13lZxja wx-icon" + (n.hidden ? " wx-hidden" : ""),
            children: /* @__PURE__ */ c("i", { className: "wx-v13lZxja wxi-eye" })
          }
        ),
        /* @__PURE__ */ c("span", { children: n.text })
      ]
    }
  );
}
function so({ columns: n = null, api: e, children: t }) {
  K(() => {
    xn("table-header", Gn);
  }, []);
  function r(d) {
    for (let f = d.header.length - 1; f >= 0; f--) {
      const h = d.header[f].text;
      if (h) return h;
    }
    return d.id;
  }
  function u(d) {
    const f = d.action;
    f && e.exec("hide-column", { id: f.id, mode: !f.hidden });
  }
  function x(d) {
    return d;
  }
  const a = he(e, "_columns"), b = S(() => {
    if (e) {
      const d = Array.isArray(a) ? a : [];
      return (n ? d.filter((h) => n[h.id]) : d).map((h) => {
        const s = r(h);
        return {
          id: h.id,
          text: s,
          type: "table-header",
          hidden: h.hidden
        };
      });
    } else
      return [];
  }, [e, n, a]);
  return /* @__PURE__ */ c(
    Tt,
    {
      dataKey: "headerId",
      options: b,
      onClick: u,
      at: "point",
      resolver: x,
      children: typeof t == "function" ? t() : t
    }
  );
}
function io(n) {
  const { content: e = null, api: t, children: r } = n, u = I(null), x = I(null), [a, b] = O(), [d, f] = O(), [h, s] = O();
  function v(y) {
    for (; y; ) {
      if (y.getAttribute) {
        const T = y.getAttribute("data-row-id"), D = y.getAttribute("data-col-id");
        if (T && t && D) {
          const R = t.getColumn(D);
          return { id: T, col: R, target: y };
        }
      }
      y = y.parentNode;
    }
    return { id: null, col: null, target: null };
  }
  K(() => {
    const y = x.current;
    if (y && a && h) {
      let T = y.getBoundingClientRect(), D = h.left, R = h.top;
      T.right >= a.right && (D = a.width - T.width - 5), T.bottom >= a.bottom && (R = h.top - (T.bottom - a.bottom + 2)), (D !== h.left || R !== h.top) && s({ ...h, left: D, top: R });
    }
  }, [a, h]);
  const E = I(), C = 300, l = (y) => {
    clearTimeout(E.current), E.current = setTimeout(() => {
      y();
    }, C);
  };
  function m(y) {
    const { id: T, target: D, col: R } = v(y.target);
    if (s(null), !T) {
      clearTimeout(E.current);
      return;
    }
    const i = y.clientX;
    l(() => {
      let k = "";
      if (T) {
        const H = g(T);
        f(H), k = w(R, H);
      }
      const U = D.getBoundingClientRect(), J = u.current;
      if (!J) return;
      const te = J.getBoundingClientRect();
      b(te);
      const V = U.top + U.height - te.top, j = i - te.left;
      s({ top: V, left: j, col: R, text: k });
    });
  }
  function g(y) {
    return t.getRow(y);
  }
  function w(y, T) {
    return T = T || d, typeof y.tooltip == "function" ? y.tooltip(T) : Ke(T, y) || "";
  }
  return /* @__PURE__ */ Q("div", { className: "wx-RYccULtD wx-area", ref: u, onMouseMove: m, children: [
    h && h.col.tooltip !== !1 && (e || h.text) ? /* @__PURE__ */ c(
      "div",
      {
        className: "wx-RYccULtD tooltip",
        role: "alert",
        tabIndex: 0,
        ref: x,
        style: { top: h.top + "px", left: h.left + "px" },
        children: e ? /* @__PURE__ */ c(e, { data: d }) : h.text
      }
    ) : null,
    r
  ] });
}
function co({
  api: n,
  options: e = [...an],
  at: t = "point",
  resolver: r,
  dataKey: u,
  filter: x,
  css: a,
  children: b,
  handler: d,
  onClick: f
}) {
  const h = Z(Je.i18n), s = S(() => h ? h.getGroup("grid") : Dt(et).getGroup("grid"), [h]), v = he(n, "selectedRows"), E = he(n, "data"), C = he(n, "reorder"), l = S(() => {
    const R = T(e), i = vt(R);
    return m(i), i;
  }, [e, C, E]);
  function m(R) {
    R.forEach((i) => {
      i.text && (i.text = s(i.text)), i.subtext && (i.subtext = s(i.subtext)), i.data && (i.data = m(i.data));
    });
  }
  const g = S(() => {
    const R = [];
    return l.forEach((i) => {
      switch (i.id) {
        case "move-item:up":
        case "move-item:down":
        case "paste-row": {
          if (!i.isDisabled || !n) {
            R.push(i);
            return;
          }
          const k = i.isDisabled(
            i.id === "paste-row" ? n : v,
            E
          );
          R.push({
            ...i,
            disabled: k
          });
          break;
        }
        default: {
          R.push(i);
          break;
        }
      }
    }), R;
  }, [l, v, E, n]);
  function w(R) {
    return v.includes(R) || n.exec("select-row", { id: R }), R;
  }
  const y = (R) => {
    const i = R.action;
    i && Ct(n, i.id), f && f(R);
  };
  function T(R) {
    return C ? R : R.filter(({ id: i }) => !(i === "move-item:up" || i === "move-item:down"));
  }
  const D = r !== void 0 ? r : w;
  return /* @__PURE__ */ c(
    Tt,
    {
      css: `wx-KKGLK8TH wx-table-menu ${a}`,
      at: t,
      dataKey: u,
      handler: d,
      options: g,
      resolver: D,
      filter: x,
      onClick: y,
      children: b
    }
  );
}
function ao({
  api: n,
  items: e = [...dn],
  onClick: t,
  ...r
}) {
  const x = Z(Je.i18n)?.getGroup("grid") || Dt(et).getGroup("grid"), a = he(n, "selectedRows"), b = he(n, "data"), d = he(n, "history"), f = he(n, "reorder"), h = he(n, "undo"), s = [
    "open-editor",
    "delete-row",
    "copy-row",
    "cut-row",
    "paste-row",
    "move-item:up",
    "move-item:down"
  ], v = ["undo", "redo"], E = S(() => {
    const m = w(e), g = vt(m);
    return y(g), g;
    function w(T) {
      return h && f ? T : T.filter(({ id: D }) => !(!h && (D === "undo" || D === "redo") || !f && (D === "move-item:up" || D === "move-item:down")));
    }
    function y(T) {
      T.forEach((D) => {
        D.text && (D.text = x(D.text)), D.menuText && (D.menuText = x(D.menuText)), D.items && (D.items = y(D.items));
      });
    }
  }, [e, h, f, x]), C = S(() => {
    const m = [], g = a?.length;
    return E.forEach((w) => {
      const y = w.id;
      if (y === "add-row")
        m.push(w);
      else if (s.includes(y)) {
        if (!g) return;
        m.push({
          ...w,
          disabled: w.isDisabled && w.isDisabled(
            y === "paste-row" ? n : a,
            b
          )
        });
      } else v.includes(y) ? m.push({
        ...w,
        disabled: w.isDisabled(d)
      }) : m.push(w);
    }), m;
  }, [E, a, b, d, n, s, v]);
  return /* @__PURE__ */ c(mn, { items: C, onClick: (m) => {
    const g = m.item;
    g && Ct(n, g.id), t && t(m);
  }, ...r });
}
function uo({ fonts: n = !0, children: e }) {
  return e ? /* @__PURE__ */ c(dt, { fonts: n, children: e }) : /* @__PURE__ */ c(dt, { fonts: n });
}
function fo({ fonts: n = !0, children: e }) {
  return e ? /* @__PURE__ */ c(ut, { fonts: n, children: e }) : /* @__PURE__ */ c(ut, { fonts: n });
}
function ho({ fonts: n = !0, children: e }) {
  return e ? /* @__PURE__ */ c(ft, { fonts: n, children: e }) : /* @__PURE__ */ c(ft, { fonts: n });
}
pn(wn);
export {
  co as ContextMenu,
  lo as Grid,
  so as HeaderMenu,
  uo as Material,
  ao as Toolbar,
  io as Tooltip,
  fo as Willow,
  ho as WillowDark,
  wo as defaultMenuOptions,
  xo as defaultToolbarButtons,
  mo as getEditorConfig
};
