// @univerjs/presets/lib/es/index.js
import { IUndoRedoService as h, IAuthzIoService as v, IMentionIOService as y, Univer as A, LogLevel as d } from "@univerjs/core";
export * from "@univerjs/core";
import { FUniver as I } from "@univerjs/core/lib/facade";
export * from "@univerjs/core/lib/facade";
export * from "@univerjs/design";
export * from "@univerjs/themes";
function w(p) {
  const { presets: a, plugins: l, collaboration: c, override: n = [], ...g } = p;
  c && (n.push([h, null]), n.push([v, null]), n.push([y, null]));
  const s = new A({
    logLevel: d.WARN,
    ...g,
    override: n
  }), o = /* @__PURE__ */ new Map();
  a == null || a.forEach((e) => {
    const r = Array.isArray(e) ? e[0] : e, { plugins: u } = r;
    u.forEach((i) => {
      const [t, m] = Array.isArray(i) ? [i[0], i[1]] : [i];
      o.has(t.pluginName) && o.delete(t.pluginName), o.set(t.pluginName, { plugin: t, options: m });
    });
  }), l == null || l.forEach((e) => {
    const [r, u] = Array.isArray(e) ? [e[0], e[1]] : [e];
    if (o.has(r.pluginName))
      throw new Error(`Plugin ${r.pluginName} already registered by presets or other ways! Repeated registration may cause potential problems, please check your code.`);
    o.set(r.pluginName, { plugin: r, options: u });
  }), o.forEach(({ plugin: e, options: r }) => {
    s.registerPlugin(e, r);
  });
  const f = I.newAPI(s);
  return {
    univer: s,
    univerAPI: f
  };
}
export {
  w as createUniver
};
