// @univerjs/preset-sheets-node-core/lib/index.js
import { UniverDocsPlugin as m } from "@univerjs/docs";
export * from "@univerjs/docs";
import { UniverFormulaEnginePlugin as n } from "@univerjs/engine-formula";
export * from "@univerjs/engine-formula";
import { UniverRPCNodeMainPlugin as p } from "@univerjs/rpc-node";
export * from "@univerjs/rpc-node";
import { UniverSheetsPlugin as u } from "@univerjs/sheets";
export * from "@univerjs/sheets";
import { UniverSheetsDataValidationPlugin as f } from "@univerjs/sheets-data-validation";
export * from "@univerjs/sheets-data-validation";
import { UniverSheetsDrawingPlugin as l } from "@univerjs/sheets-drawing";
export * from "@univerjs/sheets-drawing";
import { UniverSheetsFilterPlugin as g } from "@univerjs/sheets-filter";
export * from "@univerjs/sheets-filter";
import { UniverSheetsFormulaPlugin as s } from "@univerjs/sheets-formula";
export * from "@univerjs/sheets-formula";
import { UniverSheetsHyperLinkPlugin as x } from "@univerjs/sheets-hyper-link";
export * from "@univerjs/sheets-hyper-link";
import { UniverSheetsSortPlugin as P } from "@univerjs/sheets-sort";
export * from "@univerjs/sheets-sort";
import { UniverSheetsThreadCommentPlugin as v } from "@univerjs/sheets-thread-comment";
export * from "@univerjs/sheets-thread-comment";
import { UniverThreadCommentPlugin as U } from "@univerjs/thread-comment";
export * from "@univerjs/thread-comment";
import "@univerjs/sheets/lib/facade";
import "@univerjs/sheets-formula/lib/facade";
import "@univerjs/sheets-data-validation/lib/facade";
import "@univerjs/engine-formula/lib/facade";
import "@univerjs/sheets-filter/lib/facade";
import "@univerjs/sheets-hyper-link/lib/facade";
import "@univerjs/sheets-numfmt/lib/facade";
import "@univerjs/sheets-sort/lib/facade";
import "@univerjs/sheets-thread-comment/lib/facade";
function j(e = {}) {
  const { workerSrc: i, formula: r } = e, o = !!i;
  return {
    plugins: [
      o ? [p, { workerSrc: i }] : null,
      [n, {
        notExecuteFormula: o,
        function: r == null ? void 0 : r.function
      }],
      U,
      m,
      u,
      [s, {
        notExecuteFormula: o,
        description: r == null ? void 0 : r.description,
        initialFormulaComputing: r == null ? void 0 : r.initialFormulaComputing
      }],
      f,
      g,
      x,
      l,
      P,
      v
    ].filter((t) => !!t)
  };
}
export {
  j as UniverSheetsNodeCorePreset
};
