// @svar-ui/react-toolbar/dist/index.es.js
import { jsx as t, jsxs as b, Fragment as J } from "react/jsx-runtime";
import { useMemo as R, useCallback as X, useState as j, useRef as N, useEffect as I } from "react";
import { Dropdown as H, Button as M } from "@svar-ui/react-core";
import { uid as _ } from "@svar-ui/lib-dom";
import { useWritableProp as Y } from "@svar-ui/lib-react";
const K = {};
function ee(n) {
  return K[n] || n;
}
function k(n, e) {
  K[n] = e;
}
function U({ menu: n = !1 }) {
  return /* @__PURE__ */ t("div", { className: `wx-z1qpqrvg wx-separator${n ? "-menu" : ""}`, children: " " });
}
function Z() {
  return /* @__PURE__ */ t("div", { className: "wx-1IhFzpJV wx-spacer" });
}
const ne = ({ key: n, text: e, ...r }) => r;
function P(n) {
  const { item: e = {}, menu: r = !1, values: c, onClick: i, onChange: s } = n, w = R(
    () => ee(e.comp || "label"),
    [e]
  ), d = X(() => {
    e && e.handler && e.handler(e), i && i({ item: e });
  }, [e, i]), v = R(() => e && e.key && c ? c[e.key] : void 0, [e, c]), x = X(
    ({ value: u }) => {
      e && e.handler && e.handler(e, u), s && s({ value: u, item: e });
    },
    [e, s]
  ), h = R(() => r ? e ? e.menuText || e.text : void 0 : e ? e.text : void 0, [r, e]);
  if (e && e.comp == "spacer")
    return /* @__PURE__ */ t(Z, {});
  if (e && e.comp == "separator")
    return /* @__PURE__ */ t(U, { menu: r });
  {
    const u = w, f = [
      "wx-tb-element",
      e && e.css ? e.css : "",
      e && e.spacer ? "wx-spacer" : "",
      r ? "wx-menu" : ""
    ].filter(Boolean).join(" ");
    return /* @__PURE__ */ t(
      "div",
      {
        className: "wx-KVAsgMam " + f,
        "data-id": e ? e.id : void 0,
        children: /* @__PURE__ */ t(
          u,
          {
            value: v,
            onChange: x,
            onClick: d,
            text: h,
            menu: r,
            ...ne(e)
          }
        )
      }
    );
  }
}
function B({
  item: n,
  values: e = null,
  menu: r = !1,
  onChange: c,
  onClick: i
}) {
  const [s, w] = j(!0), d = () => w(!0), v = () => w(!1), x = (u) => {
    d(), i && i(u);
  }, h = [
    "wx-wSVFAGym",
    "wx-tb-group",
    n.css || "",
    n.layout == "column" ? "wx-column" : "",
    n.collapsed && !r ? "wx-group-collapsed" : ""
  ].filter(Boolean).join(" ");
  return /* @__PURE__ */ t("div", { className: h, children: n.collapsed && !r ? /* @__PURE__ */ b(J, { children: [
    /* @__PURE__ */ b("div", { className: "wx-wSVFAGym wx-collapsed", onClick: v, children: [
      n.icon ? /* @__PURE__ */ t("i", { className: `wx-wSVFAGym icon ${n.icon}` }) : null,
      n.text ? /* @__PURE__ */ t("div", { className: "wx-wSVFAGym wx-label-text", children: n.text }) : null,
      n.text && !n.icon ? /* @__PURE__ */ t("i", { className: "wx-wSVFAGym wx-label-arrow wxi-angle-down" }) : null
    ] }),
    s ? null : /* @__PURE__ */ t(H, { width: "", oncancel: d, children: /* @__PURE__ */ t("div", { className: "wx-wSVFAGym wx-drop-group", children: /* @__PURE__ */ t(
      B,
      {
        item: { ...n, text: "", collapsed: !1 },
        values: e,
        menu: r,
        onChange: c,
        onClick: x
      }
    ) }) })
  ] }) : /* @__PURE__ */ b(J, { children: [
    /* @__PURE__ */ t("div", { className: "wx-wSVFAGym wx-tb-body", children: n.items.map(
      (u, f) => u.items ? /* @__PURE__ */ t(
        B,
        {
          item: u,
          values: e,
          onClick: x,
          onChange: c
        },
        u.id || f
      ) : /* @__PURE__ */ t(
        P,
        {
          item: u,
          values: e,
          onClick: x,
          onChange: c
        },
        u.id || f
      )
    ) }),
    n.text ? /* @__PURE__ */ t("div", { className: "wx-wSVFAGym wx-label", children: n.text }) : null
  ] }) });
}
function te({ items: n = [], css: e, values: r, width: c, onClick: i, onChange: s }) {
  const [w, d] = j(void 0), v = N(null);
  function x() {
    d(null);
  }
  function h() {
    d(!0);
  }
  function u(f) {
    x(), i && i(f);
  }
  return /* @__PURE__ */ b(
    "div",
    {
      className: `wx-Yo6BuX0p wx-menu ${e || ""}`,
      ref: v,
      "data-id": "$menu",
      children: [
        /* @__PURE__ */ t(M, { icon: "wxi-dots-h", onClick: h }),
        w ? /* @__PURE__ */ t(H, { width: `${c}px`, onCancel: x, children: /* @__PURE__ */ t("div", { className: "wx-Yo6BuX0p wx-drop-menu", children: n.map(
          (f, C) => f.items ? /* @__PURE__ */ t(
            B,
            {
              item: f,
              values: r,
              menu: !0,
              onClick: u,
              onChange: s
            },
            f.id || C
          ) : /* @__PURE__ */ t(
            P,
            {
              item: f,
              values: r,
              menu: !0,
              onClick: u,
              onChange: s
            },
            f.id || C
          )
        ) }) }) : null
      ]
    }
  );
}
function re(n) {
  return n.forEach((e) => {
    e.id || (e.id = _());
  }), n;
}
function pe(n) {
  const {
    items: e,
    menuCss: r = "",
    css: c = "",
    values: i,
    overflow: s = "menu",
    onClick: w,
    onChange: d
  } = n, [v, x] = Y(e || []), [h, u] = Y(i || null), f = R(() => re(v), [v]), C = N(null), F = N(-1), [S, T] = j([]), $ = N(f);
  I(() => {
    $.current = f;
  }, [v]);
  const O = N(s);
  I(() => {
    O.current = s;
  }, [s]);
  const q = N(S);
  I(() => {
    q.current = S;
  }, [S]);
  const V = N(!1);
  function A(l) {
    h && (h[l.item.key] = l.value, u({ ...h })), d && d(l);
  }
  function D() {
    const l = C.current;
    if (!l) return 0;
    const a = l.children, o = $.current || [];
    let m = 0;
    for (let g = 0; g < o.length; g++)
      o[g].comp !== "spacer" && (m += a[g].clientWidth, o[g].comp === "separator" && (m += 8));
    return m;
  }
  function L() {
    const l = C.current, a = $.current || [];
    if (l) {
      for (let o = a.length - 1; o >= 0; o--)
        if (a[o].items && !a[o].collapsed) {
          a[o].collapsed = !0, a[o].$width = l.children[o].offsetWidth, V.current = !0, x([...a]);
          return;
        }
    }
  }
  function Q(l) {
    const a = C.current, o = $.current || [];
    if (a) {
      for (let m = 0; m < o.length; m++)
        if (o[m].collapsed && o[m].$width) {
          o[m].$width - a.children[m].offsetWidth < l + 10 && (o[m].collapsed = !1, V.current = !0), x([...o]);
          return;
        }
    }
  }
  function z() {
    const l = C.current;
    if (!l) return;
    const a = $.current || [], o = O.current;
    if (o === "wrap") return;
    const m = l.clientWidth;
    if (l.scrollWidth > m) {
      if (o === "collapse") return L();
      const y = l.children;
      let G = 0;
      for (let p = 0; p < a.length; p++) {
        if (G += y[p].clientWidth, a[p].comp === "separator" && (G += 8), G > m - 40) {
          if (F.current === p) return;
          F.current = p;
          const E = [];
          for (let W = p; W < a.length; W++)
            E.push(a[W]), y[W].style.visibility = "hidden";
          p > 0 && a[p - 1].comp === "separator" && (y[p - 1].style.visibility = "hidden"), T(E);
          break;
        }
        y[p].style.visibility = "";
      }
    } else {
      const y = m - D();
      if (y <= 0) return;
      if (o === "collapse") return Q(y);
      if ((q.current || []).length) {
        F.current = null;
        const G = l.children;
        for (let p = 0; p < a.length; p++)
          G[p].style.visibility = "";
        T([]);
      }
    }
  }
  return I(() => {
    V.current && (V.current = !1, z());
  }, [v]), I(() => {
    const l = new ResizeObserver(() => z());
    return C.current && l.observe(C.current), () => {
      l.disconnect();
    };
  }, []), /* @__PURE__ */ b(
    "div",
    {
      className: `wx-VdPSJj8y wx-toolbar ${c || ""} ${s === "wrap" ? "wx-wrap" : ""}`,
      ref: C,
      children: [
        f.map(
          (l) => l.items ? /* @__PURE__ */ t(
            B,
            {
              item: l,
              values: h,
              onClick: w,
              onChange: A
            },
            l.id
          ) : /* @__PURE__ */ t(
            P,
            {
              item: l,
              values: h,
              onClick: w,
              onChange: A
            },
            l.id
          )
        ),
        !!S.length && /* @__PURE__ */ t(
          te,
          {
            items: S,
            css: r,
            values: h,
            onClick: w,
            onChange: A
          }
        )
      ]
    }
  );
}
function le(n) {
  const { icon: e, text: r = "", css: c, type: i, disabled: s, menu: w, onClick: d } = n;
  return w ? /* @__PURE__ */ b("div", { className: "wx-HXpG4gnx wx-item", onClick: d, children: [
    /* @__PURE__ */ t("i", { className: `wx-HXpG4gnx ${e || "wxi-empty"} ${c || ""}` }),
    r
  ] }) : /* @__PURE__ */ t(
    M,
    {
      icon: e,
      type: i,
      css: c,
      text: r,
      disabled: s,
      onClick: d
    }
  );
}
function ce(n) {
  const { text: e, value: r, children: c } = n;
  return c ? /* @__PURE__ */ t("div", { className: "wx-PTEZGYcj wx-label", children: c() }) : /* @__PURE__ */ t("div", { className: "wx-PTEZGYcj wx-label", children: r || e });
}
function oe(n) {
  const { icon: e, text: r, css: c, type: i, disabled: s, menu: w, onClick: d } = n;
  return w ? /* @__PURE__ */ b("div", { className: "wx-3cuSqONJ wx-item", onClick: d, children: [
    e ? /* @__PURE__ */ t("i", { className: `wx-3cuSqONJ ${e || ""} ${c || ""}` }) : null,
    r
  ] }) : /* @__PURE__ */ t(
    M,
    {
      icon: e,
      type: i,
      css: c,
      title: r,
      disabled: s,
      onClick: d
    }
  );
}
function se({ id: n = "", text: e = "", css: r = "", icon: c = "", onClick: i }) {
  function s() {
    i && i({ id: n });
  }
  return /* @__PURE__ */ b("div", { className: `wx-U0Bx7pIR wx-label ${r}`, onClick: s, children: [
    c ? /* @__PURE__ */ t("i", { className: "wx-U0Bx7pIR " + c }) : null,
    e
  ] });
}
k("button", le);
k("separator", U);
k("spacer", Z);
k("label", ce);
k("item", se);
k("icon", oe);
export {
  pe as Toolbar,
  k as registerToolbarItem
};
