// @pdfme/pdf-lib/dist/node/src/core/objects/PDFNumber.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const index_1 = require("../../utils/index");
const PDFObject_1 = __importDefault(require("./PDFObject"));
class PDFNumber extends PDFObject_1.default {
    constructor(value) {
        super();
        Object.defineProperty(this, "numberValue", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "stringValue", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.numberValue = value;
        this.stringValue = (0, index_1.numberToString)(value);
    }
    asNumber() {
        return this.numberValue;
    }
    /** @deprecated in favor of [[PDFNumber.asNumber]] */
    value() {
        return this.numberValue;
    }
    clone() {
        return PDFNumber.of(this.numberValue);
    }
    toString() {
        return this.stringValue;
    }
    sizeInBytes() {
        return this.stringValue.length;
    }
    copyBytesInto(buffer, offset) {
        offset += (0, index_1.copyStringIntoBuffer)(this.stringValue, buffer, offset);
        return this.stringValue.length;
    }
}
Object.defineProperty(PDFNumber, "of", {
    enumerable: true,
    configurable: true,
    writable: true,
    value: (value) => new PDFNumber(value)
});
exports.default = PDFNumber;
