// @pdfme/pdf-lib/dist/node/src/core/objects/PDFRawStream.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const PDFStream_1 = __importDefault(require("./PDFStream"));
const utils_1 = require("../../utils");
class PDFRawStream extends PDFStream_1.default {
    constructor(dict, contents, transform) {
        super(dict);
        Object.defineProperty(this, "contents", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "transform", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.contents = contents;
        this.transform = transform;
    }
    asUint8Array() {
        return this.contents.slice();
    }
    clone(context) {
        return PDFRawStream.of(this.dict.clone(context), this.contents.slice());
    }
    getContentsString() {
        return (0, utils_1.arrayAsString)(this.contents);
    }
    getContents() {
        return this.contents;
    }
    getContentsSize() {
        return this.contents.length;
    }
}
Object.defineProperty(PDFRawStream, "of", {
    enumerable: true,
    configurable: true,
    writable: true,
    value: (dict, contents, transform) => new PDFRawStream(dict, contents, transform)
});
exports.default = PDFRawStream;
