// @svar-ui/react-filter/dist/index.es.js
import { jsx as n, jsxs as k, Fragment as ee } from "react/jsx-runtime";
import at, { createContext as st, useContext as j, useState as D, useRef as b, useEffect as T, useMemo as V, useCallback as K, Fragment as Ie, forwardRef as dt, useImperativeHandle as ft } from "react";
import { context as ne, RichSelect as fe, DateRangePicker as Be, DatePicker as Le, Text as we, Combo as pt, Button as te, Checkbox as mt, Locale as be, Portal as ht, Popup as xt, Material as ke, Willow as Ge, WillowDark as Pe } from "@svar-ui/react-core";
import { DropDownMenu as gt, ContextMenu as wt } from "@svar-ui/react-menu";
import { dateToString as ye, delegateClick as yt } from "@svar-ui/lib-dom";
import { useStore as oe, useStoreWithCounter as We, writable as bt } from "@svar-ui/lib-react";
import { en as ve } from "@svar-ui/filter-locales";
import { getFilters as Te, getFilter as ze, DataStore as vt, createFilterRule as Ct } from "@svar-ui/filter-store";
import { createArrayFilter as jt, createFilter as Ut, createFilterRule as Xt, getFilter as Yt, getFilters as Jt, getOptions as Kt, getOptionsMap as Zt } from "@svar-ui/filter-store";
import { EventBusRouter as Nt } from "@svar-ui/lib-state";
const ue = st(null);
function Rt(l) {
  const {
    fields: r = null,
    fieldsSelector: g = !0,
    field: f = null,
    buttons: w = !0,
    options: h = null,
    includes: F = null,
    type: v = "text",
    filter: o = "",
    value: a = "",
    format: x = null,
    predicate: p = null
  } = l, y = l.onApply || l.onapply, A = l.onCancel || l.oncancel, N = l.onChange || l.onchange, s = j(ne.i18n), d = s.getRaw(), _ = d.filter?.dateFormat || d.formats?.dateFormat, L = ye(_, d.calendar), I = s.getGroup("filter"), m = 1, G = 2, q = 4, M = 8, z = 16, U = 32, t = 64, [u, i] = D(), C = b(), [P, re] = D(), H = b(), [pe, qe] = D(), X = b(), [le, me] = D(), S = b(), [Ce, Me] = D([]), E = b([]), [He, Ne] = D(), Z = b(), [, Re] = D(), J = b(), [, Fe] = D(), ae = b(), [B, Qe] = D(!1), [Ae, je] = D(null);
  async function Ue(e) {
    if (!h || typeof h != "function") return;
    const c = await h(e);
    je(c || []);
  }
  const se = b(0), he = b(null);
  function _e(e) {
    return e ? e.toString() : "";
  }
  function $(e, c, R) {
    e(R), c.current = R;
  }
  function O(e, c) {
    switch (e) {
      case m:
        if (H.current !== c)
          return $(re, H, c), e;
        break;
      case z:
        if (X.current !== c)
          return $(qe, X, c), e;
        break;
      case G:
        if (C.current !== c)
          return $(i, C, c), e;
        break;
      case q:
        if (S.current !== c)
          return $(me, S, c), e;
        break;
      case M:
        if (_e(E.current) != _e(c))
          return $(Me, E, c || []), e;
        break;
      case t:
        if (Z.current !== c)
          return $(Ne, Z, c || []), e;
        break;
    }
    return 0;
  }
  function W(e) {
    se.current = se.current | e, he.current || (he.current = setTimeout(() => {
      const c = se.current;
      he.current = null, se.current = 0, tt(c);
    }));
  }
  const [Xe, $e] = D(), xe = b(), [Ye, De] = D([]), ce = b([]), [, Je] = D(), Se = b(), [Ke, Ze] = D(), et = b();
  function tt(e) {
    if (e & m && r) {
      $(Ne, Z, null);
      const c = r.find((Q) => Q.id === H.current), R = S.current instanceof Date || S.current && typeof S.current == "object", Y = c.type === "tuple" || X.current == "tuple";
      if (R && c.type !== "date" || !R && c.type == "date" || isNaN(S.current) && c.type == "number" || Y)
        $(me, S, null);
      else if (c.type === "text" && typeof S.current != "string") {
        const Q = S.current || S.current === 0 ? S.current.toString() : "";
        $(me, S, Q);
      }
      $(Re, J, c.format), $(Fe, ae, c.predicate), e = e | O(z, c.type || "text"), Ue(H.current);
    }
    if (e & z) {
      const c = Te(X.current).map((R) => ({
        id: R.id,
        label: I(R.label || R.id)
      }));
      $(Ze, et, c), c.some((R) => R?.id === C.current) || $(i, C, null);
    }
    if (e & z || e & G) {
      let c = ze(C.current, X.current);
      c || (c = Te(X.current).find((R) => R.default), e = e | O(G, c.id)), $(Je, Se, c), typeof S.current == "object" && !(S.current instanceof Date) && !c.range && (e = e | O(
        q,
        X.current === "date" ? null : ""
      ));
    }
    if (e & G || e & q || e & t)
      if (Z.current) {
        const c = S.current || S.current === 0 ? (Q) => Se.current.handler(Q, S.current || "") : null, R = C.current && c ? Z.current.filter(c) : Z.current;
        $(De, ce, R);
        let Y = E.current;
        E.current.length && (Y = E.current.filter((Q) => R.includes(Q))), Y.length ? Y.length !== E.current.length && (e = e | O(M, Y)) : e = e | O(M, [...R]);
      } else
        $(De, ce, []);
    if (e & M) {
      const c = E.current.length === ce.current.length;
      $($e, xe, c);
    }
    if (e & U) {
      const c = Ve();
      N && N({ value: c });
    }
  }
  function nt() {
    const e = Ve();
    y && y({ value: e });
  }
  function rt() {
    A && A();
  }
  function lt() {
    const e = !xe.current;
    $($e, xe, e), W(
      O(
        M,
        e ? [...ce.current] : []
      ) | U
    );
  }
  function ct(e) {
    const { inputValue: c, value: R } = e, Y = R ? [...E.current, c] : E.current.filter((Q) => Q != c);
    W(O(M, Y) | U);
  }
  function ie({ value: e }) {
    e === "$empty" && (e = ""), W(O(q, e) | U);
  }
  function it({ value: e }) {
    W(O(m, e) | U);
  }
  function ot({ value: e }) {
    W(O(G, e) | U);
  }
  function Ve() {
    const e = {
      filter: C.current,
      value: S.current,
      type: X.current
    };
    return ae.current && (e.predicate = ae.current), H.current && (e.field = H.current), E.current && E.current.length && E.current.length !== ce.current.length ? e.includes = [...E.current] : e.includes = [], e;
  }
  function Oe(e) {
    return J.current && typeof J.current == "function" ? J.current(e) : e instanceof Date ? J.current ? ye(J.current, d.calendar)(e) : L(e) : typeof e == "string" ? e : e.toString();
  }
  function ut(e) {
    let c = e ? e.map((R) => ({ id: R, label: Oe(R) })) : [];
    return [{ id: "$empty", label: "", emptyLabel: I("None") }].concat(c);
  }
  const ge = b(null);
  return T(() => {
    H.current && ge.current && setTimeout(() => {
      const e = ge.current, c = e && e.querySelector("input");
      c && c.focus();
    }, 1);
  }, [P]), T(() => {
    B && W(O(m, f));
  }, [f, B]), T(() => {
    B && W(O(z, v));
  }, [v, B]), T(() => {
    B && W(O(G, o));
  }, [o, B]), T(() => {
    B && W(O(q, a));
  }, [a, B]), T(() => {
    B && W(O(M, F));
  }, [F, B]), T(() => {
    const e = Array.isArray(h) ? h : Ae;
    B && W(O(t, e));
  }, [h, Ae, B]), T(() => {
    x && !r && $(Re, J, x), p && !r && $(Fe, ae, p);
  }, [x, p, r]), T(() => {
    Qe(!0);
  }, []), /* @__PURE__ */ n(ne.fieldId.Provider, { value: null, children: /* @__PURE__ */ k("div", { className: "wx-3z8r9Oys wx-filter-editor", children: [
    r && g ? /* @__PURE__ */ n(fe, { onChange: it, options: r, value: P }) : null,
    /* @__PURE__ */ k("div", { className: "wx-3z8r9Oys wx-wrapper", children: [
      /* @__PURE__ */ n("div", { className: "wx-3z8r9Oys wx-cell", children: /* @__PURE__ */ n(
        fe,
        {
          onChange: ot,
          options: Ke,
          value: u,
          placeholder: I("Click to select")
        }
      ) }),
      /* @__PURE__ */ n("div", { className: "wx-3z8r9Oys wx-cell", ref: ge, children: pe === "date" ? u == "between" || u == "notBetween" ? /* @__PURE__ */ n(
        Be,
        {
          format: _,
          value: le,
          buttons: ["done", "clear", "today"],
          onChange: ie
        }
      ) : /* @__PURE__ */ n(Le, { format: _, value: le, onChange: ie }) : pe === "number" ? /* @__PURE__ */ n(we, { value: le, onChange: ie, type: "number" }) : pe === "tuple" ? /* @__PURE__ */ n(
        pt,
        {
          value: le,
          options: ut(He),
          onChange: ie,
          children: ({ option: e }) => e.emptyLabel || e.label
        }
      ) : /* @__PURE__ */ n(we, { value: le, onChange: ie }) })
    ] }),
    /* @__PURE__ */ n(te, { onClick: lt, children: I(Xe ? "Unselect all" : "Select all") }),
    /* @__PURE__ */ n("div", { className: "wx-3z8r9Oys wx-list", role: "listbox", children: Ye.map((e, c) => /* @__PURE__ */ n(
      "div",
      {
        className: "wx-3z8r9Oys wx-item",
        tabIndex: c ? -1 : 0,
        role: "option",
        children: /* @__PURE__ */ n(
          mt,
          {
            label: Oe(e),
            inputValue: e,
            value: Ce && Ce.includes(e),
            onChange: ct
          }
        )
      },
      c
    )) }),
    w ? /* @__PURE__ */ k("div", { className: "wx-3z8r9Oys wx-wrapper", children: [
      /* @__PURE__ */ n(te, { type: "secondary", onClick: rt, children: I("Cancel") }),
      /* @__PURE__ */ n(te, { type: "primary", onClick: nt, children: I("Apply") })
    ] }) : null
  ] }) });
}
function Ft(l) {
  return /* @__PURE__ */ n(be, { words: ve, optional: !0, children: /* @__PURE__ */ n(Rt, { ...l }) });
}
function Ee(l) {
  const { rule: r, type: g } = l, f = V(
    () => r.includes && r.includes.length ? r.includes : null,
    [r]
  ), w = V(() => r.filter || "contains", [r]), h = V(() => r.value || "", [r]), F = V(() => r.field, [r]), v = V(() => r.id, [r]), o = j(ue), a = oe(o, "fields"), x = oe(o, "options"), p = K(() => {
    o.exec("edit-rule", {});
  }, [o]), y = K(
    ({ value: N }) => {
      o.exec("update-rule", { id: v, rule: N }), p();
    },
    [o, v]
  ), A = K(
    ({ value: N }) => {
      o.exec("change-rule", { id: v, rule: N });
    },
    [o, v]
  );
  return /* @__PURE__ */ n("div", { className: "wx-uOIGm7Ce wx-panel", children: /* @__PURE__ */ n(
    Ft,
    {
      fieldsSelector: g !== "simple",
      fields: a,
      field: F,
      options: x,
      includes: f,
      filter: w,
      value: h,
      onApply: y,
      onCancel: p,
      onChange: A
    }
  ) });
}
function At({
  filter: l,
  field: r,
  type: g = "list",
  action: f = "menu"
}) {
  const w = (_) => _, h = j(ne.i18n), F = b();
  F.current === void 0 && (F.current = h.getRaw());
  const v = F.current, o = b();
  o.current === void 0 && (o.current = r?.format || v.filter?.dateFormat || v.formats.dateFormat);
  const a = o.current, x = V(
    () => typeof a == "function" ? a : r?.type === "date" ? ye(a, v.calendar) : w,
    [r]
  ), p = b();
  p.current === void 0 && (p.current = h.getGroup("filter"));
  const y = p.current, [A, N] = D(), [s, d] = D();
  return T(() => {
    d(l.value), N(l.filter ? ze(l.filter, l.type) : null);
  }, [l]), /* @__PURE__ */ k("div", { className: `wx-5GeVl88l wx-rule wx-${g}`, "data-id": l.id, children: [
    /* @__PURE__ */ k("span", { className: "wx-5GeVl88l wx-filter", children: [
      /* @__PURE__ */ n("span", { className: "wx-5GeVl88l wx-field", children: r.label }),
      l.includes && l.includes.length ? /* @__PURE__ */ k(ee, { children: [
        " " + y("in") + " ",
        l.includes.map((_, L) => /* @__PURE__ */ k(at.Fragment, { children: [
          /* @__PURE__ */ n("span", { className: "wx-5GeVl88l wx-value", children: x(_) }),
          L < l.includes.length - 1 ? ", " : null
        ] }, _))
      ] }) : l.filter && (s || s === 0) ? /* @__PURE__ */ k(ee, { children: [
        " " + (y(A.short || A.label) || A.id) + " ",
        r.type === "date" ? s.start ? /* @__PURE__ */ k(ee, { children: [
          /* @__PURE__ */ n("span", { className: "wx-5GeVl88l wx-value", children: x(s.start) }),
          " " + y("and") + " ",
          /* @__PURE__ */ n("span", { className: "wx-5GeVl88l wx-value", children: x(s.end) })
        ] }) : /* @__PURE__ */ n("span", { className: "wx-5GeVl88l wx-value", children: x(s) }) : /* @__PURE__ */ n("span", { className: "wx-5GeVl88l wx-value", children: x(s) })
      ] }) : null
    ] }),
    /* @__PURE__ */ n(
      "i",
      {
        className: `wx-5GeVl88l wxi-${f === "menu" ? "dots-v" : f} wx-menu-icon`,
        role: "button",
        "data-action": f
      }
    )
  ] });
}
function _t({ mode: l = "and", id: r }) {
  const f = j(ne.i18n).getGroup("filter"), w = j(ue);
  function h() {
    w.exec("toggle-glue", { id: r });
  }
  return /* @__PURE__ */ n(
    "div",
    {
      className: `wx-XRdDajFX wx-glue wx-${l}`,
      role: "button",
      onClick: h,
      children: f(l)
    }
  );
}
function de(l) {
  const { group: r, type: g, onShowMenu: f } = l, w = j(ue), [h, F] = We(w, "value"), v = oe(w, "fields"), o = oe(w, "_editor"), a = V(() => h && h.getBranch(r.id) || [], [F, r.id]), x = V(() => g == "simple" ? "line" : g, [g]), p = V(
    () => r.$level == 1 ? "top" : "inner",
    [r.$level]
  ), y = V(
    () => ({
      dblclick: (m) => w.exec("edit-rule", { id: m }),
      menu: (m, G) => f && f({ id: m, ev: G }),
      delete: (m) => {
        w.exec("edit-rule", {}), w.exec("delete-rule", { id: m });
      }
    }),
    [w, f]
  ), A = K(
    (m) => v.find((G) => G.id == m),
    [v]
  ), N = b(null), [s, d] = D(), [_, L] = D();
  T(() => {
    const m = N.current;
    return m ? yt(m, y) : void 0;
  }, [y]), T(() => {
    if (!(o && _ === o.id)) {
      if (o && g !== "list" && a.find((m) => m.id === o.id)) {
        const m = N.current && N.current.querySelector(`[data-id='${o.id}']`);
        if (m) {
          d(m.getBoundingClientRect()), L(o.id);
          return;
        }
      }
      d(null), L(null);
    }
  }, [o, g, a, _]);
  function I() {
    d(null), L(null), w.exec("edit-rule", {});
  }
  return /* @__PURE__ */ k(ee, { children: [
    /* @__PURE__ */ n(
      "div",
      {
        ref: N,
        className: `wx-3udP2tTA wx-group wx-${p} wx-${x}`,
        children: a.map((m, G) => /* @__PURE__ */ k(Ie, { children: [
          g === "list" && o && o.id == m.id ? /* @__PURE__ */ n(Ee, { rule: m }) : m.data ? /* @__PURE__ */ n(de, { type: g, group: m, onShowMenu: f }) : /* @__PURE__ */ n("div", { className: "wx-3udP2tTA wx-rule-wrapper", children: /* @__PURE__ */ n(
            At,
            {
              action: g == "simple" ? "delete" : "menu",
              type: x,
              filter: m,
              field: A(m.field)
            }
          ) }),
          g !== "simple" && G < a.length - 1 ? /* @__PURE__ */ n("div", { className: "wx-3udP2tTA wx-glue-wrapper", children: /* @__PURE__ */ n(_t, { mode: r.glue, id: r.id }) }) : null
        ] }, m.id))
      }
    ),
    o && s ? /* @__PURE__ */ n(ht, { children: /* @__PURE__ */ n(
      xt,
      {
        top: s.top + s.height,
        left: s.left + s.width / 2,
        onCancel: I,
        children: /* @__PURE__ */ n("div", { className: "wx-3udP2tTA wx-editor-wrapper", children: /* @__PURE__ */ n(Ee, { type: g, rule: h.byId(o.id) }) })
      }
    ) }) : null
  ] });
}
function $t({ type: l }) {
  const r = j(ue), g = j(ne.i18n), f = V(() => g.getGroup("filter"), [g]), w = oe(r, "fields"), [h, F] = We(r, "value"), v = h.getBranch(0)[0];
  function o(s) {
    r.exec("add-rule", { rule: { $temp: !0, ...s }, edit: !0 });
  }
  const a = V(() => {
    const s = {};
    return h.forEach((d) => s[d.field] = !0), w.filter((d) => !s[d.id]).map((d) => ({ id: d.id, text: d.label }));
  }, [F, w]), x = K(
    (s) => {
      const d = s.action;
      d && o({ field: d.id });
    },
    [o]
  ), p = V(
    () => [
      {
        id: "edit-rule",
        text: f("Edit"),
        icon: "wxi-edit-outline"
      },
      {
        id: "add-rule",
        text: f("Add filter"),
        icon: "wxi-filter-plus-outline",
        resolver: (s) => ({
          after: s,
          rule: { $temp: !0 },
          edit: !0
        })
      },
      {
        id: "add-group",
        text: f("Add group"),
        icon: "wxi-filter-multiple-outline",
        resolver: (s) => ({
          after: s,
          rule: { $temp: !0 },
          edit: !0
        })
      },
      { type: "separator" },
      {
        id: "delete-rule",
        text: f("Delete"),
        icon: "wxi-delete-outline"
      }
    ],
    [f]
  ), y = b(null), A = K(
    (s) => {
      const { context: d, action: _ } = s;
      if (_) {
        const L = _.resolver ? _.resolver(d) : { id: d };
        r.exec(_.id, L);
      }
    },
    [r]
  ), N = K(({ ev: s, id: d }) => {
    y.current && typeof y.current.show == "function" && y.current.show(s, d);
  }, []);
  return /* @__PURE__ */ k(ee, { children: [
    /* @__PURE__ */ n("div", { className: `wx-VDDi7d7g wx-filter-builder wx-${l}`, children: l === "list" ? /* @__PURE__ */ k(ee, { children: [
      /* @__PURE__ */ n("div", { className: `wx-VDDi7d7g wx-toolbar wx-${l}`, children: /* @__PURE__ */ n(te, { type: "primary", onClick: o, children: f("Add filter") }) }),
      /* @__PURE__ */ n(de, { type: l, group: v, onShowMenu: N })
    ] }) : l === "line" ? /* @__PURE__ */ k("div", { className: `wx-VDDi7d7g wx-toolbar wx-${l}`, children: [
      /* @__PURE__ */ n("div", { className: "wx-VDDi7d7g wx-filters", children: /* @__PURE__ */ n(de, { type: l, group: v, onShowMenu: N }) }),
      /* @__PURE__ */ n("div", { className: "wx-VDDi7d7g wx-button", children: /* @__PURE__ */ n(te, { type: "primary", onClick: o, children: f("Add filter") }) })
    ] }) : l === "simple" ? /* @__PURE__ */ k("div", { className: `wx-VDDi7d7g wx-toolbar wx-${l}`, children: [
      /* @__PURE__ */ n("div", { className: "wx-VDDi7d7g wx-button", children: /* @__PURE__ */ n(
        gt,
        {
          options: a,
          onClick: x,
          children: /* @__PURE__ */ n(te, { disabled: !a.length, type: "primary", children: f("Add filter") })
        }
      ) }),
      /* @__PURE__ */ n("div", { className: "wx-VDDi7d7g wx-filters", children: /* @__PURE__ */ n(de, { type: l, group: v, onShowMenu: N }) })
    ] }) : null }),
    l !== "simple" ? /* @__PURE__ */ n(wt, { options: p, onClick: A, ref: y }) : null
  ] });
}
const Dt = (l) => l.split("-").map((r) => r ? r.charAt(0).toUpperCase() + r.slice(1) : "").join(""), Lt = dt(function({
  value: r = { glue: "and", rules: [] },
  fields: g = [],
  options: f = null,
  type: w = "list",
  init: h = null,
  ...F
}, v) {
  const o = b();
  o.current = F;
  const a = V(() => new vt(bt), []), x = V(() => a.in, [a]), p = b(null);
  p.current === null && (p.current = new Nt((N, s) => {
    const d = "on" + Dt(N);
    o.current && o.current[d] && o.current[d](s);
  }), x.setNext(p.current));
  const y = V(
    () => ({
      getState: a.getState.bind(a),
      getReactiveState: a.getReactive.bind(a),
      getStores: () => ({ data: a }),
      exec: x.exec,
      setNext: (N) => (p.current = p.current.setNext(N), p.current),
      intercept: x.intercept.bind(x),
      on: x.on.bind(x),
      detach: x.detach.bind(x),
      getValue: a.getValue.bind(a)
    }),
    [a, x]
  ), A = b(!0);
  return T(() => {
    a.init({
      value: r,
      fields: g,
      options: f
    }), A.current && h && (h(y), A.current = !1);
  }, [a, r, g, f, h, y]), A.current && (a.init({
    value: r,
    fields: g,
    options: f
  }), h && h(y), A.current = !1), ft(
    v,
    () => ({
      ...y
    }),
    [y]
  ), /* @__PURE__ */ n(be, { words: ve, optional: !0, children: /* @__PURE__ */ n(ue.Provider, { value: y, children: /* @__PURE__ */ n($t, { type: w }) }) });
});
function St(l) {
  const { fields: r = [], debounce: g = 300, onChange: f } = l, [w, h] = D(), [F, v] = D({}), o = j(ne.i18n), a = o.getRaw(), x = a.filter && a.filter.dateFormat || a.formats && a.formats.dateFormat, p = o.getGroup("filter");
  function y(t) {
    const u = t.options.map((i) => typeof i == "string" || typeof i == "number" ? { id: i, label: i } : i);
    return u.unshift({ id: "$empty", label: p("None") }), u;
  }
  function A(t) {
    return typeof t == "string" ? { id: t, type: "text", filter: "contains" } : {
      ...t,
      filter: t.filter || (t.options ? "equal" : "contains"),
      options: t.type !== "date" && t.options && y(t)
    };
  }
  function N(t) {
    return t.map((u) => u.type === "all" || u.type === "dynamic" ? {
      ...u,
      by: u.by.map((i) => A(i))
    } : A(u));
  }
  function s(t) {
    return t.map((u) => ({ id: u.id, label: u.id }));
  }
  const d = V(() => N(r), [r]), _ = b(null);
  T(() => {
    const t = {};
    let u = w;
    d.forEach((i) => {
      i.type === "dynamic" ? (i.by.forEach((C) => {
        typeof C.value < "u" && (t[C.id] = C.value);
      }), u = i.by[0].id) : i.type === "all" ? t[i.type] = i.value ?? "" : typeof i.value < "u" && (t[i.id] = i.value);
    }), _.current = t, v(t), h(u);
  }, [d]);
  function L(t, u, i) {
    return t.map((C) => {
      let P = {
        field: C.type,
        label: C.label,
        placeholder: C.placeholder
      };
      if (C.type === "dynamic") {
        P.dynamicField = !0, P.field = u;
        const re = C.by.find((H) => H.id === P.field);
        re && (P = {
          ...re,
          ...P,
          placeholder: re.placeholder ?? P.placeholder
        });
      } else C.type !== "all" && (P = { ...P, ...C, field: C.id });
      return C.by && (P.optionsBy = s(C.by)), P.value = i[P.field], { type: "text", filter: "contains", ...P };
    });
  }
  const I = V(
    () => L(d, w, F),
    [d, w, F]
  );
  function m(t, u) {
    if (t.field === "all")
      return Ct(
        t.optionsBy,
        "contains",
        u[t.field]
      );
    const i = u[t.field];
    return i ? {
      field: t.field,
      filter: t.filter,
      type: t.type,
      value: i
    } : null;
  }
  function G(t) {
    const u = I.map((i) => m(i, t)).filter((i) => !!i);
    return u.length === 1 && u[0]?.glue === "or" ? u[0] : {
      glue: "and",
      rules: u
    };
  }
  const q = b(null);
  function M(t, u) {
    q.current && clearTimeout(q.current), q.current = setTimeout(() => {
      f && f({ value: G(u) });
    }, t);
  }
  T(() => {
    _.current && _.current !== F && M(1, F);
  }, [F]);
  const z = (t, u) => {
    v((i) => ({
      ...i,
      [u]: t === "$empty" ? "" : t
    }));
  }, U = (t) => {
    v((u) => (h(t), w ? { ...u, [w]: "" } : u));
  };
  return /* @__PURE__ */ n("div", { className: "wx-gWqAfosQ wx-filter-bar", children: I.map((t, u) => /* @__PURE__ */ k(Ie, { children: [
    t.label ? /* @__PURE__ */ n("div", { className: "wx-gWqAfosQ wx-label", children: t.label }) : null,
    t.dynamicField ? /* @__PURE__ */ n("div", { className: "wx-gWqAfosQ wx-select", children: /* @__PURE__ */ n(
      fe,
      {
        value: t.field,
        options: t.optionsBy,
        onChange: ({ value: i }) => U(i)
      }
    ) }) : null,
    t.options ? /* @__PURE__ */ n("div", { className: "wx-gWqAfosQ wx-select", children: /* @__PURE__ */ n(
      fe,
      {
        value: t.value,
        placeholder: t.placeholder ?? `${p("filter by")} ${t.field}`,
        options: t.options,
        onChange: ({ value: i }) => z(i, t.field)
      }
    ) }) : t.type === "text" || t.type === "number" ? /* @__PURE__ */ n("div", { className: "wx-gWqAfosQ wx-text", children: /* @__PURE__ */ n(
      we,
      {
        value: t.value,
        placeholder: t.placeholder ?? `${p("filter by")} ${t.field} (${p(t.filter)})`,
        onChange: ({ value: i, input: C }) => {
          C && z(i, t.field);
        },
        type: t.type
      }
    ) }) : t.type === "date" ? /* @__PURE__ */ n("div", { className: "wx-gWqAfosQ wx-date", children: t.filter === "between" || t.filter === "notBetween" ? /* @__PURE__ */ n(
      Be,
      {
        value: t.value,
        format: x,
        buttons: ["done", "clear", "today"],
        placeholder: t.placeholder ?? `${p("filter by")} ${t.field} (${p(t.filter)})`,
        onChange: ({ value: i }) => z(i, t.field)
      }
    ) : /* @__PURE__ */ n(
      Le,
      {
        value: t.value,
        format: x,
        placeholder: t.placeholder ?? `${p("filter by")} ${t.field} (${p(t.filter)})`,
        onChange: ({ value: i }) => z(i, t.field)
      }
    ) }) : null
  ] }, u)) });
}
function Wt(l = {}) {
  return /* @__PURE__ */ n(be, { words: ve, optional: !0, children: /* @__PURE__ */ n(St, { ...l }) });
}
function zt({ fonts: l = !0, children: r }) {
  return r ? /* @__PURE__ */ n(ke, { fonts: l, children: r }) : /* @__PURE__ */ n(ke, { fonts: l });
}
function qt({ fonts: l = !0, children: r }) {
  return r ? /* @__PURE__ */ n(Ge, { fonts: l, children: r }) : /* @__PURE__ */ n(Ge, { fonts: l });
}
function Mt(l) {
  const { fonts: r = !0, children: g } = l;
  return g ? /* @__PURE__ */ n(Pe, { fonts: r, children: g }) : /* @__PURE__ */ n(Pe, { fonts: r });
}
export {
  Wt as FilterBar,
  Lt as FilterBuilder,
  Ft as FilterEditor,
  zt as Material,
  qt as Willow,
  Mt as WillowDark,
  jt as createArrayFilter,
  Ut as createFilter,
  Xt as createFilterRule,
  Yt as getFilter,
  Jt as getFilters,
  Kt as getOptions,
  Zt as getOptionsMap
};
