// regexp-util/lib/or.js
import { Base } from './base.js';
import { Charset } from './charset.js';
import { wrap } from './util.js';
export class Or extends Base {
    constructor(...inputs) {
        super();
        const strings = [];
        const charsetInputs = [];
        for (const input of inputs) {
            if (input instanceof Or) {
                strings.push(...input.strings);
                charsetInputs.push(input.charset);
            }
            else if (typeof input !== 'string' ||
                (input.length === 1 && !isSpecialChar(input))) {
                charsetInputs.push(input);
            }
            else {
                strings.push(input);
            }
        }
        this.strings = Object.keys(dictionaryify(strings));
        this.charset = new Charset(...charsetInputs);
    }
    union(...inputs) {
        return new Or(this.charset, ...this.strings, ...inputs);
    }
    subtract(...inputs) {
        const stringDictionary = dictionaryify(this.strings);
        const charset = new Charset(...this.charset.data);
        const charsetInputs = [];
        for (const input of inputs) {
            if (input instanceof Or) {
                for (const str of input.strings) {
                    delete stringDictionary[str];
                }
                charsetInputs.push(input.charset);
            }
            else if (typeof input !== 'string' ||
                (input.length === 1 && !isSpecialChar(input))) {
                charsetInputs.push(input);
            }
            else {
                delete stringDictionary[input];
            }
        }
        return new Or(charset.subtract(...charsetInputs), ...Object.keys(stringDictionary));
    }
    _isEmpty() {
        return this.charset.isEmpty() && this.strings.length === 0;
    }
    _toString() {
        const parts = [];
        if (this.charset.data.length !== 0) {
            parts.push(this.charset.toString());
        }
        if (this.strings.length !== 0) {
            parts.push(...this.strings);
        }
        return wrap(parts.join('|'));
    }
}
export const or = (...inputs) => new Or(...inputs);
function isSpecialChar(char) {
    switch (char) {
        case '^':
        case '$':
            return true;
        default:
            return false;
    }
}
function dictionaryify(array) {
    const dictionary = {};
    for (const value of array) {
        dictionary[value] = true;
    }
    return dictionary;
}
