// @pdfme/pdf-lib/dist/node/src/core/document/PDFHeader.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const CharCodes_1 = __importDefault(require("../syntax/CharCodes"));
const utils_1 = require("../../utils");
class PDFHeader {
    constructor(major, minor) {
        Object.defineProperty(this, "major", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "minor", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.major = String(major);
        this.minor = String(minor);
    }
    toString() {
        const bc = (0, utils_1.charFromCode)(129);
        return `%PDF-${this.major}.${this.minor}\n%${bc}${bc}${bc}${bc}`;
    }
    sizeInBytes() {
        return 12 + this.major.length + this.minor.length;
    }
    copyBytesInto(buffer, offset) {
        const initialOffset = offset;
        buffer[offset++] = CharCodes_1.default.Percent;
        buffer[offset++] = CharCodes_1.default.P;
        buffer[offset++] = CharCodes_1.default.D;
        buffer[offset++] = CharCodes_1.default.F;
        buffer[offset++] = CharCodes_1.default.Dash;
        offset += (0, utils_1.copyStringIntoBuffer)(this.major, buffer, offset);
        buffer[offset++] = CharCodes_1.default.Period;
        offset += (0, utils_1.copyStringIntoBuffer)(this.minor, buffer, offset);
        buffer[offset++] = CharCodes_1.default.Newline;
        buffer[offset++] = CharCodes_1.default.Percent;
        buffer[offset++] = 129;
        buffer[offset++] = 129;
        buffer[offset++] = 129;
        buffer[offset++] = 129;
        return offset - initialOffset;
    }
}
Object.defineProperty(PDFHeader, "forVersion", {
    enumerable: true,
    configurable: true,
    writable: true,
    value: (major, minor) => new PDFHeader(major, minor)
});
exports.default = PDFHeader;
