// @univerjs/preset-docs-advanced/lib/es/index.js
import { UniverDocsExchangeClientPlugin as l } from "@univerjs-pro/docs-exchange-client";
export * from "@univerjs-pro/docs-exchange-client";
import { UniverDocsPrintPlugin as p } from "@univerjs-pro/docs-print";
export * from "@univerjs-pro/docs-print";
import { UniverExchangeClientPlugin as s } from "@univerjs-pro/exchange-client";
export * from "@univerjs-pro/exchange-client";
import { UniverLicensePlugin as a } from "@univerjs-pro/license";
export * from "@univerjs-pro/license";
import "@univerjs-pro/exchange-client/lib/facade";
function f(i = {
  license: "",
  universerEndpoint: ""
}) {
  const {
    license: n,
    universerEndpoint: r,
    exchangeClientOptions: o
  } = i, e = r != null ? r : `${window.location.protocol}//${window.location.host}`;
  return {
    plugins: [
      [a, { license: n }],
      p,
      [s, {
        uploadFileServerUrl: `${e}/universer-api/stream/file/upload`,
        getTaskServerUrl: `${e}/universer-api/exchange/task/{taskID}`,
        signUrlServerUrl: `${e}/universer-api/file/{fileID}/sign-url`,
        importServerUrl: `${e}/universer-api/exchange/{type}/import`,
        exportServerUrl: `${e}/universer-api/exchange/{type}/export`,
        downloadEndpointUrl: `${e}/`,
        options: o
      }],
      l
    ].filter((t) => !!t)
  };
}
export {
  f as UniverDocsAdvancedPreset
};
