// @univerjs/preset-sheets-collaboration/lib/index.js
import { UniverCollaborationPlugin as l } from "@univerjs-pro/collaboration";
export * from "@univerjs-pro/collaboration";
import { UniverCollaborationClientPlugin as s } from "@univerjs-pro/collaboration-client";
export * from "@univerjs-pro/collaboration-client";
import { BrowserCollaborationSocketService as a, UniverCollaborationClientUIPlugin as p } from "@univerjs-pro/collaboration-client-ui";
export * from "@univerjs-pro/collaboration-client-ui";
import { UniverEditHistoryLoaderPlugin as c } from "@univerjs-pro/edit-history-loader";
export * from "@univerjs-pro/edit-history-loader";
import "@univerjs-pro/collaboration-client/lib/facade";
import "@univerjs-pro/collaboration-client-ui/lib/facade";
export * from "@univerjs-pro/edit-history-viewer";
function u(e) {
  const o = new URL(e, window.location.origin);
  switch (o.protocol) {
    case "https:":
      o.protocol = "wss:";
      break;
    case "http:":
      o.protocol = "ws:";
      break;
  }
  return o.toString();
}
function h(e = {}) {
  const {
    universerEndpoint: o,
    univerContainerId: i = "app",
    enableOfflineEditing: t = !0,
    enableSingleActiveInstanceLock: n = !0
  } = e, r = o != null ? o : `${window.location.protocol}//${window.location.host}`;
  return {
    plugins: [
      l,
      [s, {
        socketService: a,
        enableOfflineEditing: t,
        enableSingleActiveInstanceLock: n,
        enableAuthServer: !0,
        authzUrl: `${r}/universer-api/authz`,
        snapshotServerUrl: `${r}/universer-api/snapshot`,
        collabSubmitChangesetUrl: `${r}/universer-api/comb`,
        collabWebSocketUrl: u(`${r}/universer-api/comb/connect`),
        loginUrlKey: `${r}/universer-api/oidc/authpage`,
        uploadFileServerUrl: `${r}/universer-api/stream/file/upload`,
        signUrlServerUrl: `${r}/universer-api/file/{fileID}/sign-url`,
        downloadEndpointUrl: `${r}/`,
        wsSessionTicketUrl: `${r}/universer-api/user/session-ticket`,
        sendChangesetTimeout: 200
      }],
      [p, {}],
      [c, {
        univerContainerId: i,
        historyListServerUrl: `${r}/universer-api/history`
      }]
    ]
  };
}
export {
  h as UniverSheetsCollaborationPreset
};
