// @svar-ui/react-menu/dist/index.es.js
import { jsxs as N, jsx as o, Fragment as L } from "react/jsx-runtime";
import { useRef as I, useCallback as v, useMemo as Z, useState as p, useEffect as $, Fragment as Y, forwardRef as S, useImperativeHandle as E } from "react";
import { calculatePosition as j, clickOutside as B, id as T } from "@svar-ui/lib-dom";
import { Portal as F } from "@svar-ui/react-core";
function K(r, n) {
  return r.map((t) => {
    const e = n(t);
    return t.data && t.data.length && (e.data = K(t.data, n)), e;
  });
}
function q(r, n) {
  const t = [];
  return r.forEach((e) => {
    if (e.data) {
      const a = q(e.data, n);
      a.length && t.push({ ...e, data: a });
    } else
      n(e) && t.push(e);
  }), t;
}
let W = 1;
function G(r) {
  return K(r, (n) => {
    const t = { ...n, id: n.id || W++ };
    return t.type && (t.comp = t.type), t;
  });
}
const U = {};
function J(r) {
  return U[r] || r;
}
function st(r, n) {
  U[r] = n;
}
function _({ onClick: r, onShow: n, option: t }) {
  const e = I(null), a = v(() => {
    n(t.data ? t.id : !1, e.current);
  }, [n, t]), f = Z(() => t && t.comp ? J(t.comp) : null, [t]);
  return /* @__PURE__ */ N(
    "div",
    {
      ref: e,
      className: `wx-cDCz9rZQ wx-option ${t.css || ""} ${t.disabled ? "wx-disabled" : ""}`,
      "data-id": t.id,
      onMouseEnter: a,
      onClick: r,
      children: [
        t.icon ? /* @__PURE__ */ o("i", { className: `wx-cDCz9rZQ wx-icon ${t.icon}` }) : null,
        t.comp ? f ? /* @__PURE__ */ o(f, { item: t, option: t }) : null : /* @__PURE__ */ N("span", { className: "wx-cDCz9rZQ wx-value", children: [
          " ",
          t.text,
          " "
        ] }),
        t.subtext ? /* @__PURE__ */ o("span", { className: "wx-cDCz9rZQ wx-subtext", children: t.subtext }) : null,
        t.data ? /* @__PURE__ */ o("i", { className: "wx-cDCz9rZQ wx-sub-icon wxi-angle-right" }) : null
      ]
    }
  );
}
function H({
  options: r,
  left: n = 0,
  top: t = 0,
  at: e = "bottom",
  parent: a = null,
  mount: f = null,
  context: x = null,
  css: d = "",
  onClick: c
}) {
  const [m, w] = p(-1e4), [h, C] = p(-1e4), [g, s] = p(20), [l, b] = p(), P = I(null), [Q, z] = p(!1), [R, A] = p(null), D = v(() => {
    const i = j(P.current, a, e, n, t);
    i && (w(i.x), C(i.y), s(i.z), b(i.width));
  }, [a, e, n, t]);
  $(() => {
    f && f(D);
  }, []);
  const u = v(() => {
    z(!1);
  }, []), k = v(() => {
    c && c({ action: null, option: null });
  }, [c]), M = v((i, O) => {
    z(i), A(O);
  }, []), y = Z(() => G(r), [r]);
  return $(() => {
    D();
  }, [a, D]), $(() => {
    if (P.current)
      return B(P.current, { callback: k, modal: !0 }).destroy;
  }, [k]), /* @__PURE__ */ o(
    "div",
    {
      ref: P,
      "data-wx-menu": "true",
      className: `wx-XMmAGqVx wx-menu ${d}`,
      style: {
        position: "absolute",
        top: h + "px",
        left: m + "px",
        width: l,
        zIndex: g
      },
      onMouseLeave: u,
      children: y.map((i) => /* @__PURE__ */ N(Y, { children: [
        i.comp === "separator" ? /* @__PURE__ */ o("div", { className: "wx-XMmAGqVx wx-separator" }) : /* @__PURE__ */ o(
          _,
          {
            option: i,
            onShow: M,
            onClick: (O) => {
              if (!i.data && !O.defaultPrevented) {
                const X = { context: x, action: i, option: i, event: O };
                i.handler && i.handler(X), c && c(X), O.stopPropagation();
              }
            }
          }
        ),
        i.data && Q === i.id ? /* @__PURE__ */ o(
          H,
          {
            css: d,
            options: i.data,
            at: "right-overlap",
            parent: R,
            context: x,
            onClick: c
          }
        ) : null
      ] }, i.id))
    }
  );
}
const V = S(function(n, t) {
  const {
    options: e,
    at: a = "bottom",
    resolver: f = null,
    dataKey: x = "contextId",
    filter: d = null,
    css: c = "",
    children: m,
    onClick: w
  } = n, [h, C] = p(null), [g, s] = p(null), [l, b] = p(0), [P, Q] = p(0), z = Z(() => h !== null && d ? q(e, (u) => d(u, h)) : e, [h, d, e]), R = v(
    (u) => {
      s(null), w && w(u);
    },
    [w]
  ), A = v((u, k) => {
    let M = null;
    for (; u && u.dataset && !M; )
      M = u.dataset[k], u = u.parentNode;
    return M ? T(M) : null;
  }, []), D = v(
    (u, k) => {
      if (!u) {
        s(null);
        return;
      }
      if (u.defaultPrevented) return;
      const M = u.target;
      if (M && M.dataset && M.dataset.menuIgnore) return;
      b(u.clientX + 1), Q(u.clientY + 1);
      let y = typeof k < "u" ? k : A(M, x);
      f && (y = f(y, u), !y) || (C(y), s(M), u.preventDefault());
    },
    [x, A, f]
  );
  return E(t, () => ({ show: D }), [D]), /* @__PURE__ */ N(L, { children: [
    m ? /* @__PURE__ */ o("span", { onClick: D, "data-menu-ignore": "true", children: typeof m == "function" ? m() : m }) : null,
    g ? /* @__PURE__ */ o(F, { children: /* @__PURE__ */ o(
      H,
      {
        css: c,
        at: a,
        top: P,
        left: l,
        parent: g,
        context: h,
        onClick: R,
        options: z
      },
      g
    ) }) : null
  ] });
});
function rt(r) {
  const { css: n = "", menuCss: t = "", options: e, onClick: a } = r, f = Z(() => G(e), [e]), [x, d] = p(!1), [c, m] = p([]), w = I(null);
  function h(s) {
    d(null), a && a(s);
  }
  function C(s, l, b) {
    l.data && l.data.length ? x && b ? d(null) : (m(l.data), d(l.id), w.current.show(s, l)) : (w.current.show(null), b ? (a && a({ action: l, option: l }), d(null)) : d(-1));
  }
  function g(s, l) {
    x && C(s, l);
  }
  return /* @__PURE__ */ N(L, { children: [
    /* @__PURE__ */ o("div", { className: `wx-UfhPCLL4 wx-menubar ${n}`, children: f.map((s) => /* @__PURE__ */ o(
      "button",
      {
        className: `wx-UfhPCLL4 wx-option ${x === s.id ? "wx-active" : ""} ${s.disabled ? "wx-disabled" : ""}`,
        onMouseEnter: (l) => g(l, s),
        onClick: (l) => C(l, s, !0),
        children: s.text
      },
      s.id
    )) }),
    /* @__PURE__ */ o(
      V,
      {
        css: t,
        onClick: h,
        options: c,
        ref: w
      }
    )
  ] });
}
const at = S(function(n, t) {
  const { options: e, at: a = "bottom", css: f = "", children: x, onClick: d } = n, [c, m] = p(null);
  function w(l) {
    m(null), d && d(l);
  }
  const h = v((l) => {
    m(l.target), l.preventDefault();
  }, []);
  E(t, () => ({ show: h }), [h]);
  function C(l) {
    let b = l.target;
    for (; !b.dataset.menuIgnore; )
      m(b), b = b.parentNode;
  }
  const g = I(0), s = I(c);
  return $(() => {
    s.current !== c && (g.current += 1, s.current = c);
  }, [c]), /* @__PURE__ */ N(L, { children: [
    /* @__PURE__ */ o("span", { onClick: C, "data-menu-ignore": "true", children: x }),
    c ? /* @__PURE__ */ o(F, { children: /* @__PURE__ */ o(
      H,
      {
        css: f,
        at: a,
        parent: c,
        options: e,
        onClick: w
      },
      g.current
    ) }) : null
  ] });
}), ct = S(function(n, t) {
  const {
    options: e,
    at: a = "bottom",
    resolver: f = null,
    dataKey: x = "contextId",
    filter: d = null,
    css: c = "",
    children: m,
    onClick: w
  } = n, h = I(null), C = v((g, s) => {
    h.current.show(g, s);
  }, []);
  return E(
    t,
    () => ({
      show: C
    }),
    [C]
  ), /* @__PURE__ */ N(L, { children: [
    m ? /* @__PURE__ */ o("span", { onContextMenu: C, "data-menu-ignore": "true", children: m }) : null,
    /* @__PURE__ */ o(
      V,
      {
        css: c,
        at: a,
        options: e,
        resolver: f,
        dataKey: x,
        filter: d,
        ref: h,
        onClick: w
      }
    )
  ] });
});
export {
  V as ActionMenu,
  ct as ContextMenu,
  at as DropDownMenu,
  H as Menu,
  rt as MenuBar,
  st as registerMenuItem
};
