// @univerjs/sheets-formula-ui/lib/index.js
var Zr = Object.defineProperty;
var zr = (e, t, n) => t in e ? Zr(e, t, { enumerable: !0, configurable: !0, writable: !0, value: n }) : e[t] = n;
var j = (e, t, n) => zr(e, typeof t != "symbol" ? t + "" : t, n);
import { CommandType as Oe, ICommandService as me, createIdentifier as Xr, IContextService as ln, IUniverInstanceService as ne, Rectangle as nr, DOCS_NORMAL_EDITOR_UNIT_ID_KEY as un, DOCS_FORMULA_BAR_EDITOR_UNIT_ID_KEY as rr, DEFAULT_EMPTY_DOCUMENT_VALUE as or, isRealNum as Qr, CellValueType as wn, getCellValueType as Jr, Inject as q, Disposable as Ke, ObjectMatrix as Fe, Range as eo, Tools as dn, LocaleService as dt, isICellData as to, isFormulaString as Ae, isFormulaId as lt, generateRandomId as Kt, Direction as de, Injector as Rt, UniverInstanceType as W, ThemeService as It, ILogService as no, toDisposable as sr, ColorKit as hn, RxDisposable as ro, InterceptorEffectEnum as oo, FOCUSING_DOC as so, FOCUSING_UNIVER_EDITOR as io, DisposableCollection as qe, RANGE_TYPE as ve, getBodySlice as kn, EDITOR_ACTIVATED as An, createInternalEditorID as co, BuildTextUtils as ao, IConfigService as ir, RichTextBuilder as lo, DependentOn as uo, Plugin as ho, merge as fo, registerDependencies as go, touchDependencies as mo } from "@univerjs/core";
import { SheetPasteCommand as po, PREDEFINED_HOOK_NAME as ct, IEditorBridgeService as fn, SetCellEditVisibleOperation as cr, HoverManagerService as So, CellAlertManagerService as vo, CellAlertType as Co, IAutoFillService as _o, APPLY_TYPE as Ro, DATA_TYPE as Fn, ISheetClipboardService as Io, COPY_TYPE as ar, SheetSkeletonManagerService as Et, attachSelectionWithCoord as rn, SelectionControl as lr, SELECTION_SHAPE_DEPTH as Eo, useActiveWorkbook as bo, getCurrentRangeDisable$ as Ye, PASTE_SPECIAL_MENU_ID as yo, whenFormulaEditorActivated as bt, whenSheetEditorFocused as To, SheetsUIPart as No, BaseSelectionRenderService as Oo, getCoordByOffset as Dn, checkInHeaderRanges as $n, getAllSelection as xo, genNormalSelectionStyle as ur, getSheetObject as Mo, MoveSelectionCommand as Ln, JumpOver as Pn, ExpandSelectionCommand as Un, EMBEDDING_FORMULA_EDITOR as wo, IMarkSelectionService as ko, RANGE_SELECTOR_COMPONENT_KEY as Ao, EMBEDDING_FORMULA_EDITOR_COMPONENT_KEY as Fo } from "@univerjs/sheets-ui";
import { sequenceNodeType as Q, serializeRange as Ce, FormulaDataModel as yt, LexerTreeBuilder as _e, ErrorType as ue, extractFormulaError as dr, SetFormulaCalculationResultMutation as Do, SetArrayFormulaDataMutation as $o, SetFormulaCalculationStopMutation as Lo, FunctionType as hr, matchToken as je, deserializeRangeWithSheetWithCache as Po, matchRefDrawToken as Uo, isFormulaLexerToken as Vo, deserializeRangeWithSheet as ht, serializeRangeToRefString as Wo, serializeRangeWithSheet as Ct, serializeRangeWithSpreadsheet as Ho, generateStringWithSequence as Bo, operatorToken as fr, UniverFormulaEnginePlugin as qo } from "@univerjs/engine-formula";
import { Subject as mt, debounceTime as gn, combineLatestWith as jo, map as gr, switchMap as Ko, of as Vn, Observable as Yo, BehaviorSubject as mr, throttleTime as Go, filter as pr, distinctUntilChanged as Zo, merge as zo } from "rxjs";
import { IEditorService as Ge, DocSelectionRenderService as Sr, ReplaceTextRunsCommand as Wn, MoveSelectionOperation as Xo, MoveCursorOperation as Qo, useKeyboardEvent as Jo, useResize as es, DocBackScrollRenderController as ts, RichTextEditor as ns } from "@univerjs/docs-ui";
import { DeviceInputEventType as xe, IRenderManagerService as Re, ScrollTimerType as Qt, SHEET_VIEWPORT_KEY as Hn, Vector2 as Bn } from "@univerjs/engine-render";
import { SheetsSelectionsService as mn, getSheetCommandTarget as vr, getCellAtRowCol as rs, SetSelectionsOperation as pn, SheetInterceptorService as Sn, ReorderRangeCommand as os, SetRangeValuesMutation as Dt, SetRangeValuesUndoMutationFactory as ss, BEFORE_CELL_EDIT as is, SetWorksheetRowAutoHeightMutation as cs, INTERCEPTOR_POINT as as, WorksheetSetCellValuePermission as Ze, WorksheetEditPermission as ze, RangeProtectionPermissionEditPoint as Xe, WorkbookEditablePermission as Qe, IRefSelectionsService as Yt, SelectionMoveType as Jt, convertSelectionDataToRange as ls, setEndForRange as us, REF_SELECTIONS_ENABLED as qn, SetWorksheetActiveOperation as jn } from "@univerjs/sheets";
import { InsertFunctionCommand as ds, TriggerCalculationController as hs, IDescriptionService as Gt, QuickSumCommand as fs, ImageFormulaCellInterceptorController as gs, UniverSheetsFormulaPlugin as ms } from "@univerjs/sheets-formula";
import { ISidebarService as vn, IZenZoneService as ps, useDependency as T, useObservable as Se, ProgressBar as Ss, MenuItemType as Je, getMenuHiddenObservable as ft, IClipboardInterfaceService as Kn, RibbonFormulasGroup as Yn, KeyCode as $, MetaKeys as P, IMenuManagerService as vs, IShortcutService as Zt, IUIPartsService as Cr, ComponentManager as _r, connectInjector as Rr, useEvent as te, RectPopup as on, IContextMenuService as Cs, useUpdateEffect as _s, BuiltInUIPart as Rs } from "@univerjs/ui";
import { jsx as M, jsxs as H, Fragment as Is } from "react/jsx-runtime";
import { useCallback as $t, useState as V, useRef as z, createElement as et, forwardRef as $e, useEffect as L, useMemo as ie, useLayoutEffect as Lt, useImperativeHandle as Es } from "react";
import { clsx as re, scrollbarClassName as ut, borderLeftClassName as bs, Select as ys, Input as Ir, borderClassName as Cn, Button as at, borderTopClassName as Ts, Tooltip as Ns, Dialog as Os } from "@univerjs/design";
import { DocSelectionManagerService as Er } from "@univerjs/docs";
import { debounceTime as xs } from "rxjs/operators";
const _n = {
  id: "sheet.command.paste-formula",
  type: Oe.COMMAND,
  handler: async (e) => e.get(me).executeCommand(po.id, {
    value: ct.SPECIAL_PASTE_FORMULA
  })
}, gt = {
  id: "formula-ui.operation.select-editor-formula",
  type: Oe.OPERATION,
  handler: (e, t) => !0
};
var Ms = Object.getOwnPropertyDescriptor, ws = (e, t, n, o) => {
  for (var r = o > 1 ? void 0 : o ? Ms(t, n) : t, s = e.length - 1, i; s >= 0; s--)
    (i = e[s]) && (r = i(r) || r);
  return r;
}, ks = (e, t) => (n, o) => t(n, o, e);
const As = "FORMULA_PROMPT_ACTIVATED", zt = Xr("formula-ui.prompt-service");
let sn = class {
  constructor(e) {
    j(this, "_search$", new mt());
    j(this, "_help$", new mt());
    j(this, "_navigate$", new mt());
    j(this, "_accept$", new mt());
    j(this, "_acceptFormulaName$", new mt());
    j(this, "search$", this._search$.asObservable());
    j(this, "help$", this._help$.asObservable());
    j(this, "navigate$", this._navigate$.asObservable());
    j(this, "accept$", this._accept$.asObservable());
    j(this, "acceptFormulaName$", this._acceptFormulaName$.asObservable());
    j(this, "_searching", !1);
    j(this, "_helping", !1);
    j(this, "_sequenceNodes", []);
    j(this, "_isLockedOnSelectionChangeRefString", !1);
    j(this, "_isLockedOnSelectionInsertRefString", !1);
    this._contextService = e;
  }
  dispose() {
    this._search$.complete(), this._help$.complete(), this._navigate$.complete(), this._accept$.complete(), this._acceptFormulaName$.complete(), this._sequenceNodes = [];
  }
  search(e) {
    this._contextService.setContextValue(As, e.visible), this._searching = e.visible, this._search$.next(e);
  }
  isSearching() {
    return this._searching;
  }
  help(e) {
    this._helping = e.visible, this._help$.next(e);
  }
  isHelping() {
    return this._helping;
  }
  navigate(e) {
    this._navigate$.next(e);
  }
  accept(e) {
    this._accept$.next(e);
  }
  acceptFormulaName(e) {
    this._acceptFormulaName$.next(e);
  }
  getSequenceNodes() {
    return [...this._sequenceNodes];
  }
  setSequenceNodes(e) {
    this._sequenceNodes = e;
  }
  clearSequenceNodes() {
    this._sequenceNodes = [];
  }
  getCurrentSequenceNode(e) {
    return this._sequenceNodes[this.getCurrentSequenceNodeIndex(e)];
  }
  getCurrentSequenceNodeByIndex(e) {
    return this._sequenceNodes[e];
  }
  /**
   * Query the text coordinates in the sequenceNodes and determine the actual insertion index.
   * @param strIndex
   */
  getCurrentSequenceNodeIndex(e) {
    let t = 0;
    const n = this._sequenceNodes[0];
    for (let o = 0, r = this._sequenceNodes.length; o < r; o++) {
      const s = this._sequenceNodes[o];
      if (typeof s == "string")
        t++;
      else {
        const { endIndex: i } = s;
        t = i;
      }
      if (e <= t)
        return typeof n == "string" && e !== 0 ? o + 1 : o;
    }
    return this._sequenceNodes.length;
  }
  /**
   * Synchronize the reference text based on the changes of the selection.
   * @param nodeIndex
   * @param refString
   */
  updateSequenceRef(e, t) {
    const n = this._sequenceNodes[e];
    if (typeof n == "string" || n.nodeType !== Q.REFERENCE)
      return;
    const o = t.length - n.token.length, r = { ...n };
    r.token = t, r.endIndex += o, this._sequenceNodes[e] = r;
    for (let s = e + 1, i = this._sequenceNodes.length; s < i; s++) {
      const c = this._sequenceNodes[s];
      if (typeof c == "string")
        continue;
      const l = { ...c };
      l.startIndex += o, l.endIndex += o, this._sequenceNodes[s] = l;
    }
  }
  /**
   * When the cursor is on the right side of a formula token,
   * you can add reference text to the formula by drawing a selection.
   * @param index
   * @param refString
   */
  insertSequenceRef(e, t) {
    const n = t.length, o = this.getCurrentSequenceNodeIndex(e);
    this._sequenceNodes.splice(o, 0, {
      token: t,
      startIndex: e,
      endIndex: e + n - 1,
      nodeType: Q.REFERENCE
    });
    for (let r = o + 1, s = this._sequenceNodes.length; r < s; r++) {
      const i = this._sequenceNodes[r];
      if (typeof i == "string")
        continue;
      const c = { ...i };
      c.startIndex += n, c.endIndex += n, this._sequenceNodes[r] = c;
    }
  }
  /**
   * Insert a string at the cursor position in the text corresponding to the sequenceNodes.
   * @param index
   * @param content
   */
  insertSequenceString(e, t) {
    const n = this.getCurrentSequenceNodeIndex(e), o = t.split("");
    this._sequenceNodes.splice(n, 0, ...o);
    const r = o.length;
    for (let s = n + r, i = this._sequenceNodes.length; s < i; s++) {
      const c = this._sequenceNodes[s];
      if (typeof c == "string")
        continue;
      const l = { ...c };
      l.startIndex += r, l.endIndex += r, this._sequenceNodes[s] = l;
    }
  }
  enableLockedSelectionChange() {
    this._isLockedOnSelectionChangeRefString = !0;
  }
  disableLockedSelectionChange() {
    this._isLockedOnSelectionChangeRefString = !1;
  }
  isLockedSelectionChange() {
    return this._isLockedOnSelectionChangeRefString;
  }
  enableLockedSelectionInsert() {
    this._isLockedOnSelectionInsertRefString = !0;
  }
  disableLockedSelectionInsert() {
    this._isLockedOnSelectionInsertRefString = !1;
  }
  isLockedSelectionInsert() {
    return this._isLockedOnSelectionInsertRefString;
  }
};
sn = ws([
  ks(0, ln)
], sn);
const Fs = {
  id: "formula-ui.operation.help-function",
  type: Oe.OPERATION,
  handler: async (e, t) => (e.get(zt).help(t), !0)
}, pe = {
  id: "formula-ui.operation.insert-function",
  type: Oe.OPERATION,
  // eslint-disable-next-line
  handler: async (e, t) => {
    var C, _;
    const n = e.get(mn), o = e.get(Ge), r = n.getCurrentSelections();
    if (!r || !r.length)
      return !1;
    const s = vr(e.get(ne));
    if (!s) return !1;
    const { worksheet: i, unitId: c, subUnitId: l } = s, m = i.getCellMatrix(), { value: h } = t, f = e.get(me);
    e.get(fn);
    const d = [], u = [];
    let a = null, v = 0, p = 0, S = "";
    if (r.length === 1 && (Ls(r[0].range) || Ps(r[0].range) && Zn(m, r[0].range))) {
      const { range: E, primary: g } = r[0], b = (C = g == null ? void 0 : g.actualRow) != null ? C : E.startRow, y = (_ = g == null ? void 0 : g.actualColumn) != null ? _ : E.startColumn;
      a = E, v = b, p = y;
      const A = Gn(m, b, y);
      A && (S = Ce(A));
    } else
      r.some((E) => {
        var y, A;
        const { range: g, primary: b } = E;
        if (Zn(m, g)) {
          const F = (y = b == null ? void 0 : b.actualRow) != null ? y : g.startRow, N = (A = b == null ? void 0 : b.actualColumn) != null ? A : g.startColumn, O = Gn(m, F, N);
          if (!O)
            return a = g, v = F, p = N, !0;
          const k = Ce(O), I = `=${h}(${k})`;
          d.push({
            range: g,
            primary: {
              row: F,
              column: N
            },
            formula: I
          });
        } else {
          const { startRow: F, startColumn: N, endRow: O, endColumn: k } = g;
          if (F === O) {
            const I = Us(m, F, k, i.getColumnCount() - 1), x = I === k ? k - 1 : k, R = Ce({
              startRow: F,
              endRow: O,
              startColumn: N,
              endColumn: x
            }), w = `=${h}(${R})`;
            u.push({
              range: g,
              primary: {
                row: F,
                column: I
              },
              formula: w
            });
          } else {
            let I = -1;
            for (let R = N; R <= k; R++) {
              const w = Vs(m, R, O, i.getRowCount() - 1);
              I = Math.max(I, w);
            }
            const x = I === O ? O - 1 : O;
            for (let R = N; R <= k; R++) {
              const w = Ce({
                startRow: F,
                endRow: x,
                startColumn: R,
                endColumn: R
              }), D = `=${h}(${w})`;
              u.push({
                range: g,
                primary: {
                  row: I,
                  column: R
                },
                formula: D
              });
            }
          }
        }
        return !1;
      });
    if (a) {
      const E = rs(v, p, i), g = {
        range: nr.clone(a),
        primary: {
          startRow: E.startRow,
          startColumn: E.startColumn,
          endRow: E.endRow,
          endColumn: E.endColumn,
          actualRow: v,
          actualColumn: p,
          isMerged: E.isMerged,
          isMergedMainCell: E.startRow === v && E.startColumn === p
        }
      }, b = {
        unitId: c,
        subUnitId: l,
        selections: [g]
      };
      await f.executeCommand(pn.id, b);
      const y = o.getEditor(un), A = o.getEditor(rr);
      f.syncExecuteCommand(cr.id, {
        visible: !0,
        unitId: c,
        eventType: xe.Dblclick
      });
      const F = `=${h}(${S}`;
      y == null || y.replaceText(F), A == null || A.replaceText(F, !1);
    }
    return d.length === 0 && u.length === 0 ? !1 : f.executeCommand(ds.id, {
      list: d,
      listOfRangeHasNumber: u
    });
  }
};
function Gn(e, t, n) {
  const o = Ds(e, t, n);
  if (o !== t)
    return {
      startRow: o,
      endRow: t - 1,
      startColumn: n,
      endColumn: n
    };
  const r = $s(e, t, n);
  return r !== n ? {
    startRow: t,
    endRow: t,
    startColumn: r,
    endColumn: n - 1
  } : null;
}
function Ds(e, t, n) {
  let o = !1;
  if (t === 0) return t;
  for (let r = t - 1; r >= 0; r--) {
    const s = e.getValue(r, n);
    if (_t(s) && !o) {
      if (r === 0) return 0;
      o = !0;
    } else {
      if (o && !_t(s))
        return r + 1;
      if (o && r === 0)
        return 0;
    }
  }
  return t;
}
function $s(e, t, n) {
  let o = !1;
  if (n === 0) return n;
  for (let r = n - 1; r >= 0; r--) {
    const s = e.getValue(t, r);
    if (_t(s) && !o) {
      if (r === 0) return 0;
      o = !0;
    } else {
      if (o && !_t(s))
        return r + 1;
      if (o && r === 0)
        return 0;
    }
  }
  return n;
}
function _t(e) {
  if (e != null && e.p) {
    const t = e == null ? void 0 : e.p.body;
    if (t == null)
      return !1;
    const n = t.dataStream, r = n.substring(n.length - 2, n.length) === or ? n.substring(0, n.length - 2) : n;
    return Qr(r);
  }
  return e && (e.t === wn.NUMBER || Jr(e) === wn.NUMBER);
}
function Ls(e) {
  return e.startRow === e.endRow && e.startColumn === e.endColumn;
}
function Ps(e) {
  return e.startRow !== e.endRow && e.startColumn !== e.endColumn;
}
function Zn(e, t) {
  for (let n = t.startRow; n <= t.endRow; n++)
    for (let o = t.startColumn; o <= t.endColumn; o++)
      if (_t(e.getValue(n, o)))
        return !1;
  return !0;
}
function Us(e, t, n, o) {
  for (let r = n; r <= o; r++)
    if (!e.getValue(t, r))
      return r;
  return o;
}
function Vs(e, t, n, o) {
  for (let r = n; r <= o; r++)
    if (!e.getValue(r, t))
      return r;
  return o;
}
const br = "SHEET_FORMULA_UI_PLUGIN", yr = `${br}_MORE_FUNCTIONS_COMPONENT`, Rn = {
  id: "formula-ui.operation.more-functions",
  type: Oe.OPERATION,
  handler: async (e) => (e.get(vn).open({
    header: { title: "formula.insert.tooltip" },
    children: { label: yr }
  }), !0)
}, Tr = {
  id: "formula-ui.operation.change-ref-to-absolute",
  type: Oe.OPERATION,
  handler: async (e) => !0
}, Ws = {
  id: "formula-ui.operation.search-function",
  type: Oe.OPERATION,
  handler: async (e, t) => (e.get(zt).search(t), !0)
};
var Hs = Object.getOwnPropertyDescriptor, Bs = (e, t, n, o) => {
  for (var r = o > 1 ? void 0 : o ? Hs(t, n) : t, s = e.length - 1, i; s >= 0; s--)
    (i = e[s]) && (r = i(r) || r);
  return r;
}, kt = (e, t) => (n, o) => t(n, o, e);
let Pt = class extends Ke {
  constructor(e, t, n, o) {
    super(), this._sheetInterceptorService = e, this._univerInstanceService = t, this._formulaDataModel = n, this._lexerTreeBuilder = o, this._initialize();
  }
  _initialize() {
    this.disposeWithMe(this._sheetInterceptorService.interceptCommand({
      getMutations: (e) => e.id === os.id ? this._reorderFormula(e.params) : {
        redos: [],
        undos: []
      }
    }));
  }
  _reorderFormula(e) {
    const t = [], n = [], { unitId: o, subUnitId: r, range: s, order: i } = e, c = this._univerInstanceService.getUniverSheetInstance(o), l = c == null ? void 0 : c.getSheetBySheetId(r);
    if (!l)
      return {
        redos: t,
        undos: n
      };
    const m = l.getCellMatrix(), h = new Fe(), f = new Fe();
    let d = !1;
    return eo.foreach(s, (u, a) => {
      let v = u;
      i.hasOwnProperty(u) && (v = i[u]);
      const p = m.getValue(v, a);
      if (p != null && p.f || p != null && p.si) {
        d = !0;
        const S = this._formulaDataModel.getFormulaStringByCell(v, a, r, o), C = this._lexerTreeBuilder.moveFormulaRefOffset(
          S,
          0,
          u - v
        ), _ = dn.deepClone(p);
        _.f = C, _.si = null, h.setValue(u, a, _);
      } else
        h.setValue(u, a, p);
      f.setValue(u, a, m.getValue(u, a));
    }), d ? (t.push({
      id: Dt.id,
      params: {
        unitId: o,
        subUnitId: r,
        cellValue: h.getMatrix()
      }
    }), n.push({
      id: Dt.id,
      params: {
        unitId: o,
        subUnitId: r,
        cellValue: f.getMatrix()
      }
    }), {
      redos: t,
      undos: n
    }) : {
      redos: t,
      undos: n
    };
  }
};
Pt = Bs([
  kt(0, q(Sn)),
  kt(1, q(ne)),
  kt(2, q(yt)),
  kt(3, q(_e))
], Pt);
const Nr = "sheets-formula-ui.base.config", zn = {};
var qs = Object.getOwnPropertyDescriptor, js = (e, t, n, o) => {
  for (var r = o > 1 ? void 0 : o ? qs(t, n) : t, s = e.length - 1, i; s >= 0; s--)
    (i = e[s]) && (r = i(r) || r);
  return r;
}, pt = (e, t) => (n, o) => t(n, o, e);
const en = "SHEET_FORMULA_ALERT", Ks = {
  [ue.DIV_BY_ZERO]: "divByZero",
  [ue.NAME]: "name",
  [ue.VALUE]: "value",
  [ue.NUM]: "num",
  [ue.NA]: "na",
  [ue.CYCLE]: "cycle",
  [ue.REF]: "ref",
  [ue.SPILL]: "spill",
  [ue.CALC]: "calc",
  [ue.ERROR]: "error",
  [ue.CONNECT]: "connect",
  [ue.NULL]: "null"
};
let cn = class extends Ke {
  constructor(e, t, n, o, r, s) {
    super(), this._context = e, this._hoverManagerService = t, this._cellAlertManagerService = n, this._localeService = o, this._formulaDataModel = r, this._zenZoneService = s, this._init();
  }
  _init() {
    this._initCellAlertPopup(), this._initZenService();
  }
  _initCellAlertPopup() {
    this.disposeWithMe(this._hoverManagerService.currentCell$.pipe(gn(100)).subscribe((e) => {
      var t, n, o, r, s;
      if (e) {
        const c = this._context.unit.getActiveSheet();
        if (!c) return this._hideAlert();
        const l = c.getCell(e.location.row, e.location.col), m = (r = (o = (n = (t = this._formulaDataModel.getArrayFormulaCellData()) == null ? void 0 : t[e.location.unitId]) == null ? void 0 : n[e.location.subUnitId]) == null ? void 0 : o[e.location.row]) == null ? void 0 : r[e.location.col];
        if (to(l)) {
          const h = dr(l, !!m);
          if (!h) {
            this._hideAlert();
            return;
          }
          const f = this._cellAlertManagerService.currentAlert.get(en), d = (s = f == null ? void 0 : f.alert) == null ? void 0 : s.location;
          if (d && d.row === e.location.row && d.col === e.location.col && d.subUnitId === e.location.subUnitId && d.unitId === e.location.unitId) {
            this._hideAlert();
            return;
          }
          this._cellAlertManagerService.showAlert({
            type: Co.ERROR,
            title: this._localeService.t("formula.error.title"),
            message: this._localeService.t(`formula.error.${Ks[h]}`),
            location: e.location,
            width: 200,
            height: 74,
            key: en
          });
          return;
        }
      }
      this._hideAlert();
    }));
  }
  _initZenService() {
    this.disposeWithMe(this._zenZoneService.visible$.subscribe((e) => {
      e && this._hideAlert();
    }));
  }
  _hideAlert() {
    this._cellAlertManagerService.removeAlert(en);
  }
};
cn = js([
  pt(1, q(So)),
  pt(2, q(vo)),
  pt(3, q(dt)),
  pt(4, q(yt)),
  pt(5, ps)
], cn);
var Ys = Object.getOwnPropertyDescriptor, Gs = (e, t, n, o) => {
  for (var r = o > 1 ? void 0 : o ? Ys(t, n) : t, s = e.length - 1, i; s >= 0; s--)
    (i = e[s]) && (r = i(r) || r);
  return r;
}, Xn = (e, t) => (n, o) => t(n, o, e);
let Ut = class extends Ke {
  constructor(e, t) {
    super(), this._autoFillService = e, this._lexerTreeBuilder = t, this._registerAutoFill();
  }
  _registerAutoFill() {
    const e = {
      type: Fn.FORMULA,
      priority: 1001,
      match: (t) => Ae(t == null ? void 0 : t.f) || lt(t == null ? void 0 : t.si),
      isContinue: (t, n) => t.type === Fn.FORMULA,
      applyFunctions: {
        [Ro.COPY]: (t, n, o, r, s) => {
          const { data: i, index: c } = t;
          return this._fillCopyFormula(i, n, o, c, r, s);
        }
      }
    };
    this._autoFillService.registerRule(e);
  }
  _fillCopyFormula(e, t, n, o, r, s) {
    var m, h;
    const i = zs(r), c = [], l = /* @__PURE__ */ new Map();
    for (let f = 1; f <= t; f++) {
      const d = (f - 1) % e.length, u = o[d], a = dn.deepClone(e[d]);
      if (a) {
        const v = ((m = e[d]) == null ? void 0 : m.f) || "", p = ((h = e[d]) == null ? void 0 : h.si) || "", S = Ae(v);
        if (lt(p))
          a.si = p, a.f = null, a.v = null, a.p = null, a.t = null, c.push(a);
        else if (S) {
          let _ = l.get(d);
          if (_)
            a.si = _, a.f = null, a.v = null, a.p = null, a.t = null;
          else {
            _ = Kt(6), l.set(d, _);
            const { offsetX: E, offsetY: g } = Zs(i, t, n, s, u), b = this._lexerTreeBuilder.moveFormulaRefOffset(
              v,
              E,
              g
            );
            a.si = _, a.f = b, a.v = null, a.p = null, a.t = null;
          }
          c.push(a);
        }
      }
    }
    return c;
  }
};
Ut = Gs([
  Xn(0, _o),
  Xn(1, q(_e))
], Ut);
function Zs(e, t, n, o, r) {
  const { source: s, target: i } = o, { rows: c } = i, { rows: l } = s;
  let m = 0, h = 0;
  switch (n) {
    case de.UP:
      h = c[r] - l[r];
      break;
    case de.RIGHT:
      m = e;
      break;
    case de.DOWN:
      h = c[r] - l[r];
      break;
    case de.LEFT:
      m = -e * t;
      break;
  }
  return { offsetX: m, offsetY: h };
}
function zs(e) {
  let t = 0;
  for (const n in e)
    e[n].forEach((o) => {
      t += o.data.length;
    });
  return t;
}
var Xs = Object.getOwnPropertyDescriptor, Qs = (e, t, n, o) => {
  for (var r = o > 1 ? void 0 : o ? Xs(t, n) : t, s = e.length - 1, i; s >= 0; s--)
    (i = e[s]) && (r = i(r) || r);
  return r;
}, St = (e, t) => (n, o) => t(n, o, e);
const Js = "default-paste-formula";
let Vt = class extends Ke {
  constructor(e, t, n, o, r) {
    super(), this._currentUniverSheet = e, this._lexerTreeBuilder = t, this._sheetClipboardService = n, this._injector = o, this._formulaDataModel = r, this._initialize();
  }
  _initialize() {
    this._registerClipboardHook();
  }
  _registerClipboardHook() {
    this.disposeWithMe(this._sheetClipboardService.addClipboardHook(this._pasteFormulaHook())), this.disposeWithMe(this._sheetClipboardService.addClipboardHook(this._pasteWithFormulaHook()));
  }
  _pasteFormulaHook() {
    return {
      id: ct.SPECIAL_PASTE_FORMULA,
      priority: 10,
      specialPasteInfo: { label: "specialPaste.formula" },
      onPasteCells: (e, t, n, o) => this._onPasteCells(e, t, n, o, !0)
    };
  }
  _pasteWithFormulaHook() {
    return {
      id: Js,
      priority: 10,
      onPasteCells: (e, t, n, o) => this._onPasteCells(e, t, n, o, !1)
    };
  }
  _onPasteCells(e, t, n, o, r) {
    var d;
    if ([
      ct.SPECIAL_PASTE_FORMAT,
      ct.SPECIAL_PASTE_COL_WIDTH
    ].includes(o.pasteType))
      return {
        undos: [],
        redos: []
      };
    const i = this._currentUniverSheet.getCurrentUnitForType(W.UNIVER_SHEET), c = t.unitId || i.getUnitId(), l = t.subUnitId || ((d = i.getActiveSheet()) == null ? void 0 : d.getSheetId());
    if (!c || !l)
      return {
        undos: [],
        redos: []
      };
    const m = t.range, h = n, f = {
      copyType: o.copyType || ar.COPY,
      copyRange: e == null ? void 0 : e.range,
      pasteType: o.pasteType
    };
    return this._injector.invoke((u) => ei(
      c,
      l,
      m,
      h,
      u,
      f,
      this._lexerTreeBuilder,
      this._formulaDataModel,
      r,
      e
    ));
  }
};
Vt = Qs([
  St(0, ne),
  St(1, q(_e)),
  St(2, Io),
  St(3, q(Rt)),
  St(4, q(yt))
], Vt);
function ei(e, t, n, o, r, s, i, c, l = !1, m) {
  const h = [], f = [], d = ti(e, t, n, o, s, i, c, m);
  if (!d.hasValue())
    return {
      undos: [],
      redos: []
    };
  const u = {
    unitId: e,
    subUnitId: t,
    cellValue: d.getData()
  };
  h.push({
    id: Dt.id,
    params: u
  });
  const a = ss(
    r,
    u
  );
  return f.push({
    id: Dt.id,
    params: a
  }), {
    undos: f,
    redos: h
  };
}
function ti(e, t, n, o, r, s, i, c) {
  return c ? r.pasteType === ct.SPECIAL_PASTE_VALUE ? ri(e, t, n, o, i, c) : r.pasteType === ct.SPECIAL_PASTE_FORMULA ? oi(e, t, n, o, s, i, c) : si(e, t, n, o, r.copyType, s, i, c) : ni(e, t, n, o, i);
}
function ni(e, t, n, o, r) {
  const s = new Fe(), i = r.getSheetFormulaData(e, t);
  return o.forValue((c, l, m) => {
    var u;
    const h = n.rows[c], f = n.cols[l], d = {};
    Ae(m.v) ? (d.v = null, d.f = `${m.v}`, d.si = null, d.p = null, s.setValue(h, f, d)) : (u = i == null ? void 0 : i[h]) != null && u[f] && (d.v = m.v, d.f = null, d.si = null, d.p = null, s.setValue(h, f, d));
  }), s;
}
function ri(e, t, n, o, r, s) {
  var m, h;
  const i = new Fe(), c = (h = (m = r.getArrayFormulaCellData()) == null ? void 0 : m[s.unitId]) == null ? void 0 : h[s.subUnitId], l = r.getSheetFormulaData(e, t);
  return o.forValue((f, d, u) => {
    var _, E;
    const a = s.range.rows[f % s.range.rows.length], v = s.range.cols[d % s.range.cols.length], p = n.rows[f], S = n.cols[d], C = {};
    if (Ae(u.f) || lt(u.si))
      C.v = u.v, C.f = null, C.si = null, C.p = null, i.setValue(p, S, C);
    else if ((_ = c == null ? void 0 : c[a]) != null && _[v]) {
      const g = c[a][v];
      C.v = g.v, C.f = null, C.si = null, C.p = null, i.setValue(p, S, C);
    } else if ((E = l == null ? void 0 : l[p]) != null && E[S]) {
      if (C.v = u.v, C.f = null, C.si = null, C.p = null, u.p) {
        const g = Or(u);
        g && (C.v = g);
      }
      i.setValue(p, S, C);
    }
  }), i;
}
function oi(e, t, n, o, r, s, i) {
  const c = new Fe(), l = /* @__PURE__ */ new Map();
  return o.forValue((m, h, f) => {
    const d = n.rows[m], u = n.cols[h], a = {};
    if (lt(f.si)) {
      if (i.unitId !== e || i.subUnitId !== t) {
        const v = s.getFormulaStringByCell(
          i.range.rows[m % i.range.rows.length],
          i.range.cols[h % i.range.cols.length],
          i.subUnitId,
          i.unitId
        ), p = n.cols[h] - i.range.cols[h % i.range.cols.length], S = n.rows[m] - i.range.rows[m % i.range.rows.length], C = r.moveFormulaRefOffset(v || "", p, S);
        a.si = null, a.f = C;
      } else
        a.si = f.si, a.f = null;
      a.v = null, a.p = null, c.setValue(d, u, a);
    } else if (Ae(f.f)) {
      const v = `${m % i.range.rows.length}_${h % i.range.cols.length}`;
      let p = l.get(v);
      if (p)
        a.si = p, a.f = null;
      else {
        p = Kt(6), l.set(v, p);
        const S = n.cols[h] - i.range.cols[h % i.range.cols.length], C = n.rows[m] - i.range.rows[m % i.range.rows.length], _ = r.moveFormulaRefOffset(f.f || "", S, C);
        a.si = p, a.f = _;
      }
      a.v = null, a.p = null, c.setValue(d, u, a);
    } else {
      if (a.v = f.v, a.f = null, a.si = null, a.p = null, f.p) {
        const v = Or(f);
        v && (a.v = v);
      }
      c.setValue(d, u, a);
    }
  }), c;
}
function si(e, t, n, o, r, s, i, c) {
  const l = new Fe(), m = /* @__PURE__ */ new Map(), h = i.getSheetFormulaData(e, t), f = [];
  return r === ar.CUT ? o.forValue((d, u, a) => {
    const v = n.rows[d], p = n.cols[u], S = {};
    if (lt(a.si)) {
      if (Ae(a.f))
        f.push(a.si), S.f = a.f, S.si = a.si;
      else if (f.includes(a.si))
        S.f = null, S.si = a.si;
      else {
        const C = i.getFormulaStringByCell(
          c.range.rows[d % c.range.rows.length],
          c.range.cols[u % c.range.cols.length],
          c.subUnitId,
          c.unitId
        );
        S.f = C, S.si = null;
      }
      S.v = null, S.p = null, l.setValue(v, p, S);
    } else Ae(a.f) && (S.f = a.f, S.si = null, S.v = null, S.p = null, l.setValue(v, p, S));
  }) : o.forValue((d, u, a) => {
    var C;
    const v = n.rows[d], p = n.cols[u], S = {};
    if (lt(a.si)) {
      if (c.unitId !== e || c.subUnitId !== t) {
        const _ = i.getFormulaStringByCell(
          c.range.rows[d % c.range.rows.length],
          c.range.cols[u % c.range.cols.length],
          c.subUnitId,
          c.unitId
        ), E = n.cols[u] - c.range.cols[u % c.range.cols.length], g = n.rows[d] - c.range.rows[d % c.range.rows.length], b = s.moveFormulaRefOffset(_ || "", E, g);
        S.si = null, S.f = b;
      } else
        S.si = a.si, S.f = null;
      S.v = null, S.p = null, l.setValue(v, p, S);
    } else if (Ae(a.f)) {
      const _ = `${d % c.range.rows.length}_${u % c.range.cols.length}`;
      let E = m.get(_);
      if (E)
        S.si = E, S.f = null;
      else {
        E = Kt(6), m.set(_, E);
        const g = n.cols[u] - c.range.cols[u % c.range.cols.length], b = n.rows[d] - c.range.rows[d % c.range.rows.length], y = s.moveFormulaRefOffset(a.f || "", g, b);
        S.si = E, S.f = y;
      }
      S.v = null, S.p = null, l.setValue(v, p, S);
    } else (C = h == null ? void 0 : h[v]) != null && C[p] && (S.v = a.v, S.f = null, S.si = null, S.p = a.p, l.setValue(v, p, S));
  }), f.length > 0 && new Fe(h).forValue((d, u, a) => {
    if (!(c.range.rows.includes(d) && c.range.cols.includes(u)) && !(n.rows.includes(d) && n.cols.includes(u)) && f.includes(a == null ? void 0 : a.si)) {
      const v = i.getFormulaStringByCell(
        d,
        u,
        c.subUnitId,
        c.unitId
      );
      l.setValue(d, u, {
        f: v,
        si: null,
        v: null,
        p: null
      });
    }
  }), l;
}
function Or(e) {
  if (e != null && e.p) {
    const t = e == null ? void 0 : e.p.body;
    if (t == null)
      return;
    const n = t.dataStream;
    return n.substring(n.length - 2, n.length) === or ? n.substring(0, n.length - 2) : n;
  }
}
var ii = Object.getOwnPropertyDescriptor, ci = (e, t, n, o) => {
  for (var r = o > 1 ? void 0 : o ? ii(t, n) : t, s = e.length - 1, i; s >= 0; s--)
    (i = e[s]) && (r = i(r) || r);
  return r;
}, He = (e, t) => (n, o) => t(n, o, e);
let Wt = class extends Ke {
  constructor(t, n, o, r, s, i, c, l) {
    super();
    j(this, "_previousShape");
    j(this, "_skeleton");
    this._context = t, this._sheetInterceptorService = n, this._formulaDataModel = o, this._themeService = r, this._renderManagerService = s, this._sheetSkeletonManagerService = i, this._commandService = c, this._logService = l, this._initSkeletonChangeListener(), this._initInterceptorEditorStart(), this._commandExecutedListener();
  }
  _initSkeletonChangeListener() {
    this.disposeWithMe(
      this._sheetSkeletonManagerService.currentSkeleton$.subscribe((t) => {
        var n, o;
        if (t == null)
          this._logService.debug("[FormulaEditorShowController]: should not receive currentSkeleton$ as null!");
        else {
          const { skeleton: r } = t, s = (o = (n = this._skeleton) == null ? void 0 : n.worksheet) == null ? void 0 : o.getSheetId();
          if (this._changeRuntime(r), s !== r.worksheet.getSheetId())
            this._removeArrayFormulaRangeShape();
          else {
            const { unitId: i, sheetId: c } = t;
            this._updateArrayFormulaRangeShape(i, c);
          }
        }
      })
    );
  }
  _changeRuntime(t) {
    this._skeleton = t;
  }
  _initInterceptorEditorStart() {
    this.disposeWithMe(
      sr(
        this._sheetInterceptorService.writeCellInterceptor.intercept(is, {
          handler: (t, n, o) => {
            var a, v, p, S;
            const { row: r, col: s, unitId: i, subUnitId: c, worksheet: l } = n, m = this._formulaDataModel.getArrayFormulaRange(), h = this._formulaDataModel.getArrayFormulaCellData();
            if (this._removeArrayFormulaRangeShape(), t == null)
              return o(t);
            let f = null;
            const d = this._formulaDataModel.getFormulaStringByCell(r, s, c, i);
            if (d !== null && (f = { f: d }), t.v != null && t.v !== "" && ((p = (v = (a = h[i]) == null ? void 0 : a[c]) == null ? void 0 : v[r]) == null ? void 0 : p[s]) == null)
              return f ? { ...t, ...f } : o(t);
            const u = (S = m == null ? void 0 : m[i]) == null ? void 0 : S[c];
            return u != null && (f = this._displayArrayFormulaRangeShape(u, r, s, i, c, l, f)), f ? { ...t, ...f } : o(t);
          }
        })
      )
    );
  }
  _commandExecutedListener() {
    this.disposeWithMe(this._commandService.onCommandExecuted((t, n) => {
      (t.id === Do.id || t.id === $o.id && n && n.remove) && this._removeArrayFormulaRangeShape();
    })), this.disposeWithMe(
      this._commandService.beforeCommandExecuted((t) => {
        cs.id === t.id && requestIdleCallback(() => {
          const n = t.params, { unitId: o, subUnitId: r, rowsAutoHeightInfo: s } = n;
          this._refreshArrayFormulaRangeShapeByRow(o, r, s);
        });
      })
    );
  }
  _displayArrayFormulaRangeShape(t, n, o, r, s, i, c) {
    return new Fe(t).forValue((l, m, h) => {
      if (h == null)
        return !0;
      const { startRow: f, startColumn: d, endRow: u, endColumn: a } = h;
      if (l === n && m === o)
        return this._createArrayFormulaRangeShape(h, r), !1;
      if (n >= f && n <= u && o >= d && o <= a) {
        const v = i.getCell(f, d);
        return (v == null ? void 0 : v.v) === ue.SPILL || (v == null ? void 0 : v.f) == null ? void 0 : (c == null && (c = {
          f: v.f,
          isInArrayFormulaRange: !0
        }), this._createArrayFormulaRangeShape(h, r), !1);
      }
    }), c;
  }
  _createArrayFormulaRangeShape(t, n) {
    const o = this._renderManagerService.getRenderById(n), r = this._sheetSkeletonManagerService.getCurrentSkeleton();
    if (!o || !r) return;
    const { scene: s } = o;
    if (!s) return;
    const i = {
      range: t,
      primary: null,
      style: {
        strokeWidth: 1,
        stroke: this._themeService.getColorFromTheme("primary.600"),
        fill: new hn(this._themeService.getColorFromTheme("white")).setAlpha(0).toString(),
        widgets: {}
      }
    }, c = rn(i, r), { rowHeaderWidth: l, columnHeaderHeight: m } = r, h = new lr(s, Eo.FORMULA_EDITOR_SHOW, this._themeService, {
      highlightHeader: !1,
      rowHeaderWidth: l,
      columnHeaderHeight: m
    });
    h.updateRangeBySelectionWithCoord(c), h.setEvent(!1), this._previousShape = h;
  }
  _removeArrayFormulaRangeShape() {
    this._previousShape != null && (this._previousShape.dispose(), this._previousShape = null);
  }
  _refreshArrayFormulaRangeShape(t, n) {
    if (this._previousShape) {
      const { startRow: o, endRow: r, startColumn: s, endColumn: i } = this._previousShape.getRange(), c = { startRow: o, endRow: r, startColumn: s, endColumn: i };
      this._removeArrayFormulaRangeShape(), this._createArrayFormulaRangeShape(c, t);
    }
  }
  _checkCurrentSheet(t, n) {
    const o = this._sheetSkeletonManagerService.getCurrentSkeleton();
    if (!o) return !1;
    const r = o.worksheet;
    return r ? r.unitId === t && r.getSheetId() === n : !1;
  }
  _updateArrayFormulaRangeShape(t, n) {
    this._checkCurrentSheet(t, n) && this._previousShape && this._refreshArrayFormulaRangeShape(t);
  }
  _refreshArrayFormulaRangeShapeByRow(t, n, o) {
    if (!this._checkCurrentSheet(t, n) || !this._previousShape) return;
    const { startRow: r, endRow: s, startColumn: i, endColumn: c } = this._previousShape.getRange();
    for (let l = 0; l < o.length; l++) {
      const { row: m } = o[l];
      if (r >= m) {
        const h = {
          startRow: r,
          endRow: s,
          startColumn: i,
          endColumn: c
        };
        this._refreshArrayFormulaRangeShape(t, h);
        break;
      }
    }
  }
};
Wt = ci([
  He(1, q(Sn)),
  He(2, q(yt)),
  He(3, q(It)),
  He(4, Re),
  He(5, q(Et)),
  He(6, me),
  He(7, no)
], Wt);
var ai = Object.getOwnPropertyDescriptor, li = (e, t, n, o) => {
  for (var r = o > 1 ? void 0 : o ? ai(t, n) : t, s = e.length - 1, i; s >= 0; s--)
    (i = e[s]) && (r = i(r) || r);
  return r;
}, Qn = (e, t) => (n, o) => t(n, o, e);
const ui = {
  tl: {
    size: 6,
    color: "#409f11"
  }
};
let Ht = class extends ro {
  constructor(e, t) {
    super(), this._sheetInterceptorService = e, this._formulaDataModel = t, this.disposeWithMe(this._sheetInterceptorService.intercept(
      as.CELL_CONTENT,
      {
        effect: oo.Style,
        handler: (n, o, r) => {
          var c, l, m, h;
          const s = (h = (m = (l = (c = this._formulaDataModel.getArrayFormulaCellData()) == null ? void 0 : c[o.unitId]) == null ? void 0 : l[o.subUnitId]) == null ? void 0 : m[o.row]) == null ? void 0 : h[o.col];
          return !dr(n, !!s) || !n || (n === o.rawData && (n = { ...o.rawData }), n.markers = {
            ...n == null ? void 0 : n.markers,
            ...ui
          }), r(n);
        },
        priority: 10
      }
    ));
  }
};
Ht = li([
  Qn(0, q(Sn)),
  Qn(1, q(yt))
], Ht);
function di() {
  const e = T(hs), t = T(me), n = Se(e.progress$), o = $t(() => {
    t.executeCommand(Lo.id);
  }, [t]), r = $t(() => {
    e.clearProgress();
  }, [e]);
  return /* @__PURE__ */ M(Ss, { progress: n, onTerminate: o, onClearProgress: r });
}
function hi(e, t) {
  return Object.keys(e).filter((n) => isNaN(Number(n)) && n !== "DefinedName" && n !== "Table").map((n) => ({
    label: t.t(`formula.functionType.${n.toLocaleLowerCase()}`),
    value: `${e[n]}`
  }));
}
function xr(e) {
  if (!e.require && !e.repeat)
    return `[${e.name}]`;
  if (e.require && !e.repeat)
    return e.name;
  if (!e.require && e.repeat)
    return `[${e.name},...]`;
  if (e.require && e.repeat)
    return `${e.name},...`;
}
function Mr(e) {
  const { prefix: t, value: n } = e;
  return /* @__PURE__ */ H("div", { children: [
    /* @__PURE__ */ H("span", { children: [
      t,
      "("
    ] }),
    n && n.map((o, r) => /* @__PURE__ */ H("span", { children: [
      /* @__PURE__ */ M("span", { children: xr(o) }),
      r === n.length - 1 ? "" : ","
    ] }, r)),
    ")"
  ] });
}
function vt(e) {
  const { className: t, value: n, title: o } = e;
  return /* @__PURE__ */ H("div", { className: "univer-mb-2 univer-text-xs", children: [
    /* @__PURE__ */ M(
      "div",
      {
        className: re("univer-mb-2 univer-font-medium univer-text-gray-500 dark:!univer-text-gray-300", t),
        children: o
      }
    ),
    /* @__PURE__ */ M(
      "div",
      {
        className: "univer-break-all univer-text-gray-900 dark:!univer-text-white",
        children: n
      }
    )
  ] });
}
function fi(e) {
  const { functionInfo: t, onChange: n } = e;
  if (!t) return null;
  const [o, r] = V([]), [s, i] = V(t.functionParameter), [c, l] = V(-1);
  return /* @__PURE__ */ H("div", { children: [
    /* @__PURE__ */ M("div", { className: re("univer-h-[364px] univer-overflow-y-auto", ut), children: s.map((m, h) => /* @__PURE__ */ H("div", { children: [
      /* @__PURE__ */ M("div", { className: "univer-text-sm", children: m.name }),
      /* @__PURE__ */ M("div", { className: "univer-mb-2 univer-mt-1" })
    ] }, h)) }),
    /* @__PURE__ */ M("div", { className: re("univer-flex-1 univer-p-3", bs), children: /* @__PURE__ */ M(
      vt,
      {
        title: c === -1 ? /* @__PURE__ */ M(Mr, { prefix: t.functionName, value: s }) : s[c].name,
        value: c === -1 ? t.description : s[c].detail
      }
    ) })
  ] });
}
function tt({ ref: e, ...t }) {
  const { icon: n, id: o, className: r, extend: s, ...i } = t, c = `univerjs-icon univerjs-icon-${o} ${r || ""}`.trim(), l = z(`_${pi()}`);
  return wr(n, `${o}`, {
    defIds: n.defIds,
    idSuffix: l.current
  }, {
    ref: e,
    className: c,
    ...i
  }, s);
}
function wr(e, t, n, o, r) {
  return et(e.tag, {
    key: t,
    ...gi(e, n, r),
    ...o
  }, (mi(e, n).children || []).map((s, i) => wr(s, `${t}-${e.tag}-${i}`, n, void 0, r)));
}
function gi(e, t, n) {
  const o = { ...e.attrs };
  n != null && n.colorChannel1 && o.fill === "colorChannel1" && (o.fill = n.colorChannel1), e.tag === "mask" && o.id && (o.id = o.id + t.idSuffix), Object.entries(o).forEach(([s, i]) => {
    s === "mask" && typeof i == "string" && (o[s] = i.replace(/url\(#(.*)\)/, `url(#$1${t.idSuffix})`));
  });
  const { defIds: r } = t;
  return !r || r.length === 0 || (e.tag === "use" && o["xlink:href"] && (o["xlink:href"] = o["xlink:href"] + t.idSuffix), Object.entries(o).forEach(([s, i]) => {
    typeof i == "string" && (o[s] = i.replace(/url\(#(.*)\)/, `url(#$1${t.idSuffix})`));
  })), o;
}
function mi(e, t) {
  var o;
  const { defIds: n } = t;
  return !n || n.length === 0 ? e : e.tag === "defs" && ((o = e.children) != null && o.length) ? {
    ...e,
    children: e.children.map((r) => typeof r.attrs.id == "string" && n && n.includes(r.attrs.id) ? {
      ...r,
      attrs: {
        ...r.attrs,
        id: r.attrs.id + t.idSuffix
      }
    } : r)
  } : e;
}
function pi() {
  return Math.random().toString(36).substring(2, 8);
}
tt.displayName = "UniverIcon";
const Si = {
  tag: "svg",
  attrs: {
    xmlns: "http://www.w3.org/2000/svg",
    fill: "none",
    viewBox: "0 0 16 16",
    width: "1em",
    height: "1em"
  },
  children: [{
    tag: "path",
    attrs: {
      fill: "currentColor",
      d: "M14.1544 3.75557C14.3887 3.98988 14.3887 4.36978 14.1544 4.6041L6.51409 12.2444C6.40157 12.3569 6.24896 12.4201 6.08983 12.4201C5.9307 12.4201 5.77808 12.3569 5.66556 12.2444L1.84541 8.42425C1.6111 8.18993 1.6111 7.81003 1.84541 7.57572C2.07973 7.34141 2.45963 7.34141 2.69394 7.57572L6.08983 10.9716L13.3059 3.75557C13.5402 3.52126 13.9201 3.52126 14.1544 3.75557Z",
      fillRule: "evenodd",
      clipRule: "evenodd"
    }
  }]
}, kr = $e(function(t, n) {
  return et(tt, Object.assign({}, t, {
    id: "check-mark-icon",
    ref: n,
    icon: Si
  }));
});
kr.displayName = "CheckMarkIcon";
const vi = {
  tag: "svg",
  attrs: {
    xmlns: "http://www.w3.org/2000/svg",
    fill: "none",
    viewBox: "0 0 16 16",
    width: "1em",
    height: "1em"
  },
  children: [{
    tag: "path",
    attrs: {
      fill: "currentColor",
      d: "M3.71274 2.86421C3.47843 2.6299 3.09853 2.6299 2.86421 2.86421C2.6299 3.09853 2.6299 3.47843 2.86421 3.71274L7.15154 8.00007L2.86421 12.2874C2.6299 12.5217 2.6299 12.9016 2.86421 13.1359C3.09853 13.3702 3.47843 13.3702 3.71274 13.1359L8.00007 8.84859L12.2874 13.1359C12.5217 13.3702 12.9016 13.3702 13.1359 13.1359C13.3702 12.9016 13.3702 12.5217 13.1359 12.2874L8.84859 8.00007L13.1359 3.71274C13.3702 3.47843 13.3702 3.09853 13.1359 2.86421C12.9016 2.6299 12.5217 2.6299 12.2874 2.86421L8.00007 7.15154L3.71274 2.86421Z"
    }
  }]
}, Ar = $e(function(t, n) {
  return et(tt, Object.assign({}, t, {
    id: "close-icon",
    ref: n,
    icon: vi
  }));
});
Ar.displayName = "CloseIcon";
const Ci = {
  tag: "svg",
  attrs: {
    xmlns: "http://www.w3.org/2000/svg",
    fill: "none",
    viewBox: "0 0 16 16",
    width: "1em",
    height: "1em"
  },
  children: [
    {
      tag: "path",
      attrs: {
        fill: "currentColor",
        d: "M5.3313 1.4667C5.3313 1.13533 5.59993 0.866699 5.9313 0.866699H10.069C10.4004 0.866699 10.669 1.13533 10.669 1.4667C10.669 1.79807 10.4004 2.0667 10.069 2.0667H5.9313C5.59993 2.0667 5.3313 1.79807 5.3313 1.4667Z"
      }
    },
    {
      tag: "path",
      attrs: {
        fill: "currentColor",
        d: "M1.09985 3.64443C1.09985 3.31306 1.36848 3.04443 1.69985 3.04443H14.2999C14.6312 3.04443 14.8999 3.31306 14.8999 3.64443C14.8999 3.9758 14.6312 4.24443 14.2999 4.24443H1.69985C1.36848 4.24443 1.09985 3.9758 1.09985 3.64443Z"
      }
    },
    {
      tag: "path",
      attrs: {
        fill: "currentColor",
        d: "M6.12398 8.30171C6.35829 8.0674 6.73819 8.0674 6.97251 8.30171L8.00007 9.32928L9.02764 8.30171C9.26195 8.0674 9.64185 8.0674 9.87617 8.30171C10.1105 8.53603 10.1105 8.91593 9.87617 9.15024L8.8486 10.1778L9.87617 11.2054C10.1105 11.4397 10.1105 11.8196 9.87617 12.0539C9.64185 12.2882 9.26195 12.2882 9.02764 12.0539L8.00007 11.0263L6.97251 12.0539C6.73819 12.2882 6.35829 12.2882 6.12398 12.0539C5.88966 11.8196 5.88966 11.4397 6.12398 11.2054L7.15154 10.1778L6.12398 9.15024C5.88966 8.91593 5.88966 8.53603 6.12398 8.30171Z"
      }
    },
    {
      tag: "path",
      attrs: {
        fill: "currentColor",
        d: "M4.75332 5.22217C3.86966 5.22217 3.15332 5.93851 3.15332 6.82217V12.5331C3.15332 13.9691 4.31738 15.1332 5.75332 15.1332H10.2465C11.6825 15.1332 12.8465 13.9691 12.8465 12.5331V6.82217C12.8465 5.93851 12.1302 5.22217 11.2465 5.22217H4.75332ZM4.35332 6.82217C4.35332 6.60125 4.53241 6.42217 4.75332 6.42217H11.2465C11.4674 6.42217 11.6465 6.60125 11.6465 6.82217V12.5331C11.6465 13.3063 11.0197 13.9332 10.2465 13.9332H5.75332C4.98012 13.9332 4.35332 13.3063 4.35332 12.5331V6.82217Z",
        fillRule: "evenodd",
        clipRule: "evenodd"
      }
    }
  ]
}, Fr = $e(function(t, n) {
  return et(tt, Object.assign({}, t, {
    id: "delete-icon",
    ref: n,
    icon: Ci
  }));
});
Fr.displayName = "DeleteIcon";
const _i = {
  tag: "svg",
  attrs: {
    xmlns: "http://www.w3.org/2000/svg",
    fill: "none",
    viewBox: "0 0 16 16",
    width: "1em",
    height: "1em"
  },
  children: [{
    tag: "path",
    attrs: {
      fill: "currentColor",
      d: "M8.6 1.99991C8.60001 1.66854 8.33138 1.39991 8.00001 1.3999C7.66864 1.3999 7.40001 1.66853 7.4 1.9999L7.39996 7.3999H1.9999C1.66853 7.3999 1.3999 7.66853 1.3999 7.9999C1.3999 8.33127 1.66853 8.5999 1.9999 8.5999H7.39995L7.3999 13.9999C7.3999 14.3313 7.66853 14.5999 7.9999 14.5999C8.33127 14.5999 8.5999 14.3313 8.5999 13.9999L8.59995 8.5999H13.9999C14.3313 8.5999 14.5999 8.33127 14.5999 7.9999C14.5999 7.66853 14.3313 7.3999 13.9999 7.3999H8.59996L8.6 1.99991Z"
    }
  }]
}, Dr = $e(function(t, n) {
  return et(tt, Object.assign({}, t, {
    id: "increase-icon",
    ref: n,
    icon: _i
  }));
});
Dr.displayName = "IncreaseIcon";
const Ri = {
  tag: "svg",
  attrs: {
    xmlns: "http://www.w3.org/2000/svg",
    fill: "none",
    viewBox: "0 0 16 16",
    width: "1em",
    height: "1em"
  },
  children: [{
    tag: "path",
    attrs: {
      fill: "currentColor",
      d: "M5.90913 3.57564C6.14345 3.34132 6.52335 3.34132 6.75766 3.57564L10.7577 7.57564C10.992 7.80995 10.992 8.18985 10.7577 8.42417L6.75766 12.4242C6.52335 12.6585 6.14345 12.6585 5.90913 12.4242C5.67482 12.1899 5.67482 11.81 5.90913 11.5756L9.48487 7.9999L5.90913 4.42417C5.67482 4.18985 5.67482 3.80995 5.90913 3.57564Z",
      fillRule: "evenodd",
      clipRule: "evenodd"
    }
  }]
}, $r = $e(function(t, n) {
  return et(tt, Object.assign({}, t, {
    id: "more-icon",
    ref: n,
    icon: Ri
  }));
});
$r.displayName = "MoreIcon";
const Ii = {
  tag: "svg",
  attrs: {
    xmlns: "http://www.w3.org/2000/svg",
    fill: "none",
    viewBox: "0 0 16 16",
    width: "1em",
    height: "1em"
  },
  children: [{
    tag: "path",
    attrs: {
      fill: "currentColor",
      d: "M12.6185 12.4423C12.5907 12.2749 12.7773 12.15 12.9343 12.2308L15.4242 13.5126C15.6102 13.6084 15.5544 13.8745 15.3439 13.8955L14.2456 14.184L13.4521 15.1286C13.3495 15.2939 13.085 15.2463 13.0534 15.0568L12.6185 12.4423Z"
    }
  }, {
    tag: "path",
    attrs: {
      fill: "currentColor",
      d: "M1 3.6C1 2.16406 2.16406 1 3.6 1H12.3C13.7359 1 14.9 2.16406 14.9 3.6V5.81156C14.9003 5.81881 14.9004 5.82609 14.9004 5.8334C14.9004 5.84071 14.9003 5.84799 14.9 5.85524V10.045C14.9003 10.0522 14.9004 10.0595 14.9004 10.0668C14.9004 10.3982 14.6318 10.6668 14.3004 10.6668H11.1668C10.8907 10.6668 10.6668 10.8907 10.6668 11.1668V14.3C10.6668 14.6314 10.3982 14.9 10.0668 14.9L10.05 14.8998L3.6 14.9C2.16406 14.9 1 13.7359 1 12.3V3.6ZM13.2 5.2334C13.4761 5.2334 13.7 5.00954 13.7 4.7334V3.6C13.7 2.8268 13.0732 2.2 12.3 2.2H11.1668C10.8907 2.2 10.6668 2.42386 10.6668 2.7V4.7334C10.6668 5.00954 10.8907 5.2334 11.1668 5.2334H13.2ZM10.6668 6.9334C10.6668 6.65726 10.8907 6.4334 11.1668 6.4334H13.2C13.4761 6.4334 13.7 6.65726 13.7 6.9334V8.9668C13.7 9.24294 13.4761 9.4668 13.2 9.4668H11.1668C10.8907 9.4668 10.6668 9.24294 10.6668 8.9668V6.9334ZM8.9668 5.2334C9.24294 5.2334 9.4668 5.00954 9.4668 4.7334V2.7C9.4668 2.42386 9.24294 2.2 8.9668 2.2H6.9334C6.65726 2.2 6.4334 2.42386 6.4334 2.7V4.7334C6.4334 5.00954 6.65726 5.2334 6.9334 5.2334L8.9668 5.2334ZM6.4334 6.9334C6.4334 6.65726 6.65726 6.4334 6.9334 6.4334L8.9668 6.4334C9.24294 6.4334 9.4668 6.65726 9.4668 6.9334V8.9668C9.4668 9.24294 9.24294 9.4668 8.9668 9.4668L6.9334 9.4668C6.65726 9.4668 6.4334 9.24294 6.4334 8.9668V6.9334ZM4.7334 5.2334C5.00954 5.2334 5.2334 5.00954 5.2334 4.7334V2.7C5.2334 2.42386 5.00954 2.2 4.7334 2.2H3.6C2.8268 2.2 2.2 2.8268 2.2 3.6V4.7334C2.2 5.00954 2.42386 5.2334 2.7 5.2334H4.7334ZM2.2 6.9334C2.2 6.65726 2.42386 6.4334 2.7 6.4334H4.7334C5.00954 6.4334 5.2334 6.65725 5.2334 6.9334V8.9668C5.2334 9.24294 5.00954 9.4668 4.7334 9.4668H2.7C2.42386 9.4668 2.2 9.24294 2.2 8.9668V6.9334ZM5.2334 11.1668C5.2334 10.8907 5.00954 10.6668 4.7334 10.6668H2.7C2.42386 10.6668 2.2 10.8907 2.2 11.1668V12.3C2.2 13.0732 2.8268 13.7 3.6 13.7H4.7334C5.00954 13.7 5.2334 13.4761 5.2334 13.2V11.1668ZM9.4668 11.1668C9.4668 10.8907 9.24294 10.6668 8.9668 10.6668H6.9334C6.65726 10.6668 6.4334 10.8907 6.4334 11.1668V13.2C6.4334 13.4761 6.65726 13.7 6.9334 13.7H8.9668C9.24294 13.7 9.4668 13.4761 9.4668 13.2V11.1668Z",
      fillRule: "evenodd",
      clipRule: "evenodd"
    }
  }]
}, Lr = $e(function(t, n) {
  return et(tt, Object.assign({}, t, {
    id: "select-range-icon",
    ref: n,
    icon: Ii
  }));
});
Lr.displayName = "SelectRangeIcon";
function Ei(e) {
  const { onChange: t } = e, n = "-1", [o, r] = V(""), [s, i] = V([]), [c, l] = V(0), [m, h] = V(n), [f, d] = V(0), [u, a] = V(null), v = T(Gt), p = T(dt), S = T(vn), C = Se(S.sidebarOptions$), _ = hi(hr, p);
  _.unshift({
    label: p.t("formula.moreFunctions.allFunctions"),
    value: n
  });
  const E = p.t("formula.prompt.required"), g = p.t("formula.prompt.optional");
  L(() => {
    A(n);
  }, []), L(() => {
    y(0);
  }, [s]), L(() => {
    C != null && C.visible && (r(""), i([]), l(0), h(n), d(0), a(null), A(n));
  }, [C]);
  const b = (I) => {
    if (o.trim() === "") return I;
    const x = new RegExp(`(${o.toLocaleUpperCase()})`);
    return I.split(x).filter(Boolean).map((w, D) => w.match(x) ? /* @__PURE__ */ M("span", { className: "univer-text-red-500", children: w }, D) : w);
  }, y = (I) => {
    if (s.length === 0) {
      a(null);
      return;
    }
    d(I);
    const x = v.getFunctionInfo(s[I].name);
    if (!x) {
      a(null);
      return;
    }
    a(x), t(x);
  };
  function A(I) {
    h(I);
    const x = v.getSearchListByType(+I);
    i(x);
  }
  function F(I) {
    r(I);
    const x = v.getSearchListByName(I);
    i(x);
  }
  function N(I) {
    if (I.stopPropagation(), I.key === "ArrowDown") {
      const x = c + 1;
      l(x === s.length ? 0 : x);
    } else if (I.key === "ArrowUp") {
      const x = c - 1;
      l(x === -1 ? s.length - 1 : x);
    } else I.key === "Enter" && y(c);
  }
  const O = (I) => {
    l(I);
  }, k = () => {
    l(-1);
  };
  return /* @__PURE__ */ H("div", { children: [
    /* @__PURE__ */ H("div", { className: "univer-flex univer-items-center univer-justify-between univer-gap-2", children: [
      /* @__PURE__ */ M(ys, { value: m, options: _, onChange: A }),
      /* @__PURE__ */ M(
        Ir,
        {
          placeholder: p.t("formula.moreFunctions.searchFunctionPlaceholder"),
          onKeyDown: N,
          value: o,
          onChange: F,
          size: "small",
          allowClear: !0
        }
      )
    ] }),
    s.length > 0 && /* @__PURE__ */ M(
      "ul",
      {
        className: re("univer-mb-0 univer-mt-2 univer-box-border univer-max-h-72 univer-w-full univer-select-none univer-list-none univer-overflow-y-auto univer-rounded univer-p-3 univer-outline-none", Cn, ut),
        onKeyDown: N,
        tabIndex: -1,
        children: s.map(({ name: I }, x) => /* @__PURE__ */ H(
          "li",
          {
            className: re("univer-relative univer-box-border univer-cursor-pointer univer-rounded univer-px-7 univer-py-1 univer-text-sm univer-text-gray-900 univer-transition-colors dark:!univer-text-white", {
              "univer-bg-gray-200 dark:!univer-bg-gray-600": c === x
            }),
            onMouseEnter: () => O(x),
            onMouseLeave: k,
            onClick: () => y(x),
            children: [
              f === x && /* @__PURE__ */ M(
                kr,
                {
                  className: "univer-absolute univer-left-1.5 univer-top-1/2 univer-inline-flex -univer-translate-y-1/2 univer-text-base univer-text-primary-600"
                }
              ),
              /* @__PURE__ */ M("span", { className: "univer-block", children: b(I) })
            ]
          },
          x
        ))
      }
    ),
    u && /* @__PURE__ */ H("div", { className: re("univer-mx-0 univer-my-2 univer-overflow-y-auto", ut), children: [
      /* @__PURE__ */ M(vt, { title: u.functionName, value: u.description }),
      /* @__PURE__ */ M(
        vt,
        {
          title: p.t("formula.moreFunctions.syntax"),
          value: /* @__PURE__ */ M(Mr, { prefix: u.functionName, value: u.functionParameter })
        }
      ),
      /* @__PURE__ */ M(
        vt,
        {
          title: p.t("formula.prompt.helpExample"),
          value: `${u.functionName}(${u.functionParameter.map((I) => I.example).join(",")})`
        }
      ),
      u.functionParameter && u.functionParameter.map((I) => /* @__PURE__ */ M(
        vt,
        {
          title: I.name,
          value: `${I.require ? E : g} ${I.detail}`
        },
        I.name
      ))
    ] })
  ] });
}
function bi() {
  const e = bo(), [t, n] = V(!0), [o, r] = V(!1), [s, i] = V(null);
  T(fn);
  const c = T(dt), l = T(Ge), m = T(ne), h = T(me);
  function f() {
    n(!t), r(!o);
  }
  function d() {
    const u = vr(m);
    if (!u) return;
    h.executeCommand(cr.id, {
      visible: !0,
      unitId: u.unitId,
      eventType: xe.Dblclick
    });
    const a = l.getEditor(un), v = l.getEditor(rr), p = `=${s == null ? void 0 : s.functionName}(`;
    a == null || a.replaceText(p), v == null || v.replaceText(p, !1);
  }
  return /* @__PURE__ */ H(
    "div",
    {
      "data-u-comp": "sheets-formula-functions-panel",
      className: "univer-box-border univer-flex univer-h-full univer-flex-col univer-justify-between univer-py-2",
      children: [
        t && /* @__PURE__ */ M(Ei, { onChange: i }),
        o && /* @__PURE__ */ M(fi, { functionInfo: s, onChange: () => {
        } }),
        /* @__PURE__ */ H("div", { className: "univer-flex univer-justify-end", children: [
          o && /* @__PURE__ */ M(
            at,
            {
              variant: "primary",
              onClick: f,
              className: "univer-mb-5 univer-ml-4 univer-mr-0 univer-mt-0",
              children: c.t("formula.moreFunctions.next")
            }
          ),
          o && /* @__PURE__ */ M(at, { onClick: f, className: "univer-mb-5 univer-ml-4 univer-mr-0 univer-mt-0", children: c.t("formula.moreFunctions.prev") }),
          t && !!e && /* @__PURE__ */ M(
            at,
            {
              variant: "primary",
              onClick: d,
              className: "univer-mb-5 univer-ml-4 univer-mr-0 univer-mt-0",
              children: c.t("formula.moreFunctions.confirm")
            }
          )
        ] })
      ]
    }
  );
}
function yi(e) {
  return {
    id: pe.id,
    title: "SUM",
    icon: "SumIcon",
    type: Je.BUTTON,
    params: {
      value: "SUM"
    },
    hidden$: ft(e, W.UNIVER_SHEET),
    disabled$: Ye(e, {
      workbookTypes: [Qe],
      worksheetTypes: [ze, Ze],
      rangeTypes: [Xe]
    })
  };
}
function Ti(e) {
  return {
    id: pe.id,
    title: "COUNT",
    icon: "CntIcon",
    type: Je.BUTTON,
    params: {
      value: "COUNT"
    },
    hidden$: ft(e, W.UNIVER_SHEET),
    disabled$: Ye(e, {
      workbookTypes: [Qe],
      worksheetTypes: [ze, Ze],
      rangeTypes: [Xe]
    })
  };
}
function Ni(e) {
  return {
    id: pe.id,
    title: "AVERAGE",
    icon: "AvgIcon",
    type: Je.BUTTON,
    params: {
      value: "AVERAGE"
    },
    hidden$: ft(e, W.UNIVER_SHEET),
    disabled$: Ye(e, {
      workbookTypes: [Qe],
      worksheetTypes: [ze, Ze],
      rangeTypes: [Xe]
    })
  };
}
function Oi(e) {
  return {
    id: pe.id,
    title: "MAX",
    icon: "MaxIcon",
    type: Je.BUTTON,
    params: {
      value: "MAX"
    },
    hidden$: ft(e, W.UNIVER_SHEET),
    disabled$: Ye(e, {
      workbookTypes: [Qe],
      worksheetTypes: [ze, Ze],
      rangeTypes: [Xe]
    })
  };
}
function xi(e) {
  return {
    id: pe.id,
    title: "MIN",
    icon: "MinIcon",
    type: Je.BUTTON,
    params: {
      value: "MIN"
    },
    hidden$: ft(e, W.UNIVER_SHEET),
    disabled$: Ye(e, {
      workbookTypes: [Qe],
      worksheetTypes: [ze, Ze],
      rangeTypes: [Xe]
    })
  };
}
function Mi(e) {
  return {
    id: Rn.id,
    title: "formula.insert.more",
    tooltip: "formula.insert.tooltip",
    type: Je.BUTTON,
    hidden$: ft(e, W.UNIVER_SHEET),
    disabled$: Ye(e, {
      workbookTypes: [Qe],
      worksheetTypes: [ze, Ze],
      rangeTypes: [Xe]
    })
  };
}
function wi(e) {
  return e.get(ne).getCurrentTypeOfUnit$(W.UNIVER_SHEET).pipe(
    Ko((o) => o ? e.get(Kn) ? new Yo((s) => s.next(!e.get(Kn).supportClipboard)) : Vn(!0) : Vn(!0))
  );
}
function ki(e) {
  return {
    id: _n.id,
    type: Je.BUTTON,
    title: "formula.operation.pasteFormula",
    disabled$: wi(e).pipe(
      jo(Ye(e, {
        workbookTypes: [Qe],
        rangeTypes: [Xe],
        worksheetTypes: [Ze, ze]
      })),
      gr(([t, n]) => t || n)
    )
  };
}
const Ai = {
  [Yn.BASIC]: {
    [`${pe.id}.sum`]: {
      order: 0,
      menuItemFactory: yi
    },
    [`${pe.id}.count`]: {
      order: 1,
      menuItemFactory: Ti
    },
    [`${pe.id}.average`]: {
      order: 2,
      menuItemFactory: Ni
    },
    [`${pe.id}.max`]: {
      order: 3,
      menuItemFactory: Oi
    },
    [`${pe.id}.min`]: {
      order: 4,
      menuItemFactory: xi
    }
  },
  [Yn.OTHERS]: {
    [Rn.id]: {
      order: 0,
      menuItemFactory: Mi
    }
  },
  [yo]: {
    [_n.id]: {
      order: 4,
      menuItemFactory: ki
    }
  }
}, Fi = "meta_key_ctrl_And_Shift";
function Di(e) {
  return e.getContextValue(so) && e.getContextValue(io);
}
const Xt = [
  $.ARROW_DOWN,
  $.ARROW_UP,
  $.ARROW_LEFT,
  $.ARROW_RIGHT
], $i = [...Xt, $.ENTER, $.TAB, $.ESC];
function Li() {
  const e = [];
  for (const t of $i)
    e.push({
      id: gt.id,
      binding: t,
      preconditions: (n) => bt(n),
      staticParameters: {
        eventType: xe.Keyboard,
        keycode: t
      }
    });
  return e;
}
function Pi() {
  const e = [];
  for (const t of Xt)
    e.push({
      id: gt.id,
      binding: t | P.SHIFT,
      preconditions: (n) => bt(n),
      staticParameters: {
        eventType: xe.Keyboard,
        keycode: t,
        metaKey: P.SHIFT
      }
    });
  return e;
}
function Ui() {
  const e = [];
  for (const t of Xt)
    e.push({
      id: gt.id,
      binding: t | P.CTRL_COMMAND,
      preconditions: (n) => bt(n),
      staticParameters: {
        eventType: xe.Keyboard,
        keycode: t,
        metaKey: P.CTRL_COMMAND
      }
    });
  return e;
}
function Vi() {
  const e = [];
  for (const t of Xt)
    e.push({
      id: gt.id,
      binding: t | P.SHIFT | P.CTRL_COMMAND,
      preconditions: (n) => bt(n),
      staticParameters: {
        eventType: xe.Keyboard,
        keycode: t,
        metaKey: Fi
      }
    });
  return e;
}
const Wi = {
  id: Tr.id,
  binding: $.F4,
  preconditions: (e) => bt(e)
};
function Hi() {
  const e = [];
  for (const t of [$.ENTER, $.TAB, $.ARROW_DOWN, $.ARROW_UP])
    e.push({
      id: gt.id,
      binding: t,
      preconditions: (n) => Di(n),
      staticParameters: {
        eventType: xe.Keyboard,
        keycode: t,
        isSingleEditor: !0
      }
    });
  return e;
}
const Bi = {
  id: fs.id,
  binding: P.ALT | $.EQUAL,
  preconditions: To,
  mac: P.CTRL_COMMAND | P.ALT | $.EQUAL,
  description: "shortcut.sheets-formula-ui.quick-sum",
  group: "4_sheet-edit"
};
var qi = Object.getOwnPropertyDescriptor, ji = (e, t, n, o) => {
  for (var r = o > 1 ? void 0 : o ? qi(t, n) : t, s = e.length - 1, i; s >= 0; s--)
    (i = e[s]) && (r = i(r) || r);
  return r;
}, Be = (e, t) => (n, o) => t(n, o, e);
let Bt = class extends Ke {
  constructor(e, t, n, o, r, s, i) {
    super(), this._injector = e, this._menuManagerService = t, this._commandService = n, this._shortcutService = o, this._uiPartsService = r, this._renderManagerService = s, this._componentManager = i, this._initialize();
  }
  _initialize() {
    this._registerCommands(), this._registerMenus(), this._registerShortcuts(), this._registerComponents(), this._registerRenderModules();
  }
  _registerMenus() {
    this._menuManagerService.mergeMenu(Ai);
  }
  _registerCommands() {
    [
      _n,
      pe,
      Rn,
      Ws,
      Fs,
      gt,
      Tr
    ].forEach((e) => this.disposeWithMe(this._commandService.registerCommand(e)));
  }
  _registerShortcuts() {
    [
      ...Li(),
      ...Pi(),
      ...Ui(),
      ...Vi(),
      ...Hi(),
      Bi,
      Wi
    ].forEach((e) => {
      this.disposeWithMe(this._shortcutService.registerShortcut(e));
    });
  }
  _registerComponents() {
    this.disposeWithMe(this._uiPartsService.registerComponent(No.FORMULA_AUX, () => Rr(di, this._injector))), this._componentManager.register(yr, bi);
  }
  _registerRenderModules() {
    this.disposeWithMe(this._renderManagerService.registerRenderModule(W.UNIVER_SHEET, [Wt]));
  }
};
Bt = ji([
  Be(0, q(Rt)),
  Be(1, vs),
  Be(2, me),
  Be(3, Zt),
  Be(4, Cr),
  Be(5, Re),
  Be(6, q(_r))
], Bt);
var Ki = Object.getOwnPropertyDescriptor, Yi = (e, t, n, o) => {
  for (var r = o > 1 ? void 0 : o ? Ki(t, n) : t, s = e.length - 1, i; s >= 0; s--)
    (i = e[s]) && (r = i(r) || r);
  return r;
}, tn = (e, t) => (n, o) => t(n, o, e);
let qt = class extends Ke {
  constructor(e, t, n) {
    super(), this._imageFormulaCellInterceptorController = e, this._renderManagerService = t, this._univerInstanceService = n, this._imageFormulaCellInterceptorController.registerRefreshRenderFunction(() => {
      const o = this._univerInstanceService.getCurrentUnitOfType(W.UNIVER_SHEET);
      if (!o) return;
      const r = this._renderManagerService.getRenderById(o.getUnitId());
      if (!r) return;
      r.with(Et).reCalculate();
      const s = r.mainComponent;
      s && s.makeDirty();
    });
  }
};
qt = Yi([
  tn(0, q(gs)),
  tn(1, Re),
  tn(2, ne)
], qt);
class Pr {
  constructor() {
    j(this, "_currentSelector$", new mr(null));
    j(this, "currentSelector$", this._currentSelector$.asObservable());
  }
  showRangeSelectorDialog(t) {
    const n = t.callback, o = new Promise((r) => {
      t.callback = (s) => {
        r(s), n(s);
      };
    });
    return this._currentSelector$.next(t), o;
  }
}
var Gi = Object.getOwnPropertyDescriptor, Zi = (e, t, n, o) => {
  for (var r = o > 1 ? void 0 : o ? Gi(t, n) : t, s = e.length - 1, i; s >= 0; s--)
    (i = e[s]) && (r = i(r) || r);
  return r;
}, it = (e, t) => (n, o) => t(n, o, e);
let De = class extends Oo {
  constructor(t, n, o, r, s, i, c) {
    super(
      n,
      o,
      r,
      s,
      i
    );
    j(this, "_workbookSelections");
    j(this, "_eventDisposables");
    this._context = t, this._contextService = i, this._refSelectionsService = c, this._workbookSelections = this._refSelectionsService.getWorkbookSelections(this._context.unitId), this._initSelectionChangeListener(), this._initSkeletonChangeListener(), this._initUserActionSyncListener(), this._setSelectionStyle(zi(this._themeService)), this._remainLastEnabled = !0, this._highlightHeader = !1;
  }
  getLocation() {
    return this._skeleton.getLocation();
  }
  setRemainLastEnabled(t) {
    this._remainLastEnabled = t;
  }
  /**
   * This is set to true when you need to add a new selection.
   * @param {boolean} enabled
   * @memberof RefSelectionsRenderService
   */
  setSkipLastEnabled(t) {
    this._skipLastEnabled = t;
  }
  clearLastSelection() {
    const t = this._selectionControls[this._selectionControls.length - 1];
    t && (t.dispose(), this._selectionControls.pop());
  }
  /**
   * Call this method and user will be able to select on the canvas to update selections.
   */
  enableSelectionChanging() {
    return this._disableSelectionChanging(), this._eventDisposables = this._initCanvasEventListeners(), sr(() => this._disableSelectionChanging());
  }
  _disableSelectionChanging() {
    var t;
    (t = this._eventDisposables) == null || t.dispose(), this._eventDisposables = null;
  }
  disableSelectionChanging() {
    this._disableSelectionChanging();
  }
  _initCanvasEventListeners() {
    const t = this._getSheetObject(), { spreadsheetRowHeader: n, spreadsheetColumnHeader: o, spreadsheet: r, spreadsheetLeftTopPlaceholder: s } = t, { scene: i } = this._context, c = new qe();
    return c.add(r == null ? void 0 : r.onPointerDown$.subscribeEvent((l, m) => {
      this.inRefSelectionMode() && (this._onPointerDown(l, r.zIndex + 1, ve.NORMAL, this._getActiveViewport(l)), l.button !== 2 && m.stopPropagation());
    })), c.add(
      n == null ? void 0 : n.onPointerDown$.subscribeEvent((l, m) => {
        if (!this.inRefSelectionMode()) return;
        const h = this._sheetSkeletonManagerService.getCurrent().skeleton, { row: f } = Dn(l.offsetX, l.offsetY, i, h);
        $n(this._workbookSelections.getCurrentSelections(), f, ve.ROW) || (this._onPointerDown(l, (r.zIndex || 1) + 1, ve.ROW, this._getActiveViewport(l), Qt.Y), l.button !== 2 && m.stopPropagation());
      })
    ), c.add(o == null ? void 0 : o.onPointerDown$.subscribeEvent((l, m) => {
      if (!this.inRefSelectionMode()) return;
      const h = this._sheetSkeletonManagerService.getCurrent().skeleton, { column: f } = Dn(l.offsetX, l.offsetY, i, h);
      $n(this._workbookSelections.getCurrentSelections(), f, ve.COLUMN) || (this._onPointerDown(l, (r.zIndex || 1) + 1, ve.COLUMN, this._getActiveViewport(l), Qt.X), l.button !== 2 && m.stopPropagation());
    })), c.add(s == null ? void 0 : s.onPointerDown$.subscribeEvent((l, m) => {
      if (this._reset(), !this.inRefSelectionMode()) return;
      const h = this._sheetSkeletonManagerService.getCurrent().skeleton, f = xo(h);
      this._addSelectionControlByModelData(f), this._selectionMoveStart$.next(this.getSelectionDataWithStyle());
      const d = i.onPointerUp$.subscribeEvent(() => {
        d.unsubscribe(), this._selectionMoveEnd$.next(this.getSelectionDataWithStyle());
      });
      l.button !== 2 && m.stopPropagation();
    })), c;
  }
  /**
   * Add a selection in spreadsheet, create a new SelectionControl and then update this control by range derives from selection.
   * For ref selection, create selectionShapeExtension to handle user action.
   * @param {ISelectionWithCoord} selectionWithStyle
   */
  _addSelectionControlByModelData(t) {
    var i;
    const n = this._skeleton, o = (i = t.style) != null ? i : ur(this._themeService), r = this._scene;
    return t.style = o, this.newSelectionControl(r, n, t);
  }
  _initSelectionChangeListener() {
    this.disposeWithMe(this._refSelectionsService.selectionSet$.subscribe((t) => {
      this._reset(), this._skeleton && this.resetSelectionsByModelData(t || []);
    }));
  }
  /**
   * Update selectionModel in this._workbookSelections by user action in spreadsheet area.
   */
  _initUserActionSyncListener() {
    this.disposeWithMe(this.selectionMoveStart$.subscribe((t) => {
      this._updateSelections(t, Jt.MOVE_START);
    })), this.disposeWithMe(this.selectionMoving$.subscribe((t) => {
      this._updateSelections(t, Jt.MOVING);
    })), this.disposeWithMe(this.selectionMoveEnd$.subscribe((t) => {
      this._updateSelections(t, Jt.MOVE_END);
    }));
  }
  _updateSelections(t, n) {
    const r = this._context.unit.getActiveSheet().getSheetId();
    t.length !== 0 && this._workbookSelections.setSelections(
      r,
      t.map((s) => ls(s)),
      n
    );
  }
  _initSkeletonChangeListener() {
    this.disposeWithMe(this._sheetSkeletonManagerService.currentSkeleton$.subscribe((t) => {
      var i;
      if (!t)
        return;
      const { skeleton: n } = t, { scene: o } = this._context, r = o.getViewport(Hn.VIEW_MAIN);
      this._skeleton && ((i = this._skeleton.worksheet) == null ? void 0 : i.getSheetId()) !== n.worksheet.getSheetId() && this._reset(), this._changeRuntime(n, o, r);
      const s = this._workbookSelections.getCurrentSelections();
      this.resetSelectionsByModelData(s);
    }));
  }
  _getActiveViewport(t) {
    const n = this._getSheetObject();
    return n == null ? void 0 : n.scene.getActiveViewportByCoord(Bn.FromArray([t.offsetX, t.offsetY]));
  }
  _getSheetObject() {
    return Mo(this._context.unit, this._context);
  }
  /**
   * Handle pointer down event, bind pointermove & pointerup handler.
   * then trigger selectionMoveStart$.
   *
   * @param evt
   * @param _zIndex
   * @param rangeType
   * @param viewport
   * @param scrollTimerType
   */
  // eslint-disable-next-line complexity, max-lines-per-function
  _onPointerDown(t, n = 0, o = ve.NORMAL, r, s = Qt.ALL) {
    var N;
    this._rangeType = o;
    const i = this._skeleton, c = this._scene;
    if (!c || !i)
      return;
    r && (this._activeViewport = r);
    const { offsetX: l, offsetY: m } = t, h = c.getViewport(Hn.VIEW_MAIN);
    if (!h) return;
    const f = c.getCoordRelativeToViewport(Bn.FromArray([l, m])), { x: d, y: u } = f;
    this._startViewportPosX = d, this._startViewportPosY = u;
    const a = c.getScrollXYInfoByViewport(f), { scaleX: v, scaleY: p } = c.getAncestorScale(), S = this._skeleton.getCellByOffset(d, u, v, p, a);
    if (!S) return;
    switch (o) {
      case ve.NORMAL:
        break;
      case ve.ROW:
        S.startColumn = 0, S.endColumn = this._skeleton.getColumnCount() - 1;
        break;
      case ve.COLUMN:
        S.startRow = 0, S.endRow = this._skeleton.getRowCount() - 1;
        break;
      case ve.ALL:
        S.startRow = 0, S.startColumn = 0, S.endRow = this._skeleton.getRowCount() - 1, S.endColumn = this._skeleton.getColumnCount() - 1;
    }
    const C = { range: S, primary: S, style: null };
    C.range.rangeType = o;
    const _ = rn(C, this._skeleton);
    this._startRangeWhenPointerDown = { ..._.rangeWithCoord };
    const E = { ..._.rangeWithCoord, rangeType: o };
    let g = this.getActiveSelectionControl();
    const b = this.getSelectionControls();
    for (const O of b) {
      if (t.button === 2 && nr.contains(O.model, E)) {
        g = O;
        return;
      }
      if (O.model.isEqual(E)) {
        g = O;
        break;
      }
    }
    this._checkClearPreviousControls(t);
    const y = g == null ? void 0 : g.model.currentCell, A = t.shiftKey && y, F = this._remainLastEnabled && !t.ctrlKey && !t.shiftKey && !this._skipLastEnabled && !this._singleSelectionEnabled;
    A && y ? this._makeSelectionByTwoCells(
      y,
      E,
      i,
      o,
      g
      // Get updated in this method
    ) : F && g ? g.updateRangeBySelectionWithCoord(_) : g = this.newSelectionControl(c, i, C);
    for (let O = 0; O < this.getSelectionControls().length - 1; O++)
      this.getSelectionControls()[O].clearHighlight();
    this._selectionMoveStart$.next(this.getSelectionDataWithStyle()), c.disableObjectsEvent(), this._clearUpdatingListeners(), this._addEndingListeners(), (N = c.getTransformer()) == null || N.clearSelectedObjects(), this._setupPointerMoveListener(h, g, o, s, d, u), this._escapeShortcutDisposable = this._shortcutService.forceEscape(), this._scenePointerUpSub = c.onPointerUp$.subscribeEvent(() => {
      var O;
      this._clearUpdatingListeners(), this._selectionMoveEnd$.next(this.getSelectionDataWithStyle()), (O = this._escapeShortcutDisposable) == null || O.dispose(), this._escapeShortcutDisposable = null;
    });
  }
  /**
   * Diff between normal selection, no highlightHeader for ref selections.
   * @param scene
   * @param skeleton
   * @param selectionWithCoord
   * @returns {SelectionControl} selectionControl just created
   */
  newSelectionControl(t, n, o) {
    const r = this.getSelectionControls().length, { rowHeaderWidth: s, columnHeaderHeight: i } = n, c = new lr(t, r, this._themeService, {
      highlightHeader: this._highlightHeader,
      enableAutoFill: !1,
      rowHeaderWidth: s,
      columnHeaderHeight: i
    }), l = rn(o, n);
    return c.updateRangeBySelectionWithCoord(l), this._selectionControls.push(c), c.setControlExtension({
      skeleton: n,
      scene: t,
      themeService: this._themeService,
      injector: this._injector,
      selectionHooks: {
        selectionMoveEnd: () => {
          this._selectionMoveEnd$.next(this.getSelectionDataWithStyle());
        }
      }
    }), c;
  }
};
De = Zi([
  it(1, q(Rt)),
  it(2, q(It)),
  it(3, Zt),
  it(4, q(Et)),
  it(5, ln),
  it(6, Yt)
], De);
function zi(e) {
  const t = ur(e);
  return t.widgets = { tl: !0, tc: !0, tr: !0, ml: !0, mr: !0, bl: !0, bc: !0, br: !0 }, t;
}
const In = (e, t, n = !0) => {
  let o = -1;
  return e.reduce((r, s, i) => {
    if (r.isFinish)
      return r;
    const c = r.currentIndex;
    if (typeof s != "string")
      r.currentIndex += s.token.length;
    else {
      const l = s.length;
      r.currentIndex += l;
    }
    return (n ? r.currentIndex === t : t > c && t <= r.currentIndex) && (o = i, r.isFinish = !0), r;
  }, { currentIndex: 0, isFinish: !1 }), o;
}, Ur = (e, t) => {
  const n = e[t];
  let o = -1;
  if (!n || typeof n == "string" || n.nodeType !== Q.REFERENCE) return -1;
  for (let r = 0; r <= t; r++) {
    const s = e[r];
    typeof s != "string" && s.nodeType === Q.REFERENCE && o++;
  }
  return o;
}, Xi = (e, t = 100) => {
  L(() => {
    let n = null;
    const o = () => {
      n === null && (n = window.setTimeout(() => {
        e(), n = null;
      }, t));
    };
    return window.addEventListener("scroll", o), window.addEventListener("resize", o), () => {
      n !== null && clearTimeout(n), window.removeEventListener("scroll", o), window.removeEventListener("resize", o);
    };
  }, [e, t]);
};
function Vr(e, t, n) {
  const o = T(Ge), r = ie(() => new mr({ left: -999, top: -999, right: -999, bottom: -999 }), []), s = T(vn), i = T(ne), c = te(() => {
    var _;
    const l = o.getEditor(e);
    if (!l)
      return;
    const m = l.getBoundingClientRect(), { marginTop: h = 0, marginBottom: f = 0 } = l.getDocumentData().documentStyle, d = l.getSkeleton();
    if (!d) return;
    const u = (_ = d.getSkeletonData()) == null ? void 0 : _.pages[0].height;
    let { left: a, top: v, right: p, bottom: S } = m;
    v = v + h, S = u ? v + u : S - f;
    const C = r.getValue();
    if (!(C.left === a && C.top === v && C.right === p && C.bottom === S))
      return r.next({ left: a - 1, right: p + 1, top: v - 1, bottom: S + 1 }), m;
  });
  return L(() => {
    t && c();
  }, [e, o, i.unitAdded$, c, t, ...n != null ? n : []]), Xi(c), L(() => {
    const l = s.scrollEvent$.pipe(Go(100)).subscribe(c);
    return () => {
      l.unsubscribe();
    };
  }, []), [r, c];
}
const nt = (e) => {
  const t = z(e);
  return t.current = e, t;
}, Qi = (e, t, n) => {
  const o = T(zt), r = T(Gt), s = T(_e), [i, c] = V(), [l, m] = V(-1), [h, f] = V(!0), d = nt(h), u = z(t);
  u.current = t;
  const a = () => {
    c(void 0), m(-1), f(!1);
  };
  return L(() => {
    const v = s.sequenceNodesBuilder(t.slice(1));
    o.setSequenceNodes(v != null ? v : []);
  }, [t]), L(() => {
    if (n && e) {
      const v = n.selectionChange$.pipe(gn(50)).subscribe((S) => {
        if (S.textRanges.length === 1) {
          const [C] = S.textRanges;
          if (C.collapsed && d.current) {
            const { startOffset: _ } = C, E = o.getCurrentSequenceNodeIndex(_ - 2), g = o.getCurrentSequenceNodeByIndex(E), b = o.getCurrentSequenceNodeByIndex(E + 1);
            if (g)
              if (typeof g != "string" && g.nodeType === 3 && !r.hasDefinedNameDescription(g.token.trim()) && b === je.OPEN_BRACKET) {
                const y = r.getFunctionInfo(g.token);
                c(y), m(-1);
                return;
              } else {
                const y = s.getFunctionAndParameter(`${u.current}A`, _ - 1);
                if (y) {
                  const { functionName: A, paramIndex: F } = y, N = r.getFunctionInfo(A);
                  c(N), m(F);
                  return;
                }
              }
          }
        }
        c(void 0), m(-1);
      }), p = n.selectionChange$.pipe(
        pr((S) => S.textRanges.length === 1),
        gr((S) => S.textRanges[0].startOffset),
        Zo()
      ).subscribe(() => {
        f(!0);
      });
      return () => {
        v.unsubscribe(), p.unsubscribe();
      };
    }
  }, [n, e]), L(() => {
    e || a();
  }, [e]), {
    functionInfo: i,
    paramIndex: l,
    reset: a
  };
}, Ji = ({ onClick: e }) => /* @__PURE__ */ M(
  "div",
  {
    className: "univer-z-[15] univer-box-border univer-h-[18px] univer-cursor-pointer univer-overflow-visible univer-whitespace-nowrap univer-rounded-l univer-border univer-border-r-0 univer-border-gray-600 univer-bg-primary-600 univer-p-0.5 univer-text-xs univer-font-bold univer-leading-[13px] univer-text-white",
    onClick: e,
    children: "?"
  }
), nn = ({ className: e, title: t, value: n }) => /* @__PURE__ */ H("div", { className: "univer-my-2", children: [
  /* @__PURE__ */ M(
    "div",
    {
      className: re("univer-mb-2 univer-text-sm univer-font-medium univer-text-gray-900 dark:!univer-text-white", e),
      children: t
    }
  ),
  /* @__PURE__ */ M(
    "div",
    {
      className: "univer-whitespace-pre-wrap univer-break-words univer-text-xs univer-text-gray-500",
      children: n
    }
  )
] }), ec = (e) => {
  const { prefix: t, value: n, active: o, onClick: r } = e;
  return /* @__PURE__ */ H("div", { children: [
    /* @__PURE__ */ H("span", { children: [
      t,
      "("
    ] }),
    n && n.map((s, i) => /* @__PURE__ */ H("span", { children: [
      /* @__PURE__ */ M(
        "span",
        {
          className: o === i ? "univer-text-primary-500" : "",
          onClick: () => r(i),
          children: xr(s)
        }
      ),
      i === n.length - 1 ? "" : ","
    ] }, s.name)),
    ")"
  ] });
}, Jn = () => {
};
function tc(e) {
  const { onParamsSwitch: t = Jn, onClose: n = Jn, isFocus: o, editor: r, formulaText: s } = e, { functionInfo: i, paramIndex: c, reset: l } = Qi(o, s, r), m = T(fn), h = !Se(m.helpFunctionVisible$), [f, d] = V(!0), u = T(dt), a = u.t("formula.prompt.required"), v = u.t("formula.prompt.optional"), p = r.getEditorId(), [S] = Vr(p, !!i, [i, c]);
  function C(g) {
    t && t(g);
  }
  const _ = te((g) => {
    m.helpFunctionVisible$.next(!g);
  }), E = () => {
    _(!0), n();
  };
  return i ? h ? /* @__PURE__ */ M(on, { portal: !0, anchorRect$: S, direction: "left-center", children: /* @__PURE__ */ M(Ji, { onClick: () => _(!1) }) }, "hidden") : /* @__PURE__ */ M(on, { portal: !0, onClickOutside: () => l(), anchorRect$: S, direction: "vertical", children: /* @__PURE__ */ H(
    "div",
    {
      className: re("univer-m-0 univer-box-border univer-w-[250px] univer-select-none univer-list-none univer-rounded-lg univer-bg-white univer-leading-5 univer-shadow-md univer-outline-none dark:!univer-bg-gray-900", Cn),
      children: [
        /* @__PURE__ */ H(
          "div",
          {
            className: re("univer-wrap-anywhere univer-box-border univer-flex univer-items-center univer-justify-between univer-px-4 univer-py-3 univer-text-xs univer-font-medium univer-text-gray-900 dark:!univer-text-white", Ts),
            children: [
              /* @__PURE__ */ M(
                ec,
                {
                  prefix: i.functionName,
                  value: i.functionParameter,
                  active: c,
                  onClick: C
                }
              ),
              /* @__PURE__ */ H("div", { className: "univer-flex", children: [
                /* @__PURE__ */ M(
                  "div",
                  {
                    className: "univer-ml-2 univer-flex univer-h-6 univer-w-6 univer-cursor-pointer univer-items-center univer-justify-center univer-rounded univer-bg-transparent univer-p-0 univer-text-xs univer-text-gray-500 univer-outline-none univer-transition-colors hover:univer-bg-gray-200 dark:hover:!univer-bg-gray-600",
                    style: { transform: f ? "rotateZ(-90deg)" : "rotateZ(90deg)" },
                    onClick: () => d(!f),
                    children: /* @__PURE__ */ M($r, {})
                  }
                ),
                /* @__PURE__ */ M(
                  "div",
                  {
                    className: "univer-ml-2 univer-flex univer-h-6 univer-w-6 univer-cursor-pointer univer-items-center univer-justify-center univer-rounded univer-bg-transparent univer-p-0 univer-text-xs univer-text-gray-600 univer-outline-none univer-transition-colors hover:univer-bg-gray-300 dark:!univer-text-gray-200 dark:hover:!univer-bg-gray-600",
                    onClick: E,
                    children: /* @__PURE__ */ M(Ar, {})
                  }
                )
              ] })
            ]
          }
        ),
        /* @__PURE__ */ M(
          "div",
          {
            className: re("univer-box-border univer-max-h-[350px] univer-overflow-y-auto univer-px-4 univer-pb-3 univer-pt-0", ut),
            style: {
              height: f ? "unset" : 0,
              padding: f ? "revert-layer" : 0
            },
            children: /* @__PURE__ */ H("div", { className: "univer-mt-3", children: [
              /* @__PURE__ */ M(
                nn,
                {
                  title: u.t("formula.prompt.helpExample"),
                  value: `${i.functionName}(${i.functionParameter.map((g) => g.example).join(",")})`
                }
              ),
              /* @__PURE__ */ M(
                nn,
                {
                  title: u.t("formula.prompt.helpAbstract"),
                  value: i.description
                }
              ),
              i && i.functionParameter && i.functionParameter.map((g, b) => /* @__PURE__ */ M(
                nn,
                {
                  className: c === b ? "univer-text-primary-500" : "",
                  title: g.name,
                  value: `${g.require ? a : v} ${g.detail}`
                },
                b
              ))
            ] })
          }
        )
      ]
    }
  ) }, "show") : null;
}
const nc = (e) => {
  const t = T(Ge);
  return te((o) => {
    var r, s;
    if (e) {
      t.focus(e.getEditorId());
      const i = [...e.getSelectionRanges()];
      if (dn.isDefine(o))
        e.setSelectionRanges([{ startOffset: o, endOffset: o }]);
      else if (!i.length && !e.docSelectionRenderService.isOnPointerEvent) {
        const c = (s = (r = e.getDocumentData().body) == null ? void 0 : r.dataStream) != null ? s : `\r
`, l = Math.max(c.length - 2, 0);
        e.setSelectionRanges([{ startOffset: l, endOffset: l }]);
      } else
        e.setSelectionRanges(i);
    }
  });
};
function rc(e) {
  var r, s;
  const n = e.get(ne).getCurrentUniverDocInstance();
  return n != null && n.getBody() ? { dataStream: (s = (r = n.getBody()) == null ? void 0 : r.dataStream) != null ? s : "", offset: 0 } : void 0;
}
var Ne = /* @__PURE__ */ ((e) => (e[e.NOT_SELECT = 0] = "NOT_SELECT", e[e.NEED_ADD = 1] = "NEED_ADD", e[e.CAN_EDIT = 2] = "CAN_EDIT", e[e.EDIT_OTHER_SHEET_REFERENCE = 3] = "EDIT_OTHER_SHEET_REFERENCE", e[e.EDIT_OTHER_WORKBOOK_REFERENCE = 4] = "EDIT_OTHER_WORKBOOK_REFERENCE", e))(Ne || {});
function oc(e) {
  var y;
  const { editorId: t, isFocus: n, disableOnClick: o, unitId: r, subUnitId: s } = e, i = T(Re), c = T(ne), l = i.getRenderById(r), m = i.getRenderById(t), h = m == null ? void 0 : m.with(Sr), f = T(Er), d = T(Rt), [u, a] = V(
    0
    /* NOT_SELECT */
  ), v = T(_e), p = z(!0), S = l == null ? void 0 : l.with(De), C = nt(u), _ = c.getUnit(r, W.UNIVER_SHEET), E = _ == null ? void 0 : _.getSheetBySheetId(s), g = te((A) => {
    S && S.setSkipLastEnabled(
      A === 1 || A === 3 || A === 4
      /* EDIT_OTHER_WORKBOOK_REFERENCE */
    ), C.current = A, a(A);
  }), b = te(() => {
    var ce, ae, he;
    const A = c.getCurrentUnitOfType(W.UNIVER_SHEET);
    if (!A) return;
    const F = A.getActiveSheet(), N = h == null ? void 0 : h.getActiveTextRange(), O = N != null && N.collapsed ? N.startOffset : -1, k = rc(d);
    if (!k) return;
    const I = (ce = k == null ? void 0 : k.dataStream) == null ? void 0 : ce.slice(0, -2), x = ((ae = v.sequenceNodesBuilder(I)) != null ? ae : []).map((K) => typeof K == "object" ? K.nodeType === Q.REFERENCE ? {
      ...K,
      range: Po(K.token)
    } : {
      ...K,
      range: void 0
    } : K), R = I[O - 1], w = I[O], D = x.find((K) => typeof K == "object" && K.nodeType === Q.REFERENCE && O === K.endIndex + 2), U = R && Uo(R) && (!w || Vo(w) && w !== je.OPEN_BRACKET), B = !!D;
    if ((I == null ? void 0 : I.substring(0, 1)) === "=" && (U || B))
      if (B) {
        if (p.current)
          return;
        const { sheetName: K, unitId: J } = D.range, rt = (he = c.getCurrentUnitOfType(W.UNIVER_SHEET)) == null ? void 0 : he.getUnitId();
        J && J !== rt ? g(
          4
          /* EDIT_OTHER_WORKBOOK_REFERENCE */
        ) : !K && F.getSheetId() === (E == null ? void 0 : E.getSheetId()) || K === F.getName() ? g(
          2
          /* CAN_EDIT */
        ) : g(
          3
          /* EDIT_OTHER_SHEET_REFERENCE */
        );
      } else
        p.current = !1, g(
          1
          /* NEED_ADD */
        );
    else
      g(
        0
        /* NOT_SELECT */
      );
  });
  return L(() => {
    const A = f.textSelection$.pipe(pr((F) => F.unitId === t)).subscribe(() => {
      b();
    });
    return () => A.unsubscribe();
  }, [b, f.textSelection$, t]), L(() => {
    n || (g(
      0
      /* NOT_SELECT */
    ), p.current = !0);
  }, [n, g]), L(() => {
    var F;
    if (!o) return;
    const A = (F = m == null ? void 0 : m.mainComponent) == null ? void 0 : F.onPointerDown$.subscribeEvent(() => {
      g(
        0
        /* NOT_SELECT */
      ), p.current = !0;
    });
    return () => A == null ? void 0 : A.unsubscribe();
  }, [o, (y = m == null ? void 0 : m.mainComponent) == null ? void 0 : y.onPointerDown$, g]), L(() => {
    if (!n) return;
    const A = _ == null ? void 0 : _.activeSheet$.subscribe(() => {
      b();
    }), F = c.getCurrentTypeOfUnit$(W.UNIVER_SHEET).subscribe(() => {
      b();
    });
    return () => {
      A == null || A.unsubscribe(), F == null || F.unsubscribe();
    };
  }, [b, n, _ == null ? void 0 : _.activeSheet$, c.getCurrentTypeOfUnit$]), { isSelecting: u, isSelectingRef: C };
}
const sc = () => {
  const e = T(_e);
  return $t((n) => e.sequenceNodesBuilder(n) || [], [e]);
};
function ic(e, t, n) {
  const o = new hn(t).setAlpha(0.05).toRgbString();
  return {
    id: n,
    strokeWidth: 1,
    stroke: t,
    fill: o,
    widgets: { tl: !0, tc: !0, tr: !0, ml: !0, mr: !0, bl: !0, bc: !0, br: !0 },
    widgetSize: 6,
    widgetStrokeWidth: 1,
    widgetStroke: e.getColorFromTheme("white")
  };
}
function Wr(e) {
  var _, E, g;
  const {
    unitId: t,
    subUnitId: n,
    currentWorkbook: o,
    refSelections: r,
    editor: s,
    refSelectionsService: i,
    refSelectionsRenderService: c,
    sheetSkeletonManagerService: l,
    themeService: m,
    univerInstanceService: h
  } = e, f = o.getUnitId(), d = h.getUnit(t, W.UNIVER_SHEET), u = d == null ? void 0 : d.getActiveSheet(), a = [];
  if (!d || !u) {
    i.setSelections(a);
    return;
  }
  const v = u.getSheetId(), p = (b) => {
    var y;
    return (y = d == null ? void 0 : d.getSheetBySheetName(b)) == null ? void 0 : y.getSheetId();
  };
  if (!((_ = l == null ? void 0 : l.getWorksheetSkeleton(v)) == null ? void 0 : _.skeleton)) return;
  const C = [];
  for (let b = 0, y = r.length; b < y; b++) {
    const A = r[b], { themeColor: F, token: N, refIndex: O, endIndex: k } = A, I = ht(N), { unitId: x, sheetName: R, range: w } = I, D = p(R);
    if (!D && R || f !== t && x !== f || x && x !== f || D && D !== v || !D && v !== n)
      continue;
    const U = us(w, u.getRowCount(), u.getColumnCount());
    U.unitId = t, U.sheetId = v, a.push({
      range: U,
      primary: null,
      style: ic(m, F, O.toString())
    }), C.push(k);
  }
  if (s) {
    const b = (g = (E = s.getSelectionRanges()) == null ? void 0 : E[0]) == null ? void 0 : g.startOffset, y = C.findIndex((A) => A + 2 === b);
    y !== -1 ? c == null || c.setActiveSelectionIndex(y) : c == null || c.resetActiveSelectionIndex();
  }
  return a;
}
function cc(e, t) {
  const n = T(ne), o = T(It), r = T(Yt), s = T(Re), i = Se(ie(() => n.getCurrentTypeOfUnit$(W.UNIVER_SHEET), [n])), c = i ? s.getRenderById(i.getUnitId()) : null, l = c == null ? void 0 : c.with(De), m = c == null ? void 0 : c.with(Et), h = te((f, d) => {
    const u = n.getCurrentUnitOfType(W.UNIVER_SHEET);
    if (!u || l != null && l.selectionMoving) return;
    const a = Wr({
      unitId: e,
      subUnitId: t,
      currentWorkbook: u,
      refSelections: f,
      editor: d,
      refSelectionsService: r,
      refSelectionsRenderService: l,
      sheetSkeletonManagerService: m,
      themeService: o,
      univerInstanceService: n
    });
    if (!a) return;
    ((l == null ? void 0 : l.getSelectionControls()) || []).length === a.length ? l == null || l.resetSelectionsByModelData(a) : r.setSelections(a);
  });
  return L(() => () => {
    l == null || l.resetActiveSelectionIndex();
  }, [l]), h;
}
function Hr(e = "") {
  const t = T(Gt), n = ac(), o = T(me), r = ie(() => e.length, [e]);
  return te((i, c, l = !0, m) => {
    const h = i.getDocumentData(), f = i.getEditorId();
    if (!h)
      return [];
    const d = h.body;
    if (!d)
      return [];
    const u = d.dataStream.slice(0, d.dataStream.length - 2), a = { dataStream: "", ...h.body };
    if (!u.startsWith(e)) return [];
    if (c == null || c.length === 0)
      return a.textRuns = [], o.syncExecuteCommand(Wn.id, {
        unitId: f,
        body: kn(a, 0, a.dataStream.length - 2)
      }), [];
    {
      const { textRuns: v, refSelections: p } = lc(t, n, c);
      r && v.forEach((_) => {
        _.ed = _.ed + r, _.st = _.st + r;
      }), a.textRuns = [{ st: 0, ed: 1, ts: { fs: 11 } }, ...v];
      const S = c.reduce((_, E) => typeof E == "string" ? `${_}${E}` : `${_}${E.token}`, "");
      a.dataStream = `${e}${S}\r
`;
      let C;
      if (l) {
        C = i.getSelectionRanges();
        const _ = a.dataStream.length - 2 + r;
        C.forEach((E) => {
          E.startOffset = Math.max(0, Math.min(E.startOffset, _)), E.endOffset = Math.max(0, Math.min(E.endOffset, _));
        });
      }
      return o.syncExecuteCommand(Wn.id, {
        unitId: f,
        body: kn(a, 0, a.dataStream.length - 2),
        textRanges: m != null ? m : C
      }), p;
    }
  });
}
function ac() {
  const e = T(It), t = e.getCurrentTheme();
  return ie(() => {
    const o = [
      e.getColorFromTheme("loop-color.1"),
      e.getColorFromTheme("loop-color.2"),
      e.getColorFromTheme("loop-color.3"),
      e.getColorFromTheme("loop-color.4"),
      e.getColorFromTheme("loop-color.5"),
      e.getColorFromTheme("loop-color.6"),
      e.getColorFromTheme("loop-color.7"),
      e.getColorFromTheme("loop-color.8"),
      e.getColorFromTheme("loop-color.9"),
      e.getColorFromTheme("loop-color.10"),
      e.getColorFromTheme("loop-color.11"),
      e.getColorFromTheme("loop-color.12")
    ].map((c) => e.isValidThemeColor(c) ? e.getColorFromTheme(c) : c), r = e.getColorFromTheme("blue.700"), s = e.getColorFromTheme("jiqing.800"), i = e.getColorFromTheme("black");
    return { formulaRefColors: o, numberColor: r, stringColor: s, plainTextColor: i };
  }, [t]);
}
function lc(e, t, n) {
  const { formulaRefColors: o, numberColor: r, stringColor: s, plainTextColor: i } = t, c = [], l = [], m = /* @__PURE__ */ new Map();
  let h = 0;
  for (let f = 0, d = n.length; f < d; f++) {
    const u = n[f];
    if (typeof u == "string") {
      const _ = c[c.length - 1], E = _ ? _.ed : 0, g = E + u.length;
      c.push({
        st: E,
        ed: g,
        ts: {
          cl: {
            rgb: i
          },
          fs: 11
        }
      });
      continue;
    }
    if (e.hasDefinedNameDescription(u.token.trim())) {
      c.push({
        st: u.startIndex,
        ed: u.endIndex + 1,
        ts: {
          cl: {
            rgb: i
          },
          fs: 11
        }
      });
      continue;
    }
    const { startIndex: a, endIndex: v, nodeType: p, token: S } = u;
    let C = "";
    if (p === Q.REFERENCE) {
      if (m.has(S))
        C = m.get(S);
      else {
        const _ = h % o.length;
        C = o[_], m.set(S, C), h++;
      }
      l.push({
        refIndex: f,
        themeColor: C,
        token: S,
        startIndex: u.startIndex,
        endIndex: u.endIndex,
        index: l.length
      });
    } else p === Q.NUMBER ? C = r : (p === Q.STRING || p === Q.ARRAY) && (C = s);
    C && C.length > 0 ? c.push({
      st: a,
      ed: v + 1,
      ts: {
        cl: {
          rgb: C
        },
        fs: 11
      }
    }) : c.push({
      st: a,
      ed: v + 1,
      ts: {
        cl: {
          rgb: i
        },
        fs: 11
      }
    });
  }
  return { textRuns: c, refSelections: l };
}
const uc = (e, t, n, o) => {
  const r = T(me), s = T(Zt), i = z(t);
  i.current = t;
  const c = z(o);
  c.current = o, L(() => {
    if (!n || !e)
      return;
    const m = `sheet.formula-embedding-editor.${n.getEditorId()}`, h = new qe(), f = (a, v) => {
      if (c.current) {
        c.current(a, v);
        return;
      }
      let p = de.LEFT;
      a === $.ARROW_DOWN ? p = de.DOWN : a === $.ARROW_UP ? p = de.UP : a === $.ARROW_RIGHT && (p = de.RIGHT), v === P.SHIFT ? r.executeCommand(Xo.id, {
        direction: p
      }) : r.executeCommand(Qo.id, {
        direction: p
      });
    }, d = (a, v) => {
      let p = de.DOWN;
      a === $.ARROW_DOWN ? p = de.DOWN : a === $.ARROW_UP ? p = de.UP : a === $.ARROW_LEFT ? p = de.LEFT : a === $.ARROW_RIGHT && (p = de.RIGHT), i.current ? v === P.CTRL_COMMAND ? r.executeCommand(Ln.id, {
        direction: p,
        jumpOver: Pn.moveGap,
        extra: "formula-editor",
        fromCurrentSelection: i.current === Ne.NEED_ADD || i.current === Ne.EDIT_OTHER_SHEET_REFERENCE
      }) : v === P.SHIFT ? r.executeCommand(Un.id, {
        direction: p,
        extra: "formula-editor"
      }) : v === (P.CTRL_COMMAND | P.SHIFT) ? r.executeCommand(Un.id, {
        direction: p,
        jumpOver: Pn.moveGap,
        extra: "formula-editor"
      }) : r.executeCommand(Ln.id, {
        direction: p,
        extra: "formula-editor",
        fromCurrentSelection: i.current === Ne.NEED_ADD || i.current === Ne.EDIT_OTHER_SHEET_REFERENCE
      }) : f(a, v);
    };
    return h.add(r.registerCommand({
      id: m,
      type: Oe.OPERATION,
      handler(a, v) {
        const { keyCode: p, metaKey: S } = v;
        d(p, S);
      }
    })), [
      { keyCode: $.ARROW_DOWN },
      { keyCode: $.ARROW_LEFT },
      { keyCode: $.ARROW_RIGHT },
      { keyCode: $.ARROW_UP },
      { keyCode: $.ARROW_DOWN, metaKey: P.SHIFT },
      { keyCode: $.ARROW_LEFT, metaKey: P.SHIFT },
      { keyCode: $.ARROW_RIGHT, metaKey: P.SHIFT },
      { keyCode: $.ARROW_UP, metaKey: P.SHIFT },
      { keyCode: $.ARROW_DOWN, metaKey: P.CTRL_COMMAND },
      { keyCode: $.ARROW_LEFT, metaKey: P.CTRL_COMMAND },
      { keyCode: $.ARROW_RIGHT, metaKey: P.CTRL_COMMAND },
      { keyCode: $.ARROW_UP, metaKey: P.CTRL_COMMAND },
      { keyCode: $.ARROW_DOWN, metaKey: P.CTRL_COMMAND | P.SHIFT },
      { keyCode: $.ARROW_LEFT, metaKey: P.CTRL_COMMAND | P.SHIFT },
      { keyCode: $.ARROW_RIGHT, metaKey: P.CTRL_COMMAND | P.SHIFT },
      { keyCode: $.ARROW_UP, metaKey: P.CTRL_COMMAND | P.SHIFT }
    ].map(({ keyCode: a, metaKey: v }) => ({
      id: m,
      binding: v ? a | v : a,
      preconditions: () => !0,
      priority: 900,
      staticParameters: {
        eventType: xe.Keyboard,
        keyCode: a,
        metaKey: v
      }
    })).forEach((a) => {
      h.add(s.registerShortcut(a));
    }), () => {
      h.dispose();
    };
  }, [r, n, e, s]);
}, dc = (e, t, n, o, r = !0) => {
  var u;
  const s = T(Re), i = T(ln), c = T(Cs), l = T(Yt), m = T(ne), h = Se(ie(() => m.getCurrentTypeOfUnit$(W.UNIVER_SHEET), [m])), f = s.getRenderById((u = h == null ? void 0 : h.getUnitId()) != null ? u : ""), d = f == null ? void 0 : f.with(De);
  Lt(() => {
    if (e)
      return i.setContextValue(An, !0), r && c.disable(), () => {
        const a = m.getCurrentUnitOfType(W.UNIVER_DOC);
        (a == null ? void 0 : a.getUnitId()) === o && i.setContextValue(An, !1), r && c.enable(), l.clear();
      };
  }, [i, e, l, r, o]), Lt(() => {
    if (e && t) {
      const a = d == null ? void 0 : d.enableSelectionChanging();
      return i.setContextValue(qn, !0), () => {
        i.setContextValue(qn, !1), a == null || a.dispose();
      };
    }
  }, [i, e, d, t]), L(() => {
    e && (d == null || d.setSkipLastEnabled(!1));
  }, [e, d]);
}, hc = (e, t, n) => {
  const o = T(ne), r = T(mn);
  return $t(() => {
    if (e) {
      const i = [...r.getWorkbookSelections(t).getSelectionsOfWorksheet(n)], c = o.getCurrentUnitForType(W.UNIVER_SHEET), l = c == null ? void 0 : c.getActiveSheet();
      (c == null ? void 0 : c.getUnitId()) !== t && o.setCurrentUnitForType(t), l && l.getSheetId() === n && r.setSelections(i);
    }
  }, [e, r, n, t, o]);
}, fc = (e) => e.reduce((t, n) => typeof n == "string" ? t + n.length : t + n.token.length, 0), an = (e) => e.map((t) => typeof t == "string" ? t : t.token).join(""), At = (e, t = !1, n = "", o = !1) => !t && !o ? e.map((r) => Ce(r.range)) : e.map((r) => o ? Wo(r) : r.sheetName !== "" && r.sheetName !== n ? Ct(r.sheetName, r.range) : Ce(r.range)), gc = (e) => {
  var h, f, d;
  const { editor: t, lexerTreeBuilder: n } = e, o = t == null ? void 0 : t.getSelectionRanges();
  if ((o == null ? void 0 : o.length) !== 1)
    return;
  const s = o[0].startOffset - 1, i = ((f = (h = t == null ? void 0 : t.getDocumentData().body) == null ? void 0 : h.dataStream) != null ? f : `\r
`).slice(0, -2), c = (d = n.sequenceNodesBuilder(i.slice(1))) != null ? d : [], l = In(c, s, !1), m = Ur(c, l);
  return {
    nodeIndex: l,
    updatingRefIndex: m,
    sequenceNodes: c,
    offset: s
  };
}, mc = (() => {
}), pc = (e, t, n, o, r, s, i, c, l, m = mc) => {
  var O;
  const h = T(Re), f = T(ne), d = T(me), u = T(Er), a = T(It), v = T(_e), p = f.getUnit(o), S = te((k, I) => {
    var x, R, w;
    return (w = (R = (x = f.getUnit(k)) == null ? void 0 : x.getSheetBySheetId(I)) == null ? void 0 : R.getName()) != null ? w : "";
  }), C = ie(() => S(o, r), [S, r, o]), _ = Se(p == null ? void 0 : p.activeSheet$), E = nt({ activeSheet: _, sheetName: C }), g = Se(ie(() => f.getCurrentTypeOfUnit$(W.UNIVER_SHEET), [f])), b = h.getRenderById((O = g == null ? void 0 : g.getUnitId()) != null ? O : ""), y = b == null ? void 0 : b.with(De), A = b == null ? void 0 : b.with(Et), F = T(Yt), N = te((k, I) => {
    var B, ce, ae, he, K, J, rt, ot, Tt, Ie;
    const x = gc({ editor: l, lexerTreeBuilder: v });
    if (!x) return;
    const { nodeIndex: R, updatingRefIndex: w, sequenceNodes: D, offset: U } = x;
    if (n.current === Ne.NEED_ADD)
      if (U !== 0) {
        if (R === -1 && D.length)
          return;
        const Y = k[k.length - 1], X = D.splice(R + 1), Ee = (B = Y.sheetId) != null ? B : r, fe = {
          range: Y,
          unitId: (ce = Y.unitId) != null ? ce : g.getUnitId(),
          sheetName: S((ae = Y.unitId) != null ? ae : g.getUnitId(), Ee)
        }, ee = Ee !== r, le = (g == null ? void 0 : g.getUnitId()) !== o, be = At([fe], i && (ee || le), C, le);
        D.push({ token: be[0], nodeType: Q.REFERENCE });
        const Le = [...D, ...X], Me = an(Le);
        m(Me, fc(D), I);
      } else {
        const Y = k[k.length - 1], X = (he = Y.sheetId) != null ? he : r, Ee = {
          range: Y,
          unitId: (K = Y.unitId) != null ? K : g.getUnitId(),
          sheetName: S((J = Y.unitId) != null ? J : g.getUnitId(), X)
        }, fe = X !== r, ee = (g == null ? void 0 : g.getUnitId()) !== o, le = At([Ee], i && (fe || ee), C, ee);
        D.unshift({ token: le[0], nodeType: Q.REFERENCE });
        const be = an(D);
        m(be, le[0].length, I);
      }
    else if (n.current === Ne.EDIT_OTHER_SHEET_REFERENCE || n.current === Ne.EDIT_OTHER_WORKBOOK_REFERENCE) {
      const Y = k.pop();
      if (!Y) return;
      const X = D[R];
      if (typeof X == "object" && X.nodeType === Q.REFERENCE) {
        const Ee = X.token;
        (g == null ? void 0 : g.getUnitId()) !== o ? X.token = Ho((rt = g == null ? void 0 : g.getUnitId()) != null ? rt : "", C, Y) : X.token = C === (_ == null ? void 0 : _.getName()) ? Ce(Y) : Ct(_.getName(), Y);
        const ee = U + (X.token.length - Ee.length);
        m(Bo(D), ee, I);
      }
    } else {
      const Y = [...k];
      if (w !== -1) {
        const G = Y.pop();
        G && Y.splice(w, 0, G);
      }
      let X = 0;
      const Ee = D.map((G) => {
        var ge, st, Pe, Ue;
        if (typeof G == "string")
          return G;
        if (G.nodeType === Q.REFERENCE) {
          const Ve = ht(G.token);
          if (Ve.sheetName || (Ve.sheetName = C), (Ve.unitId || o) !== (g == null ? void 0 : g.getUnitId()) || i && ((ge = E.current.activeSheet) == null ? void 0 : ge.getName()) !== Ve.sheetName)
            return G.token;
          const oe = Y[X];
          if (X++, !oe)
            return "";
          const ye = (st = oe.sheetId) != null ? st : r, Nt = {
            range: oe,
            unitId: (Pe = oe.unitId) != null ? Pe : g.getUnitId(),
            sheetName: S((Ue = oe.unitId) != null ? Ue : g.getUnitId(), ye)
          }, Ot = (g == null ? void 0 : g.getUnitId()) !== o;
          return At([Nt], i && (ye !== r || Ot), C, Ot)[0];
        }
        return G.token;
      });
      let fe = "", ee;
      Ee.forEach((G, ge) => {
        fe += G, ge === R && (ee = fe.length);
      });
      const le = [];
      for (let G = X; G <= k.length - 1; G++) {
        const ge = k[G], st = (ot = ge.sheetId) != null ? ot : r, Pe = {
          range: ge,
          unitId: (Tt = ge.unitId) != null ? Tt : g.getUnitId(),
          sheetName: S((Ie = ge.unitId) != null ? Ie : g.getUnitId(), st)
        }, Ue = (g == null ? void 0 : g.getUnitId()) !== o, oe = At([Pe], i && (st !== r || Ue), C, Ue);
        le.push(oe[0]);
      }
      const be = D[D.length - 1], Le = be && (typeof be == "string" ? !1 : be.nodeType === Q.REFERENCE), Me = `${fe}${le.length && Le ? "," : ""}${le.join(",")}`;
      m(Me, !le.length && ee ? ee : Me.length, I);
    }
  });
  L(() => {
    if (y && e) {
      let k = !0;
      const I = (R, w) => {
        if (k) {
          k = !1;
          return;
        }
        N(R.map((D) => D.rangeWithCoord), w);
      }, x = new qe();
      return x.add(y.selectionMoving$.subscribe((R) => {
        I(R, !1);
      })), x.add(y.selectionMoveEnd$.subscribe((R) => {
        I(R, !0);
      })), () => {
        x.dispose();
      };
    }
  }, [e, N, y]), L(() => {
    if (t && y && l) {
      const k = new qe(), I = () => {
        k.dispose(), y.getSelectionControls().forEach((w, D) => {
          k.add(
            w.selectionScaling$.subscribe((U) => {
              const B = y.getSelectionDataWithStyle().map((ae) => ae.rangeWithCoord), ce = B[D];
              U.sheetId = ce.sheetId, U.unitId = ce.unitId, B[D] = U, N(B, !1);
            })
          ), k.add(
            w.selectionMoving$.subscribe((U) => {
              const B = y.getSelectionDataWithStyle().map((ae) => ae.rangeWithCoord), ce = B[D];
              U.sheetId = ce.sheetId, U.unitId = ce.unitId, B[D] = U, N(B, !0);
            })
          );
        });
      }, x = zo(
        l.input$,
        F.selectionSet$,
        y.selectionMoveEnd$
      ).pipe(
        xs(50)
      ).subscribe(() => {
        I();
      });
      return () => {
        x.unsubscribe(), k.dispose();
      };
    }
  }, [l, t, N, y, F.selectionSet$]), y == null || y.getSelectionDataWithStyle(), L(() => {
    if (c) {
      const k = d.onCommandExecuted((I) => {
        var R;
        if (I.id !== pn.id)
          return;
        const x = I.params;
        if (x.extra === "formula-editor" && x.selections.length) {
          const w = x.selections[x.selections.length - 1];
          if (w) {
            const D = n.current === Ne.NEED_ADD, U = ((R = y == null ? void 0 : y.getSelectionDataWithStyle()) != null ? R : []).map((B) => B.rangeWithCoord);
            D ? U.push(w.range) : U[U.length - 1] = w.range, N(U, !0);
          }
        }
      });
      return () => {
        k.dispose();
      };
    }
  }, [d, l, n, v, c, N, y]), L(() => {
    if (!l)
      return;
    const k = u.textSelection$.subscribe((I) => {
      I.unitId === l.getEditorId() && Wr({
        unitId: o,
        subUnitId: r,
        refSelections: s.current,
        editor: l,
        refSelectionsService: F,
        refSelectionsRenderService: y,
        sheetSkeletonManagerService: A,
        themeService: a,
        univerInstanceService: f,
        currentWorkbook: g
      });
    });
    return () => k.unsubscribe();
  }, [u.textSelection$, l, s, y, F, A, r, a, o, f]);
}, Sc = (e, t, n, o, r, s) => {
  const i = T(me), c = T(Ge), m = T(Re).getRenderById(t), h = T(ne), f = m == null ? void 0 : m.with(De);
  L(() => {
    if (e && f)
      if (n) {
        const d = () => {
          const v = f.getSelectionControls().length;
          for (let p = 1; p <= v; p++)
            f.clearLastSelection();
          return setTimeout(() => {
            s();
          }, 30);
        }, u = i.onCommandExecuted((v) => {
          v.id === jn.id && d();
        }), a = h.getCurrentTypeOfUnit$(W.UNIVER_SHEET).subscribe((v) => {
          d();
        });
        return () => {
          u.dispose(), a.unsubscribe();
        };
      } else {
        const d = i.beforeCommandExecuted((u) => {
          if (u.id === jn.id) {
            o(!1), r(), s();
            const a = c.getEditor(un);
            a == null || a.focus();
          }
        });
        return () => {
          d.dispose();
        };
      }
  }, [e, f]);
}, vc = (e, t, n) => {
  const o = T(_e), r = z(!0);
  L(() => {
    if (e) {
      const s = setTimeout(() => {
        r.current = !1;
      }, 500);
      return () => {
        clearTimeout(s);
      };
    }
  }, [e]), L(() => {
    if (!r.current && t) {
      const s = o.checkIfAddBracket(n);
      t(s === 0 && n.startsWith(fr.EQUALS), `${n}`);
    }
  }, [n, t]);
}, Cc = (e, t = [], n) => {
  const o = T(Gt), [r, s] = V([]), [i, c] = V(""), l = z(-1), m = nt({ nodes: t }), h = () => {
    s([]), c(""), l.current = -1;
  };
  return L(() => {
    if (n && e) {
      const d = n.input$.pipe(gn(300)).subscribe(() => {
        const u = n.getSelectionRanges();
        if (u.length === 1) {
          const a = m.current.nodes, v = u[0];
          if (v.collapsed) {
            const p = In(a, v.startOffset - 1, !1);
            l.current = p;
            const S = a[p];
            if (S && typeof S != "string" && S.nodeType === Q.FUNCTION) {
              l.current = p;
              const C = S.token, _ = o.getSearchListByNameFirstLetter(C);
              s(_), c(C);
              return;
            }
          }
        }
        l.current = -1, c(""), s((a) => a != null && a.length ? [] : a);
      });
      return () => {
        d.unsubscribe();
      };
    }
  }, [n, e]), L(() => {
    e || h();
  }, [e]), {
    searchList: r,
    searchText: i,
    handlerFormulaReplace: (d, u) => {
      const a = [...m.current.nodes];
      if (l.current !== -1) {
        const v = a.splice(l.current + 1), p = a.pop() || "";
        let S = (typeof p == "string" ? p.length : p.token.length) - d.length;
        return a.push(d), v[0] !== je.OPEN_BRACKET && u !== hr.DefinedName && (a.push(je.OPEN_BRACKET), S--), { text: an([...a, ...v]), offset: S };
      }
    },
    reset: h
  };
}, _c = () => {
}, Rc = $e(Ic);
function Ic(e, t) {
  const { isFocus: n, sequenceNodes: o, onSelect: r, editor: s, onClose: i = _c } = e, c = s.getEditorId(), l = T(Zt), m = T(me), { searchList: h, searchText: f, handlerFormulaReplace: d, reset: u } = Cc(n, o, s), a = ie(() => !!h.length, [h]), v = z(void 0), [p, S] = V(0), C = z(!1), [_] = Vr(c, a, [f, h]), E = nt({ searchList: h, active: p }), g = (N, O) => {
    const k = d(N, O);
    k && (u(), r(k));
  };
  function b(N) {
    C.current && S(N);
  }
  function y() {
    C.current && S(-1);
  }
  L(() => {
    if (!h.length)
      return;
    const N = `sheet.formula-embedding-editor.search_function.${c}`, O = new qe(), k = (I) => {
      const { searchList: x, active: R } = E.current;
      switch (I) {
        case $.ARROW_UP: {
          S((w) => {
            const D = Math.max(0, w - 1);
            return A(D), D;
          });
          break;
        }
        case $.ARROW_DOWN: {
          S((w) => {
            const D = Math.min(x.length - 1, w + 1);
            return A(D), D;
          });
          break;
        }
        case $.TAB:
        case $.ENTER: {
          const w = x[R];
          g(w.name, w.functionType);
          break;
        }
        case $.ESC: {
          u(), i();
          break;
        }
      }
    };
    return O.add(m.registerCommand({
      id: N,
      type: Oe.OPERATION,
      handler(I, x) {
        const { keyCode: R } = x;
        k(R);
      }
    })), [$.ARROW_UP, $.ARROW_DOWN, $.ENTER, $.ESC, $.TAB].map((I) => ({
      id: N,
      binding: I,
      preconditions: () => !0,
      priority: 1e3,
      staticParameters: {
        eventType: xe.Keyboard,
        keyCode: I
      }
    })).forEach((I) => {
      O.add(l.registerShortcut(I));
    }), () => {
      O.dispose();
    };
  }, [h]);
  function A(N) {
    const O = v.current;
    if (!O) return;
    const k = O.children[N];
    if (!k) return;
    const x = O.getBoundingClientRect().top, R = O.offsetHeight, w = k.getBoundingClientRect(), D = w.top, U = w.height;
    if (D >= 0 && D > x && D - x + U <= R)
      return;
    const B = k.offsetTop - (R - U) / 2;
    O.scrollTo({
      top: B,
      behavior: "smooth"
    });
  }
  const F = ie(() => {
    let N = "";
    return () => {
      clearTimeout(N), C.current = !0, N = setTimeout(() => {
        C.current = !1;
      }, 300);
    };
  }, []);
  return h.length > 0 && a && /* @__PURE__ */ M(on, { portal: !0, anchorRect$: _, direction: "vertical", children: /* @__PURE__ */ M(
    "ul",
    {
      ref: (N) => {
        v.current = N, t && (t.current = N);
      },
      "data-u-comp": "sheets-formula-editor",
      className: re("univer-m-0 univer-box-border univer-max-h-[400px] univer-w-[250px] univer-list-none univer-overflow-y-auto univer-rounded-lg univer-bg-white univer-p-2 univer-leading-5 univer-shadow-md univer-outline-none dark:!univer-bg-gray-900", Cn, ut),
      children: h.map((N, O) => /* @__PURE__ */ H(
        "li",
        {
          className: re("univer-box-border univer-cursor-pointer univer-rounded univer-px-2 univer-py-1 univer-text-gray-900 univer-transition-colors dark:!univer-text-white", {
            "univer-bg-gray-200 dark:!univer-bg-gray-600": p === O
          }),
          onMouseEnter: () => b(O),
          onMouseLeave: y,
          onMouseMove: F,
          onClick: () => {
            g(N.name, N.functionType), s && s.focus();
          },
          children: [
            /* @__PURE__ */ H("span", { className: "univer-block univer-truncate univer-text-xs", children: [
              /* @__PURE__ */ M("span", { className: "univer-text-red-500", children: N.name.substring(0, f.length) }),
              /* @__PURE__ */ M("span", { children: N.name.slice(f.length) })
            ] }),
            /* @__PURE__ */ M(
              "span",
              {
                className: "univer-block univer-text-xs univer-text-gray-400",
                children: N.desc
              }
            )
          ]
        },
        N.name
      ))
    }
  ) });
}
const Ec = (e) => e.startsWith(fr.EQUALS) ? e.slice(1) : "", er = () => {
}, bc = $e((e, t) => {
  var yn, Tn, Nn, On;
  const {
    errorText: n,
    initValue: o,
    unitId: r,
    subUnitId: s,
    isFocus: i = !0,
    isSupportAcrossSheet: c = !1,
    onFocus: l = er,
    onBlur: m = er,
    onChange: h,
    onVerify: f,
    className: d,
    editorId: u,
    moveCursor: a = !0,
    onFormulaSelectingChange: v,
    keyboardEventConfig: p,
    onMoveInEditor: S,
    resetSelectionOnBlur: C = !0,
    autoScrollbar: _ = !0,
    isSingle: E = !0,
    disableSelectionOnClick: g = !1,
    autofocus: b = !0,
    disableContextMenu: y,
    style: A
  } = e, F = T(Ge), N = z(null), O = te(h);
  Es(t, () => ({
    isClickOutSide: (Z) => N.current ? !N.current.contains(Z.target) : !1
  }));
  const k = te(v), I = z(null), x = z(void 0), R = x.current, [w, D] = V(i), U = z(null), B = ie(() => u != null ? u : co(`${wo}-${Kt(4)}`), []), ce = ie(() => n !== void 0, [n]), ae = T(ne), he = ae.getUnit(B);
  Se(he == null ? void 0 : he.change$);
  const K = sc(), J = ao.transform.getPlainText((Tn = (yn = he == null ? void 0 : he.getBody()) == null ? void 0 : yn.dataStream) != null ? Tn : ""), rt = nt(J), ot = ie(() => Ec(J), [J]), Tt = ie(() => K(ot), [ot, K]), { isSelecting: Ie, isSelectingRef: Y } = oc({ unitId: r, subUnitId: s, editorId: B, isFocus: w, disableOnClick: g }), X = z(""), fe = T(Re).getRenderById(B), ee = fe == null ? void 0 : fe.with(Sr), le = ee == null ? void 0 : ee.isFocusing, be = ie(() => ae.getCurrentTypeOfUnit$(W.UNIVER_DOC), [ae]), Le = Se(be), Me = (Le == null ? void 0 : Le.getUnitId()) === B, G = z([]), ge = Ie, Pe = (On = (Nn = T(ir).getConfig(Nr)) == null ? void 0 : Nn.functionScreenTips) != null ? On : !0;
  _s(() => {
    O(J);
  }, [J, O]);
  const Ue = Hr("="), Ve = cc(r, s), oe = te((Z, se = !0, we, Te) => {
    if (!x.current) return;
    X.current = Z;
    const xt = Z[0] === "=" ? Z.slice(1) : "", ke = K(xt), Kr = ke.reduce((We, wt) => typeof wt == "object" ? `${We}${wt.token}` : `${We}${wt}`, ""), Mt = Ue(
      x.current,
      Kr === xt ? ke : [],
      se,
      Te
    );
    if (G.current = Mt, we) {
      const We = Te != null ? Te : R == null ? void 0 : R.getSelectionRanges();
      if ((We == null ? void 0 : We.length) !== 1)
        return;
      const Yr = We[0].startOffset - 1, Gr = In(ke, Yr, !1), xn = Ur(ke, Gr);
      if (xn >= 0) {
        const Mn = Mt.splice(xn, 1)[0];
        Mn && Mt.push(Mn);
      }
      Ve(w ? Mt : [], x.current);
    }
  });
  L(() => {
    w && oe(J, !1, !0);
  }, [w]), L(() => {
    if (w) {
      if (X.current === J) return;
      oe(J, !1, !0);
    }
  }, [J]), vc(w, f, J);
  const ye = nc(R), Nt = hc(w, r, s);
  L(() => {
    var Z;
    k(Ie, (Z = ee == null ? void 0 : ee.isFocusing) != null ? Z : !0);
  }, [k, Ie]), Jo(w, p, R), Lt(() => {
    let Z;
    if (U.current) {
      Z = F.register({
        autofocus: b,
        editorUnitId: B,
        initialSnapshot: {
          id: B,
          body: {
            dataStream: `${o}\r
`,
            textRuns: [],
            customBlocks: [],
            customDecorations: [],
            customRanges: []
          },
          documentStyle: {}
        }
      }, U.current);
      const se = F.getEditor(B);
      x.current = se, oe(o, !1, !0);
    }
    return () => {
      Z == null || Z.dispose();
    };
  }, []), Lt(() => {
    i ? (D(i), ye()) : (C && (R == null || R.blur(), Nt()), D(i));
  }, [i, R, ye, Nt, C]);
  const { checkScrollBar: Ot } = es(R, E, _);
  dc(w, !!(Ie && Me), r, B, y), uc(!!(w && le && a), ge, R, S);
  const En = te((Z, se, we) => {
    if (!le)
      return;
    const Te = se !== -1 ? [{ startOffset: se + 1, endOffset: se + 1, collapsed: !0 }] : void 0;
    oe(`=${Z}`, !0, we, Te), we && (ye(), se !== -1 && setTimeout(() => {
      const xt = { startOffset: se + 1, endOffset: se + 1 }, ke = R == null ? void 0 : R.render.with(ts);
      ke == null || ke.scrollToRange({ ...xt, collapsed: !0 });
    }, 50), Ot());
  });
  pc(
    w && !!(Ie && Me),
    w,
    Y,
    r,
    s,
    G,
    c,
    !!ge,
    R,
    En
  ), Sc(w && !!(Ie && Me), r, c, D, m, () => {
    oe(rt.current, !1, !0);
  });
  const bn = (Z) => {
    if (Z) {
      const se = R == null ? void 0 : R.getSelectionRanges();
      if (se && se.length === 1) {
        const we = se[0];
        if (we.collapsed) {
          const Te = Z.offset;
          setTimeout(() => {
            R == null || R.setSelectionRanges([{ startOffset: we.startOffset - Te, endOffset: we.endOffset - Te }]);
          }, 30);
        }
      }
      ye(), oe(`=${Z.text}`);
    }
  }, jr = () => {
    D(!0), l(), ye();
  };
  return /* @__PURE__ */ H("div", { className: d, children: [
    /* @__PURE__ */ M(
      "div",
      {
        ref: N,
        className: re("univer-relative univer-box-border univer-flex univer-h-full univer-w-full univer-items-center univer-justify-around univer-gap-2 univer-rounded-none univer-p-0 univer-ring-1", {
          "univer-ring-primary-500": w,
          "univer-ring-red-500": ce
        }),
        children: /* @__PURE__ */ M(
          "div",
          {
            ref: U,
            className: "univer-relative univer-h-full univer-w-full",
            onMouseUp: jr
          }
        )
      }
    ),
    n !== void 0 && /* @__PURE__ */ M("div", { className: "univer-my-1 univer-text-xs univer-text-red-500", children: n }),
    Pe && R && ot !== "" && /* @__PURE__ */ M(
      tc,
      {
        editor: R,
        isFocus: w,
        formulaText: J,
        onClose: () => ye()
      }
    ),
    Pe && !!R && /* @__PURE__ */ M(
      Rc,
      {
        isFocus: w,
        sequenceNodes: Tt,
        onSelect: bn,
        ref: I,
        editor: R
      }
    )
  ] });
});
function yc(e, t, n, o) {
  const r = T(_e), s = Hr(""), i = Se(e == null ? void 0 : e.getDocumentDataModel().change$), [c, l] = V([]), m = T(ko), h = z(""), f = T(ne);
  return L(() => {
    if (!e) return;
    const d = e.getDocumentDataModel().getPlainText();
    if (h.current === d)
      return;
    h.current = d;
    const u = r.sequenceNodesBuilder(d);
    l(u != null ? u : []);
  }, [i, e, r]), L(() => {
    var a, v;
    if (!e) return;
    if (!t) {
      const p = e.getDocumentData();
      e.setDocumentData({
        ...p,
        body: {
          ...p.body,
          dataStream: (v = (a = p.body) == null ? void 0 : a.dataStream) != null ? v : "",
          textRuns: []
        }
      });
      return;
    }
    const d = s(e, c, !1), u = new qe();
    return d.forEach((p) => {
      const S = ht(p.token), C = f.getCurrentUnitForType(W.UNIVER_SHEET), _ = C == null ? void 0 : C.getActiveSheet();
      if (!S.sheetName && o !== (_ == null ? void 0 : _.getSheetId()) || S.sheetName && (_ == null ? void 0 : _.getName()) !== S.sheetName)
        return;
      const E = new hn(p.themeColor).toRgb(), g = m.addShape({
        range: S.range,
        style: {
          stroke: p.themeColor,
          fill: `rgba(${E.r}, ${E.g}, ${E.b}, 0.1)`,
          strokeDash: 12
        },
        primary: null
      });
      g && u.add(() => m.removeShape(g));
    }), () => {
      u.dispose();
    };
  }, [e, t, s, m, c]), { sequenceNodes: c };
}
function Tc(e) {
  const t = T(mn), { supportAcrossSheet: n = !1, keepSheetReference: o = !1, unitId: r, subUnitId: s, onChange: i } = e, l = T(ne).getUnit(r, W.UNIVER_SHEET), m = te(i), h = te((f, d) => {
    const u = l == null ? void 0 : l.getActiveSheet();
    if (!u || !n && u.getSheetId() !== s || !(f != null && f.length)) return;
    const a = o ? u.getName() : u.getSheetId() === s ? "" : u.getName(), v = f.map((p) => ({
      range: p.range,
      unitId: r,
      sheetName: a
    }));
    m(v, d);
  });
  L(() => {
    const f = new qe();
    return f.add(t.selectionMoveStart$.subscribe((d) => {
      h(d, !0);
    })), f.add(t.selectionMoving$.subscribe((d) => {
      h(d, !1);
    })), f.add(t.selectionMoveEnd$.subscribe((d) => {
      h(d, !1);
    })), () => {
      f.dispose();
    };
  }, [h, t.selectionMoveEnd$, t.selectionMoveStart$, t.selectionMoving$]);
}
const tr = (e) => !e.some((n) => {
  if (typeof n == "string") {
    if (n !== je.COMMA)
      return !0;
  } else if (n.nodeType !== Q.REFERENCE)
    return !0;
  return !1;
}), Nc = (e) => {
  if (e.endColumn < e.startColumn) {
    const t = e.endColumn;
    e.endColumn = e.startColumn, e.startColumn = t;
  }
  if (e.endRow < e.startRow) {
    const t = e.endRow;
    e.endRow = e.startRow, e.startRow = t;
  }
  return e;
};
function Oc(e) {
  const {
    visible: t,
    initialValue: n,
    unitId: o,
    subUnitId: r,
    maxRangeCount: s = 1 / 0,
    supportAcrossSheet: i,
    keepSheetReference: c,
    onConfirm: l,
    onClose: m,
    onShowBySelection: h
  } = e, f = T(dt), d = T(_e), [u, a] = V([]), [v, p] = V(0), S = z(null);
  L(() => {
    if (t && n.length) {
      const g = n.map((b) => b.sheetName ? Ct(b.sheetName, b.range) : Ce(b.range));
      a(g), p(g.length - 1);
    } else
      a([""]), p(0);
  }, [t]);
  const C = (g, b) => {
    const y = [...u];
    y[g] = b, a(y);
  }, _ = () => {
    a([...u, ""]), p(u.length);
  }, E = (g) => {
    u.splice(g, 1), a([...u]);
  };
  return Tc({
    unitId: o,
    subUnitId: r,
    supportAcrossSheet: i,
    keepSheetReference: c,
    onChange: (g, b) => {
      if (!t && h != null && h(g))
        return;
      const y = new Set(u), A = g.map((O) => O.sheetName ? Ct(O.sheetName, O.range) : Ce(O.range)), F = A.filter((O) => !y.has(O));
      if (!F.length) return;
      const N = [...u];
      if (A.length > 1) {
        b || N.splice(v, 1), N.push(...F);
        const O = N.slice(0, s);
        a(O), p(O.length - 1), requestAnimationFrame(() => {
          var k;
          (k = S.current) == null || k.scrollTo({ top: S.current.scrollHeight });
        });
      } else {
        N.splice(v, 1, ...F);
        const O = N.slice(0, s);
        a(O), p(v + F.length - 1);
      }
    }
  }), /* @__PURE__ */ M(
    Os,
    {
      width: "328px",
      open: t,
      title: f.t("rangeSelector.title"),
      draggable: !0,
      mask: !1,
      maskClosable: !1,
      footer: /* @__PURE__ */ H("footer", { className: "univer-flex univer-gap-2", children: [
        /* @__PURE__ */ M(at, { onClick: m, children: f.t("rangeSelector.cancel") }),
        /* @__PURE__ */ M(
          at,
          {
            variant: "primary",
            onClick: () => {
              l(
                u.filter((g) => {
                  const b = d.sequenceNodesBuilder(g);
                  return b && b.length === 1 && typeof b[0] != "string" && b[0].nodeType === Q.REFERENCE;
                }).map((g) => ht(g)).map((g) => ({ ...g, range: Nc(g.range) }))
              );
            },
            children: f.t("rangeSelector.confirm")
          }
        )
      ] }),
      onClose: m,
      children: /* @__PURE__ */ H(
        "div",
        {
          ref: S,
          className: re("-univer-mx-6 univer-max-h-60 univer-overflow-y-auto univer-px-6", ut),
          children: [
            u.map((g, b) => /* @__PURE__ */ H(
              "div",
              {
                className: "univer-mb-2 univer-flex univer-items-center univer-gap-4",
                children: [
                  /* @__PURE__ */ M(
                    Ir,
                    {
                      className: re("univer-w-full", {
                        "univer-border-primary-600": v === b
                      }),
                      placeholder: f.t("rangeSelector.placeHolder"),
                      onFocus: () => p(b),
                      value: g,
                      onChange: (y) => C(b, y)
                    }
                  ),
                  u.length > 1 && /* @__PURE__ */ M(
                    Fr,
                    {
                      className: "univer-cursor-pointer",
                      onClick: () => E(b)
                    }
                  )
                ]
              },
              b
            )),
            u.length < s && /* @__PURE__ */ M("div", { children: /* @__PURE__ */ H(at, { variant: "link", onClick: _, children: [
              /* @__PURE__ */ M(Dr, {}),
              /* @__PURE__ */ M("span", { children: f.t("rangeSelector.addAnotherRange") })
            ] }) })
          ]
        }
      )
    }
  );
}
function xc(e) {
  return e.split(je.COMMA).filter((t) => !!t).map((t) => ht(t));
}
function Mc(e) {
  return e.map((t) => t.sheetName ? Ct(t.sheetName, t.range) : Ce(t.range)).join(je.COMMA);
}
function Br(e) {
  const [t, n] = V(null), {
    onVerify: o,
    selectorRef: r,
    unitId: s,
    subUnitId: i,
    maxRangeCount: c,
    supportAcrossSheet: l,
    keepSheetReference: m,
    autoFocus: h,
    onChange: f,
    onRangeSelectorDialogVisibleChange: d,
    onClickOutside: u,
    onFocusChange: a,
    forceShowDialogWhenSelectionChanged: v,
    hideEditor: p,
    resetRange: S
  } = e, [C, _] = V(h != null ? h : !1), [E, g] = V(!1), [b, y] = V([]), A = T(dt), F = T(Ge), { sequenceNodes: N } = yc(t, C, s, i), O = nt(N), k = T(me), I = te(() => {
    t == null || t.setSelectionRanges([]), t == null || t.blur(), F.blur();
  }), x = te(() => {
    var R;
    I(), y(xc((R = t == null ? void 0 : t.getDocumentDataModel().getPlainText()) != null ? R : "")), g(!0);
  });
  return L(() => {
    r && (r.current = {
      get editor() {
        return t;
      },
      focus() {
        F.focus(t.getEditorId());
      },
      blur: I,
      verify: () => tr(O.current),
      showDialog: (R) => {
        I(), y(R), g(!0);
      },
      hideDialog: () => {
        y([]), g(!1);
      },
      getValue: () => {
        var R;
        return (R = t == null ? void 0 : t.getDocumentDataModel().getPlainText()) != null ? R : "";
      }
    });
  }, [I, t, F, r, O]), L(() => {
    var R;
    o == null || o(tr(N), (R = t == null ? void 0 : t.getDocumentDataModel().getPlainText()) != null ? R : "");
  }, [N]), L(() => {
    d == null || d(E);
  }, [E]), L(() => {
    if (E && S)
      return () => {
        const R = {
          unitId: s,
          subUnitId: i,
          selections: S
        };
        k.executeCommand(pn.id, R);
      };
  }, [E]), /* @__PURE__ */ H(Is, { children: [
    p ? null : /* @__PURE__ */ M(
      ns,
      {
        isSingle: !0,
        ...e,
        onFocusChange: (R, w) => {
          _(R), a == null || a(R, w);
        },
        editorRef: n,
        onClickOutside: () => {
          _(!1), I(), u == null || u();
        },
        icon: /* @__PURE__ */ M(Ns, { title: A.t("rangeSelector.buttonTooltip"), placement: "bottom", children: /* @__PURE__ */ M(
          Lr,
          {
            className: "univer-cursor-pointer dark:!univer-text-gray-300",
            onClick: x
          }
        ) })
      }
    ),
    /* @__PURE__ */ M(
      Oc,
      {
        initialValue: b,
        unitId: s,
        subUnitId: i,
        visible: E,
        maxRangeCount: c,
        onConfirm: (R) => {
          const w = Mc(R), D = lo.newEmptyData();
          D.body.dataStream = w, t == null || t.replaceText(w, !1), f == null || f(D, w), g(!1), y([]), requestAnimationFrame(() => {
            I();
          });
        },
        onClose: () => {
          g(!1), y([]);
        },
        supportAcrossSheet: l,
        keepSheetReference: m,
        onShowBySelection: (R) => C || v ? (y(R), g(!0), !1) : !0
      }
    )
  ] });
}
const wc = () => {
  var o, r;
  const e = T(Pr), t = Se(e.currentSelector$), n = z(null);
  return L(() => {
    var s, i;
    if (t)
      return (i = n.current) == null || i.showDialog((s = t.initialValue) != null ? s : []), () => {
        var c;
        (c = n.current) == null || c.hideDialog();
      };
  }, [t]), /* @__PURE__ */ M(
    Br,
    {
      unitId: (o = t == null ? void 0 : t.unitId) != null ? o : "",
      subUnitId: (r = t == null ? void 0 : t.subUnitId) != null ? r : "",
      hideEditor: !0,
      selectorRef: n,
      onChange: (s, i) => {
        var c;
        t == null || t.callback((c = i == null ? void 0 : i.split(",").map((l) => ht(l))) != null ? c : []);
      }
    }
  );
};
var kc = Object.defineProperty, Ac = Object.getOwnPropertyDescriptor, Fc = (e, t, n) => t in e ? kc(e, t, { enumerable: !0, configurable: !0, writable: !0, value: n }) : e[t] = n, Dc = (e, t, n, o) => {
  for (var r = o > 1 ? void 0 : o ? Ac(t, n) : t, s = e.length - 1, i; s >= 0; s--)
    (i = e[s]) && (r = i(r) || r);
  return r;
}, Ft = (e, t) => (n, o) => t(n, o, e), qr = (e, t, n) => Fc(e, typeof t != "symbol" ? t + "" : t, n);
let jt = class extends ho {
  constructor(e = zn, t, n, o, r) {
    super(), this._config = e, this._injector = t, this._renderManagerService = n, this._configService = o, this._uiPartsService = r;
    const { menu: s, ...i } = fo(
      zn,
      this._config
    );
    s && this._configService.setConfig("menu", s, { merge: !0 }), this._configService.setConfig(Nr, i, { merge: !0 });
  }
  onStarting() {
    go(this._injector, [
      [zt, { useClass: sn }],
      [Pr],
      [Bt],
      [Ut],
      [Vt],
      [Wt],
      [Ht],
      [Pt],
      [qt]
    ]), this._initUIPart();
  }
  onReady() {
    [
      [De]
    ].forEach((e) => {
      this.disposeWithMe(this._renderManagerService.registerRenderModule(W.UNIVER_SHEET, e));
    });
  }
  onRendered() {
    [
      [cn]
    ].forEach((e) => {
      this.disposeWithMe(this._renderManagerService.registerRenderModule(W.UNIVER_SHEET, e));
    }), mo(this._injector, [
      [Bt],
      // FormulaProgressBar relies on TriggerCalculationController, but it is necessary to ensure that the formula calculation is done after rendered.
      [Vt],
      [Ht],
      [qt]
    ]);
  }
  onSteady() {
    this._injector.get(Ut), this._injector.get(Pt);
  }
  _initUIPart() {
    const e = this._injector.get(_r);
    this.disposeWithMe(e.register(Ao, Br)), this.disposeWithMe(e.register(Fo, bc)), this.disposeWithMe(this._uiPartsService.registerComponent(Rs.GLOBAL, () => Rr(wc, this._injector)));
  }
};
qr(jt, "pluginName", br);
qr(jt, "type", W.UNIVER_SHEET);
jt = Dc([
  uo(qo, ms),
  Ft(1, q(Rt)),
  Ft(2, Re),
  Ft(3, ir),
  Ft(4, Cr)
], jt);
export {
  As as FORMULA_PROMPT_ACTIVATED,
  bc as FormulaEditor,
  Pt as FormulaReorderController,
  Pr as GlobalRangeSelectorService,
  Fs as HelpFunctionOperation,
  pe as InsertFunctionOperation,
  Rn as MoreFunctionsOperation,
  Br as RangeSelector,
  De as RefSelectionsRenderService,
  Tr as ReferenceAbsoluteOperation,
  Ws as SearchFunctionOperation,
  gt as SelectEditorFormulaOperation,
  _n as SheetOnlyPasteFormulaCommand,
  jt as UniverSheetsFormulaUIPlugin
};
