// @ant-design/pro-flow/es/Inspector/index.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
var _excluded = ["open", "width", "mask", "style", "rootStyle", "className", "onClose", "children"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
// import { DraggablePanel } from '@ant-design/pro-editor';

import { Drawer } from 'antd';
import React from 'react';
import useStyles from "./index.style";
import { jsx as _jsx } from "react/jsx-runtime";
// const VisibleMinWidth = 300; // 画布最小可视宽度
var Inspector = function Inspector(_ref) {
  var open = _ref.open,
    width = _ref.width,
    _ref$mask = _ref.mask,
    mask = _ref$mask === void 0 ? false : _ref$mask,
    _ref$style = _ref.style,
    style = _ref$style === void 0 ? {} : _ref$style,
    _ref$rootStyle = _ref.rootStyle,
    rootStyle = _ref$rootStyle === void 0 ? {} : _ref$rootStyle,
    className = _ref.className,
    _ref$onClose = _ref.onClose,
    onClose = _ref$onClose === void 0 ? function () {} : _ref$onClose,
    children = _ref.children,
    props = _objectWithoutProperties(_ref, _excluded);
  var _useStyles = useStyles(),
    styles = _useStyles.styles,
    cx = _useStyles.cx;

  // const [dragedWith, setDragedWidth] = useState<string | number | undefined>(width);
  // const onSizeDragging = (delta: NumberSize, size?: Size) => {
  //   const reactFlowWidth = document.querySelector('.react-flow')?.offsetWidth; // 画布宽度
  //   const drawerWidthNum = size?.width?.split('px')?.[0]
  //     ? Number(size?.width?.split('px')?.[0])
  //     : 0; // 侧边抽屉动态宽度
  //   if (drawerWidthNum && drawerWidthNum <= reactFlowWidth - VisibleMinWidth) {
  //     setDragedWidth(size?.width);
  //   }
  // };

  return /*#__PURE__*/_jsx(Drawer, _objectSpread(_objectSpread({
    open: open,
    onClose: onClose,
    width: width,
    mask: mask,
    className: className ? cx(_defineProperty(_defineProperty({}, styles.dataFlowDrawer, true), "className", true)) : styles.dataFlowDrawer,
    getContainer: false,
    style: _objectSpread(_objectSpread({}, style), {}, {
      padding: '0 0'
    }),
    rootStyle: _objectSpread(_objectSpread({}, rootStyle), {}, {
      position: 'absolute'
    })
  }, props), {}, {
    children: children
  }));
};
export default Inspector;
