// @ant-design/pro-flow/es/FlowEditor/container/index.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
var _excluded = ["devtools", "onNodesInit", "beforeConnect", "onConnect", "afterConnect", "defaultViewport", "contextMenuEnabled", "flowProps", "nodeTypes", "children", "miniMap", "background"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
import { forwardRef } from 'react';
import { HotkeysProvider } from 'react-hotkeys-hook';
import App from "./FlowEditor";
import { FlowEditorProvider } from "./FlowEditorProvider";
import StoreUpdater from "./StoreUpdater";
import { jsx as _jsx } from "react/jsx-runtime";
import { jsxs as _jsxs } from "react/jsx-runtime";
var FlowEditor = /*#__PURE__*/forwardRef(function (_ref, ref) {
  var devtools = _ref.devtools,
    onNodesInit = _ref.onNodesInit,
    beforeConnect = _ref.beforeConnect,
    onConnect = _ref.onConnect,
    afterConnect = _ref.afterConnect,
    defaultViewport = _ref.defaultViewport,
    contextMenuEnabled = _ref.contextMenuEnabled,
    flowProps = _ref.flowProps,
    nodeTypes = _ref.nodeTypes,
    children = _ref.children,
    miniMap = _ref.miniMap,
    background = _ref.background,
    res = _objectWithoutProperties(_ref, _excluded);
  return /*#__PURE__*/_jsx(HotkeysProvider, {
    initiallyActiveScopes: ['flow-editor'],
    children: /*#__PURE__*/_jsxs(FlowEditorProvider, {
      showDevtools: devtools,
      children: [/*#__PURE__*/_jsx(App, {
        ref: ref,
        nodeTypes: nodeTypes,
        miniMap: miniMap,
        background: background,
        onNodesInit: onNodesInit,
        beforeConnect: beforeConnect,
        onConnect: onConnect,
        afterConnect: afterConnect,
        contextMenuEnabled: contextMenuEnabled,
        defaultViewport: defaultViewport,
        flowProps: flowProps,
        children: children
      }), /*#__PURE__*/_jsx(StoreUpdater, _objectSpread({}, res))]
    })
  });
});
export default FlowEditor;
