// @univerjs/sheets-conditional-formatting-ui/lib/index.js
var ar = Object.defineProperty;
var lr = (t, e, n) => e in t ? ar(t, e, { enumerable: !0, configurable: !0, writable: !0, value: n }) : t[e] = n;
var Ie = (t, e, n) => lr(t, typeof e != "symbol" ? e + "" : e, n);
import { CommandType as ge, ICommandService as te, IUniverInstanceService as Z, ColorKit as gn, BooleanNumber as se, LocaleService as re, UniverInstanceType as A, Tools as jt, get as cr, set as ur, createInterceptorKey as mn, InterceptorManager as dr, Inject as P, Disposable as le, Injector as me, Rectangle as Ce, generateRandomId as hr, ObjectMatrix as Ne, Range as ye, toDisposable as fn, InterceptorEffectEnum as vr, DependentOn as pn, IConfigService as Cn, Plugin as yn, merge as bn, registerDependencies as gr, touchDependencies as Vt } from "@univerjs/core";
import { getSheetCommandTarget as be, SheetsSelectionsService as je, RemoveSheetMutation as mr, SetWorksheetActiveOperation as Bt, setEndForRange as fr, SetSelectionsOperation as pr, checkRangesEditablePermission as gt, SheetInterceptorService as Wt, ClearSelectionFormatCommand as Cr, ClearSelectionAllCommand as yr, RangeMergeUtil as br, rangeToDiscreteRange as Ir, findAllRectangle as mt, createTopMatrixFromMatrix as ft, SheetPermissionCheckController as Sr, WorksheetEditPermission as In, WorksheetSetCellStylePermission as Sn, RangeProtectionPermissionEditPoint as _n, WorkbookEditablePermission as wn, RefRangeService as _r, handleDefaultRangeChangeWithEffectRefCommands as wr, INTERCEPTOR_POINT as Rr, AFTER_CELL_EDIT as Tr } from "@univerjs/sheets";
import { ConditionalFormattingRuleModel as Q, CFSubRuleType as F, CFRuleType as L, AddConditionalRuleMutation as oe, CFNumberOperator as U, getColorScaleFromValue as xr, iconMap as Me, DEFAULT_BG_COLOR as Er, DEFAULT_FONT_COLOR as Nr, defaultDataBarPositiveColor as Ye, defaultDataBarNativeColor as Ke, CFValueType as $, createDefaultValueByValueType as Xe, removeUndefinedAttr as Mr, createDefaultValue as pt, CFTextOperator as H, CFTimePeriodOperator as he, iconGroup as Vr, getOppositeOperator as At, compareWithNumber as kr, EMPTY_ICON_TYPE as Fr, SetCfCommand as Lr, AddCfCommand as Rn, SetConditionalRuleMutation as Se, DeleteConditionalRuleMutation as _e, MoveConditionalRuleMutation as Tn, ClearWorksheetCfCommand as xn, DeleteCfCommand as zt, MoveCfCommand as Ur, ClearRangeCfCommand as Pr, createDefaultRule as Te, setConditionalRuleMutationUndoFactory as De, DeleteConditionalRuleMutationUndoFactory as Ae, ConditionalFormattingViewModel as rt, SHEET_CONDITIONAL_FORMATTING_PLUGIN as it, AddConditionalRuleMutationUndoFactory as En, isRangesEqual as Or, ConditionalFormattingService as Nn, DEFAULT_PADDING as $r, DEFAULT_WIDTH as Dr, UniverSheetsConditionalFormattingPlugin as Mn, CONDITIONAL_FORMATTING_VIEWPORT_CACHE_LENGTH as Ar } from "@univerjs/sheets-conditional-formatting";
import { useDependency as q, useSidebarClick as Et, ILayoutService as Hr, useScrollYOverContainer as jr, ISidebarService as Vn, useObservable as Br, ComponentManager as Wr, getMenuHiddenObservable as qr, MenuItemType as Gr, RibbonDataGroup as Zr, IMenuManagerService as zr } from "@univerjs/ui";
import { jsx as c, jsxs as k, Fragment as Ht } from "react/jsx-runtime";
import { useRef as ve, createElement as ce, forwardRef as ue, useMemo as J, useState as N, useEffect as X } from "react";
import { Dropdown as qt, clsx as Y, ColorPicker as Yr, borderClassName as Fe, Select as ae, InputNumber as Ve, RadioGroup as Kr, Radio as Yt, Checkbox as Ct, Input as Xr, Button as Kt, Tooltip as Xt, ReactGridLayout as Jr } from "@univerjs/design";
import { serializeRange as kn, deserializeRangeWithSheet as Qr } from "@univerjs/engine-formula";
import { FormulaEditor as Nt, RangeSelector as ei } from "@univerjs/sheets-formula-ui";
import { useHighlightRange as ti, ISheetClipboardService as ni, COPY_TYPE as Jt, PREDEFINED_HOOK_NAME as kt, virtualizeDiscreteRanges as Fn, getRepeatRange as ri, SheetSkeletonManagerService as Ln, IAutoFillService as ii, APPLY_TYPE as Ft, getAutoFillRepeatRange as oi, getCurrentRangeDisable$ as si, IFormatPainterService as ai, FormatPainterStatus as Lt } from "@univerjs/sheets-ui";
import { Observable as Ee, debounceTime as Qt, merge as Un } from "rxjs";
import { IRenderManagerService as Pn } from "@univerjs/engine-render";
import { bufferTime as li, filter as en, debounceTime as tn } from "rxjs/operators";
const On = {
  type: ge.COMMAND,
  id: "sheet.command.add-average-conditional-rule",
  handler(t, e) {
    if (!e)
      return !1;
    const { ranges: n, style: r, stopIfTrue: i, operator: o } = e, s = t.get(Q), d = t.get(te), a = t.get(Z), h = be(a);
    if (!h) return !1;
    const { unitId: l, subUnitId: u } = h, v = s.createCfId(l, u), p = {
      ranges: n,
      cfId: v,
      stopIfTrue: !!i,
      rule: {
        type: L.highlightCell,
        subType: F.average,
        operator: o,
        style: r
      }
    };
    return d.executeCommand(oe.id, { unitId: l, subUnitId: u, rule: p });
  }
}, $n = {
  type: ge.COMMAND,
  id: "sheet.command.add-color-scale-conditional-rule",
  handler(t, e) {
    if (!e)
      return !1;
    const { ranges: n, config: r, stopIfTrue: i } = e, o = t.get(Q), s = t.get(te), d = t.get(Z), a = be(d);
    if (!a) return !1;
    const { unitId: h, subUnitId: l } = a, u = o.createCfId(h, l), v = {
      ranges: n,
      cfId: u,
      stopIfTrue: !!i,
      rule: {
        type: L.colorScale,
        config: r
      }
    };
    return s.executeCommand(oe.id, { unitId: h, subUnitId: l, rule: v });
  }
}, Dn = {
  type: ge.COMMAND,
  id: "sheet.command.add-data-bar-conditional-rule",
  handler(t, e) {
    if (!e)
      return !1;
    const { ranges: n, min: r, max: i, nativeColor: o, positiveColor: s, isGradient: d, stopIfTrue: a, isShowValue: h } = e, l = t.get(Q), u = t.get(Z), v = be(u);
    if (!v) return !1;
    const { unitId: p, subUnitId: y } = v, f = t.get(te), w = l.createCfId(p, y), I = {
      ranges: n,
      cfId: w,
      stopIfTrue: !!a,
      rule: {
        type: L.dataBar,
        isShowValue: h,
        config: {
          min: r,
          max: i,
          nativeColor: o,
          positiveColor: s,
          isGradient: d
        }
      }
    };
    return f.executeCommand(oe.id, { unitId: p, subUnitId: y, rule: I });
  }
}, An = {
  type: ge.COMMAND,
  id: "sheet.command.add-duplicate-values-conditional-rule",
  handler(t, e) {
    if (!e)
      return !1;
    const { ranges: n, style: r, stopIfTrue: i } = e, o = t.get(Q), s = t.get(te), d = t.get(Z), a = be(d);
    if (!a) return !1;
    const { unitId: h, subUnitId: l } = a, u = o.createCfId(h, l), v = {
      ranges: n,
      cfId: u,
      stopIfTrue: !!i,
      rule: {
        type: L.highlightCell,
        subType: F.duplicateValues,
        style: r
      }
    };
    return s.executeCommand(oe.id, { unitId: h, subUnitId: l, rule: v });
  }
}, Hn = {
  type: ge.COMMAND,
  id: "sheet.command.add-number-conditional-rule",
  handler(t, e) {
    if (!e)
      return !1;
    const { ranges: n, style: r, stopIfTrue: i, operator: o, value: s } = e, d = t.get(Q), a = t.get(te), h = t.get(Z), l = be(h);
    if (!l) return !1;
    const { unitId: u, subUnitId: v } = l, p = d.createCfId(u, v);
    let y;
    if ([U.between, U.notBetween].includes(o)) {
      const f = s;
      if (f.length !== 2 || !Array.isArray(f))
        return !1;
      y = {
        ranges: n,
        cfId: p,
        stopIfTrue: !!i,
        rule: {
          type: L.highlightCell,
          subType: F.number,
          operator: o,
          style: r,
          value: f
        }
      };
    } else {
      const f = s;
      if (typeof f != "number")
        return !1;
      y = {
        ranges: n,
        cfId: p,
        stopIfTrue: !!i,
        rule: {
          type: L.highlightCell,
          subType: F.number,
          operator: o,
          style: r,
          value: f
        }
      };
    }
    return a.executeCommand(oe.id, { unitId: u, subUnitId: v, rule: y });
  }
}, jn = {
  type: ge.COMMAND,
  id: "sheet.command.add-rank-conditional-rule",
  handler(t, e) {
    if (!e)
      return !1;
    const { ranges: n, style: r, stopIfTrue: i, isPercent: o, isBottom: s, value: d } = e, a = t.get(Q), h = t.get(Z), l = t.get(te), u = be(h);
    if (!u) return !1;
    const { unitId: v, subUnitId: p } = u, y = a.createCfId(v, p), f = {
      ranges: n,
      cfId: y,
      stopIfTrue: !!i,
      rule: {
        type: L.highlightCell,
        subType: F.rank,
        isPercent: o,
        isBottom: s,
        style: r,
        value: d
      }
    };
    return l.executeCommand(oe.id, { unitId: v, subUnitId: p, rule: f });
  }
}, Bn = {
  type: ge.COMMAND,
  id: "sheet.command.add-text-conditional-rule",
  handler(t, e) {
    if (!e)
      return !1;
    const { ranges: n, style: r, stopIfTrue: i, operator: o, value: s } = e, d = t.get(Q), a = t.get(Z), h = be(a);
    if (!h) return !1;
    const { unitId: l, subUnitId: u } = h, v = t.get(te), p = d.createCfId(l, u), y = {
      ranges: n,
      cfId: p,
      stopIfTrue: !!i,
      rule: {
        type: L.highlightCell,
        subType: F.text,
        operator: o,
        style: r,
        value: s
      }
    };
    return v.executeCommand(oe.id, { unitId: l, subUnitId: u, rule: y });
  }
}, Wn = {
  type: ge.COMMAND,
  id: "sheet.command.add-time-period-conditional-rule",
  handler(t, e) {
    if (!e)
      return !1;
    const { ranges: n, style: r, stopIfTrue: i, operator: o } = e, s = t.get(Q), d = t.get(Z), a = t.get(te), h = be(d);
    if (!h) return !1;
    const { unitId: l, subUnitId: u } = h, v = s.createCfId(l, u), p = {
      ranges: n,
      cfId: v,
      stopIfTrue: !!i,
      rule: {
        type: L.highlightCell,
        subType: F.timePeriod,
        operator: o,
        style: r
      }
    };
    return a.executeCommand(oe.id, { unitId: l, subUnitId: u, rule: p });
  }
}, qn = {
  type: ge.COMMAND,
  id: "sheet.command.add-uniqueValues-conditional-rule",
  handler(t, e) {
    if (!e)
      return !1;
    const { ranges: n, style: r, stopIfTrue: i } = e, o = t.get(Q), s = t.get(Z), d = t.get(te), a = be(s);
    if (!a) return !1;
    const { unitId: h, subUnitId: l } = a, u = o.createCfId(h, l), v = {
      ranges: n,
      cfId: u,
      stopIfTrue: !!i,
      rule: {
        type: L.highlightCell,
        subType: F.uniqueValues,
        style: r
      }
    };
    return d.executeCommand(oe.id, { unitId: h, subUnitId: l, rule: v });
  }
};
function de({ ref: t, ...e }) {
  const { icon: n, id: r, className: i, extend: o, ...s } = e, d = `univerjs-icon univerjs-icon-${r} ${i || ""}`.trim(), a = ve(`_${di()}`);
  return Gn(n, `${r}`, {
    defIds: n.defIds,
    idSuffix: a.current
  }, {
    ref: t,
    className: d,
    ...s
  }, o);
}
function Gn(t, e, n, r, i) {
  return ce(t.tag, {
    key: e,
    ...ci(t, n, i),
    ...r
  }, (ui(t, n).children || []).map((o, s) => Gn(o, `${e}-${t.tag}-${s}`, n, void 0, i)));
}
function ci(t, e, n) {
  const r = { ...t.attrs };
  n != null && n.colorChannel1 && r.fill === "colorChannel1" && (r.fill = n.colorChannel1), t.tag === "mask" && r.id && (r.id = r.id + e.idSuffix), Object.entries(r).forEach(([o, s]) => {
    o === "mask" && typeof s == "string" && (r[o] = s.replace(/url\(#(.*)\)/, `url(#$1${e.idSuffix})`));
  });
  const { defIds: i } = e;
  return !i || i.length === 0 || (t.tag === "use" && r["xlink:href"] && (r["xlink:href"] = r["xlink:href"] + e.idSuffix), Object.entries(r).forEach(([o, s]) => {
    typeof s == "string" && (r[o] = s.replace(/url\(#(.*)\)/, `url(#$1${e.idSuffix})`));
  })), r;
}
function ui(t, e) {
  var r;
  const { defIds: n } = e;
  return !n || n.length === 0 ? t : t.tag === "defs" && ((r = t.children) != null && r.length) ? {
    ...t,
    children: t.children.map((i) => typeof i.attrs.id == "string" && n && n.includes(i.attrs.id) ? {
      ...i,
      attrs: {
        ...i.attrs,
        id: i.attrs.id + e.idSuffix
      }
    } : i)
  } : t;
}
function di() {
  return Math.random().toString(36).substring(2, 8);
}
de.displayName = "UniverIcon";
const hi = {
  tag: "svg",
  attrs: {
    xmlns: "http://www.w3.org/2000/svg",
    fill: "none",
    viewBox: "0 0 16 16",
    width: "1em",
    height: "1em"
  },
  children: [{
    tag: "path",
    attrs: {
      fill: "currentColor",
      d: "M4.65791 1.30005C3.77355 1.30005 3.05664 2.01696 3.05664 2.90132V12.8588C3.05664 13.8755 3.88086 14.6998 4.89759 14.6998H9.1016C11.2233 14.6998 12.9433 12.9798 12.9433 10.8581C12.9433 9.13925 11.8145 7.68407 10.2578 7.1934C10.8806 6.56856 11.2655 5.70659 11.2655 4.75472C11.2655 2.84676 9.71883 1.30005 7.81087 1.30005H4.65791ZM4.25664 2.90132C4.25664 2.6797 4.4363 2.50005 4.65791 2.50005H7.81087C9.05609 2.50005 10.0655 3.5095 10.0655 4.75472C10.0655 5.99993 9.05609 7.00938 7.81087 7.00938H4.25664V2.90132ZM4.25664 12.8588V8.21636H9.1016C10.5606 8.21636 11.7433 9.39909 11.7433 10.8581C11.7433 12.317 10.5606 13.4998 9.1016 13.4998H4.89759C4.5436 13.4998 4.25664 13.2128 4.25664 12.8588Z",
      fillRule: "evenodd",
      clipRule: "evenodd"
    }
  }]
}, Zn = ue(function(e, n) {
  return ce(de, Object.assign({}, e, {
    id: "bold-icon",
    ref: n,
    icon: hi
  }));
});
Zn.displayName = "BoldIcon";
const vi = {
  tag: "svg",
  attrs: {
    xmlns: "http://www.w3.org/2000/svg",
    fill: "none",
    viewBox: "0 0 16 16",
    width: "1em",
    height: "1em"
  },
  children: [
    {
      tag: "path",
      attrs: {
        fill: "currentColor",
        d: "M5.3313 1.4667C5.3313 1.13533 5.59993 0.866699 5.9313 0.866699H10.069C10.4004 0.866699 10.669 1.13533 10.669 1.4667C10.669 1.79807 10.4004 2.0667 10.069 2.0667H5.9313C5.59993 2.0667 5.3313 1.79807 5.3313 1.4667Z"
      }
    },
    {
      tag: "path",
      attrs: {
        fill: "currentColor",
        d: "M1.09985 3.64443C1.09985 3.31306 1.36848 3.04443 1.69985 3.04443H14.2999C14.6312 3.04443 14.8999 3.31306 14.8999 3.64443C14.8999 3.9758 14.6312 4.24443 14.2999 4.24443H1.69985C1.36848 4.24443 1.09985 3.9758 1.09985 3.64443Z"
      }
    },
    {
      tag: "path",
      attrs: {
        fill: "currentColor",
        d: "M6.12398 8.30171C6.35829 8.0674 6.73819 8.0674 6.97251 8.30171L8.00007 9.32928L9.02764 8.30171C9.26195 8.0674 9.64185 8.0674 9.87617 8.30171C10.1105 8.53603 10.1105 8.91593 9.87617 9.15024L8.8486 10.1778L9.87617 11.2054C10.1105 11.4397 10.1105 11.8196 9.87617 12.0539C9.64185 12.2882 9.26195 12.2882 9.02764 12.0539L8.00007 11.0263L6.97251 12.0539C6.73819 12.2882 6.35829 12.2882 6.12398 12.0539C5.88966 11.8196 5.88966 11.4397 6.12398 11.2054L7.15154 10.1778L6.12398 9.15024C5.88966 8.91593 5.88966 8.53603 6.12398 8.30171Z"
      }
    },
    {
      tag: "path",
      attrs: {
        fill: "currentColor",
        d: "M4.75332 5.22217C3.86966 5.22217 3.15332 5.93851 3.15332 6.82217V12.5331C3.15332 13.9691 4.31738 15.1332 5.75332 15.1332H10.2465C11.6825 15.1332 12.8465 13.9691 12.8465 12.5331V6.82217C12.8465 5.93851 12.1302 5.22217 11.2465 5.22217H4.75332ZM4.35332 6.82217C4.35332 6.60125 4.53241 6.42217 4.75332 6.42217H11.2465C11.4674 6.42217 11.6465 6.60125 11.6465 6.82217V12.5331C11.6465 13.3063 11.0197 13.9332 10.2465 13.9332H5.75332C4.98012 13.9332 4.35332 13.3063 4.35332 12.5331V6.82217Z",
        fillRule: "evenodd",
        clipRule: "evenodd"
      }
    }
  ]
}, vt = ue(function(e, n) {
  return ce(de, Object.assign({}, e, {
    id: "delete-icon",
    ref: n,
    icon: vi
  }));
});
vt.displayName = "DeleteIcon";
const gi = {
  tag: "svg",
  attrs: {
    xmlns: "http://www.w3.org/2000/svg",
    fill: "none",
    viewBox: "0 0 16 16",
    width: "1em",
    height: "1em"
  },
  children: [{
    tag: "path",
    attrs: {
      fill: "currentColor",
      d: "M8.6 1.99991C8.60001 1.66854 8.33138 1.39991 8.00001 1.3999C7.66864 1.3999 7.40001 1.66853 7.4 1.9999L7.39996 7.3999H1.9999C1.66853 7.3999 1.3999 7.66853 1.3999 7.9999C1.3999 8.33127 1.66853 8.5999 1.9999 8.5999H7.39995L7.3999 13.9999C7.3999 14.3313 7.66853 14.5999 7.9999 14.5999C8.33127 14.5999 8.5999 14.3313 8.5999 13.9999L8.59995 8.5999H13.9999C14.3313 8.5999 14.5999 8.33127 14.5999 7.9999C14.5999 7.66853 14.3313 7.3999 13.9999 7.3999H8.59996L8.6 1.99991Z"
    }
  }]
}, zn = ue(function(e, n) {
  return ce(de, Object.assign({}, e, {
    id: "increase-icon",
    ref: n,
    icon: gi
  }));
});
zn.displayName = "IncreaseIcon";
const mi = {
  tag: "svg",
  attrs: {
    xmlns: "http://www.w3.org/2000/svg",
    fill: "none",
    viewBox: "0 0 16 16",
    width: "1em",
    height: "1em"
  },
  children: [{
    tag: "path",
    attrs: {
      fill: "currentColor",
      d: "M9.80385 1.40005H11.9997C12.3311 1.40005 12.5997 1.66868 12.5997 2.00005C12.5997 2.33143 12.3311 2.60005 11.9997 2.60005H10.1185L7.12251 13.4001H9.33324C9.66461 13.4001 9.93324 13.6687 9.93324 14.0001C9.93324 14.3314 9.66461 14.6001 9.33324 14.6001H6.34785C6.33847 14.6003 6.32905 14.6003 6.31962 14.6001H3.9999C3.66853 14.6001 3.3999 14.3314 3.3999 14.0001C3.3999 13.6687 3.66853 13.4001 3.9999 13.4001H5.87719L8.87322 2.60005H6.66641C6.33504 2.60005 6.06641 2.33143 6.06641 2.00005C6.06641 1.66868 6.33504 1.40005 6.66641 1.40005H9.52916C9.61698 1.37929 9.71064 1.3781 9.80385 1.40005Z"
    }
  }]
}, Yn = ue(function(e, n) {
  return ce(de, Object.assign({}, e, {
    id: "italic-icon",
    ref: n,
    icon: mi
  }));
});
Yn.displayName = "ItalicIcon";
const fi = {
  tag: "svg",
  attrs: {
    xmlns: "http://www.w3.org/2000/svg",
    fill: "none",
    viewBox: "0 0 16 16",
    width: "1em",
    height: "1em"
  },
  children: [{
    tag: "path",
    attrs: {
      fill: "currentColor",
      d: "M11.3536 6.14645C11.5488 6.34171 11.5488 6.65829 11.3536 6.85355L8.35355 9.85355C8.15829 10.0488 7.84171 10.0488 7.64645 9.85355L4.64645 6.85355C4.45118 6.65829 4.45118 6.34171 4.64645 6.14645C4.84171 5.95118 5.15829 5.95118 5.35355 6.14645L8 8.79289L10.6464 6.14645C10.8417 5.95118 11.1583 5.95118 11.3536 6.14645Z",
      fillRule: "evenodd",
      clipRule: "evenodd"
    }
  }]
}, Gt = ue(function(e, n) {
  return ce(de, Object.assign({}, e, {
    id: "more-down-icon",
    ref: n,
    icon: fi
  }));
});
Gt.displayName = "MoreDownIcon";
const pi = {
  tag: "svg",
  attrs: {
    xmlns: "http://www.w3.org/2000/svg",
    fill: "none",
    viewBox: "0 0 16 16",
    width: "1em",
    height: "1em"
  },
  children: [
    {
      tag: "path",
      attrs: {
        fill: "currentColor",
        d: "M6 5C6.55228 5 7 4.55228 7 4C7 3.44772 6.55228 3 6 3C5.44772 3 5 3.44772 5 4C5 4.55228 5.44772 5 6 5Z"
      }
    },
    {
      tag: "path",
      attrs: {
        fill: "currentColor",
        d: "M6 9C6.55228 9 7 8.55229 7 8C7 7.44772 6.55228 7 6 7C5.44772 7 5 7.44772 5 8C5 8.55229 5.44772 9 6 9Z"
      }
    },
    {
      tag: "path",
      attrs: {
        fill: "currentColor",
        d: "M7 12C7 12.5523 6.55228 13 6 13C5.44772 13 5 12.5523 5 12C5 11.4477 5.44772 11 6 11C6.55228 11 7 11.4477 7 12Z"
      }
    },
    {
      tag: "path",
      attrs: {
        fill: "currentColor",
        d: "M10 5C10.5523 5 11 4.55228 11 4C11 3.44772 10.5523 3 10 3C9.44771 3 9 3.44772 9 4C9 4.55228 9.44771 5 10 5Z"
      }
    },
    {
      tag: "path",
      attrs: {
        fill: "currentColor",
        d: "M11 8C11 8.55229 10.5523 9 10 9C9.44771 9 9 8.55229 9 8C9 7.44772 9.44771 7 10 7C10.5523 7 11 7.44772 11 8Z"
      }
    },
    {
      tag: "path",
      attrs: {
        fill: "currentColor",
        d: "M10 13C10.5523 13 11 12.5523 11 12C11 11.4477 10.5523 11 10 11C9.44771 11 9 11.4477 9 12C9 12.5523 9.44771 13 10 13Z"
      }
    }
  ]
}, Kn = ue(function(e, n) {
  return ce(de, Object.assign({}, e, {
    id: "sequence-icon",
    ref: n,
    icon: pi
  }));
});
Kn.displayName = "SequenceIcon";
const Ci = {
  tag: "svg",
  attrs: {
    xmlns: "http://www.w3.org/2000/svg",
    fill: "none",
    viewBox: "0 0 16 16",
    width: "1em",
    height: "1em"
  },
  children: [{
    tag: "path",
    attrs: {
      fill: "currentColor",
      d: "M8.00033 1.20996C6.43345 1.20996 5.23596 1.88077 4.64804 2.94304C4.06338 3.99943 4.15123 5.32084 4.90099 6.47062C5.08199 6.7482 5.45374 6.82648 5.73131 6.64548C6.00888 6.46448 6.08716 6.09273 5.90616 5.81516C5.36418 4.98402 5.35768 4.13897 5.69797 3.52412C6.035 2.91515 6.78176 2.40996 8.00033 2.40996C9.19423 2.40996 10.0017 2.93409 10.5236 3.46755C10.7871 3.73682 10.9756 4.00712 11.0974 4.20903C11.158 4.30951 11.2013 4.39164 11.2284 4.44625C11.2419 4.47351 11.2514 4.49377 11.2568 4.50588L11.2621 4.51771C11.3901 4.82216 11.7407 4.96686 12.046 4.83992C12.3519 4.71267 12.4968 4.36149 12.3695 4.05552L11.8231 4.28278C12.3695 4.05552 12.3697 4.0558 12.3695 4.05552L12.3691 4.05437L12.3685 4.05298L12.367 4.04953L12.3629 4.03993L12.3498 4.01036C12.3388 3.98619 12.3234 3.95326 12.3033 3.91282C12.2632 3.83203 12.2043 3.72072 12.1249 3.58908C11.9666 3.32674 11.7235 2.97805 11.3814 2.62834C10.6921 1.92381 9.59192 1.20996 8.00033 1.20996Z"
    }
  }, {
    tag: "path",
    attrs: {
      fill: "currentColor",
      d: "M2.11953 8.5999H7.88935C10.1109 9.46549 10.5895 10.7649 10.4268 11.6869C10.2475 12.703 9.26939 13.5898 8.00009 13.5898C6.63323 13.5898 5.77631 13.0502 5.25534 12.528C4.99021 12.2623 4.80823 11.9965 4.69365 11.7993C4.63659 11.7011 4.59698 11.6211 4.5727 11.5686C4.56059 11.5424 4.55236 11.5231 4.54774 11.512L4.54352 11.5016C4.4267 11.193 4.0823 11.0363 3.77279 11.1516C3.46228 11.2673 3.30438 11.6129 3.4201 11.9234L3.9649 11.7203C3.4201 11.9234 3.41998 11.9231 3.4201 11.9234L3.4206 11.9247L3.42118 11.9263L3.42262 11.9301L3.42661 11.9403C3.42977 11.9484 3.43394 11.9588 3.43914 11.9713C3.44956 11.9965 3.46418 12.0304 3.48334 12.0719C3.52161 12.1547 3.57828 12.2682 3.65608 12.4021C3.81124 12.6692 4.0534 13.0223 4.40584 13.3756C5.12 14.0914 6.27196 14.7898 8.00009 14.7898C9.82574 14.7898 11.3236 13.5101 11.6085 11.8955C11.812 10.7422 11.3901 9.55402 10.2665 8.5999H13.8804C14.2117 8.5999 14.4803 8.33127 14.4803 7.9999C14.4803 7.66853 14.2117 7.3999 13.8804 7.3999H8.01411C8.0048 7.39968 7.99551 7.39969 7.98625 7.3999H2.11953C1.78816 7.3999 1.51953 7.66853 1.51953 7.9999C1.51953 8.33127 1.78816 8.5999 2.11953 8.5999Z"
    }
  }]
}, Xn = ue(function(e, n) {
  return ce(de, Object.assign({}, e, {
    id: "strikethrough-icon",
    ref: n,
    icon: Ci
  }));
});
Xn.displayName = "StrikethroughIcon";
const yi = {
  tag: "svg",
  attrs: {
    xmlns: "http://www.w3.org/2000/svg",
    fill: "none",
    viewBox: "0 0 16 16",
    width: "1em",
    height: "1em"
  },
  children: [{
    tag: "path",
    attrs: {
      fill: "currentColor",
      d: "M4.46416 2.03237C4.46416 1.701 4.19553 1.43237 3.86416 1.43237C3.53279 1.43237 3.26416 1.701 3.26416 2.03237V7.33338C3.26416 10.0313 5.35834 12.2667 7.9999 12.2667C10.6415 12.2667 12.7356 10.0313 12.7356 7.33338V2.03237C12.7356 1.701 12.467 1.43237 12.1356 1.43237C11.8043 1.43237 11.5356 1.701 11.5356 2.03237V7.33338C11.5356 9.42194 9.92656 11.0667 7.9999 11.0667C6.07325 11.0667 4.46416 9.42194 4.46416 7.33338V2.03237Z"
    }
  }, {
    tag: "path",
    attrs: {
      fill: "currentColor",
      d: "M2.66641 13.4663C2.33504 13.4663 2.06641 13.7349 2.06641 14.0663C2.06641 14.3977 2.33504 14.6663 2.66641 14.6663H13.3331C13.6644 14.6663 13.9331 14.3977 13.9331 14.0663C13.9331 13.7349 13.6644 13.4663 13.3331 13.4663H2.66641Z"
    }
  }]
}, Jn = ue(function(e, n) {
  return ce(de, Object.assign({}, e, {
    id: "underline-icon",
    ref: n,
    icon: yi
  }));
});
Jn.displayName = "UnderlineIcon";
const bi = {
  tag: "svg",
  attrs: {
    xmlns: "http://www.w3.org/2000/svg",
    fill: "none",
    viewBox: "0 0 17 16",
    width: "1em",
    height: "1em"
  },
  children: [{
    tag: "path",
    attrs: {
      fill: "colorChannel1",
      d: "M1.98682 13.4992C1.98682 12.5603 2.74793 11.7992 3.68682 11.7992H14.2868C15.2257 11.7992 15.9868 12.5603 15.9868 13.4992V13.4992C15.9868 14.4381 15.2257 15.1992 14.2868 15.1992H3.68682C2.74793 15.1992 1.98682 14.4381 1.98682 13.4992V13.4992Z"
    }
  }, {
    tag: "path",
    attrs: {
      fill: "currentColor",
      d: "M10.3014 1.70389C9.85268 0.479568 8.12109 0.479557 7.67238 1.70389L4.72235 9.75326C4.60832 10.0644 4.76811 10.4091 5.07924 10.5231C5.39038 10.6371 5.73504 10.4773 5.84907 10.1662L6.99975 7.02646C7.03588 7.03324 7.07314 7.03679 7.11123 7.03679H10.9778L12.1247 10.1662C12.2387 10.4773 12.5834 10.6371 12.8945 10.5231C13.2057 10.4091 13.3654 10.0644 13.2514 9.75326L10.3014 1.70389ZM10.538 5.83679L9.17467 2.11682C9.11057 1.94192 8.8632 1.94192 8.7991 2.11682L7.43576 5.83679H10.538Z",
      fillRule: "evenodd",
      clipRule: "evenodd"
    }
  }]
}, Qn = ue(function(e, n) {
  return ce(de, Object.assign({}, e, {
    id: "font-color-double-icon",
    ref: n,
    icon: bi
  }));
});
Qn.displayName = "FontColorDoubleIcon";
const Ii = {
  tag: "svg",
  attrs: {
    xmlns: "http://www.w3.org/2000/svg",
    fill: "none",
    viewBox: "0 0 17 16",
    width: "1em",
    height: "1em"
  },
  children: [
    {
      tag: "path",
      attrs: {
        fill: "currentColor",
        d: "M8.01281 1.36643C8.79386 0.585377 10.0602 0.585378 10.8412 1.36643L12.9223 3.44752C13.7034 4.22857 13.7034 5.4949 12.9223 6.27595L9.36445 9.83383C8.5834 10.6149 7.31707 10.6149 6.53602 9.83383L4.45493 7.75273C3.67388 6.97168 3.67388 5.70535 4.45493 4.9243L8.01281 1.36643ZM9.9927 2.21495C9.68028 1.90253 9.17375 1.90253 8.86133 2.21495L5.30346 5.77283L5.29671 5.77966L10.839 6.66224L12.0738 5.42742C12.3862 5.115 12.3862 4.60847 12.0738 4.29605L9.9927 2.21495Z",
        fillRule: "evenodd",
        clipRule: "evenodd"
      }
    },
    {
      tag: "path",
      attrs: {
        fill: "currentColor",
        d: "M14.5179 9.48875C14.5179 9.99175 14.1101 10.3995 13.607 10.3995C13.1041 10.3995 12.6963 9.99175 12.6963 9.48875C12.6963 9.1773 13.0455 8.59966 13.3114 8.20487C13.4549 7.99177 13.7591 7.99177 13.9027 8.20486C14.1687 8.59965 14.5179 9.1773 14.5179 9.48875Z"
      }
    },
    {
      tag: "path",
      attrs: {
        fill: "colorChannel1",
        d: "M1.98682 13.4992C1.98682 12.5603 2.74793 11.7992 3.68682 11.7992H14.2868C15.2257 11.7992 15.9868 12.5603 15.9868 13.4992V13.4992C15.9868 14.4381 15.2257 15.1992 14.2868 15.1992H3.68682C2.74793 15.1992 1.98682 14.4381 1.98682 13.4992V13.4992Z"
      }
    }
  ]
}, er = ue(function(e, n) {
  return ce(de, Object.assign({}, e, {
    id: "paint-bucket-double-icon",
    ref: n,
    icon: Ii
  }));
});
er.displayName = "PaintBucketDoubleIcon";
const Si = {
  tag: "svg",
  attrs: {
    xmlns: "http://www.w3.org/2000/svg",
    fill: "none",
    viewBox: "0 0 17 17",
    width: "1em",
    height: "1em"
  },
  children: [{
    tag: "path",
    attrs: {
      fill: "currentColor",
      d: "M2.27177 13.714L13.5855 2.40025L14.434 3.24878L3.1203 14.5625L2.27177 13.714Z",
      fillRule: "evenodd",
      clipRule: "evenodd"
    }
  }, {
    tag: "path",
    attrs: {
      fill: "colorChannel1",
      d: "M4.2603 1.82495L2.89617 1.82495C2.23343 1.82495 1.69617 2.36221 1.69617 3.02495L1.69617 4.38908L3.02617 4.38908L3.02617 3.15495L4.2603 3.15495L4.2603 1.82495ZM15.0092 4.38908L15.0092 3.02495C15.0092 2.36221 14.4719 1.82495 13.8092 1.82495L12.4451 1.82495L12.4451 3.15495L13.6792 3.15495L13.6792 4.38908L15.0092 4.38908ZM6.98856 1.82495L6.98856 3.15495L9.71681 3.15495L9.71681 1.82495L6.98856 1.82495ZM15.0092 7.11734L13.6792 7.11734L13.6792 9.84559L15.0092 9.84559L15.0092 7.11734ZM15.0092 12.5739L13.6792 12.5739L13.6792 13.808L12.4451 13.808L12.4451 15.138L13.8092 15.138C14.4719 15.138 15.0092 14.6007 15.0092 13.938L15.0092 12.5739ZM9.71681 15.138L9.71681 13.808L6.98856 13.808L6.98856 15.138L9.71681 15.138ZM4.2603 15.138L4.2603 13.808L3.02617 13.808L3.02617 12.5739L1.69617 12.5739L1.69617 13.938C1.69617 14.6007 2.23343 15.138 2.89617 15.138L4.2603 15.138ZM1.69617 9.8456L3.02617 9.8456L3.02617 7.11734L1.69617 7.11734L1.69617 9.8456Z",
      fillRule: "evenodd",
      clipRule: "evenodd"
    }
  }]
}, ot = ue(function(e, n) {
  return ce(de, Object.assign({}, e, {
    id: "slash-double-icon",
    ref: n,
    icon: Si
  }));
});
ot.displayName = "SlashDoubleIcon";
const ke = (t) => {
  const { color: e, onChange: n, disable: r = !1, Icon: i = er, className: o } = t, s = J(() => new gn(e), [e]), d = () => {
    const a = {
      className: Y("univer-fill-primary-600", r && o),
      extend: { colorChannel1: s.isValid ? e : "" }
    };
    return /* @__PURE__ */ c(i, { ...a });
  };
  return r ? d() : /* @__PURE__ */ c(
    qt,
    {
      overlay: /* @__PURE__ */ c("div", { className: "univer-rounded-lg univer-p-4", children: /* @__PURE__ */ c(Yr, { value: e, onChange: n }) }),
      children: /* @__PURE__ */ c(
        "span",
        {
          className: Y("univer-flex univer-cursor-pointer univer-items-center univer-rounded univer-p-1 hover:univer-bg-gray-100 dark:hover:!univer-bg-gray-700", o),
          children: d()
        }
      )
    }
  );
}, Be = (t) => {
  var o, s, d, a, h;
  const e = t.rule;
  if (!e) return null;
  const n = J(() => {
    if (e.type === L.colorScale) {
      const l = e.config.map((p, y) => ({ color: new gn(p.color), value: y })), u = l.length - 1;
      return new Array(5).fill("").map((p, y, f) => y * u / (f.length - 1)).map((p) => xr(l, p));
    }
    return null;
  }, [e]), r = J(() => {
    if (e.type === L.iconSet)
      return e.config.map((l) => {
        const u = Me[l.iconType];
        return u && u[Number(l.iconId)];
      });
  }, [e]), i = "univer-pointer-events-none univer-flex univer-h-5 univer-min-w-[72px] univer-items-center univer-justify-center univer-text-xs";
  switch (e.type) {
    case L.dataBar: {
      const { isGradient: l } = e.config, u = l ? `linear-gradient(to right, ${e.config.positiveColor || Ye}, rgb(255 255 255))` : e.config.positiveColor, v = l ? `linear-gradient(to right,  rgb(255 255 255),${e.config.nativeColor || Ke})` : e.config.nativeColor;
      return /* @__PURE__ */ k("div", { className: i, children: [
        /* @__PURE__ */ c(
          "div",
          {
            className: "univer-h-full univer-w-1/2 univer-border univer-border-solid",
            style: {
              background: v,
              borderColor: (o = e.config.nativeColor) != null ? o : Ke
            }
          }
        ),
        /* @__PURE__ */ c(
          "div",
          {
            className: "univer-h-full univer-w-1/2 univer-border univer-border-solid",
            style: {
              background: u,
              borderColor: (s = e.config.positiveColor) != null ? s : Ye
            }
          }
        )
      ] });
    }
    case L.colorScale:
      return n && /* @__PURE__ */ c("div", { className: i, children: n.map((l, u) => /* @__PURE__ */ c(
        "div",
        {
          className: "univer-h-full",
          style: { width: `${100 / n.length}%`, background: l }
        },
        u
      )) });
    case L.iconSet:
      return r && /* @__PURE__ */ c("div", { className: i, children: r.map((l, u) => l ? /* @__PURE__ */ c("img", { className: "univer-h-full", src: l, draggable: !1 }, u) : /* @__PURE__ */ c(ot, {}, u)) });
    case L.highlightCell: {
      const { ul: l, st: u, it: v, bl: p, bg: y, cl: f } = e.style, w = (l == null ? void 0 : l.s) === se.TRUE, I = (u == null ? void 0 : u.s) === se.TRUE, R = v === se.TRUE, x = p === se.TRUE, T = (d = y == null ? void 0 : y.rgb) != null ? d : Er, g = (a = f == null ? void 0 : f.rgb) != null ? a : Nr, b = {
        textDecoration: (h = `${w ? "underline" : ""} ${I ? "line-through" : ""}`.replace(/^ /, "")) != null ? h : void 0,
        backgroundColor: T,
        color: g
      };
      return /* @__PURE__ */ c(
        "div",
        {
          className: Y(i, {
            "univer-font-bold": x,
            "univer-italic": R
          }),
          style: b,
          children: "123"
        }
      );
    }
  }
}, st = Y("univer-mt-5 univer-px-1 univer-py-2 univer-rounded", Fe), xe = (t, e) => ({ label: e.t(`sheet.cf.valueType.${t}`), value: t }), Ut = (t) => {
  var y;
  const { type: e, className: n, onChange: r, id: i, value: o } = t, s = q(Z), d = s.getCurrentUnitForType(A.UNIVER_SHEET).getUnitId(), a = (y = s.getCurrentUnitForType(A.UNIVER_SHEET).getActiveSheet()) == null ? void 0 : y.getSheetId(), h = J(() => String(o).startsWith("=") ? String(o) : "=", [o]), l = J(() => [$.max, $.min, "none"].includes(e) ? { disabled: !0 } : [$.percent, $.percentile].includes(e) ? {
    min: 0,
    max: 100
  } : {
    min: Number.MIN_SAFE_INTEGER,
    max: Number.MAX_SAFE_INTEGER
  }, [e]), u = ve(null), [v, p] = N(!1);
  return Et((f) => {
    var I;
    ((I = u.current) == null ? void 0 : I.isClickOutSide(f)) && p(!1);
  }), e === $.formula ? /* @__PURE__ */ c("div", { className: "univer-ml-1 univer-w-full", children: /* @__PURE__ */ c(
    Nt,
    {
      ref: u,
      className: Y("univer-box-border univer-h-8 univer-w-full univer-cursor-pointer univer-items-center univer-rounded-lg univer-bg-white univer-pt-2 univer-transition-colors hover:univer-border-primary-600 dark:!univer-bg-gray-700 dark:!univer-text-white [&>div:first-child]:univer-px-2.5 [&>div]:univer-h-5 [&>div]:univer-ring-transparent", Fe),
      initValue: h,
      unitId: d,
      subUnitId: a,
      isFocus: v,
      onChange: (f = "") => {
        r(f || "");
      },
      onFocus: () => p(!0)
    }
  ) }) : /* @__PURE__ */ c(Ve, { className: n, value: Number(t.value) || 0, onChange: (f) => t.onChange(f || 0), ...l });
}, _i = (t) => {
  var O;
  const { interceptorManager: e } = t, n = q(re), r = ((O = t.rule) == null ? void 0 : O.type) === L.colorScale ? t.rule : void 0, i = [xe($.num, n), xe($.percent, n), xe($.percentile, n), xe($.formula, n)], o = [xe($.min, n), ...i], s = [xe("none", n), ...i], d = [xe($.max, n), ...i], [a, h] = N(() => {
    var _;
    const m = $.min;
    return r && ((_ = r.config[0]) == null ? void 0 : _.value.type) || m;
  }), [l, u] = N(() => {
    var _;
    const m = "none";
    return !r || r.config.length !== 3 ? m : ((_ = r.config[1]) == null ? void 0 : _.value.type) || m;
  }), [v, p] = N(() => {
    var _;
    const m = $.max;
    return r && ((_ = r.config[r.config.length - 1]) == null ? void 0 : _.value.type) || m;
  }), [y, f] = N(() => {
    if (!r)
      return 10;
    const _ = r.config[0];
    return (_ == null ? void 0 : _.value.value) === void 0 ? 10 : _ == null ? void 0 : _.value.value;
  }), [w, I] = N(() => {
    var j;
    if (!r || r.config.length !== 3)
      return 50;
    const _ = (j = r.config[1]) == null ? void 0 : j.value.value;
    return _ === void 0 ? 50 : _;
  }), [R, x] = N(() => {
    var j;
    if (!r)
      return 90;
    const _ = (j = r.config[r.config.length - 1]) == null ? void 0 : j.value.value;
    return _ === void 0 ? 90 : _;
  }), [T, g] = N(() => {
    var _;
    const m = "#d0d9fb";
    return r && ((_ = r.config[0]) == null ? void 0 : _.color) || m;
  }), [b, S] = N(() => {
    var _;
    const m = "#7790f3";
    return !r || r.config.length !== 3 ? m : ((_ = r.config[1]) == null ? void 0 : _.color) || m;
  }), [C, M] = N(() => {
    var _;
    const m = "#2e55ef";
    return r && ((_ = r.config[r.config.length - 1]) == null ? void 0 : _.color) || m;
  }), E = J(() => (m) => {
    const { minType: _, medianType: j, maxType: K, minValue: fe, medianValue: pe, maxValue: Le, minColor: Ue, medianColor: D, maxColor: B } = m, G = [];
    return G.push({ color: Ue, value: { type: _, value: fe } }), j !== "none" && G.push({ color: D, value: { type: j, value: pe } }), G.push({ color: B, value: { type: K, value: Le } }), { config: G.map((ee, ne) => ({ ...ee, index: ne })), type: L.colorScale };
  }, []);
  X(() => e.intercept(e.getInterceptPoints().submit, {
    handler() {
      return E({ minType: a, medianType: l, maxType: v, minValue: y, medianValue: w, maxValue: R, minColor: T, medianColor: b, maxColor: C });
    }
  }), [E, a, l, v, y, w, R, T, b, C, e]);
  const V = (m) => {
    t.onChange(E(m));
  };
  return /* @__PURE__ */ k("div", { children: [
    /* @__PURE__ */ c(
      "div",
      {
        className: "univer-mt-4 univer-text-sm univer-text-gray-600 dark:!univer-text-gray-200",
        children: n.t("sheet.cf.panel.styleRule")
      }
    ),
    /* @__PURE__ */ c("div", { className: st, children: /* @__PURE__ */ c(Be, { rule: E({ minType: a, medianType: l, maxType: v, minValue: y, medianValue: w, maxValue: R, minColor: T, medianColor: b, maxColor: C }) }) }),
    /* @__PURE__ */ c(
      "div",
      {
        className: "univer-mt-3 univer-text-xs univer-text-gray-600 dark:!univer-text-gray-200",
        children: n.t("sheet.cf.valueType.min")
      }
    ),
    /* @__PURE__ */ k("div", { className: "univer-mt-3 univer-flex univer-h-8 univer-items-center", children: [
      /* @__PURE__ */ c(
        ae,
        {
          className: "univer-flex-shrink-0",
          options: o,
          value: a,
          onChange: (m) => {
            h(m);
            const _ = Xe(m, 10);
            f(_), V({
              minType: m,
              medianType: l,
              maxType: v,
              minValue: _,
              medianValue: w,
              maxValue: R,
              minColor: T,
              medianColor: b,
              maxColor: C
            });
          }
        }
      ),
      /* @__PURE__ */ c(
        Ut,
        {
          id: "min",
          className: "univer-ml-1",
          value: y,
          type: a,
          onChange: (m) => {
            f(m), V({
              minType: a,
              medianType: l,
              maxType: v,
              minValue: m,
              medianValue: w,
              maxValue: R,
              minColor: T,
              medianColor: b,
              maxColor: C
            });
          }
        }
      ),
      /* @__PURE__ */ c(
        ke,
        {
          className: "univer-ml-1",
          color: T,
          onChange: (m) => {
            g(m), V({
              minType: a,
              medianType: l,
              maxType: v,
              minValue: y,
              medianValue: w,
              maxValue: R,
              minColor: m,
              medianColor: b,
              maxColor: C
            });
          }
        }
      )
    ] }),
    /* @__PURE__ */ c(
      "div",
      {
        className: "univer-text-gray-600dark:!univer-text-gray-200 univer-mt-3 univer-text-xs",
        children: n.t("sheet.cf.panel.medianValue")
      }
    ),
    /* @__PURE__ */ k("div", { className: "univer-mt-3 univer-flex univer-h-8 univer-items-center", children: [
      /* @__PURE__ */ c(
        ae,
        {
          className: "univer-flex-shrink-0",
          options: s,
          value: l,
          onChange: (m) => {
            u(m);
            const _ = Xe(m, 50);
            I(_), V({
              minType: a,
              medianType: m,
              maxType: v,
              minValue: y,
              medianValue: _,
              maxValue: R,
              minColor: T,
              medianColor: b,
              maxColor: C
            });
          }
        }
      ),
      /* @__PURE__ */ c(
        Ut,
        {
          id: "median",
          className: "univer-ml-1",
          value: w,
          type: l,
          onChange: (m) => {
            I(m), V({
              minType: a,
              medianType: l,
              maxType: v,
              minValue: y,
              medianValue: m,
              maxValue: R,
              minColor: T,
              medianColor: b,
              maxColor: C
            });
          }
        }
      ),
      l !== "none" && /* @__PURE__ */ c(
        ke,
        {
          className: "univer-ml-1",
          color: b,
          onChange: (m) => {
            S(m), V({
              minType: a,
              medianType: l,
              maxType: v,
              minValue: y,
              medianValue: w,
              maxValue: R,
              minColor: T,
              medianColor: m,
              maxColor: C
            });
          }
        }
      )
    ] }),
    /* @__PURE__ */ c(
      "div",
      {
        className: "univer-mt-3 univer-text-xs univer-text-gray-600 dark:!univer-text-gray-200",
        children: n.t("sheet.cf.valueType.max")
      }
    ),
    /* @__PURE__ */ k("div", { className: "univer-mt-3 univer-flex univer-h-8 univer-items-center", children: [
      /* @__PURE__ */ c(
        ae,
        {
          className: "univer-flex-shrink-0",
          options: d,
          value: v,
          onChange: (m) => {
            p(m);
            const _ = Xe(m, 90);
            x(_), V({
              minType: a,
              medianType: l,
              maxType: m,
              minValue: y,
              medianValue: w,
              maxValue: _,
              minColor: T,
              medianColor: b,
              maxColor: C
            });
          }
        }
      ),
      /* @__PURE__ */ c(
        Ut,
        {
          id: "max",
          className: "univer-ml-1",
          value: R,
          type: v,
          onChange: (m) => {
            x(m), V({
              minType: a,
              medianType: l,
              maxType: v,
              minValue: y,
              medianValue: w,
              maxValue: m,
              minColor: T,
              medianColor: b,
              maxColor: C
            });
          }
        }
      ),
      /* @__PURE__ */ c(
        ke,
        {
          className: "univer-ml-1",
          color: C,
          onChange: (m) => {
            M(m), V({
              minType: a,
              medianType: l,
              maxType: v,
              minValue: y,
              medianValue: w,
              maxValue: R,
              minColor: T,
              medianColor: b,
              maxColor: m
            });
          }
        }
      )
    ] })
  ] });
}, Pe = (t, e) => ({ label: e.t(`sheet.cf.valueType.${t}`), value: t }), nn = (t) => {
  var f;
  const { onChange: e, className: n, value: r, type: i, id: o, disabled: s = !1 } = t, d = q(Z), a = d.getCurrentUnitOfType(A.UNIVER_SHEET).getUnitId(), h = (f = d.getCurrentUnitOfType(A.UNIVER_SHEET).getActiveSheet()) == null ? void 0 : f.getSheetId(), l = ve(null), [u, v] = N(!1);
  Et((w) => {
    var R;
    ((R = l.current) == null ? void 0 : R.isClickOutSide(w)) && v(!1);
  });
  const p = ve(r), y = J(() => [$.percentile, $.percent].includes(i) ? {
    max: 100,
    min: 0
  } : {
    min: Number.MIN_SAFE_INTEGER,
    max: Number.MAX_SAFE_INTEGER
  }, [i]);
  if (i === $.formula) {
    const w = String(p.current).startsWith("=") ? String(p.current) || "" : "=";
    return /* @__PURE__ */ c("div", { className: "univer-ml-3 univer-w-full", children: /* @__PURE__ */ c(
      Nt,
      {
        ref: l,
        className: Y("univer-box-border univer-h-8 univer-w-full univer-cursor-pointer univer-items-center univer-rounded-lg univer-bg-white univer-pt-2 univer-transition-colors hover:univer-border-primary-600 dark:!univer-bg-gray-700 dark:!univer-text-white [&>div:first-child]:univer-px-2.5 [&>div]:univer-h-5 [&>div]:univer-ring-transparent", Fe),
        initValue: w,
        unitId: a,
        subUnitId: h,
        isFocus: u,
        onChange: (I = "") => {
          e(I || "");
        },
        onFocus: () => v(!0)
      }
    ) });
  }
  return /* @__PURE__ */ c(
    Ve,
    {
      className: n,
      value: Number(r) || 0,
      disabled: s,
      onChange: (w) => {
        e(w || 0);
      },
      ...y
    }
  );
}, wi = (t) => {
  var O;
  const { interceptorManager: e } = t, n = q(re), r = ((O = t.rule) == null ? void 0 : O.type) === L.dataBar ? t.rule : void 0, [i, o] = N(() => {
    var _;
    return r && (_ = r.config) != null && _.isGradient ? "1" : "0";
  }), [s, d] = N(() => {
    var m;
    return r ? ((m = r.config) == null ? void 0 : m.positiveColor) || Ye : Ye;
  }), [a, h] = N(() => {
    var m;
    return r ? ((m = r.config) == null ? void 0 : m.nativeColor) || Ke : Ke;
  }), l = [
    Pe($.num, n),
    Pe($.percent, n),
    Pe($.percentile, n),
    Pe($.formula, n)
  ], u = [Pe($.min, n), ...l], v = [Pe($.max, n), ...l], [p, y] = N(() => {
    var _;
    const m = u[0].value;
    return r && ((_ = r.config) == null ? void 0 : _.min.type) || m;
  }), [f, w] = N(() => {
    var _;
    const m = v[0].value;
    return r && ((_ = r.config) == null ? void 0 : _.max.type) || m;
  }), [I, R] = N(() => {
    var j;
    if (!r)
      return 0;
    const _ = ((j = r.config) == null ? void 0 : j.min) || {};
    return _.type === $.formula ? _.value || "=" : _.value || 0;
  }), [x, T] = N(() => {
    var j;
    if (!r)
      return 100;
    const _ = ((j = r.config) == null ? void 0 : j.max) || {};
    return _.type === $.formula ? _.value || "=" : _.value === void 0 ? 100 : _.value;
  }), [g, b] = N(() => r ? r.isShowValue === void 0 ? !0 : !!r.isShowValue : !0), S = (m) => ({ config: {
    min: { type: m.minValueType, value: m.minValue },
    max: { type: m.maxValueType, value: m.maxValue },
    isGradient: m.isGradient === "1",
    positiveColor: m.positiveColor || Ye,
    nativeColor: m.nativeColor || Ke
  }, type: L.dataBar, isShowValue: m.isShowValue });
  X(() => e.intercept(e.getInterceptPoints().submit, {
    handler() {
      return S({ isGradient: i, minValue: I, minValueType: p, maxValue: x, maxValueType: f, positiveColor: s, nativeColor: a, isShowValue: g });
    }
  }), [i, I, p, x, f, s, a, e, g]);
  const C = (m) => {
    t.onChange(S(m));
  }, M = (m) => {
    d(m), C({
      isGradient: i,
      minValue: I,
      minValueType: p,
      maxValue: x,
      maxValueType: f,
      positiveColor: m,
      nativeColor: a,
      isShowValue: g
    });
  }, E = (m) => {
    h(m), C({
      isGradient: i,
      minValue: I,
      minValueType: p,
      maxValue: x,
      maxValueType: f,
      positiveColor: s,
      nativeColor: m,
      isShowValue: g
    });
  }, V = (m) => l.map((_) => _.value).includes(m);
  return /* @__PURE__ */ k("div", { children: [
    /* @__PURE__ */ c(
      "div",
      {
        className: "univer-mt-4 univer-text-sm univer-text-gray-600 dark:!univer-text-gray-200",
        children: n.t("sheet.cf.panel.styleRule")
      }
    ),
    /* @__PURE__ */ c("div", { className: st, children: /* @__PURE__ */ c(
      Be,
      {
        rule: S({
          isGradient: i,
          minValue: I,
          minValueType: p,
          maxValue: x,
          maxValueType: f,
          positiveColor: s,
          nativeColor: a,
          isShowValue: g
        })
      }
    ) }),
    /* @__PURE__ */ k("div", { children: [
      /* @__PURE__ */ c(
        "div",
        {
          className: "univer-mt-3 univer-text-sm univer-text-gray-600 dark:!univer-text-gray-200",
          children: n.t("sheet.cf.panel.fillType")
        }
      ),
      /* @__PURE__ */ k("div", { className: "univer-ml-1 univer-mt-3 univer-flex univer-items-center", children: [
        /* @__PURE__ */ k(
          Kr,
          {
            value: i,
            onChange: (m) => {
              o(m), C({
                isGradient: m,
                minValue: I,
                minValueType: p,
                maxValue: x,
                maxValueType: f,
                positiveColor: s,
                nativeColor: a,
                isShowValue: g
              });
            },
            children: [
              /* @__PURE__ */ c(Yt, { value: "0", children: /* @__PURE__ */ c("span", { className: "univer-text-xs", children: n.t("sheet.cf.panel.pureColor") }) }),
              /* @__PURE__ */ c(Yt, { value: "1", children: /* @__PURE__ */ c("span", { className: "univer-text-xs", children: n.t("sheet.cf.panel.gradient") }) })
            ]
          }
        ),
        /* @__PURE__ */ k("div", { className: "univer-ml-6 univer-flex univer-items-center univer-text-xs", children: [
          /* @__PURE__ */ c(
            Ct,
            {
              checked: !g,
              onChange: (m) => {
                b(!m), C({
                  isGradient: m,
                  minValue: I,
                  minValueType: p,
                  maxValue: x,
                  maxValueType: f,
                  positiveColor: s,
                  nativeColor: a,
                  isShowValue: !m
                });
              }
            }
          ),
          n.t("sheet.cf.panel.onlyShowDataBar")
        ] })
      ] })
    ] }),
    /* @__PURE__ */ k("div", { children: [
      /* @__PURE__ */ c(
        "div",
        {
          className: "univer-mt-3 univer-text-sm univer-text-gray-600 dark:!univer-text-gray-200",
          children: n.t("sheet.cf.panel.colorSet")
        }
      ),
      /* @__PURE__ */ k("div", { className: "univer-ml-1 univer-mt-3 univer-flex univer-items-center", children: [
        /* @__PURE__ */ k("div", { className: "univer-flex univer-items-center", children: [
          /* @__PURE__ */ c("div", { className: "univer-text-xs", children: n.t("sheet.cf.panel.native") }),
          /* @__PURE__ */ c(
            ke,
            {
              color: a,
              onChange: E
            }
          )
        ] }),
        /* @__PURE__ */ k("div", { className: "univer-ml-3 univer-flex univer-items-center", children: [
          /* @__PURE__ */ c("div", { className: "univer-text-xs", children: n.t("sheet.cf.panel.positive") }),
          /* @__PURE__ */ c(
            ke,
            {
              color: s,
              onChange: M
            }
          )
        ] })
      ] })
    ] }),
    /* @__PURE__ */ k("div", { children: [
      /* @__PURE__ */ c(
        "div",
        {
          className: "univer-mt-3 univer-text-sm univer-text-gray-600 dark:!univer-text-gray-200",
          children: n.t("sheet.cf.valueType.min")
        }
      ),
      /* @__PURE__ */ k("div", { className: "univer-mt-3 univer-flex univer-items-center", children: [
        /* @__PURE__ */ c(
          ae,
          {
            className: "univer-w-1/2 univer-flex-shrink-0",
            options: u,
            value: p,
            onChange: (m) => {
              y(m);
              const _ = Xe(m, 10);
              R(_), C({
                isGradient: i,
                minValue: _,
                minValueType: m,
                maxValue: x,
                maxValueType: f,
                positiveColor: s,
                nativeColor: a,
                isShowValue: g
              });
            }
          }
        ),
        /* @__PURE__ */ c(
          nn,
          {
            id: "min",
            className: "univer-ml-3",
            disabled: !V(p),
            type: p,
            value: I,
            onChange: (m) => {
              R(m || 0), C({
                isGradient: i,
                minValue: m || 0,
                minValueType: p,
                maxValue: x,
                maxValueType: f,
                positiveColor: s,
                nativeColor: a,
                isShowValue: g
              });
            }
          }
        )
      ] }),
      /* @__PURE__ */ c(
        "div",
        {
          className: "univer-mt-3 univer-text-sm univer-text-gray-600 dark:!univer-text-gray-200",
          children: n.t("sheet.cf.valueType.max")
        }
      ),
      /* @__PURE__ */ k("div", { className: "univer-mt-3 univer-flex univer-items-center", children: [
        /* @__PURE__ */ c(
          ae,
          {
            className: "univer-w-1/2 univer-flex-shrink-0",
            options: v,
            value: f,
            onChange: (m) => {
              w(m);
              const _ = Xe(m, 90);
              T(_), C({
                isGradient: i,
                minValue: I,
                minValueType: p,
                maxValue: _,
                maxValueType: m,
                positiveColor: s,
                nativeColor: a,
                isShowValue: g
              });
            }
          }
        ),
        /* @__PURE__ */ c(
          nn,
          {
            className: "univer-ml-3",
            disabled: !V(f),
            id: "max",
            type: f,
            value: x,
            onChange: (m) => {
              T(m || 0), C({
                isGradient: i,
                minValue: I,
                minValueType: p,
                maxValue: m || 0,
                maxValueType: f,
                positiveColor: s,
                nativeColor: a,
                isShowValue: g
              });
            }
          }
        )
      ] })
    ] })
  ] });
}, lt = (t) => [se.FALSE, void 0].includes(t) ? se.TRUE : se.FALSE, ct = (t) => t !== se.FALSE, Zt = (t) => {
  const { style: e, onChange: n, className: r } = t, [i, o] = N(() => {
    if (e != null && e.bl)
      return e.bl;
  }), [s, d] = N(() => {
    if (e != null && e.it)
      return e.it;
  }), [a, h] = N(() => {
    if (e != null && e.ul)
      return e.ul.s;
  }), [l, u] = N(() => {
    if (e != null && e.st)
      return e.st.s;
  }), [v, p] = N(() => {
    var R;
    const I = "#2f56ef";
    return (R = e == null ? void 0 : e.cl) != null && R.rgb ? e.cl.rgb : I;
  }), [y, f] = N(() => {
    var R;
    const I = "#e8ecfc";
    return (R = e == null ? void 0 : e.bg) != null && R.rgb ? e.bg.rgb : I;
  });
  X(() => {
    const I = {
      bl: i,
      it: s
    };
    v !== void 0 && (I.cl = { rgb: v }), y !== void 0 && (I.bg = { rgb: y }), l !== void 0 && (I.st = { s: l }), a !== void 0 && (I.ul = { s: a }), n(Mr(I));
  }, [i, s, a, l, v, y]);
  const w = "univer-flex univer-cursor-pointer univer-items-center univer-rounded univer-px-1";
  return /* @__PURE__ */ k("div", { className: Y("univer-my-2.5 univer-flex univer-justify-between", r), children: [
    /* @__PURE__ */ c(
      "div",
      {
        className: Y(w, {
          "univer-bg-gray-100 dark:!univer-bg-gray-700": ct(i || se.FALSE)
        }),
        onClick: () => o(lt(i)),
        children: /* @__PURE__ */ c(Zn, {})
      }
    ),
    /* @__PURE__ */ c(
      "div",
      {
        className: Y(w, {
          "univer-bg-gray-100 dark:!univer-bg-gray-700": ct(s || se.FALSE)
        }),
        onClick: () => d(lt(s)),
        children: /* @__PURE__ */ c(Yn, {})
      }
    ),
    /* @__PURE__ */ c(
      "div",
      {
        className: Y(w, {
          "univer-bg-gray-100 dark:!univer-bg-gray-700": ct(a || se.FALSE)
        }),
        onClick: () => h(lt(a)),
        children: /* @__PURE__ */ c(Jn, {})
      }
    ),
    /* @__PURE__ */ c(
      "div",
      {
        className: Y(w, {
          "univer-bg-gray-100 dark:!univer-bg-gray-700": ct(l || se.FALSE)
        }),
        onClick: () => u(lt(l)),
        children: /* @__PURE__ */ c(Xn, {})
      }
    ),
    /* @__PURE__ */ c(ke, { color: v, onChange: p, Icon: Qn }),
    /* @__PURE__ */ c(ke, { color: y, onChange: f })
  ] });
}, Ri = (t) => {
  var T;
  const { onChange: e, interceptorManager: n } = t, r = q(re), o = q(Z).getCurrentUnitForType(A.UNIVER_SHEET), s = o.getActiveSheet(), d = ((T = t.rule) == null ? void 0 : T.type) === L.highlightCell ? t.rule : void 0, a = ve(null), [h, l] = N(!1), u = ve(null), [v, p] = N({}), [y, f] = N(() => (d == null ? void 0 : d.subType) === F.formula ? d.value : "="), [w, I] = N(void 0), R = (g) => ({
    style: g.style,
    value: y,
    type: L.highlightCell,
    subType: F.formula
  });
  X(() => n.intercept(n.getInterceptPoints().submit, {
    handler() {
      return R({ style: v });
    }
  }), [v, y, n]), X(() => n.intercept(n.getInterceptPoints().beforeSubmit, {
    handler: (b, S, C) => w || y.length === 1 || !y.startsWith("=") ? (I(r.t("sheet.cf.errorMessage.formulaError")), !1) : C(b)
  }), [w, y]);
  const x = (g) => {
    e(R(g));
  };
  return Et((g) => {
    var S;
    ((S = u.current) == null ? void 0 : S.isClickOutSide(g)) && l(!1);
  }), /* @__PURE__ */ k("div", { ref: a, children: [
    /* @__PURE__ */ c(
      "div",
      {
        className: "univer-mt-4 univer-text-sm univer-text-gray-600 dark:!univer-text-gray-200",
        children: r.t("sheet.cf.panel.styleRule")
      }
    ),
    /* @__PURE__ */ c("div", { className: "univer-mt-3", children: /* @__PURE__ */ c(
      Nt,
      {
        ref: u,
        className: Y("univer-box-border univer-h-8 univer-w-full univer-cursor-pointer univer-items-center univer-rounded-lg univer-bg-white univer-pt-2 univer-transition-colors hover:univer-border-primary-600 dark:!univer-bg-gray-700 dark:!univer-text-white [&>div:first-child]:univer-px-2.5 [&>div]:univer-h-5 [&>div]:univer-ring-transparent", Fe),
        errorText: w,
        isFocus: h,
        initValue: y,
        unitId: o.getUnitId(),
        subUnitId: s == null ? void 0 : s.getSheetId(),
        onFocus: () => {
          l(!0);
        },
        onChange: (g) => {
          f(g), x({ style: v });
        },
        onVerify: (g, b) => {
          !g || b.length === 1 ? I(r.t("sheet.cf.errorMessage.formulaError")) : I(void 0);
        }
      }
    ) }),
    /* @__PURE__ */ c("div", { className: st, children: /* @__PURE__ */ c(Be, { rule: R({ style: v }) }) }),
    /* @__PURE__ */ c(
      Zt,
      {
        style: d == null ? void 0 : d.style,
        className: "univer-mt-3",
        onChange: (g) => {
          p(g), x({ style: g });
        }
      }
    )
  ] });
}, ut = (t) => /* @__PURE__ */ k("div", { className: "univer-relative", children: [
  /* @__PURE__ */ c(
    "div",
    {
      className: "univer-absolute univer-bottom-[-13px] univer-z-[999] univer-text-[10px] univer-text-red-500",
      children: t.errorText
    }
  ),
  t.children
] }), W = (t, e) => ({ label: e.t(`sheet.cf.operator.${t}`), value: t }), Ti = (t) => {
  const { type: e, operator: n, onChange: r, value: i, interceptorManager: o } = t, s = q(re), [d, a] = N(() => typeof i == "number" ? i : 0), [h, l] = N(""), [u, v] = N(() => typeof i == "string" ? i : ""), [p, y] = N(""), [f, w] = N(() => Array.isArray(i) ? i[0] === void 0 ? 0 : i[0] : 0), [I, R] = N(""), [x, T] = N(() => Array.isArray(i) ? i[1] === void 0 ? 100 : i[1] : 100), [g, b] = N("");
  switch (X(() => {
    switch (e) {
      case F.text: {
        [H.beginsWith, H.endsWith, H.containsText, H.notContainsText, H.equal, H.notEqual].includes(n) && r(u);
        break;
      }
      case F.number: {
        [U.equal, U.notEqual, U.greaterThan, U.greaterThanOrEqual, U.lessThan, U.lessThanOrEqual].includes(n) && r(d), [U.between, U.notBetween].includes(n) && r([f, x]);
        break;
      }
    }
  }, [e]), X(() => {
    const S = o.intercept(o.getInterceptPoints().beforeSubmit, {
      handler: (C, M, E) => {
        switch (e) {
          case F.text:
            if ([H.beginsWith, H.containsText, H.endsWith, H.notEqual, H.notContainsText, H.equal].includes(n))
              return u ? E(C) : (y(s.t("sheet.cf.errorMessage.notBlank")), !1);
        }
        return E(C);
      }
    });
    return () => {
      S();
    };
  }, [e, d, u, n]), e) {
    case F.text: {
      if ([H.beginsWith, H.endsWith, H.containsText, H.notContainsText, H.equal, H.notEqual].includes(n)) {
        const S = (C) => {
          v(C), r(C);
        };
        return /* @__PURE__ */ c("div", { className: "univer-mt-3", children: /* @__PURE__ */ c(ut, { errorText: p, children: /* @__PURE__ */ c(
          Xr,
          {
            value: u,
            onChange: (C) => {
              y(""), S(C);
            }
          }
        ) }) });
      }
      break;
    }
    case F.number: {
      if ([U.equal, U.notEqual, U.greaterThan, U.greaterThanOrEqual, U.lessThan, U.lessThanOrEqual].includes(n)) {
        const S = (C) => {
          a(C || 0), r(C || 0), l("");
        };
        return /* @__PURE__ */ c("div", { className: "univer-mt-3", children: /* @__PURE__ */ c(ut, { errorText: h, children: /* @__PURE__ */ c(
          Ve,
          {
            className: "univer-w-full",
            min: Number.MIN_SAFE_INTEGER,
            max: Number.MAX_SAFE_INTEGER,
            value: d,
            onChange: S
          }
        ) }) });
      }
      if ([U.between, U.notBetween].includes(n)) {
        const S = (M) => {
          w(M || 0), r([M || 0, x]), R("");
        }, C = (M) => {
          T(M || 0), r([f, M || 0]), b("");
        };
        return /* @__PURE__ */ k("div", { className: "univer-mt-3 univer-flex univer-items-center", children: [
          /* @__PURE__ */ c(ut, { errorText: I, children: /* @__PURE__ */ c(Ve, { min: Number.MIN_SAFE_INTEGER, max: Number.MAX_SAFE_INTEGER, value: f, onChange: S }) }),
          /* @__PURE__ */ c(ut, { errorText: g, children: /* @__PURE__ */ c(
            Ve,
            {
              className: "univer-ml-3",
              min: Number.MIN_SAFE_INTEGER,
              max: Number.MAX_SAFE_INTEGER,
              value: x,
              onChange: C
            }
          ) })
        ] });
      }
    }
  }
  return null;
}, rn = (t, e) => {
  switch (t) {
    case F.text:
      return [
        W(H.containsText, e),
        W(H.notContainsText, e),
        W(H.beginsWith, e),
        W(H.endsWith, e),
        W(H.equal, e),
        W(H.notEqual, e),
        W(H.containsBlanks, e),
        W(H.notContainsBlanks, e),
        W(H.containsErrors, e),
        W(H.notContainsErrors, e)
      ];
    case F.number:
      return [
        W(U.between, e),
        W(U.notBetween, e),
        W(U.equal, e),
        W(U.notEqual, e),
        W(U.greaterThan, e),
        W(U.greaterThanOrEqual, e),
        W(U.lessThan, e),
        W(U.lessThanOrEqual, e)
      ];
    case F.timePeriod:
      return [
        W(he.yesterday, e),
        W(he.today, e),
        W(he.tomorrow, e),
        W(he.last7Days, e),
        W(he.lastWeek, e),
        W(he.thisWeek, e),
        W(he.nextWeek, e),
        W(he.lastMonth, e),
        W(he.thisMonth, e),
        W(he.nextMonth, e)
      ];
  }
}, on = (t) => {
  var T;
  const { interceptorManager: e, onChange: n } = t, r = q(re), i = ((T = t.rule) == null ? void 0 : T.type) === L.highlightCell ? t.rule : void 0, [o, s] = N(() => {
    const g = F.text;
    return i && i.subType || g;
  }), d = [{
    value: F.text,
    label: r.t("sheet.cf.subRuleType.text")
  }, {
    value: F.number,
    label: r.t("sheet.cf.subRuleType.number")
  }, {
    value: F.timePeriod,
    label: r.t("sheet.cf.subRuleType.timePeriod")
  }, {
    value: F.duplicateValues,
    label: r.t("sheet.cf.subRuleType.duplicateValues")
  }, {
    value: F.uniqueValues,
    label: r.t("sheet.cf.subRuleType.uniqueValues")
  }], a = J(() => rn(o, r), [o]), [h, l] = N(() => {
    const g = a ? a[0].value : void 0;
    return i && i.operator || g;
  }), [u, v] = N(() => {
    var S;
    return i ? (S = i.value) != null ? S : pt(i.subType, i.operator) : "";
  }), [p, y] = N({}), f = J(() => (g) => {
    var b, S, C, M, E, V, O, m, _, j, K, fe, pe, Le, Ue;
    switch (g.subType || o) {
      case F.text: {
        if ([H.beginsWith, H.endsWith, H.containsText, H.notContainsText, H.equal, H.notEqual].includes(h))
          return {
            type: L.highlightCell,
            subType: (b = g.subType) != null ? b : o,
            operator: (S = g.operator) != null ? S : h,
            style: (C = g.style) != null ? C : p,
            value: (M = g.value) != null ? M : u
          };
        break;
      }
      case F.number: {
        if ([U.equal, U.notEqual, U.greaterThan, U.greaterThanOrEqual, U.lessThan, U.lessThanOrEqual].includes(h))
          return {
            type: L.highlightCell,
            subType: (E = g.subType) != null ? E : o,
            operator: (V = g.operator) != null ? V : h,
            style: (O = g.style) != null ? O : p,
            value: (m = g.value) != null ? m : u
          };
        if ([U.between, U.notBetween].includes(h))
          return {
            type: L.highlightCell,
            subType: (_ = g.subType) != null ? _ : o,
            operator: (j = g.operator) != null ? j : h,
            style: (K = g.style) != null ? K : p,
            value: (fe = g.value) != null ? fe : u
          };
        break;
      }
    }
    return {
      type: L.highlightCell,
      subType: (pe = g.subType) != null ? pe : o,
      operator: (Le = g.operator) != null ? Le : h,
      style: (Ue = g.style) != null ? Ue : p
    };
  }, [o, h, u, p]);
  X(() => e.intercept(e.getInterceptPoints().submit, {
    handler() {
      return f({});
    }
  }), [f, e]), X(() => {
    d.some((g) => g.value === o) || s(d[0].value);
  }, [d]);
  const w = (g) => {
    const b = g, S = rn(b, r), C = S && S[0].value;
    s(b), l(C), C && v(pt(b, C)), n(f({ subType: b, operator: C }));
  }, I = (g) => {
    const b = g;
    l(b), n(f({ operator: b }));
  }, R = (g) => {
    v(g), n(f({ value: g }));
  }, x = J(() => `${o}_${h}_${Math.random()}`, [o, h]);
  return /* @__PURE__ */ k("div", { children: [
    /* @__PURE__ */ c(
      "div",
      {
        className: "univer-mt-4 univer-text-sm univer-text-gray-600 dark:!univer-text-gray-200",
        children: r.t("sheet.cf.panel.styleRule")
      }
    ),
    /* @__PURE__ */ k("div", { className: "univer-flex univer-justify-between univer-gap-4", children: [
      /* @__PURE__ */ c(
        ae,
        {
          className: "univer-mt-3 univer-w-full",
          onChange: w,
          value: o,
          options: d
        }
      ),
      (a == null ? void 0 : a.length) && /* @__PURE__ */ c(
        ae,
        {
          className: "univer-mt-3 univer-w-full",
          onChange: I,
          value: h || "",
          options: a
        }
      )
    ] }),
    /* @__PURE__ */ c(
      Ti,
      {
        value: u,
        interceptorManager: e,
        type: o,
        operator: h,
        rule: i,
        onChange: R
      },
      x
    ),
    /* @__PURE__ */ c("div", { className: st, children: /* @__PURE__ */ c(Be, { rule: f({}) }) }),
    /* @__PURE__ */ c(
      Zt,
      {
        style: i == null ? void 0 : i.style,
        className: "univer-ml-1",
        onChange: (g) => {
          y(g), n(f({ style: g }));
        }
      }
    )
  ] });
}, tr = (t, e) => (Me[t] || [])[Number(e)] || "", xi = (t) => {
  var l;
  const { error: e, type: n, onChange: r } = t, i = q(Z), o = i.getCurrentUnitForType(A.UNIVER_SHEET).getUnitId(), s = (l = i.getCurrentUnitForType(A.UNIVER_SHEET).getActiveSheet()) == null ? void 0 : l.getSheetId(), d = ve(null), [a, h] = N(!1);
  return Et((u) => {
    var p;
    ((p = d.current) == null ? void 0 : p.isClickOutSide(u)) && h(!1);
  }), /* @__PURE__ */ c("div", { className: "univer-relative", children: n !== $.formula ? /* @__PURE__ */ k(Ht, { children: [
    /* @__PURE__ */ c(
      Ve,
      {
        className: Y({
          "univer-border-red-500": e
        }),
        value: Number(t.value) || 0,
        onChange: (u) => r(u != null ? u : 0)
      }
    ),
    e && /* @__PURE__ */ c("div", { className: "univer-absolute univer-text-xs univer-text-red-500", children: e })
  ] }) : /* @__PURE__ */ c("div", { className: "univer-w-full", children: /* @__PURE__ */ c(
    Nt,
    {
      ref: d,
      className: Y("univer-box-border univer-h-8 univer-w-full univer-cursor-pointer univer-items-center univer-rounded-lg univer-bg-white univer-pt-2 univer-transition-colors hover:univer-border-primary-600 dark:!univer-bg-gray-700 dark:!univer-text-white [&>div:first-child]:univer-px-2.5 [&>div]:univer-h-5 [&>div]:univer-ring-transparent", Fe),
      initValue: String(t.value),
      unitId: o,
      subUnitId: s,
      isFocus: a,
      onChange: (u = "") => {
        r(u || "");
      },
      onFocus: () => h(!0)
    }
  ) }) });
}, sn = (t, e, n) => ({
  operator: U.greaterThan,
  value: { type: $.num, value: (n.length - 1 - e) * 10 },
  iconType: t,
  iconId: String(e)
}), Ei = ue((t, e) => {
  const { onClick: n } = t, r = q(re), i = (o) => {
    n(o);
  };
  return /* @__PURE__ */ c("div", { ref: e, className: "univer-w-80", children: Vr.map((o, s) => /* @__PURE__ */ k("div", { className: "univer-mb-3", children: [
    /* @__PURE__ */ c("div", { className: "univer-mb-1 univer-text-sm", children: r.t(o.title) }),
    /* @__PURE__ */ c("div", { className: "univer-flex univer-flex-wrap", children: o.group.map((d) => /* @__PURE__ */ c(
      "div",
      {
        className: "univer-mb-1 univer-flex univer-w-1/2 univer-items-center",
        onClick: () => {
          i(d.name);
        },
        children: /* @__PURE__ */ c(
          "a",
          {
            className: "univer-cursor-pointer univer-rounded hover:univer-bg-gray-100 dark:hover:!univer-bg-gray-700",
            children: d.list.map((a, h) => /* @__PURE__ */ c(
              "img",
              {
                className: "univer-size-5",
                src: a,
                draggable: !1
              },
              h
            ))
          }
        )
      },
      d.name
    )) })
  ] }, s)) });
}), Ni = (t) => {
  const { onClick: e } = t, n = J(() => {
    const i = [];
    for (const o in Me) {
      const s = Me[o], d = o;
      s.forEach((a, h) => {
        i.push({
          iconType: d,
          base64: a,
          iconId: String(h)
        });
      });
    }
    return i;
  }, []), r = (i) => {
    e(i.iconType, i.iconId);
  };
  return /* @__PURE__ */ k("div", { children: [
    /* @__PURE__ */ k(
      "div",
      {
        className: "univer-mb-2.5 univer-flex univer-cursor-pointer univer-items-center univer-pl-1",
        onClick: () => r({ iconType: Fr, iconId: "" }),
        children: [
          /* @__PURE__ */ c(ot, { className: "univer-size-5" }),
          /* @__PURE__ */ c("span", { className: "univer-ml-2", children: "" })
        ]
      }
    ),
    /* @__PURE__ */ c("div", { className: "univer-flex univer-w-64 univer-flex-wrap", children: n.map((i) => /* @__PURE__ */ c(
      "div",
      {
        className: "univer-mb-2 univer-mr-2 univer-flex univer-cursor-pointer univer-items-center univer-justify-center univer-rounded hover:univer-bg-gray-100 dark:hover:!univer-bg-gray-700",
        children: /* @__PURE__ */ c(
          "img",
          {
            className: "univer-size-5",
            src: i.base64,
            draggable: !1,
            onClick: () => r(i)
          }
        )
      },
      `${i.iconType}_${i.iconId}`
    )) })
  ] });
}, Mi = (t) => {
  const { onChange: e, configList: n, errorMap: r = {} } = t, i = q(re), o = [{ label: i.t(`sheet.cf.symbol.${U.greaterThan}`), value: U.greaterThan }, { label: i.t(`sheet.cf.symbol.${U.greaterThanOrEqual}`), value: U.greaterThanOrEqual }], s = [
    { label: i.t(`sheet.cf.valueType.${$.num}`), value: $.num },
    { label: i.t(`sheet.cf.valueType.${$.percent}`), value: $.percent },
    { label: i.t(`sheet.cf.valueType.${$.percentile}`), value: $.percentile },
    { label: i.t(`sheet.cf.valueType.${$.formula}`), value: $.formula }
  ], d = (u, v) => {
    e([String(v), "value", "value"], u);
  }, a = (u, v) => {
    e([String(v), "operator"], u);
    const p = pt(F.number, u);
    d(p, v);
  }, h = (u, v) => {
    e([String(v), "value", "type"], u);
    const p = n[v], y = pt(F.number, p.operator);
    d(y, v);
  };
  return J(() => n.map((u, v) => {
    const p = r[v], y = tr(u.iconType, u.iconId), f = v === n.length - 1, w = v === 0, I = n[v - 1], R = (I == null ? void 0 : I.value.type) === $.formula ? i.t("sheet.cf.valueType.formula") : I == null ? void 0 : I.value.value, x = (T, g) => {
      const b = { ...u, iconId: g, iconType: T };
      e([String(v)], b);
    };
    return /* @__PURE__ */ k(
      "div",
      {
        className: v ? "univer-mt-6" : "univer-mt-3",
        children: [
          /* @__PURE__ */ k(
            "div",
            {
              className: "univer-mt-3 univer-flex univer-items-center univer-justify-between univer-text-sm univer-text-gray-600 dark:!univer-text-gray-200",
              children: [
                /* @__PURE__ */ k(
                  "div",
                  {
                    className: "univer-w-[45%]",
                    children: [
                      i.t("sheet.cf.iconSet.icon"),
                      v + 1
                    ]
                  }
                ),
                /* @__PURE__ */ c("div", { className: "univer-w-[45%]", children: /* @__PURE__ */ k(Ht, { children: [
                  !w && !f && i.t("sheet.cf.iconSet.rule"),
                  !w && !f && /* @__PURE__ */ k(
                    "span",
                    {
                      className: "univer-font-medium univer-text-gray-600 dark:!univer-text-gray-200",
                      children: [
                        i.t("sheet.cf.iconSet.when"),
                        i.t(`sheet.cf.symbol.${At(I.operator)}`),
                        R,
                        f ? "" : ` ${i.t("sheet.cf.iconSet.and")} `
                      ]
                    }
                  )
                ] }) })
              ]
            }
          ),
          /* @__PURE__ */ k("div", { className: "univer-mt-3 univer-grid univer-grid-cols-2 univer-gap-4", children: [
            /* @__PURE__ */ c("div", { className: "univer-flex univer-items-center", children: /* @__PURE__ */ c(
              qt,
              {
                overlay: /* @__PURE__ */ c("div", { className: "univer-rounded-lg univer-p-4", children: /* @__PURE__ */ c(Ni, { onClick: x, iconId: u.iconId, iconType: u.iconType }) }),
                children: /* @__PURE__ */ k(
                  "div",
                  {
                    className: Y("univer-box-border univer-flex univer-h-8 univer-w-full univer-items-center univer-justify-between univer-rounded-md univer-bg-white univer-px-4 univer-py-2 univer-text-xs univer-text-gray-600 univer-transition-all hover:univer-border-primary-600 dark:!univer-text-gray-200", Fe),
                    children: [
                      y ? /* @__PURE__ */ c("img", { src: y, className: "univer-size-4", draggable: !1 }) : /* @__PURE__ */ c(ot, { className: "univer-size-4" }),
                      /* @__PURE__ */ c(Gt, {})
                    ]
                  }
                )
              }
            ) }),
            f ? /* @__PURE__ */ k(
              "div",
              {
                className: "univer-mt-0 univer-text-sm univer-text-gray-600 dark:!univer-text-gray-200",
                children: [
                  i.t("sheet.cf.iconSet.rule"),
                  /* @__PURE__ */ k("span", { className: "univer-font-medium", children: [
                    i.t("sheet.cf.iconSet.when"),
                    i.t(`sheet.cf.symbol.${At(I.operator)}`),
                    R,
                    f ? "" : ` ${i.t("sheet.cf.iconSet.and")} `
                  ] })
                ]
              }
            ) : /* @__PURE__ */ c(
              ae,
              {
                options: o,
                value: u.operator,
                onChange: (T) => {
                  a(T, v);
                }
              }
            )
          ] }),
          f ? /* @__PURE__ */ c("div", {}) : /* @__PURE__ */ k(Ht, { children: [
            /* @__PURE__ */ k(
              "div",
              {
                className: "univer-mt-3 univer-grid univer-grid-cols-2 univer-gap-4 univer-text-sm univer-text-gray-600 dark:!univer-text-gray-200",
                children: [
                  /* @__PURE__ */ c("div", { children: i.t("sheet.cf.iconSet.type") }),
                  /* @__PURE__ */ c("div", { children: i.t("sheet.cf.iconSet.value") })
                ]
              }
            ),
            /* @__PURE__ */ k(
              "div",
              {
                className: "univer-mt-3 univer-grid univer-grid-cols-2 univer-gap-4",
                children: [
                  /* @__PURE__ */ c(
                    ae,
                    {
                      options: s,
                      value: u.value.type,
                      onChange: (T) => {
                        h(T, v);
                      }
                    }
                  ),
                  /* @__PURE__ */ c(
                    xi,
                    {
                      id: v,
                      type: u.value.type,
                      error: p,
                      value: u.value.value || "",
                      onChange: (T) => {
                        d(T, v);
                      }
                    }
                  )
                ]
              }
            )
          ] })
        ]
      },
      v
    );
  }), [n, r]);
}, Vi = (t) => {
  var T;
  const { interceptorManager: e } = t, n = ((T = t.rule) == null ? void 0 : T.type) === L.iconSet ? t.rule : void 0, r = q(re), [i, o] = N({}), [s, d] = N(() => {
    const g = Object.keys(Me)[0];
    if (n && n.config.length) {
      const b = n.config[0].iconType;
      if (!n.config.some((C) => C.iconType !== b))
        return b;
    }
    return g;
  }), [a, h] = N(() => {
    if (n && n.config.length)
      return jt.deepClone(n == null ? void 0 : n.config);
    const g = Me[s] || [];
    return new Array(g.length).fill("").map((b, S, C) => S === C.length - 1 ? {
      operator: U.lessThanOrEqual,
      value: { type: $.num, value: Number.MAX_SAFE_INTEGER },
      iconType: s,
      iconId: String(S)
    } : sn(s, S, C));
  }), [l, u] = N(() => n ? !!n.isShowValue : !0), v = J(() => {
    const g = a.map((b) => tr(b.iconType, b.iconId));
    return /* @__PURE__ */ c("div", { className: "univer-flex univer-items-center", children: g.map((b, S) => b ? /* @__PURE__ */ c(
      "img",
      {
        className: "univer-size-5",
        src: b
      },
      S
    ) : /* @__PURE__ */ c(ot, { className: "univer-size-5" }, S)) });
  }, [a]), p = (g) => {
    if (g.reduce((S, C, M) => S.preType && !S.result || g.length - 1 === M ? S : C.value.type === $.formula ? {
      preType: $.formula,
      result: !1
    } : S.preType ? {
      result: S.preType === C.value.type,
      preType: C.value.type
    } : {
      result: !0,
      preType: C.value.type
    }, { result: !0, preType: "" }).result && [$.num, $.percent, $.percentile].includes(g[0].value.type)) {
      const S = {};
      return g.forEach((C, M, E) => {
        if (M - 1 < 0 || M === E.length - 1)
          return;
        const O = g[M - 1], m = At(O.operator);
        kr({ operator: m, value: O.value.value }, C.value.value) || (S[M] = `${r.t(`sheet.cf.form.${m}`, String(O.value.value))} `);
      }), S;
    }
    return {};
  }, y = (g, b) => {
    cr(a, g) !== b && (ur(a, g, b), h([...a]), o(p(a)));
  }, f = (g) => {
    d(g);
    const b = Me[g] || [], S = new Array(b.length).fill("").map((C, M, E) => sn(g, M, E));
    h(S), o(p(S));
  };
  X(() => {
    const g = e.intercept(e.getInterceptPoints().submit, {
      handler() {
        return { type: L.iconSet, isShowValue: l, config: a };
      }
    });
    return () => {
      g();
    };
  }, [l, a, e]), X(() => {
    const g = e.intercept(e.getInterceptPoints().beforeSubmit, {
      handler() {
        return Object.keys(i).length === 0;
      }
    });
    return () => {
      g();
    };
  }, [l, a, e, i]);
  const w = () => {
    const g = a.map((b) => ({ ...b }));
    a.forEach((b, S) => {
      const C = a.length - 1 - S, M = g[C];
      b.iconId = M.iconId, b.iconType = M.iconType;
    }), h([...a]);
  }, I = q(Hr), [R, x] = N();
  return jr(R, I.rootContainerElement), /* @__PURE__ */ k("div", { children: [
    /* @__PURE__ */ c("div", { className: "univer-mt-4 univer-text-sm univer-text-gray-600", children: r.t("sheet.cf.panel.styleRule") }),
    /* @__PURE__ */ c("div", { className: "univer-mt-3", children: /* @__PURE__ */ c(
      qt,
      {
        overlay: /* @__PURE__ */ c("div", { className: "univer-rounded-lg univer-p-4", children: /* @__PURE__ */ c(
          Ei,
          {
            ref: (g) => {
              !R && g && x(g);
            },
            iconType: s,
            onClick: f
          }
        ) }),
        children: /* @__PURE__ */ k(
          "div",
          {
            className: Y("univer-box-border univer-flex univer-h-8 univer-w-full univer-items-center univer-justify-between univer-rounded-md univer-bg-white univer-px-4 univer-py-2 univer-text-xs univer-text-gray-600 univer-transition-all hover:univer-border-primary-600", Fe),
            children: [
              v,
              /* @__PURE__ */ c(Gt, {})
            ]
          }
        )
      }
    ) }),
    /* @__PURE__ */ k("div", { className: "univer-mt-3 univer-flex univer-items-center univer-text-xs", children: [
      /* @__PURE__ */ k("div", { className: "univer-flex univer-items-center univer-text-xs", children: [
        /* @__PURE__ */ c(Ct, { onChange: w }),
        r.t("sheet.cf.iconSet.reverseIconOrder")
      ] }),
      /* @__PURE__ */ k("div", { className: "univer-ml-6 univer-flex univer-items-center univer-text-xs", children: [
        /* @__PURE__ */ c(Ct, { checked: !l, onChange: (g) => {
          u(!g);
        } }),
        r.t("sheet.cf.iconSet.onlyShowIcon")
      ] })
    ] }),
    /* @__PURE__ */ c(Mi, { errorMap: i, onChange: y, configList: a })
  ] });
}, ki = (t) => {
  var w;
  const { onChange: e, interceptorManager: n } = t, r = q(re), i = ((w = t.rule) == null ? void 0 : w.type) === L.highlightCell ? t.rule : void 0, o = [{ label: r.t("sheet.cf.panel.isNotBottom"), value: "isNotBottom" }, { label: r.t("sheet.cf.panel.isBottom"), value: "isBottom" }, { label: r.t("sheet.cf.panel.greaterThanAverage"), value: "greaterThanAverage" }, { label: r.t("sheet.cf.panel.lessThanAverage"), value: "lessThanAverage" }], [s, d] = N(() => {
    const I = o[0].value, R = i == null ? void 0 : i.type;
    if (!i)
      return I;
    switch (R) {
      case L.highlightCell:
        switch (i.subType) {
          case F.average:
            return [U.greaterThan, U.greaterThanOrEqual].includes(i.operator) ? "greaterThanAverage" : [U.lessThan, U.lessThanOrEqual].includes(i.operator) ? "lessThanAverage" : I;
          case F.rank:
            return i.isBottom ? "isBottom" : "isNotBottom";
        }
    }
    return I;
  }), [a, h] = N(() => {
    const R = i == null ? void 0 : i.type;
    if (!i)
      return 10;
    switch (R) {
      case L.highlightCell:
        switch (i.subType) {
          case F.rank:
            return i.value || 10;
        }
    }
    return 10;
  }), [l, u] = N(() => {
    const R = i == null ? void 0 : i.type;
    if (!i)
      return !1;
    switch (R) {
      case L.highlightCell:
        switch (i.subType) {
          case F.rank:
            return i.isPercent || !1;
        }
    }
    return !1;
  }), [v, p] = N({}), y = (I) => {
    const { type: R, isPercent: x, value: T, style: g } = I;
    if (R === "isNotBottom")
      return { type: L.highlightCell, subType: F.rank, isPercent: x, isBottom: !1, value: T, style: g };
    if (R === "isBottom")
      return { type: L.highlightCell, subType: F.rank, isPercent: x, isBottom: !0, value: T, style: g };
    if (R === "greaterThanAverage")
      return { type: L.highlightCell, subType: F.average, operator: U.greaterThan, style: g };
    if (R === "lessThanAverage")
      return { type: L.highlightCell, subType: F.average, operator: U.lessThan, style: g };
  };
  X(() => n.intercept(n.getInterceptPoints().submit, {
    handler() {
      return y({ type: s, isPercent: l, value: a, style: v });
    }
  }), [s, l, a, v, n]);
  const f = (I) => {
    e(y(I));
  };
  return /* @__PURE__ */ k("div", { children: [
    /* @__PURE__ */ c(
      "div",
      {
        className: "univer-mt-4 univer-text-sm univer-text-gray-600 dark:!univer-text-gray-200",
        children: r.t("sheet.cf.panel.styleRule")
      }
    ),
    /* @__PURE__ */ c(
      ae,
      {
        className: "univer-mt-3 univer-w-full",
        value: s,
        options: o,
        onChange: (I) => {
          d(I), f({ type: I, isPercent: l, value: a, style: v });
        }
      }
    ),
    ["isNotBottom", "isBottom"].includes(s) && /* @__PURE__ */ k("div", { className: "univer-mt-3 univer-flex univer-items-center", children: [
      /* @__PURE__ */ c(
        Ve,
        {
          min: 1,
          max: 1e3,
          value: a,
          onChange: (I) => {
            const R = I || 0;
            h(R), f({ type: s, isPercent: l, value: R, style: v });
          }
        }
      ),
      /* @__PURE__ */ k(
        "div",
        {
          className: "univer-ml-3 univer-flex univer-items-center univer-text-xs",
          children: [
            /* @__PURE__ */ c(
              Ct,
              {
                checked: l,
                onChange: (I) => {
                  u(!!I), f({ type: s, isPercent: !!I, value: a, style: v });
                }
              }
            ),
            r.t("sheet.cf.valueType.percent")
          ]
        }
      )
    ] }),
    /* @__PURE__ */ c("div", { className: st, children: /* @__PURE__ */ c(Be, { rule: y({ type: s, isPercent: l, value: a, style: v }) }) }),
    /* @__PURE__ */ c(
      Zt,
      {
        style: i == null ? void 0 : i.style,
        className: "univer-mt-3",
        onChange: (I) => {
          p(I), f({ type: s, isPercent: l, value: a, style: I });
        }
      }
    )
  ] });
}, Fi = mn("beforeSubmit"), Li = mn("submit"), an = (t) => t.getCurrentUnitForType(A.UNIVER_SHEET).getUnitId(), ln = (t) => {
  var e;
  return (e = t.getCurrentUnitForType(A.UNIVER_SHEET).getActiveSheet()) == null ? void 0 : e.getSheetId();
}, Ui = (t) => {
  var S, C, M;
  const e = q(re), n = q(te), r = q(Z), i = q(Q), o = q(je), s = an(r), d = ln(r), [a, h] = N(void 0), l = ve((C = (S = t.rule) == null ? void 0 : S.ranges) != null ? C : []), u = J(() => {
    var V, O, m;
    let E = (V = t.rule) == null ? void 0 : V.ranges;
    return E != null && E.length || (E = (m = (O = o.getCurrentSelections()) == null ? void 0 : O.map((_) => _.range)) != null ? m : []), l.current = E, E != null && E.length ? E.map((_) => {
      const j = kn(_);
      return j === "NaN" ? "" : j;
    }).filter((_) => !!_).join(",") : "";
  }, [t.rule]), v = [
    { label: e.t("sheet.cf.ruleType.highlightCell"), value: "1" },
    { label: e.t("sheet.cf.panel.rankAndAverage"), value: "2" },
    { label: e.t("sheet.cf.ruleType.dataBar"), value: "3" },
    { label: e.t("sheet.cf.ruleType.colorScale"), value: "4" },
    { label: e.t("sheet.cf.ruleType.formula"), value: "5" },
    { label: e.t("sheet.cf.ruleType.iconSet"), value: "6" }
  ], [p, y] = N(() => {
    var O, m;
    const E = (O = t.rule) == null ? void 0 : O.rule.type, V = v[0].value;
    if (!E)
      return V;
    switch (E) {
      case L.highlightCell: {
        switch ((m = t.rule) == null ? void 0 : m.rule.subType) {
          case F.number:
          case F.text:
          case F.duplicateValues:
          case F.uniqueValues:
          case F.timePeriod:
            return "1";
          case F.average:
          case F.rank:
            return "2";
          case F.formula:
            return "5";
        }
        break;
      }
      case L.dataBar:
        return "3";
      case L.colorScale:
        return "4";
      case L.iconSet:
        return "6";
    }
    return V;
  }), f = ve(void 0), w = J(() => new dr({ beforeSubmit: Fi, submit: Li }), []), I = J(() => {
    switch (p) {
      case "1":
        return on;
      case "2":
        return ki;
      case "3":
        return wi;
      case "4":
        return _i;
      case "5":
        return Ri;
      case "6":
        return Vi;
      default:
        return on;
    }
  }, [p]);
  X(() => {
    const E = n.onCommandExecuted((V) => {
      if (V.id === mr.id) {
        const O = V.params;
        O.subUnitId === d && O.unitId === s && t.onCancel();
      }
      V.id === Bt.id && t.onCancel();
    });
    return () => E.dispose();
  }, []);
  const R = (E) => {
    f.current = E;
  }, x = (E) => {
    const V = E.split(",").filter((O) => !!O).map(Qr).map((O) => O.range);
    l.current = V;
  }, T = () => {
    if (a)
      return;
    const V = (() => {
      const m = r.getCurrentUnitForType(A.UNIVER_SHEET).getActiveSheet();
      if (!m)
        throw new Error("No active sheet found");
      return l.current.map((K) => fr(K, m.getRowCount(), m.getColumnCount())).filter((K) => !(Number.isNaN(K.startRow) || Number.isNaN(K.startColumn)));
    })();
    if (w.fetchThroughInterceptors(w.getInterceptPoints().beforeSubmit)(!0, null)) {
      const m = w.fetchThroughInterceptors(w.getInterceptPoints().submit)(null, null);
      if (m) {
        const _ = an(r), j = ln(r);
        if (!_ || !j)
          throw new Error("No active sheet found");
        let K = {};
        t.rule && t.rule.cfId ? (K = { ...t.rule, ranges: V, rule: m }, n.executeCommand(Lr.id, { unitId: _, subUnitId: j, rule: K }), t.onCancel()) : (K = { cfId: i.createCfId(_, j), ranges: V, rule: m, stopIfTrue: !1 }, n.executeCommand(Rn.id, { unitId: _, subUnitId: j, rule: K }), t.onCancel());
      }
    }
  }, g = () => {
    t.onCancel();
  }, b = (E, V) => {
    E ? V.length < 1 ? h(e.t("sheet.cf.errorMessage.rangeError")) : h(void 0) : h(e.t("sheet.cf.errorMessage.rangeError"));
  };
  return /* @__PURE__ */ k("div", { children: [
    /* @__PURE__ */ c(
      "div",
      {
        className: "univer-mt-4 univer-text-sm univer-text-gray-600 dark:!univer-text-gray-200",
        children: e.t("sheet.cf.panel.range")
      }
    ),
    /* @__PURE__ */ k("div", { className: "univer-mt-4", children: [
      /* @__PURE__ */ c(
        ei,
        {
          unitId: s,
          subUnitId: d,
          initialValue: u,
          onChange: (E, V) => x(V),
          onVerify: b
        }
      ),
      a && /* @__PURE__ */ c("div", { className: "univer-mt-1 univer-text-xs univer-text-red-500", children: a })
    ] }),
    /* @__PURE__ */ c(
      "div",
      {
        className: "univer-mt-4 univer-text-sm univer-text-gray-600 dark:!univer-text-gray-200",
        children: e.t("sheet.cf.panel.styleType")
      }
    ),
    /* @__PURE__ */ c(
      ae,
      {
        className: "univer-mt-4 univer-w-full",
        value: p,
        options: v,
        onChange: (E) => y(E)
      }
    ),
    /* @__PURE__ */ c(
      I,
      {
        interceptorManager: w,
        rule: (M = t.rule) == null ? void 0 : M.rule,
        onChange: R
      }
    ),
    /* @__PURE__ */ k("div", { className: "univer-mt-4 univer-flex univer-justify-end", children: [
      /* @__PURE__ */ c(Kt, { onClick: g, children: e.t("sheet.cf.panel.cancel") }),
      /* @__PURE__ */ c(Kt, { className: "univer-ml-3", variant: "primary", onClick: T, children: e.t("sheet.cf.panel.submit") })
    ] })
  ] });
};
var Pi = Object.getOwnPropertyDescriptor, Oi = (t, e, n, r) => {
  for (var i = r > 1 ? void 0 : r ? Pi(e, n) : e, o = t.length - 1, s; o >= 0; o--)
    (s = t[o]) && (i = s(i) || i);
  return i;
}, $i = (t, e) => (n, r) => e(n, r, t);
let He = class extends le {
  constructor(e) {
    super();
    Ie(this, "_initLocal", () => {
    });
    Ie(this, "_findReplaceIndex", (e) => {
      const n = /\{([^}]+)?\}/g, r = [];
      let i = n.exec(e);
      for (; i; )
        r.push({
          startIndex: i.index,
          key: Number(i[1]),
          endIndex: i.index + i[0].length - 1
        }), i = n.exec(e);
      return r;
    });
    this._localeService = e, this._initLocal();
  }
  tWithReactNode(e, ...n) {
    const r = this._localeService.getLocales(), i = e.split("."), o = r && this._localeService.resolveKeyPath(r, i);
    if (typeof o == "string") {
      const s = [];
      return this._findReplaceIndex(o).forEach((d, a, h) => {
        const l = h[a - 1] || { endIndex: -1 };
        if (l.endIndex + 1 < d.startIndex) {
          const u = o.slice(l.endIndex + 1, d.startIndex);
          u && s.push(u);
        }
        if (n[d.key] && s.push(n[d.key]), a === h.length - 1) {
          const u = o.slice(d.endIndex + 1);
          u && s.push(u);
        }
      }), s;
    }
    return [];
  }
};
He = Oi([
  $i(0, P(re))
], He);
const Di = (t, e) => {
  const n = t.rule;
  switch (n.type) {
    case L.colorScale:
      return e.t("sheet.cf.ruleType.colorScale");
    case L.dataBar:
      return e.t("sheet.cf.ruleType.dataBar");
    case L.iconSet:
      return e.t("sheet.cf.ruleType.iconSet");
    case L.highlightCell:
      switch (n.subType) {
        case F.average: {
          const r = n.operator;
          return e.t(`sheet.cf.preview.describe.${r}`, e.t("sheet.cf.subRuleType.average"));
        }
        case F.duplicateValues:
          return e.t("sheet.cf.subRuleType.duplicateValues");
        case F.uniqueValues:
          return e.t("sheet.cf.subRuleType.uniqueValues");
        case F.number: {
          const r = n.operator;
          return e.t(`sheet.cf.preview.describe.${r}`, ...Array.isArray(n.value) ? n.value.map((i) => String(i)) : [String(n.value || "")]);
        }
        case F.text: {
          const r = n.operator;
          return e.t(`sheet.cf.preview.describe.${r}`, n.value || "");
        }
        case F.timePeriod: {
          const r = n.operator;
          return e.t(`sheet.cf.preview.describe.${r}`);
        }
        case F.rank:
          return n.isPercent ? n.isBottom ? e.t("sheet.cf.preview.describe.bottomNPercent", String(n.value)) : e.t("sheet.cf.preview.describe.topNPercent", String(n.value)) : n.isBottom ? e.t("sheet.cf.preview.describe.bottomN", String(n.value)) : e.t("sheet.cf.preview.describe.topN", String(n.value));
        case F.formula:
          return e.t("sheet.cf.ruleType.formula");
      }
  }
};
let cn = 0;
const Ai = (t) => {
  const { onClick: e } = t, n = q(Q), r = q(Z), i = q(je), o = q(te), s = q(re), d = q(me), a = q(Vn), h = q(He), l = Br(() => r.getCurrentTypeOfUnit$(A.UNIVER_SHEET), void 0, void 0, []), u = l.getUnitId(), v = l.getActiveSheet();
  if (!v)
    throw new Error("No active sheet found");
  const p = v.getSheetId(), [y, f] = N([]), [w, I] = N("2"), [R, x] = N(0), [T, g] = N(-1), [b, S] = N(cn), C = ve(null), M = [
    { label: s.t("sheet.cf.panel.workSheet"), value: "2" },
    { label: s.t("sheet.cf.panel.selectedRange"), value: "1" }
  ], E = () => {
    const D = n.getSubunitRules(u, p);
    if (!D || !D.length)
      return [];
    if (w === "1") {
      const B = i.getCurrentLastSelection();
      if (!B)
        return [];
      const G = B.range;
      return D.filter((ee) => ee.ranges.some((ne) => Ce.intersects(ne, G)));
    } else if (w === "2")
      return [...D];
    return [];
  }, [V, O] = N(E);
  ti(y), X(() => {
    const D = o.onCommandExecuted((B) => {
      B.id === Bt.id && x(Math.random());
    });
    return () => D.dispose();
  }), X(() => {
    O(E);
  }, [w, R, u, p]), X(() => {
    if (w === "2")
      return;
    const D = new Ee((B) => {
      const G = [pr.id, oe.id, Se.id, _e.id, Tn.id], z = o.onCommandExecuted((ee) => {
        const { id: ne, params: we } = ee, Re = r.getCurrentUnitOfType(A.UNIVER_SHEET).getUnitId();
        G.includes(ne) && we.unitId === Re && B.next(null);
      });
      return () => z.dispose();
    }).pipe(Qt(16)).subscribe(() => {
      O(E);
    });
    return () => {
      D.unsubscribe();
    };
  }, [r, w, u, p]), X(() => {
    const D = n.$ruleChange.subscribe(() => {
      x(Math.random());
    });
    return () => D.unsubscribe();
  }, [n]), X(() => {
    const D = () => {
      var ee, ne;
      const z = Math.max(0, ((ne = (ee = C.current) == null ? void 0 : ee.getBoundingClientRect().width) != null ? ne : 0) - 8);
      return cn = z, z;
    }, G = new Ee((z) => {
      const ee = a.getContainer();
      if (ee) {
        let ne = setTimeout(() => {
          z.next(void 0);
        }, 150);
        const we = () => {
          ne && clearTimeout(ne), ne = null;
        }, Re = (at) => {
          at.propertyName === "width" && (we(), z.next(void 0));
        };
        return ee.addEventListener("transitionend", Re), () => {
          we(), ee.removeEventListener("transitionend", Re);
        };
      }
    }).pipe(Qt(16)).subscribe(() => {
      S(D());
    });
    return () => {
      G.unsubscribe();
    };
  }, []);
  const m = (D) => {
    var z;
    const B = r.getCurrentUnitOfType(A.UNIVER_SHEET).getUnitId(), G = (z = r.getCurrentUnitOfType(A.UNIVER_SHEET).getActiveSheet()) == null ? void 0 : z.getSheetId();
    if (!B || !G)
      throw new Error("No active sheet found");
    o.executeCommand(zt.id, { unitId: B, subUnitId: G, cfId: D.cfId });
  }, _ = (D, B) => {
    g(B.y);
  }, j = (D, B, G) => {
    var at;
    g(-1);
    const z = r.getCurrentUnitOfType(A.UNIVER_SHEET).getUnitId(), ee = (at = r.getCurrentUnitOfType(A.UNIVER_SHEET).getActiveSheet()) == null ? void 0 : at.getSheetId();
    if (!z || !ee)
      throw new Error("No active sheet found");
    const ne = (or) => {
      const sr = V.length;
      return Math.min(sr - 1, Math.max(0, or));
    }, we = V[ne(B.y)].cfId, Re = V[ne(G.y)].cfId;
    we !== Re && o.executeCommand(Ur.id, { unitId: z, subUnitId: ee, start: { id: we, type: "self" }, end: { id: Re, type: G.y > B.y ? "after" : "before" } });
  }, K = () => {
    t.onCreate();
  }, fe = () => {
    w === "2" ? o.executeCommand(xn.id) : w === "1" && V.map((B) => ({ unitId: u, subUnitId: p, cfId: B.cfId })).forEach((B) => {
      o.executeCommand(zt.id, B);
    });
  }, pe = J(() => {
    const D = r.getCurrentUnitOfType(A.UNIVER_SHEET), B = D.getActiveSheet();
    return V.filter((G) => {
      const z = G.ranges;
      return gt(d, D.getUnitId(), B.getSheetId(), z);
    });
  }, [V]), Le = pe.map((D, B) => ({ i: D.cfId, x: 0, w: 12, y: B, h: 1, isResizable: !1 })), Ue = J(() => {
    const D = r.getCurrentUnitOfType(A.UNIVER_SHEET), B = D.getActiveSheet();
    return V.every((G) => {
      const z = G.ranges;
      return gt(d, D.getUnitId(), B.getSheetId(), z);
    });
  }, [V]);
  return /* @__PURE__ */ k("div", { children: [
    /* @__PURE__ */ k("div", { className: "univer-flex univer-items-center univer-justify-between univer-gap-2 univer-text-sm", children: [
      /* @__PURE__ */ c("div", { className: "univer-flex univer-items-center univer-gap-2", children: h.tWithReactNode(
        "sheet.cf.panel.managerRuleSelect",
        /* @__PURE__ */ c(
          ae,
          {
            className: "univer-w-36",
            options: M,
            value: w,
            onChange: (D) => {
              I(D);
            }
          }
        )
      ).map((D, B) => /* @__PURE__ */ c("span", { children: D }, B)) }),
      /* @__PURE__ */ k("div", { className: "univer-flex univer-justify-end", children: [
        /* @__PURE__ */ c(Xt, { title: s.t("sheet.cf.panel.createRule"), placement: "bottom", children: /* @__PURE__ */ c(
          "a",
          {
            className: "univer-size-5 univer-cursor-pointer",
            onClick: K,
            children: /* @__PURE__ */ c(zn, {})
          }
        ) }),
        V.length && Ue ? /* @__PURE__ */ c(Xt, { title: s.t("sheet.cf.panel.clear"), placement: "bottom", children: /* @__PURE__ */ c(
          "a",
          {
            className: "univer-size-5 univer-cursor-pointer",
            onClick: fe,
            children: /* @__PURE__ */ c(vt, { className: "univer-text-red-500" })
          }
        ) }) : /* @__PURE__ */ c("div", { children: /* @__PURE__ */ c(vt, { className: "univer-text-gray-300" }) })
      ] })
    ] }),
    /* @__PURE__ */ c("div", { ref: C, children: b > 0 && /* @__PURE__ */ c(
      Jr,
      {
        className: "[&_.react-grid-item]:univer-transition-none [&_.react-grid-placeholder]:univer-rounded [&_.react-grid-placeholder]:!univer-bg-gray-200",
        draggableHandle: ".draggableHandle",
        layout: Le,
        cols: 12,
        rowHeight: 60,
        width: b,
        margin: [0, 10],
        onDragStop: j,
        onDragStart: _,
        children: pe == null ? void 0 : pe.map((D, B) => /* @__PURE__ */ c("div", { children: /* @__PURE__ */ k(
          "div",
          {
            className: Y("univer-group univer-relative univer-flex univer-items-center univer-justify-between univer-rounded univer-py-2 univer-pl-5 univer-pr-8 hover:univer-bg-gray-100 dark:hover:!univer-bg-gray-700", {
              "univer-bg-gray-100 dark:!univer-bg-gray-700": T === B
            }),
            onMouseMove: () => {
              D.ranges !== y && f(D.ranges);
            },
            onMouseLeave: () => f([]),
            onClick: () => {
              e(D);
            },
            children: [
              /* @__PURE__ */ c(
                "div",
                {
                  className: Y("univer-absolute univer-left-0 univer-hidden univer-size-5 univer-cursor-grab univer-items-center univer-justify-center univer-rounded group-hover:univer-flex", "draggableHandle"),
                  onClick: (G) => G.stopPropagation(),
                  children: /* @__PURE__ */ c(Kn, {})
                }
              ),
              /* @__PURE__ */ k(
                "div",
                {
                  className: "univer-min-w-0 univer-max-w-full univer-flex-shrink univer-overflow-hidden",
                  children: [
                    /* @__PURE__ */ c(
                      "div",
                      {
                        className: "univer-text-sm univer-text-gray-900 dark:!univer-text-white",
                        children: Di(D, s)
                      }
                    ),
                    /* @__PURE__ */ c("div", { className: "univer-text-xs univer-text-gray-400", children: D.ranges.map((G) => kn(G)).join(",") })
                  ]
                }
              ),
              /* @__PURE__ */ c("div", { children: /* @__PURE__ */ c(Be, { rule: D.rule }) }),
              /* @__PURE__ */ c(
                "div",
                {
                  className: Y("univer-absolute univer-right-1 univer-hidden univer-size-6 univer-cursor-pointer univer-items-center univer-justify-center univer-rounded group-hover:univer-flex hover:univer-bg-gray-200", {
                    "univer-flex univer-items-center univer-justify-center": T === B
                  }),
                  onClick: (G) => {
                    G.stopPropagation(), m(D), f([]);
                  },
                  children: /* @__PURE__ */ c(vt, {})
                }
              )
            ]
          }
        ) }, `${D.cfId}`))
      }
    ) })
  ] });
}, Hi = (t) => {
  const [e, n] = N(t.rule), [r, i] = N(!!t.rule);
  return /* @__PURE__ */ c("div", { className: "univer-flex univer-h-full univer-flex-col univer-justify-between univer-py-4", children: r ? /* @__PURE__ */ c(Ui, { onCancel: () => {
    i(!1), n(void 0);
  }, rule: e }) : /* @__PURE__ */ c(Ai, { onClick: (a) => {
    n(a), i(!0);
  }, onCreate: () => {
    i(!0);
  } }) });
};
var ji = Object.getOwnPropertyDescriptor, Bi = (t, e, n, r) => {
  for (var i = r > 1 ? void 0 : r ? ji(e, n) : e, o = t.length - 1, s; o >= 0; o--)
    (s = t[o]) && (i = s(i) || i);
  return i;
}, We = (t, e) => (n, r) => e(n, r, t);
const dt = "sheet.conditional.formatting.panel";
let Je = class extends le {
  constructor(e, n, r, i, o) {
    super();
    Ie(this, "_sidebarDisposable", null);
    this._univerInstanceService = e, this._injector = n, this._componentManager = r, this._sidebarService = i, this._localeService = o, this._initPanel(), this.disposeWithMe(
      this._univerInstanceService.getCurrentTypeOfUnit$(A.UNIVER_SHEET).subscribe((s) => {
        var d;
        s || (d = this._sidebarDisposable) == null || d.dispose();
      })
    ), this.disposeWithMe(this._sidebarService.sidebarOptions$.subscribe((s) => {
      s.id === dt && (s.visible || setTimeout(() => {
        this._sidebarService.sidebarOptions$.next({ visible: !1 });
      }));
    }));
  }
  openPanel(e) {
    const n = {
      id: dt,
      header: { title: this._localeService.t("sheet.cf.title") },
      children: {
        label: dt,
        rule: e,
        key: hr(4)
      },
      onClose: () => this._sidebarDisposable = null
    };
    this._sidebarDisposable = this._sidebarService.open(n);
  }
  _initPanel() {
    this.disposeWithMe(
      this._componentManager.register(dt, Hi)
    );
  }
};
Je = Bi([
  We(0, Z),
  We(1, P(me)),
  We(2, P(Wr)),
  We(3, P(Vn)),
  We(4, P(re))
], Je);
var ie = /* @__PURE__ */ ((t) => (t[t.createRule = 1] = "createRule", t[t.viewRule = 2] = "viewRule", t[t.highlightCell = 3] = "highlightCell", t[t.rank = 4] = "rank", t[t.formula = 5] = "formula", t[t.colorScale = 6] = "colorScale", t[t.dataBar = 7] = "dataBar", t[t.icon = 8] = "icon", t[t.clearRangeRules = 9] = "clearRangeRules", t[t.clearWorkSheetRules = 10] = "clearWorkSheetRules", t))(ie || {});
const Mt = {
  id: "sheet.operation.open.conditional.formatting.panel",
  type: ge.OPERATION,
  handler: (t, e) => {
    var d;
    const n = t.get(Je), r = t.get(je), i = t.get(te), o = ((d = r.getCurrentSelections()) == null ? void 0 : d.map((a) => a.range)) || [];
    switch (e.value) {
      case 3: {
        n.openPanel({ ...Te(), ranges: o });
        break;
      }
      case 4: {
        const a = {
          ...Te,
          ranges: o,
          rule: {
            type: L.highlightCell,
            subType: F.rank
          }
        };
        n.openPanel(a);
        break;
      }
      case 5: {
        const a = {
          ...Te,
          ranges: o,
          rule: {
            type: L.highlightCell,
            subType: F.formula,
            value: "="
          }
        };
        n.openPanel(a);
        break;
      }
      case 6: {
        const a = {
          ...Te,
          ranges: o,
          rule: {
            type: L.colorScale,
            config: []
          }
        };
        n.openPanel(a);
        break;
      }
      case 7: {
        const a = {
          ...Te,
          ranges: o,
          rule: {
            type: L.dataBar,
            isShowValue: !0
          }
        };
        n.openPanel(a);
        break;
      }
      case 8: {
        const a = {
          ...Te,
          ranges: o,
          rule: {
            type: L.iconSet,
            config: [],
            isShowValue: !0
          }
        };
        n.openPanel(a);
        break;
      }
      case 2: {
        n.openPanel();
        break;
      }
      case 1: {
        n.openPanel({ ...Te(), ranges: o });
        break;
      }
      case 9: {
        i.executeCommand(Pr.id, { ranges: o });
        break;
      }
      case 10: {
        i.executeCommand(xn.id);
        break;
      }
    }
    return !0;
  }
};
var Wi = Object.getOwnPropertyDescriptor, qi = (t, e, n, r) => {
  for (var i = r > 1 ? void 0 : r ? Wi(e, n) : e, o = t.length - 1, s; o >= 0; o--)
    (s = t[o]) && (i = s(i) || i);
  return i;
}, qe = (t, e) => (n, r) => e(n, r, t);
let yt = class extends le {
  constructor(t, e, n, r, i) {
    super(), this._injector = t, this._univerInstanceService = e, this._sheetInterceptorService = n, this._selectionManagerService = r, this._conditionalFormattingRuleModel = i, this._init();
  }
  _init() {
    this.disposeWithMe(this._sheetInterceptorService.interceptCommand({
      getMutations: (t) => {
        var i;
        const e = [], n = [], r = { redos: e, undos: n };
        if ([Cr.id, yr.id].includes(t.id)) {
          const o = (i = this._selectionManagerService.getCurrentSelections()) == null ? void 0 : i.map((p) => p.range);
          if (!o)
            return r;
          const s = this._univerInstanceService.getCurrentUnitForType(A.UNIVER_SHEET), d = s.getActiveSheet();
          if (!d)
            return r;
          const a = s.getUnitId(), h = d.getSheetId(), l = this._conditionalFormattingRuleModel.getSubunitRules(a, h);
          if (!l || !l.length)
            return r;
          const { redos: u, undos: v } = un(this._injector, l, o, a, h);
          e.push(...u), n.push(...v);
        }
        return r;
      }
    })), this.disposeWithMe(this._sheetInterceptorService.interceptRanges({
      getMutations: ({ unitId: t, subUnitId: e, ranges: n }) => {
        const r = [], i = [], o = { redos: r, undos: i };
        if (!n || !n.length)
          return o;
        const s = this._conditionalFormattingRuleModel.getSubunitRules(t, e);
        if (!s || !s.length)
          return o;
        const { redos: d, undos: a } = un(this._injector, s, n, t, e);
        return r.push(...d), i.push(...a), o;
      }
    }));
  }
};
yt = qi([
  qe(0, P(me)),
  qe(1, P(Z)),
  qe(2, P(Wt)),
  qe(3, P(je)),
  qe(4, P(Q))
], yt);
function un(t, e, n, r, i) {
  const o = [], s = [];
  return e.filter((d) => n.some((a) => d.ranges.some((h) => Ce.getIntersects(h, a)))).forEach((d) => {
    const h = new br().add(...d.ranges).subtract(...n).merge();
    if (h.length) {
      const l = {
        id: Se.id,
        params: {
          unitId: r,
          subUnitId: i,
          rule: { ...d, ranges: h }
        }
      }, u = De(t, l.params);
      o.push(l), s.push(...u);
    } else {
      const l = {
        id: _e.id,
        params: {
          unitId: r,
          subUnitId: i,
          cfId: d.cfId
        }
      }, u = Ae(t, l.params);
      o.push(l), s.push(...u);
    }
  }), { redos: o, undos: s };
}
var Gi = Object.getOwnPropertyDescriptor, Zi = (t, e, n, r) => {
  for (var i = r > 1 ? void 0 : r ? Gi(e, n) : e, o = t.length - 1, s; o >= 0; o--)
    (s = t[o]) && (i = s(i) || i);
  return i;
}, Ge = (t, e) => (n, r) => e(n, r, t);
let Qe = class extends le {
  constructor(e, n, r, i, o) {
    super();
    Ie(this, "_copyInfo");
    this._sheetClipboardService = e, this._conditionalFormattingRuleModel = n, this._injector = r, this._conditionalFormattingViewModel = i, this._univerInstanceService = o, this._initClipboardHook();
  }
  _initClipboardHook() {
    this.disposeWithMe(
      this._sheetClipboardService.addClipboardHook({
        id: it,
        onBeforeCopy: (e, n, r) => this._collectConditionalRule(e, n, r),
        onPasteCells: (e, n, r, i) => {
          const { copyType: o = Jt.COPY, pasteType: s } = i, { range: d } = e || {}, { range: a } = n;
          return this._generateConditionalFormattingMutations(a, { copyType: o, pasteType: s, copyRange: d });
        }
      })
    );
  }
  _collectConditionalRule(e, n, r) {
    const i = new Ne(), o = {};
    this._copyInfo = {
      matrix: i,
      info: {
        unitId: e,
        subUnitId: n,
        cfMap: o
      }
    };
    const s = this._injector.invoke((l) => Ir(r, l, e, n));
    if (!s)
      return;
    const { rows: d, cols: a } = s, h = /* @__PURE__ */ new Set();
    d.forEach((l, u) => {
      a.forEach((v, p) => {
        const y = this._conditionalFormattingViewModel.getCellCfs(e, n, l, v);
        y && (y.forEach((f) => h.add(f.cfId)), i.setValue(u, p, y.map((f) => f.cfId)));
      });
    }), h.forEach((l) => {
      const u = this._conditionalFormattingRuleModel.getRule(e, n, l);
      u && (o[l] = u.rule);
    });
  }
  // eslint-disable-next-line max-lines-per-function
  _generateConditionalFormattingMutations(e, n) {
    const r = this._univerInstanceService.getCurrentUnitForType(A.UNIVER_SHEET), i = r.getActiveSheet(), o = r.getUnitId();
    if (!i) return { redos: [], undos: [] };
    const s = i.getSheetId();
    if (n.copyType === Jt.CUT)
      return this._copyInfo = null, { redos: [], undos: [] };
    if (!this._copyInfo || !n.copyRange)
      return { redos: [], undos: [] };
    if (![
      kt.SPECIAL_PASTE_FORMAT,
      kt.DEFAULT_PASTE,
      kt.SPECIAL_PASTE_BESIDES_BORDER
    ].includes(
      n.pasteType
    ))
      return { redos: [], undos: [] };
    const { ranges: [a, h], mapFunc: l } = Fn([n.copyRange, e]), u = ri(a, h, !0), v = {};
    ye.foreach(h, (g, b) => {
      const { row: S, col: C } = l(g, b), M = this._conditionalFormattingViewModel.getCellCfs(o, s, S, C);
      M && M.forEach((E) => {
        if (!v[E.cfId]) {
          const V = new Ne();
          v[E.cfId] = V;
          const O = this._conditionalFormattingRuleModel.getRule(o, s, E.cfId);
          O == null || O.ranges.forEach((m) => {
            ye.foreach(m, (_, j) => {
              V.setValue(_, j, 1);
            });
          });
        }
        v[E.cfId].realDeleteValue(S, C);
      });
    });
    const { matrix: p, info: y } = this._copyInfo, f = [];
    let w = this._conditionalFormattingRuleModel.createCfId(o, s);
    const I = {}, R = (g) => {
      if (I[g])
        return I[g];
      const b = y == null ? void 0 : y.cfMap[g], S = [...this._conditionalFormattingRuleModel.getSubunitRules(o, s) || [], ...f].find((C) => jt.diffValue(C.rule, b));
      if (S)
        return I[g] = S, S;
      {
        const C = {
          rule: b,
          cfId: w,
          ranges: [],
          stopIfTrue: !1
        };
        return I[g] = C, f.push(C), w = `${Number(w) + 1}`, C;
      }
    };
    u.forEach((g) => {
      p && p.forValue((b, S, C) => {
        const M = Ce.getPositionRange(
          {
            startRow: b,
            endRow: b,
            startColumn: S,
            endColumn: S
          },
          g.startRange
        ), { row: E, col: V } = l(M.startRow, M.startColumn);
        C.forEach((O) => {
          if (!v[O]) {
            const m = R(O), _ = new Ne();
            v[O] = _, m.ranges.forEach((j) => {
              ye.foreach(j, (K, fe) => {
                _.setValue(K, fe, 1);
              });
            });
          }
          v[O].setValue(E, V, 1);
        });
      });
    });
    const x = [], T = [];
    for (const g in v) {
      const b = v[g], S = mt(ft(b));
      if (!S.length) {
        const C = {
          unitId: o,
          subUnitId: s,
          cfId: g
        };
        x.push({ id: _e.id, params: C }), T.push(...Ae(this._injector, C));
      }
      if (f.some((C) => C.cfId === g)) {
        const C = R(g), M = {
          unitId: o,
          subUnitId: s,
          rule: { ...C, ranges: S }
        };
        x.push({ id: oe.id, params: M }), T.push(En(this._injector, M));
      } else {
        const C = this._conditionalFormattingRuleModel.getRule(o, s, g);
        if (!C)
          continue;
        const M = {
          unitId: o,
          subUnitId: s,
          rule: { ...C, ranges: S }
        };
        x.push({ id: Se.id, params: M }), T.push(...De(this._injector, M));
      }
    }
    return {
      redos: x,
      undos: T
    };
  }
};
Qe = Zi([
  Ge(0, P(ni)),
  Ge(1, P(Q)),
  Ge(2, P(me)),
  Ge(3, P(rt)),
  Ge(4, P(Z))
], Qe);
var zi = Object.getOwnPropertyDescriptor, Yi = (t, e, n, r) => {
  for (var i = r > 1 ? void 0 : r ? zi(e, n) : e, o = t.length - 1, s; o >= 0; o--)
    (s = t[o]) && (i = s(i) || i);
  return i;
}, Pt = (t, e) => (n, r) => e(n, r, t);
let et = class extends le {
  constructor(t, e, n) {
    super(), this._localeService = t, this._commandService = e, this._sheetPermissionCheckController = n, this._commandExecutedListener();
  }
  _commandExecutedListener() {
    this.disposeWithMe(
      this._commandService.beforeCommandExecuted((t) => {
        if (t.id === Rn.id) {
          const { unitId: e, subUnitId: n, rule: { ranges: r } } = t.params;
          this._sheetPermissionCheckController.permissionCheckWithRanges({
            workbookTypes: [wn],
            rangeTypes: [_n],
            worksheetTypes: [In, Sn]
          }, r, e, n) || this._sheetPermissionCheckController.blockExecuteWithoutPermission(this._localeService.t("permission.dialog.setStyleErr"));
        }
      })
    );
  }
};
et = Yi([
  Pt(0, P(re)),
  Pt(1, te),
  Pt(2, P(Sr))
], et);
var Ki = Object.getOwnPropertyDescriptor, Xi = (t, e, n, r) => {
  for (var i = r > 1 ? void 0 : r ? Ki(e, n) : e, o = t.length - 1, s; o >= 0; o--)
    (s = t[o]) && (i = s(i) || i);
  return i;
}, ht = (t, e) => (n, r) => e(n, r, t);
let tt = class extends le {
  constructor(t, e, n, r) {
    super(), this._conditionalFormattingRuleModel = t, this._univerInstanceService = e, this._injector = n, this._refRangeService = r, this._initRefRange();
  }
  _initRefRange() {
    const t = /* @__PURE__ */ new Map(), e = (r, i, o) => `${r}_${i}_${o}`, n = (r, i, o) => {
      const s = (a) => {
        const h = [...o.ranges], l = h.map((v) => wr(v, a)).filter((v) => !!v);
        if (Or(l, h))
          return { redos: [], undos: [] };
        if (l.length) {
          const v = { unitId: r, subUnitId: i, rule: { ...o, ranges: l } }, p = [{ id: Se.id, params: v }], y = De(this._injector, v);
          return { redos: p, undos: y };
        } else {
          const v = { unitId: r, subUnitId: i, cfId: o.cfId }, p = [{ id: _e.id, params: v }], y = Ae(this._injector, v);
          return { redos: p, undos: y };
        }
      }, d = [];
      o.ranges.forEach((a) => {
        const h = this._refRangeService.registerRefRange(a, s);
        d.push(() => h.dispose());
      }), t.set(e(r, i, o.cfId), () => d.forEach((a) => a()));
    };
    this.disposeWithMe(this._conditionalFormattingRuleModel.$ruleChange.subscribe((r) => {
      const { unitId: i, subUnitId: o, rule: s } = r, d = this._univerInstanceService.getCurrentUnitForType(A.UNIVER_SHEET), a = d.getActiveSheet();
      if (!(r.unitId !== d.getUnitId() || r.subUnitId !== (a == null ? void 0 : a.getSheetId())))
        switch (r.type) {
          case "add": {
            n(r.unitId, r.subUnitId, r.rule);
            break;
          }
          case "delete": {
            const h = t.get(e(i, o, s.cfId));
            h && h();
            break;
          }
          case "set": {
            const h = t.get(e(i, o, s.cfId));
            h && h(), n(r.unitId, r.subUnitId, r.rule);
          }
        }
    })), this.disposeWithMe(fn(() => {
      t.forEach((r) => {
        r();
      }), t.clear();
    }));
  }
};
tt = Xi([
  ht(0, P(Q)),
  ht(1, P(Z)),
  ht(2, P(me)),
  ht(3, P(_r))
], tt);
var Ji = Object.getOwnPropertyDescriptor, Qi = (t, e, n, r) => {
  for (var i = r > 1 ? void 0 : r ? Ji(e, n) : e, o = t.length - 1, s; o >= 0; o--)
    (s = t[o]) && (i = s(i) || i);
  return i;
}, Oe = (t, e) => (n, r) => e(n, r, t);
let nt = class extends le {
  constructor(e, n, r, i, o, s) {
    super();
    /**
     * When a set operation is triggered multiple times over a short period of time, it may result in some callbacks not being disposed,and caused a render cache exception.
     * The solution here is to store all the asynchronous tasks and focus on processing after the last callback
     */
    Ie(this, "_ruleChangeCacheMap", /* @__PURE__ */ new Map());
    this._sheetInterceptorService = e, this._conditionalFormattingService = n, this._univerInstanceService = r, this._renderManagerService = i, this._conditionalFormattingViewModel = o, this._conditionalFormattingRuleModel = s, this._initViewModelInterceptor(), this._initSkeleton(), this.disposeWithMe(() => {
      this._ruleChangeCacheMap.clear();
    });
  }
  _markDirtySkeleton() {
    var n, r, i;
    const e = this._univerInstanceService.getCurrentUnitForType(A.UNIVER_SHEET).getUnitId();
    (n = this._renderManagerService.getRenderById(e)) == null || n.with(Ln).reCalculate(), (i = (r = this._renderManagerService.getRenderById(e)) == null ? void 0 : r.mainComponent) == null || i.makeDirty();
  }
  _initSkeleton() {
    this.disposeWithMe(Un(this._conditionalFormattingRuleModel.$ruleChange, this._conditionalFormattingViewModel.markDirty$).pipe(
      li(16),
      en((e) => !!e.length),
      en((e) => {
        const n = this._univerInstanceService.getCurrentUnitForType(A.UNIVER_SHEET);
        if (!n) return !1;
        const r = n.getActiveSheet();
        return r ? e.filter((i) => i.unitId === n.getUnitId() && i.subUnitId === r.getSheetId()).length > 0 : !1;
      })
    ).subscribe(() => this._markDirtySkeleton()));
  }
  _initViewModelInterceptor() {
    this.disposeWithMe(this._sheetInterceptorService.intercept(Rr.CELL_CONTENT, {
      effect: vr.Style,
      handler: (e, n, r) => {
        const i = this._conditionalFormattingService.composeStyle(n.unitId, n.subUnitId, n.row, n.col);
        if (!i)
          return r(e);
        const o = n.workbook.getStyles(), s = (typeof (e == null ? void 0 : e.s) == "string" ? o.get(e == null ? void 0 : e.s) : e == null ? void 0 : e.s) || {}, d = e === n.rawData ? { ...n.rawData } : e;
        if (i.style) {
          const a = {
            ...s,
            ...i.style
          };
          Object.assign(d, { s: a });
        }
        return d.fontRenderExtension || (d.fontRenderExtension = {}, i.isShowValue !== void 0 && (d.fontRenderExtension.isSkip = !i.isShowValue)), i.dataBar && (d.dataBar = i.dataBar), i.iconSet && (d.iconSet = i.iconSet, d.fontRenderExtension.leftOffset = $r + Dr), r(d);
      },
      priority: 10
    }));
  }
};
nt = Qi([
  Oe(0, P(Wt)),
  Oe(1, P(Nn)),
  Oe(2, P(Z)),
  Oe(3, P(Pn)),
  Oe(4, P(rt)),
  Oe(5, P(Q))
], nt);
const nr = "sheets-conditional-formatting-ui.config", bt = {};
var eo = Object.defineProperty, to = Object.getOwnPropertyDescriptor, no = (t, e, n) => e in t ? eo(t, e, { enumerable: !0, configurable: !0, writable: !0, value: n }) : t[e] = n, ro = (t, e, n, r) => {
  for (var i = r > 1 ? void 0 : r ? to(e, n) : e, o = t.length - 1, s; o >= 0; o--)
    (s = t[o]) && (i = s(i) || i);
  return i;
}, Ot = (t, e) => (n, r) => e(n, r, t), rr = (t, e, n) => no(t, typeof e != "symbol" ? e + "" : e, n);
let It = class extends yn {
  constructor(t = bt, e, n, r) {
    super(), this._config = t, this._injector = e, this._commandService = n, this._configService = r;
    const { menu: i, ...o } = bn(
      {},
      bt,
      this._config
    );
    i && this._configService.setConfig("menu", i, { merge: !0 }), this._configService.setConfig(nr, o), this._initCommand(), this._injector.add([nt]), this._injector.add([tt]), this._injector.add([Qe]), this._injector.add([et]), this._injector.add([He]);
  }
  _initCommand() {
    [
      On,
      $n,
      Dn,
      An,
      Hn,
      jn,
      Bn,
      Wn,
      qn,
      Mt
    ].forEach((t) => {
      this._commandService.registerCommand(t);
    });
  }
};
rr(It, "pluginName", `${it}_MOBILE_UI_PLUGIN`);
rr(It, "type", A.UNIVER_SHEET);
It = ro([
  pn(Mn),
  Ot(1, P(me)),
  Ot(2, P(te)),
  Ot(3, Cn)
], It);
var io = Object.getOwnPropertyDescriptor, oo = (t, e, n, r) => {
  for (var i = r > 1 ? void 0 : r ? io(e, n) : e, o = t.length - 1, s; o >= 0; o--)
    (s = t[o]) && (i = s(i) || i);
  return i;
}, Ze = (t, e) => (n, r) => e(n, r, t);
let St = class extends le {
  constructor(t, e, n, r, i) {
    super(), this._injector = t, this._univerInstanceService = e, this._autoFillService = n, this._conditionalFormattingRuleModel = r, this._conditionalFormattingViewModel = i, this._initAutoFill();
  }
  // eslint-disable-next-line max-lines-per-function
  _initAutoFill() {
    const t = () => ({ redos: [], undos: [] }), e = (i, o, s, d, a) => {
      var p;
      const h = this._univerInstanceService.getCurrentUnitForType(A.UNIVER_SHEET).getUnitId(), l = (p = this._univerInstanceService.getCurrentUnitForType(A.UNIVER_SHEET).getActiveSheet()) == null ? void 0 : p.getSheetId();
      if (!h || !l)
        return;
      const u = {
        startRow: i.row,
        startColumn: i.col,
        endColumn: i.col,
        endRow: i.row
      }, v = {
        startRow: o.row,
        startColumn: o.col,
        endColumn: o.col,
        endRow: o.row
      };
      ye.foreach(s, (y, f) => {
        const w = Ce.getPositionRange(
          {
            startRow: y,
            startColumn: f,
            endColumn: f,
            endRow: y
          },
          u
        ), I = Ce.getPositionRange(
          {
            startRow: y,
            startColumn: f,
            endColumn: f,
            endRow: y
          },
          v
        ), { row: R, col: x } = a(w.startRow, w.startColumn), T = this._conditionalFormattingViewModel.getCellCfs(
          h,
          l,
          R,
          x
        ), { row: g, col: b } = a(I.startRow, I.startColumn), S = this._conditionalFormattingViewModel.getCellCfs(
          h,
          l,
          g,
          b
        );
        S && S.forEach((C) => {
          let M = d.get(C.cfId);
          if (!d.get(C.cfId)) {
            const E = this._conditionalFormattingRuleModel.getRule(h, l, C.cfId);
            if (!E)
              return;
            M = new Ne(), E.ranges.forEach((V) => {
              ye.foreach(V, (O, m) => {
                M.setValue(O, m, 1);
              });
            }), d.set(C.cfId, M);
          }
          M.realDeleteValue(g, b);
        }), T && T.forEach((C) => {
          let M = d.get(C.cfId);
          if (!d.get(C.cfId)) {
            const E = this._conditionalFormattingRuleModel.getRule(h, l, C.cfId);
            if (!E)
              return;
            M = new Ne(), E.ranges.forEach((V) => {
              ye.foreach(V, (O, m) => {
                M.setValue(O, m, 1);
              });
            }), d.set(C.cfId, M);
          }
          M.setValue(g, b, 1);
        });
      });
    }, n = (i, o) => {
      var I, R, x;
      const s = (I = this._univerInstanceService.getCurrentUnitForType(A.UNIVER_SHEET)) == null ? void 0 : I.getUnitId(), d = (x = (R = this._univerInstanceService.getCurrentUnitForType(A.UNIVER_SHEET)) == null ? void 0 : R.getActiveSheet()) == null ? void 0 : x.getSheetId(), a = /* @__PURE__ */ new Map(), h = [], l = [];
      if (!s || !d)
        return t();
      const u = Fn([i, o]), [v, p] = u.ranges, { mapFunc: y } = u, f = {
        row: v.startRow,
        col: v.startColumn
      };
      return oi(v, p).forEach((T) => {
        e(f, T.repeatStartCell, T.relativeRange, a, y);
      }), a.forEach((T, g) => {
        const b = this._conditionalFormattingRuleModel.getRule(s, d, g);
        if (!b)
          return;
        const S = mt(ft(T));
        if (S.length) {
          const C = {
            unitId: s,
            subUnitId: d,
            rule: { ...b, ranges: S }
          };
          h.push({ id: Se.id, params: C }), l.push(...De(this._injector, C));
        } else {
          const C = {
            unitId: s,
            subUnitId: d,
            cfId: b.cfId
          };
          h.push({ id: _e.id, params: C }), l.push(...Ae(this._injector, C));
        }
      }), {
        undos: l,
        redos: h
      };
    }, r = {
      id: it,
      onFillData: (i, o, s) => {
        if (s === Ft.COPY || s === Ft.ONLY_FORMAT || s === Ft.SERIES) {
          const { source: d, target: a } = i;
          return n(d, a);
        }
        return t();
      }
    };
    this.disposeWithMe(this._autoFillService.addHook(r));
  }
};
St = oo([
  Ze(0, P(me)),
  Ze(1, P(Z)),
  Ze(2, P(ii)),
  Ze(3, P(Q)),
  Ze(4, P(rt))
], St);
var so = Object.getOwnPropertyDescriptor, ao = (t, e, n, r) => {
  for (var i = r > 1 ? void 0 : r ? so(e, n) : e, o = t.length - 1, s; o >= 0; o--)
    (s = t[o]) && (i = s(i) || i);
  return i;
}, dn = (t, e) => (n, r) => e(n, r, t);
let _t = class extends le {
  constructor(t, e) {
    super(), this._sheetInterceptorService = t, this._conditionalFormattingService = e, this._initInterceptorEditorEnd();
  }
  /**
   * Process the  values after  edit
   * @private
   * @memberof NumfmtService
   */
  _initInterceptorEditorEnd() {
    this.disposeWithMe(
      fn(
        this._sheetInterceptorService.writeCellInterceptor.intercept(
          Tr,
          {
            handler: (t, e, n) => {
              var s, d, a;
              t || n(t);
              const r = this._conditionalFormattingService.composeStyle(e.unitId, e.subUnitId, e.row, e.col), i = (s = r == null ? void 0 : r.style) != null ? s : {}, o = Object.keys(i);
              if (t != null && t.p)
                return (a = (d = t.p.body) == null ? void 0 : d.textRuns) == null || a.forEach((h) => {
                  h.ts && o.forEach((l) => {
                    var u;
                    (u = h.ts) == null || delete u[l];
                  });
                }), n(t);
              {
                const h = { ...(typeof (t == null ? void 0 : t.s) == "string" ? e.workbook.getStyles().get(t.s) : t == null ? void 0 : t.s) || {} };
                o.forEach((u) => {
                  delete h[u];
                });
                const l = { ...t, s: { ...h } };
                return n(l);
              }
            }
          }
        )
      )
    );
  }
};
_t = ao([
  dn(0, P(Wt)),
  dn(1, P(Nn))
], _t);
const hn = [
  Bt.id,
  oe.id,
  Se.id,
  _e.id,
  Tn.id
], ze = [
  {
    label: {
      name: "sheet.cf.ruleType.highlightCell",
      selectable: !1
    },
    value: ie.highlightCell
  },
  {
    label: {
      name: "sheet.cf.panel.rankAndAverage",
      selectable: !1
    },
    value: ie.rank
  },
  {
    label: {
      name: "sheet.cf.ruleType.formula",
      selectable: !1
    },
    value: ie.formula
  },
  {
    label: {
      name: "sheet.cf.ruleType.colorScale",
      selectable: !1
    },
    value: ie.colorScale
  },
  {
    label: {
      name: "sheet.cf.ruleType.dataBar",
      selectable: !1
    },
    value: ie.dataBar
  },
  {
    label: {
      name: "sheet.cf.ruleType.iconSet",
      selectable: !1
    },
    value: ie.icon
  },
  {
    label: {
      name: "sheet.cf.menu.manageConditionalFormatting",
      selectable: !1
    },
    value: ie.viewRule
  },
  {
    label: {
      name: "sheet.cf.menu.createConditionalFormatting",
      selectable: !1
    },
    value: ie.createRule
  },
  {
    label: {
      name: "sheet.cf.menu.clearRangeRules",
      selectable: !1
    },
    value: ie.clearRangeRules,
    disabled: !1
  },
  {
    label: {
      name: "sheet.cf.menu.clearWorkSheetRules",
      selectable: !1
    },
    value: ie.clearWorkSheetRules
  }
], lo = (t) => {
  const e = t.get(je), n = t.get(te), r = t.get(Z), i = t.get(Q), o = new Ee((a) => Un(
    e.selectionMoveEnd$,
    e.selectionSet$,
    new Ee((h) => {
      const l = n.onCommandExecuted((u) => {
        var f;
        const { id: v, params: p } = u, y = (f = r.getCurrentUnitForType(A.UNIVER_SHEET)) == null ? void 0 : f.getUnitId();
        hn.includes(v) && p.unitId === y && h.next(null);
      });
      return () => l.dispose();
    })
  ).pipe(tn(16)).subscribe(() => {
    var f;
    const h = ((f = e.getCurrentSelections()) == null ? void 0 : f.map((w) => w.range)) || [], l = r.getCurrentUnitForType(A.UNIVER_SHEET);
    if (!l) return;
    const u = l.getActiveSheet();
    if (!u) return;
    const y = (i.getSubunitRules(l.getUnitId(), u.getSheetId()) || []).filter((w) => w.ranges.some((I) => h.some((R) => Ce.intersects(R, I)))).map((w) => w.ranges).every((w) => gt(t, l.getUnitId(), u.getSheetId(), w));
    a.next(y);
  })), s = new Ee(
    (a) => new Ee((h) => {
      const l = n.onCommandExecuted((u) => {
        var f;
        const { id: v, params: p } = u, y = (f = r.getCurrentUnitForType(A.UNIVER_SHEET)) == null ? void 0 : f.getUnitId();
        hn.includes(v) && p.unitId === y && h.next(null);
      });
      return () => l.dispose();
    }).pipe(tn(16)).subscribe(() => {
      const h = r.getCurrentUnitForType(A.UNIVER_SHEET);
      if (!h) return;
      const l = h.getActiveSheet();
      if (!l) return;
      const u = i.getSubunitRules(h.getUnitId(), l.getSheetId()) || [];
      if (!u.length)
        return a.next(!1), !1;
      const v = u.map((p) => p.ranges).every((p) => gt(t, h.getUnitId(), l.getSheetId(), p));
      a.next(v);
    })
  ), d = new Ee((a) => {
    o.subscribe((h) => {
      const l = ze.find((u) => u.value === ie.clearRangeRules);
      l && (l.disabled = !h, a.next(ze));
    }), s.subscribe((h) => {
      const l = ze.find((u) => u.value === ie.clearWorkSheetRules);
      l && (l.disabled = !h, a.next(ze));
    }), a.next(ze);
  });
  return {
    id: Mt.id,
    type: Gr.SELECTOR,
    icon: "ConditionsDoubleIcon",
    tooltip: "sheet.cf.title",
    selections: d,
    hidden$: qr(t, A.UNIVER_SHEET),
    disabled$: si(t, { workbookTypes: [wn], worksheetTypes: [Sn, In], rangeTypes: [_n] })
  };
}, co = {
  [Zr.RULES]: {
    [Mt.id]: {
      order: 1,
      menuItemFactory: lo
    }
  }
};
var uo = Object.getOwnPropertyDescriptor, ho = (t, e, n, r) => {
  for (var i = r > 1 ? void 0 : r ? uo(e, n) : e, o = t.length - 1, s; o >= 0; o--)
    (s = t[o]) && (i = s(i) || i);
  return i;
}, vn = (t, e) => (n, r) => e(n, r, t);
let wt = class extends le {
  constructor(e, n) {
    super();
    Ie(this, "_sidebarDisposable", null);
    this._injector = e, this._menuManagerService = n, this._menuManagerService.mergeMenu(co);
  }
};
wt = ho([
  vn(0, P(me)),
  vn(1, zr)
], wt);
var vo = Object.getOwnPropertyDescriptor, go = (t, e, n, r) => {
  for (var i = r > 1 ? void 0 : r ? vo(e, n) : e, o = t.length - 1, s; o >= 0; o--)
    (s = t[o]) && (i = s(i) || i);
  return i;
}, $e = (t, e) => (n, r) => e(n, r, t);
const mo = (t, e) => {
  const n = (l) => l.endRow - l.startRow + 1, r = (l) => l.endColumn - l.startColumn + 1, i = n(e) % n(t), o = r(e) % r(t), s = Math.floor(n(e) / n(t)), d = Math.floor(r(e) / r(t)), a = [], h = {
    startRow: 0,
    endRow: n(t) - 1,
    startColumn: 0,
    endColumn: r(t) - 1
  };
  if (n(e) === 1 && r(e) === 1) {
    const l = {
      startRow: e.startRow,
      endRow: e.startRow,
      startColumn: e.startColumn,
      endColumn: e.startColumn
    };
    return a.push({ repeatRelativeRange: h, startRange: l }), a;
  }
  for (let l = 0; l < s + (i ? 0.1 : 0); l++)
    for (let u = 0; u < d + (o ? 0.1 : 0); u++) {
      const v = n(t) * l, p = r(t) * u, y = {
        startRow: v + e.startRow,
        endRow: v + e.startRow,
        startColumn: p + e.startColumn,
        endColumn: p + e.startColumn
      };
      let f = h;
      l === s && i && (f = { ...f }, f.endRow = f.endRow - (n(t) - i)), u === d && o && (f = { ...f }, f.endColumn = f.endColumn - (r(t) - o)), a.push({ repeatRelativeRange: f, startRange: y });
    }
  return a;
};
let Rt = class extends le {
  constructor(e, n, r, i, o, s) {
    super();
    Ie(this, "_painterConfig", null);
    this._injector = e, this._univerInstanceService = n, this._formatPainterService = r, this._sheetsSelectionsService = i, this._conditionalFormattingRuleModel = o, this._conditionalFormattingViewModel = s, this._initFormattingPainter();
  }
  // eslint-disable-next-line max-lines-per-function
  _initFormattingPainter() {
    const e = () => ({ redos: [], undos: [] }), n = (o, s, d, a, h) => {
      const { unitId: l, subUnitId: u } = this._painterConfig, { targetUnitId: v, targetSubUnitId: p } = h, y = {
        startRow: o.row,
        startColumn: o.col,
        endColumn: o.col,
        endRow: o.row
      }, f = {
        startRow: s.row,
        startColumn: s.col,
        endColumn: s.col,
        endRow: s.row
      };
      ye.foreach(d, (w, I) => {
        const R = Ce.getPositionRange(
          {
            startRow: w,
            startColumn: I,
            endColumn: I,
            endRow: w
          },
          y
        ), x = Ce.getPositionRange(
          {
            startRow: w,
            startColumn: I,
            endColumn: I,
            endRow: w
          },
          f
        ), T = this._conditionalFormattingViewModel.getCellCfs(
          l,
          u,
          R.startRow,
          R.startColumn
        ), g = this._conditionalFormattingViewModel.getCellCfs(
          v,
          p,
          x.startRow,
          x.startColumn
        );
        g && g.forEach((b) => {
          let S = a.get(b.cfId);
          if (!a.get(b.cfId)) {
            const C = this._conditionalFormattingRuleModel.getRule(v, p, b.cfId);
            if (!C)
              return;
            S = new Ne(), C.ranges.forEach((M) => {
              ye.foreach(M, (E, V) => {
                S.setValue(E, V, 1);
              });
            }), a.set(b.cfId, S);
          }
          S.realDeleteValue(x.startRow, x.startColumn);
        }), T && T.forEach((b) => {
          const S = a.get(b.cfId);
          S && S.setValue(x.startRow, x.startColumn, 1);
        });
      });
    }, r = (o, s, d) => {
      var R;
      const { range: a, unitId: h, subUnitId: l } = this._painterConfig, u = o !== h || l !== s, v = /* @__PURE__ */ new Map(), p = [], y = [];
      if (!o || !s || !h || !l)
        return e();
      const f = (R = this._conditionalFormattingRuleModel.getSubunitRules(h, l)) != null ? R : [];
      f == null || f.forEach((x) => {
        const { ranges: T, cfId: g } = x;
        if (T.some((b) => Ce.intersects(a, b))) {
          const b = new Ne();
          u || T.forEach((S) => {
            ye.foreach(S, (C, M) => {
              b.setValue(C, M, 1);
            });
          }), v.set(g, b);
        }
      });
      const w = {
        row: a.startRow,
        col: a.startColumn
      };
      return mo(a, d).forEach((x) => {
        n(w, { row: x.startRange.startRow, col: x.startRange.startColumn }, x.repeatRelativeRange, v, { targetUnitId: o, targetSubUnitId: s });
      }), v.forEach((x, T) => {
        if (u) {
          const g = this._conditionalFormattingRuleModel.getRule(o, s, T), b = mt(ft(x));
          if (g)
            if (b.length) {
              const S = {
                unitId: o,
                subUnitId: s,
                rule: { ...g, ranges: b }
              };
              p.push({ id: Se.id, params: S }), y.push(...De(this._injector, S));
            } else {
              const S = {
                unitId: o,
                subUnitId: s,
                cfId: g.cfId
              };
              p.push({ id: _e.id, params: S }), y.push(...Ae(this._injector, S));
            }
          else if (b.length) {
            const S = this._conditionalFormattingRuleModel.getRule(h, l, T);
            if (S) {
              const C = {
                unitId: o,
                subUnitId: s,
                rule: {
                  ...jt.deepClone(S),
                  cfId: this._conditionalFormattingRuleModel.createCfId(o, s),
                  ranges: b
                }
              };
              p.push({ id: oe.id, params: C }), y.push(En(this._injector, C));
            }
          }
        } else {
          const g = this._conditionalFormattingRuleModel.getRule(h, l, T);
          if (!g)
            return;
          const b = mt(ft(x));
          if (b.length) {
            const S = {
              unitId: h,
              subUnitId: l,
              rule: { ...g, ranges: b }
            };
            p.push({ id: Se.id, params: S }), y.push(...De(this._injector, S));
          } else {
            const S = {
              unitId: h,
              subUnitId: l,
              cfId: g.cfId
            };
            p.push({ id: _e.id, params: S }), y.push(...Ae(this._injector, S));
          }
        }
      }), {
        undos: y,
        redos: p
      };
    }, i = {
      id: it,
      onStatusChange: (o) => {
        var s, d, a;
        switch (o) {
          case Lt.INFINITE:
          case Lt.ONCE: {
            const h = (s = this._univerInstanceService.getCurrentUnitForType(A.UNIVER_SHEET)) == null ? void 0 : s.getUnitId(), l = (a = (d = this._univerInstanceService.getCurrentUnitForType(A.UNIVER_SHEET)) == null ? void 0 : d.getActiveSheet()) == null ? void 0 : a.getSheetId(), u = this._sheetsSelectionsService.getCurrentLastSelection(), v = u == null ? void 0 : u.range;
            h && l && v && (this._painterConfig = { unitId: h, subUnitId: l, range: v });
            break;
          }
          case Lt.OFF: {
            this._painterConfig = null;
            break;
          }
        }
      },
      onApply: (o, s, d) => this._painterConfig ? r(o, s, d) : {
        redos: [],
        undos: []
      }
    };
    this._formatPainterService.addHook(i);
  }
};
Rt = go([
  $e(0, P(me)),
  $e(1, P(Z)),
  $e(2, P(ai)),
  $e(3, P(je)),
  $e(4, P(Q)),
  $e(5, P(rt))
], Rt);
var fo = Object.getOwnPropertyDescriptor, po = (t, e, n, r) => {
  for (var i = r > 1 ? void 0 : r ? fo(e, n) : e, o = t.length - 1, s; o >= 0; o--)
    (s = t[o]) && (i = s(i) || i);
  return i;
}, $t = (t, e) => (n, r) => e(n, r, t);
let Tt = class extends le {
  constructor(t, e, n) {
    super(), this._conditionalFormattingViewModel = t, this._univerInstanceService = e, this._renderManagerService = n, this._init();
  }
  _init() {
    const t = this._univerInstanceService.getCurrentUnitForType(A.UNIVER_SHEET), e = (n) => {
      const r = n.getUnitId(), i = this._renderManagerService.getRenderById(r);
      if (!i)
        return;
      const o = i.with(Ln);
      this.disposeWithMe(o.currentSkeleton$.subscribe((s) => {
        if (s) {
          const d = s.skeleton.rowColumnSegment, a = d.endColumn - d.startColumn + 1, l = (d.endRow - d.startRow + 1) * a * 9, u = Math.max(Ar, l);
          this._conditionalFormattingViewModel.setCacheLength(u);
        }
      }));
    };
    t && e(t), this._univerInstanceService.getCurrentTypeOfUnit$(A.UNIVER_SHEET).subscribe((n) => {
      n && e(n);
    });
  }
};
Tt = po([
  $t(0, P(rt)),
  $t(1, Z),
  $t(2, Pn)
], Tt);
var Co = Object.defineProperty, yo = Object.getOwnPropertyDescriptor, bo = (t, e, n) => e in t ? Co(t, e, { enumerable: !0, configurable: !0, writable: !0, value: n }) : t[e] = n, Io = (t, e, n, r) => {
  for (var i = r > 1 ? void 0 : r ? yo(e, n) : e, o = t.length - 1, s; o >= 0; o--)
    (s = t[o]) && (i = s(i) || i);
  return i;
}, Dt = (t, e) => (n, r) => e(n, r, t), ir = (t, e, n) => bo(t, typeof e != "symbol" ? e + "" : e, n);
let xt = class extends yn {
  constructor(t = bt, e, n, r) {
    super(), this._config = t, this._injector = e, this._commandService = n, this._configService = r;
    const { menu: i, ...o } = bn(
      {},
      bt,
      this._config
    );
    i && this._configService.setConfig("menu", i, { merge: !0 }), this._configService.setConfig(nr, o), this._initCommand();
  }
  onStarting() {
    gr(this._injector, [
      [nt],
      [tt],
      [Qe],
      [St],
      [et],
      [Je],
      [wt],
      [He],
      [_t],
      [yt],
      [Rt],
      [Tt]
    ]), Vt(this._injector, [
      [nt]
    ]);
  }
  onReady() {
    Vt(this._injector, [
      [wt],
      [Je]
    ]);
  }
  onRendered() {
    Vt(this._injector, [
      [St],
      [yt],
      [Qe],
      [_t],
      [He],
      [Rt],
      [et],
      [tt],
      [Tt]
    ]);
  }
  _initCommand() {
    [
      On,
      $n,
      Dn,
      An,
      Hn,
      jn,
      Bn,
      Wn,
      qn,
      Mt
    ].forEach((t) => {
      this._commandService.registerCommand(t);
    });
  }
};
ir(xt, "pluginName", `${it}_UI_PLUGIN`);
ir(xt, "type", A.UNIVER_SHEET);
xt = Io([
  pn(Mn),
  Dt(1, P(me)),
  Dt(2, P(te)),
  Dt(3, Cn)
], xt);
export {
  On as AddAverageCfCommand,
  $n as AddColorScaleConditionalRuleCommand,
  Dn as AddDataBarConditionalRuleCommand,
  An as AddDuplicateValuesCfCommand,
  Hn as AddNumberCfCommand,
  jn as AddRankCfCommand,
  Bn as AddTextCfCommand,
  Wn as AddTimePeriodCfCommand,
  qn as AddUniqueValuesCfCommand,
  yt as ConditionalFormattingClearController,
  Mt as OpenConditionalFormattingOperator,
  It as UniverSheetsConditionalFormattingMobileUIPlugin,
  xt as UniverSheetsConditionalFormattingUIPlugin
};
