// @svar-ui/react-editor/dist/index.es.js
import { jsxs as R, jsx as a } from "react/jsx-runtime";
import { context as Z, Field as ue, Locale as fe, ModalArea as be, SideArea as ve, Material as oe, Willow as se, WillowDark as re, Text as xe, TextArea as ke, Checkbox as Ce } from "@svar-ui/react-core";
import { useRef as X, useEffect as J, useContext as _, useMemo as T, useState as U, useCallback as Ee } from "react";
import { uid as Ne, isSame as de, deepCopy as ie } from "@svar-ui/lib-state";
import { useWritableProp as Se, useHotkeys as Te } from "@svar-ui/lib-react";
import { Toolbar as Ae } from "@svar-ui/react-toolbar";
import { setEnv as je, env as Me } from "@svar-ui/lib-dom";
const ee = {};
function ce(t) {
  return typeof t < "u" ? ee[t] || t : ee.text;
}
function z(t, e) {
  ee[t] = e;
}
const me = {
  editor: {}
};
function F(t) {
  const {
    editors: e,
    data: o,
    css: l = "",
    errors: r,
    focus: u = !1,
    onClick: n,
    children: s,
    onChange: c
  } = t, m = X(null);
  J(() => {
    if (u) {
      const f = document.activeElement;
      if (f && m.current && m.current.contains(f)) return;
      const p = m.current ? m.current.querySelector(
        "input:not([disabled]), textarea:not([disabled]), select:not([disabled])"
      ) : null;
      p && setTimeout(() => {
        typeof p.select == "function" && p.select(), typeof p.focus == "function" && p.focus();
      }, 300);
    }
  }, []);
  const i = _(Z.i18n), y = T(() => i.getGroup("editor"), [i]), g = T(
    () => e.config[0].comp === "readonly" && e.config.every((f) => !Object.keys(o).includes(f.key)),
    [e, o]
  );
  return /* @__PURE__ */ R("div", { className: "wx-s2aE1xdZ wx-sections " + l, ref: m, children: [
    s,
    g ? /* @__PURE__ */ a("div", { className: "wx-s2aE1xdZ wx-overlay", children: y("No data") }) : null,
    e.config.map((f) => {
      if (!f.hidden) {
        const { key: p, onChange: k, ...b } = f;
        if (f.comp === "readonly" || f.comp === "section") {
          const N = ce(f.comp);
          return /* @__PURE__ */ a(
            N,
            {
              fieldKey: p,
              label: f.label,
              value: o[p],
              ...b,
              onClick: n
            },
            p
          );
        } else {
          const N = ce(f.comp);
          return /* @__PURE__ */ R("div", { children: [
            /* @__PURE__ */ a(
              ue,
              {
                label: f.labelTemplate ? f.labelTemplate(o[p]) : f.label ?? "",
                required: f.required,
                children: /* @__PURE__ */ a(
                  N,
                  {
                    fieldKey: p,
                    ...b,
                    onChange: k || ((E) => {
                      c && c({
                        value: E.value,
                        key: p,
                        input: E.input
                      });
                    }),
                    label: void 0,
                    error: r && r[p],
                    value: o[p]
                  },
                  p
                )
              }
            ),
            r && r[p] && f.validationMessage ? /* @__PURE__ */ a("div", { className: "wx-s2aE1xdZ wx-message", children: f.validationMessage }) : null
          ] }, p);
        }
      }
      return null;
    })
  ] });
}
function Ve(t) {
  if (typeof t == "string" && t.includes(".")) {
    const e = t.split(".");
    return (o) => {
      let l = o;
      return e.forEach((r) => {
        l = l[r];
      }), l;
    };
  }
  return (e) => e[t];
}
function De(t) {
  if (typeof t == "string" && t.includes(".")) {
    const e = t.split(".");
    return (o, l) => {
      let r = o;
      e.forEach((u, n) => {
        n === e.length - 1 ? r[u] = l : r = r[u];
      });
    };
  }
  return (e, o) => e[t] = o;
}
function $e(t) {
  const e = t.map((n) => {
    const s = { ...n };
    return n.config && Object.assign(s, n.config), s.key = n.key || Ne(), s.setter = n.setter || De(n.key), s.getter = n.getter || Ve(n.key), s;
  }), o = (n) => {
    const s = {};
    return e.forEach((c) => {
      c.comp !== "section" && (c.getter ? s[c.key] = c.getter(n) : s[c.key] = n[c.key]);
    }), s;
  }, l = (n, s, c) => ((c.length ? c.map((i) => e.find((y) => y.key === i)) : e).forEach((i) => {
    i.setter ? i.setter(n, s[i.key]) : n[i.key] = s[i.key];
  }), n), r = (n, s) => {
    const c = o(n), m = [];
    return e.forEach((i) => {
      const y = c[i.key], g = s[i.key];
      !de(y, g) && (y !== void 0 || g) && m.push(i.key);
    }), m;
  }, u = (n, s, c) => {
    let m = 0;
    const i = {};
    return (s?.length ? s.map((g) => e.find((f) => f.key === g)) : e).forEach((g) => {
      g.required && !n[g.key] ? (i[g.key] = {
        errorType: "required"
      }, g.validationMessage = g.validationMessage ?? c("This field is required"), m++) : g.validation && !g.validation(n[g.key]) && (i[g.key] = {
        errorType: "validation"
      }, g.validationMessage = g.validationMessage ?? c("Invalid value"), m++);
    }), m > 0 ? i : null;
  };
  return {
    config: e.filter((n) => n.comp !== "hidden"),
    getValues: o,
    setValues: l,
    diff: r,
    validateValues: u
  };
}
function pe({
  values: t,
  items: e,
  css: o,
  activeBatch: l,
  autoSave: r,
  focus: u,
  readonly: n,
  topBar: s = !0,
  bottomBar: c = !0,
  layout: m = "default",
  placement: i = "inline",
  view: y,
  children: g,
  onChange: f,
  onSave: p,
  onAction: k,
  onValidation: b,
  hotkeys: N
}) {
  const $ = _(Z.i18n).getGroup("editor"), [v, V] = Se(t), [O, B] = U(null), S = T(() => {
    const h = $e(e);
    O && h.config.forEach((d) => {
      d.comp === "section" && d.key === O && (d.sectionMode === "accordion" ? d.activeSection || (h.config.forEach((D) => {
        D.comp === "section" && D.key !== d.key && (D.activeSection = !1);
      }), d.activeSection = !0) : d.activeSection = !d.activeSection);
    });
    let C = /* @__PURE__ */ new Set(), j = null;
    return h.config.forEach((d) => {
      d.sectionMode === "exclusive" && d.activeSection && (j = d.key), d.activeSection && C.add(d.key);
    }), h.config.forEach((d) => {
      d.hidden = d.hidden || l && l !== d.batch || j && d.key != j && d.section !== j || d.section && !C.has(d.section);
    }), n ? {
      ...h,
      config: h.config.map((d) => ({ ...d, comp: "readonly" })),
      diff: () => []
    } : h;
  }, [e, O, l, n]), [M, w] = U({}), [A, W] = U({}), x = v;
  J(() => {
    v !== void 0 && (w(ie(v)), W(ie(v)), x.errors && (x.errors = K()));
  }, [v]);
  const [q, I] = U([]);
  J(() => {
    v && I([]);
  }, [v]);
  function te(h) {
    return [...new Set(h)];
  }
  function K(h) {
    const C = S.validateValues(M, h, $);
    return de(C, x.errors) || b && b({ errors: C, values: M }), C;
  }
  function he(h, C) {
    if (r && !x.errors) {
      const j = S.setValues(v, C ?? A, h);
      V(j), p && p({ changes: h, values: j });
    } else
      I(h);
  }
  function ge({ value: h, key: C, input: j }) {
    let d = { ...A || {}, [C]: h };
    const D = {
      key: C,
      value: h,
      update: d
    };
    if (j && (D.input = j), f && f(D), !v) return;
    d = D.update, W(d);
    const G = S.diff(v, d), ye = S.setValues(
      { ...M || {} },
      d,
      te([...G, C])
    );
    if (w(ye), G.length) {
      const P = r ? [] : te([...G, ...Object.keys(x.errors ?? {}), C]);
      x.errors = K(P), he(G, d);
    } else {
      const P = Object.keys(x.errors ?? {});
      P.length && (x.errors = K(P)), I([]);
    }
  }
  function we() {
    if (q.length && (r || (x.errors = K()), !x.errors)) {
      p && p({
        changes: q,
        values: M
      });
      const h = S.setValues(v, A, q);
      V(h), I([]), V({ ...M });
    }
  }
  function ne({ item: h }) {
    h.id === "save" ? we() : h.id === "toggle-section" && B(h.key), k && k({ item: h, values: M, changes: q });
  }
  return /* @__PURE__ */ a(
    y,
    {
      topBar: s,
      bottomBar: c,
      placement: i,
      layout: m,
      readonly: n,
      autoSave: r,
      css: o,
      data: A,
      editors: S,
      focus: u,
      hotkeys: N,
      errors: x.errors,
      onClick: ne,
      onKeyDown: ne,
      onChange: ge,
      children: g
    }
  );
}
function Ue({
  values: t = {},
  items: e = [],
  css: o = "",
  activeBatch: l = null,
  focus: r = !1,
  autoSave: u = !0,
  readonly: n = !1,
  children: s,
  ...c
}) {
  const m = {};
  for (const i in c) {
    if (/^on[a-z]/.test(i)) {
      const y = "on" + i.charAt(2).toUpperCase() + i.slice(3);
      if (y in c)
        continue;
      m[y] = c[i];
      continue;
    }
    m[i] = c[i];
  }
  return /* @__PURE__ */ a(fe, { words: me, optional: !0, children: /* @__PURE__ */ a("div", { className: "wx-H902AF2Y wx-content", children: /* @__PURE__ */ a(
    pe,
    {
      view: F,
      values: t,
      items: e,
      css: o,
      focus: r,
      activeBatch: l,
      autoSave: u,
      readonly: n,
      ...m,
      children: s
    }
  ) }) });
}
function Oe(t) {
  const { editors: e, data: o, layout: l, errors: r, focus: u, onClick: n, onChange: s } = t, c = T(() => {
    let m = [];
    if (l === "columns" && (m = [
      { ...e, config: [] },
      { ...e, config: [] }
    ], e.config.forEach((i) => {
      const y = i.column === "left" ? 0 : 1;
      m[y].config.push(i);
    }), m[0].config.length)) {
      const i = m[0].config[0];
      i.comp === "text" && (m[0][0] = {
        ...i,
        css: "title",
        label: ""
      });
    }
    return m;
  }, [l, e]);
  return l === "columns" ? /* @__PURE__ */ R("div", { className: "wx-bNrSbszs wx-cols", children: [
    /* @__PURE__ */ a("div", { className: "wx-bNrSbszs wx-left", children: /* @__PURE__ */ a(
      F,
      {
        editors: c[0],
        data: o,
        errors: r,
        onClick: n,
        onChange: s
      }
    ) }),
    /* @__PURE__ */ a("div", { className: "wx-bNrSbszs wx-right", children: /* @__PURE__ */ a(
      F,
      {
        editors: c[1],
        data: o,
        focus: u,
        errors: r,
        onClick: n,
        onChange: s
      }
    ) })
  ] }) : /* @__PURE__ */ a(
    F,
    {
      editors: e,
      data: o,
      focus: u,
      errors: r,
      onClick: n,
      onChange: s
    }
  );
}
function le({
  items: t,
  values: e = null,
  top: o = !0,
  onClick: l,
  onChange: r
}) {
  const u = Ee(
    ({ item: n, value: s }) => {
      r && r({ key: n.key, value: s });
    },
    [r]
  );
  return t.length ? /* @__PURE__ */ a(
    "div",
    {
      className: `wx-66OW1j0R wx-editor-toolbar ${o ? "wx-topbar" : "wx-bottom"}`,
      children: /* @__PURE__ */ a(
        Ae,
        {
          items: t,
          values: e,
          onClick: l,
          onChange: u
        }
      )
    }
  ) : null;
}
const H = () => ({ comp: "spacer" }), L = (t) => ({
  comp: "button",
  text: t("Cancel"),
  id: "cancel"
}), Q = (t) => ({
  type: "primary",
  comp: "button",
  text: t("Save"),
  id: "save"
}), ae = () => ({
  comp: "icon",
  icon: "wxi-close",
  id: "close"
});
function Y(t) {
  const {
    data: e,
    editors: o,
    focus: l,
    css: r,
    topBar: u,
    bottomBar: n,
    layout: s,
    placement: c,
    errors: m,
    readonly: i,
    autoSave: y,
    children: g,
    onClick: f,
    onKeyDown: p,
    onChange: k,
    hotkeys: b
  } = t, N = _(Z.i18n), E = T(() => N.getGroup("editor"), [N]), $ = T(
    () => u === !0 && n === !0,
    [u, n]
  ), v = T(() => {
    let w = u && u.items ? u.items.map((A) => ({ ...A })) : [];
    return $ && (i ? w = [H(), ae()] : (y ? w = [H(), ae()] : c !== "modal" && (w = [H(), L(E), Q(E)]), s === "columns" && !w.length && (w = [H(), Q(E), L(E)]))), w;
  }, [u, $, i, y, c, s, E]), V = T(() => {
    let w = n && n.items ? n.items.map((A) => ({ ...A })) : [];
    return $ && (i || (c === "modal" && !y && (w = [H(), Q(E), L(E)]), s === "columns" && v.length && (w = []))), w;
  }, [n, $, i, c, y, s, v, E]), O = T(() => [...v, ...V], [v, V]), B = X(null), S = X(null);
  S.current = (w, ...A) => {
    const W = O.findIndex((I) => A.includes(I.id));
    if (W === -1) return !1;
    const x = w.target, q = O[W];
    w.key == "Escape" && (x.closest(".wx-combo") || x.closest(".wx-multicombo") || x.closest(".wx-richselect")) || w.key == "Delete" && (x.tagName === "INPUT" || x.tagName === "TEXTAREA") || (w.preventDefault(), p && p({ item: q }));
  };
  const M = T(() => b === !1 ? {} : {
    "ctrl+s": (w) => S.current(w, "save"),
    escape: (w) => S.current(w, "cancel", "close"),
    "ctrl+d": (w) => S.current(w, "delete"),
    ...b || {}
  }, [b]);
  return Te(M, B), /* @__PURE__ */ R("div", { className: r ? "wx-85HDaNoA " + r : "wx-85HDaNoA", ref: B, children: [
    /* @__PURE__ */ a(
      le,
      {
        ...u && typeof u == "object" ? u : {},
        items: v,
        values: e,
        onClick: f,
        onChange: k
      }
    ),
    /* @__PURE__ */ R(
      "div",
      {
        className: `wx-85HDaNoA wx-content${s === "columns" ? " wx-layout-columns" : ""}`,
        children: [
          g,
          /* @__PURE__ */ a(
            Oe,
            {
              editors: o,
              layout: s,
              data: e,
              focus: l,
              errors: m,
              onClick: f,
              onChange: k
            }
          ),
          /* @__PURE__ */ a(
            le,
            {
              ...n && typeof n == "object" ? n : {},
              items: V,
              values: e,
              top: !1,
              onClick: f,
              onChange: k
            }
          )
        ]
      }
    )
  ] });
}
function qe(t) {
  const { css: e, onClick: o, placement: l, ...r } = t;
  function u() {
    o && o({ item: { id: "close" } });
  }
  return l === "modal" ? /* @__PURE__ */ a(be, { children: /* @__PURE__ */ a(
    Y,
    {
      css: `wx-panel ${e}`,
      onClick: o,
      placement: l,
      ...r
    }
  ) }) : l === "sidebar" ? /* @__PURE__ */ a(ve, { onCancel: u, children: /* @__PURE__ */ a(
    Y,
    {
      css: `wx-panel ${e}`,
      onClick: o,
      placement: l,
      ...r
    }
  ) }) : /* @__PURE__ */ a(
    Y,
    {
      css: `wx-inline-form ${e}`,
      onClick: o,
      placement: l,
      ...r
    }
  );
}
function Fe(t) {
  const {
    values: e = {},
    items: o = [],
    css: l = "",
    activeBatch: r = null,
    topBar: u = !0,
    bottomBar: n = !0,
    focus: s = !1,
    autoSave: c = !1,
    layout: m = "default",
    readonly: i = !1,
    placement: y = "inline",
    children: g,
    ...f
  } = t, p = Object.keys(f).reduce((k, b) => {
    if (/^on[a-z]/.test(b)) {
      const N = "on" + b.charAt(2).toUpperCase() + b.slice(3);
      N in f ? k[b] = f[b] : k[N] = f[b];
    } else
      k[b] = f[b];
    return k;
  }, {});
  return /* @__PURE__ */ a(fe, { words: me, optional: !0, children: /* @__PURE__ */ a(
    pe,
    {
      view: qe,
      values: e,
      items: o,
      css: l,
      activeBatch: r,
      topBar: u,
      bottomBar: n,
      focus: s,
      autoSave: c,
      layout: m,
      readonly: i,
      placement: y,
      ...p,
      children: g
    }
  ) });
}
function Ze({ fonts: t = !0, children: e }) {
  return e ? /* @__PURE__ */ a(oe, { fonts: t, children: e }) : /* @__PURE__ */ a(oe, { fonts: t });
}
function _e({ fonts: t = !0, children: e }) {
  return e ? /* @__PURE__ */ a(se, { fonts: t, children: e }) : /* @__PURE__ */ a(se, { fonts: t });
}
function Le(t) {
  const { fonts: e = !0, children: o } = t;
  return o ? /* @__PURE__ */ a(re, { fonts: e, children: o }) : /* @__PURE__ */ a(re, { fonts: e });
}
function Ie({ value: t, options: e, label: o }) {
  const r = _(Z.i18n).getGroup("editor"), u = T(() => {
    let n = t;
    if (typeof t == "boolean" && (n = r(t ? "Yes" : "No")), e) {
      const s = e.find((c) => c.id === t);
      s && (n = s.label);
    }
    return n;
  }, [t, e, r]);
  return u || u === 0 ? /* @__PURE__ */ a(ue, { label: o, children: u }) : null;
}
function Re({ fieldKey: t, label: e, activeSection: o, onClick: l }) {
  return /* @__PURE__ */ R(
    "div",
    {
      className: `wx-OmgQq65I wx-section${o ? " wx-section-active" : ""}`,
      onClick: () => l && l({
        item: { id: "toggle-section", key: o ? null : t }
      }),
      children: [
        /* @__PURE__ */ a("h3", { children: e }),
        /* @__PURE__ */ a(
          "i",
          {
            className: `wx-OmgQq65I wxi-angle-${o ? "down" : "right"} wx-icon`
          }
        )
      ]
    }
  );
}
z("text", xe);
z("textarea", ke);
z("checkbox", Ce);
z("readonly", Ie);
z("section", Re);
je(Me);
export {
  Fe as Editor,
  Ue as Form,
  Ze as Material,
  _e as Willow,
  Le as WillowDark,
  z as registerEditorItem
};
