// @pdfme/converter/dist/cjs/src/index.node.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.img2pdf = exports.pdf2size = exports.pdf2img = void 0;
const pdfjsLib = __importStar(require("pdfjs-dist/legacy/build/pdf"));
// @ts-expect-error - PDFJSWorker import is not properly typed but required for functionality
const pdf_worker_js_1 = __importDefault(require("pdfjs-dist/legacy/build/pdf.worker.js"));
const canvas_1 = require("canvas");
const pdf2img_js_1 = require("./pdf2img.js");
const pdf2size_js_1 = require("./pdf2size.js");
pdfjsLib.GlobalWorkerOptions.workerSrc = pdf_worker_js_1.default;
const pdf2img = async (pdf, options = {}) => (0, pdf2img_js_1.pdf2img)(pdf, options, {
    getDocument: (pdf) => pdfjsLib.getDocument({ data: pdf, isEvalSupported: false }).promise,
    createCanvas: (width, height) => (0, canvas_1.createCanvas)(width, height),
    canvasToArrayBuffer: (canvas) => {
        // Using a more specific type for the canvas from the 'canvas' package
        const nodeCanvas = canvas;
        // Get buffer from the canvas - using the synchronous version without parameters
        // This will use the default PNG format
        const buffer = nodeCanvas.toBuffer();
        // Convert to ArrayBuffer
        return new Uint8Array(buffer).buffer;
    },
});
exports.pdf2img = pdf2img;
const pdf2size = async (pdf, options = {}) => (0, pdf2size_js_1.pdf2size)(pdf, options, {
    getDocument: (pdf) => pdfjsLib.getDocument({ data: pdf, isEvalSupported: false }).promise,
});
exports.pdf2size = pdf2size;
var img2pdf_js_1 = require("./img2pdf.js");
Object.defineProperty(exports, "img2pdf", { enumerable: true, get: function () { return img2pdf_js_1.img2pdf; } });
