// @svar-ui/react-tasklist/dist/index.es.js
import { jsxs as K, jsx as a } from "react/jsx-runtime";
import { useContext as I, useRef as y, useCallback as k, useEffect as g, useState as O, useMemo as A } from "react";
import { context as W, Checkbox as G, Button as H, Locale as L } from "@svar-ui/react-core";
import { clickOutside as S, delegateClick as U } from "@svar-ui/lib-dom";
import { tempID as _ } from "@svar-ui/lib-state";
import { useWritableProp as B } from "@svar-ui/lib-react";
import { en as F } from "@svar-ui/tasklist-locales";
function T(w) {
  const { task: e, edit: l, readonly: i, onUpdate: c, onRemove: f } = w, d = I(W.i18n).getGroup("tasklist"), x = y(null), u = y(""), h = k(() => {
    const n = x.current;
    n && (n.style.height = n.scrollHeight + 2 + "px");
  }, []), C = k(() => {
    const n = x.current;
    n && (u.current = n.value, h());
  }, [h]), D = k(
    ({ value: n }) => {
      c({
        task: e,
        status: n ? 1 : 0,
        content: u.current || e.content
      });
    },
    [c, e]
  ), p = k(
    (n) => {
      c({
        task: e,
        content: u.current,
        status: e.status,
        next: n
      });
    },
    [c, e]
  ), N = k(
    (n) => {
      n.key === "Escape" ? c({ task: e, status: e.status, content: "" }) : n.key === "Enter" && !n.shiftKey && (n.preventDefault(), p(!0));
    },
    [c, e, p]
  );
  return g(() => {
    const n = x.current;
    n && (u.current = n.value = e.content, n.focus(), h());
  }, [e, l, h]), g(() => {
    const n = x.current;
    if (n)
      return S(n, p).destroy;
  }, [l, p]), /* @__PURE__ */ K("div", { className: `wx-OQDwWK17 wx-task${e.status ? " wx-done" : ""}`, children: [
    /* @__PURE__ */ a("div", { className: "wx-OQDwWK17 wx-checkbox-wrapper", children: l === e.id && e.id === -1 ? /* @__PURE__ */ a("div", { className: "wx-OQDwWK17 wx-icon-add", children: /* @__PURE__ */ a("i", { className: "wx-OQDwWK17 wxi-plus" }) }) : /* @__PURE__ */ a(G, { onChange: D, value: e.status }) }),
    /* @__PURE__ */ a("div", { className: "wx-OQDwWK17 wx-wrapper", children: l === e.id ? /* @__PURE__ */ a(
      "textarea",
      {
        ref: x,
        className: "wx-OQDwWK17 wx-texarea",
        placeholder: d("Enter the task..."),
        onKeyDown: N,
        onInput: C
      }
    ) : /* @__PURE__ */ a("div", { className: "wx-OQDwWK17 wx-text-wrapper", "data-id": e.id, children: /* @__PURE__ */ a("span", { className: "wx-OQDwWK17 wx-text", children: e.content }) }) }),
    /* @__PURE__ */ a("div", { className: "wx-OQDwWK17 wx-icon-close", children: !i && l !== e.id ? /* @__PURE__ */ a(
      "i",
      {
        className: "wx-OQDwWK17 wxi-close",
        onClick: () => f(e.id)
      }
    ) : null })
  ] });
}
function j(w) {
  const { readonly: e = !1, onChange: l } = w, [i, c] = B(w.data || []), [f, d] = O(null), [x, u] = O(null), h = I(W.i18n), C = A(() => h.getGroup("tasklist"), [h]);
  function D() {
    d(-1), u({ id: -1, content: "", status: 0 });
  }
  function p(t) {
    const s = {
      id: _(),
      content: t,
      status: 0
    }, r = [...i, s];
    if (c(r), l) {
      const o = l({ value: r, task: s, action: "add" });
      o && typeof o == "object" && (o.then ? o.then((m) => {
        N(s.id, m);
      }) : N(s.id, o));
    }
  }
  function N(t, s) {
    c((r) => {
      const o = r.findIndex((v) => v.id === t);
      if (o === -1) return r;
      const m = [...r];
      return m[o] = { ...r[o], ...s }, m;
    });
  }
  function n(t) {
    f === t && d(null);
    const s = i.filter((r) => r.id !== t);
    c(s), t !== -1 && l && l({ value: s, id: t, action: "delete" });
  }
  function P(t, s, r) {
    let o;
    const m = i.map((v) => v.id === t ? (o = { ...v, content: s, status: r }, o) : v);
    c(m), d(null), l && l({ value: m, id: t, action: "update", task: o });
  }
  function Q(t) {
    n(t);
  }
  function b({ task: t, content: s, status: r, next: o }) {
    t.id === -1 ? (d(null), s ? (p(s), o && D()) : n(t.id)) : (s ? (s !== t.content || r !== t.status) && P(t.id, s, r) : n(t.id), d(null));
  }
  const E = y({});
  E.current.dblclick = (t) => {
    w.readonly || d(t);
  };
  const R = y(null);
  return g(() => {
    U(R.current, {
      dblclick: (t) => E.current.dblclick(t)
    });
  }, []), /* @__PURE__ */ a(W.fieldId.Provider, { value: null, children: /* @__PURE__ */ K("div", { className: "wx-kro6Nsfl wx-tasks-list", children: [
    /* @__PURE__ */ K("div", { className: "wx-kro6Nsfl wx-list", ref: R, children: [
      i.map((t) => /* @__PURE__ */ a(
        T,
        {
          task: t,
          edit: f,
          onUpdate: b,
          onRemove: Q,
          readonly: e
        },
        t.id
      )),
      f === -1 ? /* @__PURE__ */ a(
        T,
        {
          task: x,
          edit: f,
          onUpdate: b,
          onRemove: Q
        },
        f
      ) : null
    ] }),
    !e && f !== -1 ? /* @__PURE__ */ a("div", { className: "wx-kro6Nsfl wx-button", children: /* @__PURE__ */ a(H, { onClick: D, children: C("Add task") }) }) : null
  ] }) });
}
function Y(w) {
  const { onData: e, onChange: l, value: i, ...c } = w;
  let [f, d] = O(null);
  g(() => {
    e && i ? Promise.resolve(e(i)).then((u) => d(u)) : i && i.then ? i.then((u) => d(u)) : d(i || []);
  }, [e, i]);
  const x = k(
    (u) => (u.originalValue = i, l && l(u)),
    [l, i]
  );
  return /* @__PURE__ */ a(L, { words: F, optional: !0, children: f === null ? /* @__PURE__ */ a(
    j,
    {
      data: [],
      readonly: !0,
      ...c,
      onChange: x
    }
  ) : /* @__PURE__ */ a(j, { data: f, ...c, onChange: x }) });
}
export {
  Y as Tasklist
};
