// @pdfme/converter/dist/esm/src/index.node.js
import * as pdfjsLib from 'pdfjs-dist/legacy/build/pdf';
// @ts-expect-error - PDFJSWorker import is not properly typed but required for functionality
import PDFJSWorker from 'pdfjs-dist/legacy/build/pdf.worker.js';
import { createCanvas } from 'canvas';
import { pdf2img as _pdf2img } from './pdf2img.js';
import { pdf2size as _pdf2size } from './pdf2size.js';
pdfjsLib.GlobalWorkerOptions.workerSrc = PDFJSWorker;
export const pdf2img = async (pdf, options = {}) => _pdf2img(pdf, options, {
    getDocument: (pdf) => pdfjsLib.getDocument({ data: pdf, isEvalSupported: false }).promise,
    createCanvas: (width, height) => createCanvas(width, height),
    canvasToArrayBuffer: (canvas) => {
        // Using a more specific type for the canvas from the 'canvas' package
        const nodeCanvas = canvas;
        // Get buffer from the canvas - using the synchronous version without parameters
        // This will use the default PNG format
        const buffer = nodeCanvas.toBuffer();
        // Convert to ArrayBuffer
        return new Uint8Array(buffer).buffer;
    },
});
export const pdf2size = async (pdf, options = {}) => _pdf2size(pdf, options, {
    getDocument: (pdf) => pdfjsLib.getDocument({ data: pdf, isEvalSupported: false }).promise,
});
export { img2pdf } from './img2pdf.js';
