// @pdfme/pdf-lib/dist/node/src/api/form/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PDFTextField = exports.PDFSignature = exports.PDFRadioGroup = exports.PDFOptionList = exports.PDFForm = exports.PDFField = exports.PDFDropdown = exports.PDFCheckBox = exports.PDFButton = void 0;
__exportStar(require("./appearances"), exports);
var PDFButton_1 = require("./PDFButton");
Object.defineProperty(exports, "PDFButton", { enumerable: true, get: function () { return __importDefault(PDFButton_1).default; } });
var PDFCheckBox_1 = require("./PDFCheckBox");
Object.defineProperty(exports, "PDFCheckBox", { enumerable: true, get: function () { return __importDefault(PDFCheckBox_1).default; } });
var PDFDropdown_1 = require("./PDFDropdown");
Object.defineProperty(exports, "PDFDropdown", { enumerable: true, get: function () { return __importDefault(PDFDropdown_1).default; } });
var PDFField_1 = require("./PDFField");
Object.defineProperty(exports, "PDFField", { enumerable: true, get: function () { return __importDefault(PDFField_1).default; } });
var PDFForm_1 = require("./PDFForm");
Object.defineProperty(exports, "PDFForm", { enumerable: true, get: function () { return __importDefault(PDFForm_1).default; } });
var PDFOptionList_1 = require("./PDFOptionList");
Object.defineProperty(exports, "PDFOptionList", { enumerable: true, get: function () { return __importDefault(PDFOptionList_1).default; } });
var PDFRadioGroup_1 = require("./PDFRadioGroup");
Object.defineProperty(exports, "PDFRadioGroup", { enumerable: true, get: function () { return __importDefault(PDFRadioGroup_1).default; } });
var PDFSignature_1 = require("./PDFSignature");
Object.defineProperty(exports, "PDFSignature", { enumerable: true, get: function () { return __importDefault(PDFSignature_1).default; } });
var PDFTextField_1 = require("./PDFTextField");
Object.defineProperty(exports, "PDFTextField", { enumerable: true, get: function () { return __importDefault(PDFTextField_1).default; } });
