// @pdfme/pdf-lib/dist/cjs/src/core/objects/PDFBool.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const errors_1 = require("../errors");
const PDFObject_1 = __importDefault(require("./PDFObject"));
const CharCodes_1 = __importDefault(require("../syntax/CharCodes"));
const ENFORCER = {};
class PDFBool extends PDFObject_1.default {
    constructor(enforcer, value) {
        if (enforcer !== ENFORCER)
            throw new errors_1.PrivateConstructorError('PDFBool');
        super();
        Object.defineProperty(this, "value", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.value = value;
    }
    asBoolean() {
        return this.value;
    }
    clone() {
        return this;
    }
    toString() {
        return String(this.value);
    }
    sizeInBytes() {
        return this.value ? 4 : 5;
    }
    copyBytesInto(buffer, offset) {
        if (this.value) {
            buffer[offset++] = CharCodes_1.default.t;
            buffer[offset++] = CharCodes_1.default.r;
            buffer[offset++] = CharCodes_1.default.u;
            buffer[offset++] = CharCodes_1.default.e;
            return 4;
        }
        else {
            buffer[offset++] = CharCodes_1.default.f;
            buffer[offset++] = CharCodes_1.default.a;
            buffer[offset++] = CharCodes_1.default.l;
            buffer[offset++] = CharCodes_1.default.s;
            buffer[offset++] = CharCodes_1.default.e;
            return 5;
        }
    }
}
Object.defineProperty(PDFBool, "True", {
    enumerable: true,
    configurable: true,
    writable: true,
    value: new PDFBool(ENFORCER, true)
});
Object.defineProperty(PDFBool, "False", {
    enumerable: true,
    configurable: true,
    writable: true,
    value: new PDFBool(ENFORCER, false)
});
exports.default = PDFBool;
