// @pdfme/pdf-lib/dist/cjs/src/core/objects/PDFInvalidObject.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const PDFObject_1 = __importDefault(require("./PDFObject"));
class PDFInvalidObject extends PDFObject_1.default {
    constructor(data) {
        super();
        Object.defineProperty(this, "data", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.data = data;
    }
    clone() {
        return PDFInvalidObject.of(this.data.slice());
    }
    toString() {
        return `PDFInvalidObject(${this.data.length} bytes)`;
    }
    sizeInBytes() {
        return this.data.length;
    }
    copyBytesInto(buffer, offset) {
        const length = this.data.length;
        for (let idx = 0; idx < length; idx++) {
            buffer[offset++] = this.data[idx];
        }
        return length;
    }
}
Object.defineProperty(PDFInvalidObject, "of", {
    enumerable: true,
    configurable: true,
    writable: true,
    value: (data) => new PDFInvalidObject(data)
});
exports.default = PDFInvalidObject;
