// @univerjs/preset-sheets-core/lib/es/index.js
import { UniverDocsPlugin as v } from "@univerjs/docs";
export * from "@univerjs/docs";
import { UniverDocsUIPlugin as C } from "@univerjs/docs-ui";
export * from "@univerjs/docs-ui";
import { UniverFormulaEnginePlugin as b } from "@univerjs/engine-formula";
export * from "@univerjs/engine-formula";
import { UniverRenderEnginePlugin as R } from "@univerjs/engine-render";
export * from "@univerjs/engine-render";
import { UniverNetworkPlugin as y } from "@univerjs/network";
export * from "@univerjs/network";
import { UniverRPCMainThreadPlugin as k } from "@univerjs/rpc";
export * from "@univerjs/rpc";
import { UniverSheetsPlugin as w } from "@univerjs/sheets";
export * from "@univerjs/sheets";
import { UniverSheetsFormulaPlugin as M } from "@univerjs/sheets-formula";
export * from "@univerjs/sheets-formula";
import { UniverSheetsFormulaUIPlugin as A } from "@univerjs/sheets-formula-ui";
export * from "@univerjs/sheets-formula-ui";
import { UniverSheetsNumfmtPlugin as T } from "@univerjs/sheets-numfmt";
export * from "@univerjs/sheets-numfmt";
import { UniverSheetsNumfmtUIPlugin as E } from "@univerjs/sheets-numfmt-ui";
export * from "@univerjs/sheets-numfmt-ui";
import { UniverSheetsUIPlugin as I } from "@univerjs/sheets-ui";
export * from "@univerjs/sheets-ui";
import { UniverUIPlugin as H } from "@univerjs/ui";
export * from "@univerjs/ui";
import "@univerjs/network/lib/facade";
import "@univerjs/sheets/lib/facade";
import "@univerjs/ui/lib/facade";
import "@univerjs/docs-ui/lib/facade";
import "@univerjs/sheets-ui/lib/facade";
import "@univerjs/engine-formula/lib/facade";
import "@univerjs/sheets-formula/lib/facade";
import "@univerjs/sheets-numfmt/lib/facade";
import "@univerjs/sheets-formula-ui/lib/facade";
function or(m = {}) {
  var n;
  const {
    container: l = "app",
    workerURL: t,
    header: p,
    footer: u,
    toolbar: a,
    ribbonType: c,
    formulaBar: g,
    menu: f,
    contextMenu: S,
    disableAutoFocus: d,
    customFontFamily: U,
    docs: i,
    sheets: r,
    formula: o,
    disableTextFormatAlert: x,
    disableTextFormatMark: F
  } = m, e = !!t;
  return {
    plugins: [
      y,
      [v, {
        hasScroll: i == null ? void 0 : i.hasScroll
      }],
      R,
      [H, {
        container: l,
        header: p,
        toolbar: a,
        ribbonType: c,
        menu: f,
        contextMenu: S,
        disableAutoFocus: d,
        customFontFamily: U
      }],
      C,
      e ? [k, { workerURL: t }] : null,
      [b, {
        notExecuteFormula: e,
        function: o == null ? void 0 : o.function
      }],
      [w, {
        notExecuteFormula: e,
        onlyRegisterFormulaRelatedMutations: !1,
        isRowStylePrecedeColumnStyle: r == null ? void 0 : r.isRowStylePrecedeColumnStyle,
        autoHeightForMergedCells: r == null ? void 0 : r.autoHeightForMergedCells,
        freezeSync: r == null ? void 0 : r.freezeSync
      }],
      [I, {
        formulaBar: g,
        footer: u,
        maxAutoHeightCount: r == null ? void 0 : r.maxAutoHeightCount,
        clipboardConfig: r == null ? void 0 : r.clipboardConfig,
        scrollConfig: r == null ? void 0 : r.scrollConfig,
        protectedRangeShadow: (n = r == null ? void 0 : r.protectedRangeShadow) != null ? n : !0,
        protectedRangeUserSelector: r == null ? void 0 : r.protectedRangeUserSelector,
        disableForceStringAlert: r == null ? void 0 : r.disableForceStringAlert,
        disableForceStringMark: r == null ? void 0 : r.disableForceStringMark
      }],
      [T, {
        disableTextFormatAlert: x,
        disableTextFormatMark: F
      }],
      E,
      [M, {
        notExecuteFormula: e,
        description: o == null ? void 0 : o.description,
        initialFormulaComputing: o == null ? void 0 : o.initialFormulaComputing
      }],
      [A, {
        functionScreenTips: o == null ? void 0 : o.functionScreenTips
      }]
    ].filter((P) => !!P)
  };
}
export {
  or as UniverSheetsCorePreset
};
