// @univerjs/thread-comment-ui/lib/es/index.js
var He = Object.defineProperty;
var Ve = (n, e, t) => e in n ? He(n, e, { enumerable: !0, configurable: !0, writable: !0, value: t }) : n[e] = t;
var W = (n, e, t) => Ve(n, typeof e != "symbol" ? e + "" : e, t);
import { Inject as ye, IUniverInstanceService as je, Disposable as Pe, UniverInstanceType as de, CommandType as we, DependentOn as Ae, Injector as Be, ICommandService as te, IConfigService as We, Plugin as Ze, merge as Fe, mergeOverrideWithDependencies as Ge, LocaleService as he, DOCS_NORMAL_EDITOR_UNIT_ID_KEY as Ke, BuildTextUtils as ue, Tools as ze, CustomRangeType as Ye, getBodySlice as qe, UserManagerService as fe, generateRandomId as Je } from "@univerjs/core";
import { ISidebarService as _e, useDependency as S, KeyCode as pe, useObservable as Z, useConfigValue as Qe, UI_PLUGIN_CONFIG_KEY as Xe } from "@univerjs/ui";
import { BehaviorSubject as Ce, filter as et, debounceTime as tt } from "rxjs";
import { UniverThreadCommentPlugin as nt, ThreadCommentModel as Se, getDT as rt, AddCommentCommand as it, ResolveCommentCommand as ot, DeleteCommentTreeCommand as Ne, UpdateCommentCommand as st, DeleteCommentCommand as lt } from "@univerjs/thread-comment";
import { jsxs as I, jsx as l } from "react/jsx-runtime";
import { Button as le, clsx as ve, Tooltip as at, scrollbarClassName as ct, borderClassName as dt, Dropdown as ut, Select as ge } from "@univerjs/design";
import { useRef as ee, createElement as G, forwardRef as K, useState as P, useEffect as ae, useMemo as X, useImperativeHandle as vt } from "react";
import { IEditorService as mt, BreakLineCommand as ht, RichTextEditor as ft } from "@univerjs/docs-ui";
var pt = Object.getOwnPropertyDescriptor, Ct = (n, e, t, r) => {
  for (var i = r > 1 ? void 0 : r ? pt(e, t) : e, o = n.length - 1, c; o >= 0; o--)
    (c = n[o]) && (i = c(i) || i);
  return i;
}, Ie = (n, e) => (t, r) => e(t, r, n);
let Y = class extends Pe {
  constructor(e, t) {
    super();
    W(this, "_panelVisible", !1);
    W(this, "_panelVisible$", new Ce(!1));
    W(this, "_activeCommentId");
    W(this, "_activeCommentId$", new Ce(void 0));
    W(this, "panelVisible$", this._panelVisible$.asObservable());
    W(this, "activeCommentId$", this._activeCommentId$.asObservable());
    this._sidebarService = e, this._univerInstanceService = t, this._init(), this.disposeWithMe(() => {
      this._activeCommentId$.complete(), this._panelVisible$.complete();
    });
  }
  _init() {
    this.disposeWithMe(
      this._sidebarService.sidebarOptions$.subscribe((e) => {
        e.visible || this.setPanelVisible(!1);
      })
    ), this.disposeWithMe(
      this._univerInstanceService.getCurrentTypeOfUnit$(de.UNIVER_SHEET).pipe(et((e) => !e)).subscribe(() => {
        this._sidebarService.close();
      })
    );
  }
  get panelVisible() {
    return this._panelVisible;
  }
  get activeCommentId() {
    return this._activeCommentId;
  }
  setPanelVisible(e) {
    this._panelVisible = e, this._panelVisible$.next(e);
  }
  setActiveComment(e) {
    this._activeCommentId = e, this._activeCommentId$.next(e);
  }
};
Y = Ct([
  Ie(0, ye(_e)),
  Ie(1, je)
], Y);
const gt = "thread-comment-panel", It = "UNIVER_THREAD_COMMENT_UI_PLUGIN", xt = {
  id: "thread-comment-ui.operation.toggle-panel",
  type: we.OPERATION,
  handler(n) {
    const e = n.get(_e), t = n.get(Y);
    return t.panelVisible ? (e.close(), t.setPanelVisible(!1)) : (e.open({
      header: { title: "threadCommentUI.panel.title" },
      children: { label: gt },
      width: 330
    }), t.setPanelVisible(!0)), !0;
  }
}, F = {
  id: "thread-comment-ui.operation.set-active-comment",
  type: we.OPERATION,
  handler(n, e) {
    return n.get(Y).setActiveComment(e), !0;
  }
}, bt = "thread-comment-ui.config", xe = {};
var yt = Object.defineProperty, wt = Object.getOwnPropertyDescriptor, _t = (n, e, t) => e in n ? yt(n, e, { enumerable: !0, configurable: !0, writable: !0, value: t }) : n[e] = t, St = (n, e, t, r) => {
  for (var i = r > 1 ? void 0 : r ? wt(e, t) : e, o = n.length - 1, c; o >= 0; o--)
    (c = n[o]) && (i = c(i) || i);
  return i;
}, me = (n, e) => (t, r) => e(t, r, n), $e = (n, e, t) => _t(n, typeof e != "symbol" ? e + "" : e, t);
let ce = class extends Ze {
  constructor(n = xe, e, t, r) {
    super(), this._config = n, this._injector = e, this._commandService = t, this._configService = r;
    const { menu: i, ...o } = Fe(
      {},
      xe,
      this._config
    );
    i && this._configService.setConfig("menu", i, { merge: !0 }), this._configService.setConfig(bt, o);
  }
  onStarting() {
    var n;
    Ge([
      [Y]
    ], (n = this._config) == null ? void 0 : n.overrides).forEach((e) => {
      this._injector.add(e);
    }), [xt, F].forEach((e) => {
      this._commandService.registerCommand(e);
    });
  }
};
$e(ce, "pluginName", It);
$e(ce, "type", de.UNIVER_UNKNOWN);
ce = St([
  Ae(nt),
  me(1, ye(Be)),
  me(2, te),
  me(3, We)
], ce);
function z({ ref: n, ...e }) {
  const { icon: t, id: r, className: i, extend: o, ...c } = e, x = `univerjs-icon univerjs-icon-${r} ${i || ""}`.trim(), h = ee(`_${kt()}`);
  return ke(t, `${r}`, {
    defIds: t.defIds,
    idSuffix: h.current
  }, {
    ref: n,
    className: x,
    ...c
  }, o);
}
function ke(n, e, t, r, i) {
  return G(n.tag, {
    key: e,
    ...Nt(n, t, i),
    ...r
  }, ($t(n, t).children || []).map((o, c) => ke(o, `${e}-${n.tag}-${c}`, t, void 0, i)));
}
function Nt(n, e, t) {
  const r = { ...n.attrs };
  t != null && t.colorChannel1 && r.fill === "colorChannel1" && (r.fill = t.colorChannel1), n.tag === "mask" && r.id && (r.id = r.id + e.idSuffix), Object.entries(r).forEach(([o, c]) => {
    o === "mask" && typeof c == "string" && (r[o] = c.replace(/url\(#(.*)\)/, `url(#$1${e.idSuffix})`));
  });
  const { defIds: i } = e;
  return !i || i.length === 0 || (n.tag === "use" && r["xlink:href"] && (r["xlink:href"] = r["xlink:href"] + e.idSuffix), Object.entries(r).forEach(([o, c]) => {
    typeof c == "string" && (r[o] = c.replace(/url\(#(.*)\)/, `url(#$1${e.idSuffix})`));
  })), r;
}
function $t(n, e) {
  var r;
  const { defIds: t } = e;
  return !t || t.length === 0 ? n : n.tag === "defs" && ((r = n.children) != null && r.length) ? {
    ...n,
    children: n.children.map((i) => typeof i.attrs.id == "string" && t && t.includes(i.attrs.id) ? {
      ...i,
      attrs: {
        ...i.attrs,
        id: i.attrs.id + e.idSuffix
      }
    } : i)
  } : n;
}
function kt() {
  return Math.random().toString(36).substring(2, 8);
}
z.displayName = "UniverIcon";
const Ut = {
  tag: "svg",
  attrs: {
    xmlns: "http://www.w3.org/2000/svg",
    fill: "none",
    viewBox: "0 0 16 16",
    width: "1em",
    height: "1em"
  },
  children: [
    {
      tag: "path",
      attrs: {
        fill: "currentColor",
        d: "M5.3313 1.4667C5.3313 1.13533 5.59993 0.866699 5.9313 0.866699H10.069C10.4004 0.866699 10.669 1.13533 10.669 1.4667C10.669 1.79807 10.4004 2.0667 10.069 2.0667H5.9313C5.59993 2.0667 5.3313 1.79807 5.3313 1.4667Z"
      }
    },
    {
      tag: "path",
      attrs: {
        fill: "currentColor",
        d: "M1.09985 3.64443C1.09985 3.31306 1.36848 3.04443 1.69985 3.04443H14.2999C14.6312 3.04443 14.8999 3.31306 14.8999 3.64443C14.8999 3.9758 14.6312 4.24443 14.2999 4.24443H1.69985C1.36848 4.24443 1.09985 3.9758 1.09985 3.64443Z"
      }
    },
    {
      tag: "path",
      attrs: {
        fill: "currentColor",
        d: "M6.12398 8.30171C6.35829 8.0674 6.73819 8.0674 6.97251 8.30171L8.00007 9.32928L9.02764 8.30171C9.26195 8.0674 9.64185 8.0674 9.87617 8.30171C10.1105 8.53603 10.1105 8.91593 9.87617 9.15024L8.8486 10.1778L9.87617 11.2054C10.1105 11.4397 10.1105 11.8196 9.87617 12.0539C9.64185 12.2882 9.26195 12.2882 9.02764 12.0539L8.00007 11.0263L6.97251 12.0539C6.73819 12.2882 6.35829 12.2882 6.12398 12.0539C5.88966 11.8196 5.88966 11.4397 6.12398 11.2054L7.15154 10.1778L6.12398 9.15024C5.88966 8.91593 5.88966 8.53603 6.12398 8.30171Z"
      }
    },
    {
      tag: "path",
      attrs: {
        fill: "currentColor",
        d: "M4.75332 5.22217C3.86966 5.22217 3.15332 5.93851 3.15332 6.82217V12.5331C3.15332 13.9691 4.31738 15.1332 5.75332 15.1332H10.2465C11.6825 15.1332 12.8465 13.9691 12.8465 12.5331V6.82217C12.8465 5.93851 12.1302 5.22217 11.2465 5.22217H4.75332ZM4.35332 6.82217C4.35332 6.60125 4.53241 6.42217 4.75332 6.42217H11.2465C11.4674 6.42217 11.6465 6.60125 11.6465 6.82217V12.5331C11.6465 13.3063 11.0197 13.9332 10.2465 13.9332H5.75332C4.98012 13.9332 4.35332 13.3063 4.35332 12.5331V6.82217Z",
        fillRule: "evenodd",
        clipRule: "evenodd"
      }
    }
  ]
}, Ue = K(function(e, t) {
  return G(z, Object.assign({}, e, {
    id: "delete-icon",
    ref: t,
    icon: Ut
  }));
});
Ue.displayName = "DeleteIcon";
const Tt = {
  tag: "svg",
  attrs: {
    xmlns: "http://www.w3.org/2000/svg",
    fill: "none",
    viewBox: "0 0 16 16",
    width: "1em",
    height: "1em"
  },
  children: [{
    tag: "path",
    attrs: {
      fill: "currentColor",
      d: "M8.6 1.99991C8.60001 1.66854 8.33138 1.39991 8.00001 1.3999C7.66864 1.3999 7.40001 1.66853 7.4 1.9999L7.39996 7.3999H1.9999C1.66853 7.3999 1.3999 7.66853 1.3999 7.9999C1.3999 8.33127 1.66853 8.5999 1.9999 8.5999H7.39995L7.3999 13.9999C7.3999 14.3313 7.66853 14.5999 7.9999 14.5999C8.33127 14.5999 8.5999 14.3313 8.5999 13.9999L8.59995 8.5999H13.9999C14.3313 8.5999 14.5999 8.33127 14.5999 7.9999C14.5999 7.66853 14.3313 7.3999 13.9999 7.3999H8.59996L8.6 1.99991Z"
    }
  }]
}, Te = K(function(e, t) {
  return G(z, Object.assign({}, e, {
    id: "increase-icon",
    ref: t,
    icon: Tt
  }));
});
Te.displayName = "IncreaseIcon";
const Dt = {
  tag: "svg",
  attrs: {
    xmlns: "http://www.w3.org/2000/svg",
    fill: "none",
    viewBox: "0 0 16 16",
    width: "1em",
    height: "1em"
  },
  children: [
    {
      tag: "path",
      attrs: {
        fill: "currentColor",
        d: "M3 9C3.55228 9 4 8.55228 4 8C4 7.44772 3.55228 7 3 7C2.44772 7 2 7.44772 2 8C2 8.55228 2.44772 9 3 9Z"
      }
    },
    {
      tag: "path",
      attrs: {
        fill: "currentColor",
        d: "M8 9C8.55228 9 9 8.55228 9 8C9 7.44772 8.55228 7 8 7C7.44772 7 7 7.44772 7 8C7 8.55228 7.44772 9 8 9Z"
      }
    },
    {
      tag: "path",
      attrs: {
        fill: "currentColor",
        d: "M13 9C13.5523 9 14 8.55228 14 8C14 7.44772 13.5523 7 13 7C12.4477 7 12 7.44772 12 8C12 8.55228 12.4477 9 13 9Z"
      }
    }
  ]
}, De = K(function(e, t) {
  return G(z, Object.assign({}, e, {
    id: "more-horizontal-icon",
    ref: t,
    icon: Dt
  }));
});
De.displayName = "MoreHorizontalIcon";
const Et = {
  tag: "svg",
  attrs: {
    xmlns: "http://www.w3.org/2000/svg",
    fill: "none",
    viewBox: "0 0 16 16",
    width: "1em",
    height: "1em"
  },
  children: [
    {
      tag: "path",
      attrs: {
        stroke: "currentColor",
        d: "M7.48389 10.3267V12.1905C7.48389 12.7428 7.9316 13.1905 8.48389 13.1905H11.2216L12.2955 14.2644L13.3695 13.1905H14.1593C14.7116 13.1905 15.1593 12.7428 15.1593 12.1905V8.46289C15.1593 7.91061 14.7116 7.46289 14.1593 7.46289H12.2955",
        strokeLinecap: "round",
        strokeLinejoin: "round",
        strokeWidth: 1.2
      }
    },
    {
      tag: "path",
      attrs: {
        stroke: "currentColor",
        d: "M0.840332 3.73535C0.840332 2.63078 1.73576 1.73535 2.84033 1.73535H10.2955C11.4001 1.73535 12.2955 2.63078 12.2955 3.73535V8.32676C12.2955 9.43132 11.4001 10.3268 10.2955 10.3268H5.6014L4.1695 11.7587L3.05978 10.3268H2.84033C1.73576 10.3268 0.840332 9.43133 0.840332 8.32676V3.73535Z",
        strokeLinecap: "round",
        strokeLinejoin: "round",
        strokeWidth: 1.2
      }
    },
    {
      tag: "path",
      attrs: {
        stroke: "currentColor",
        d: "M6.41016 6.1311H6.76813",
        strokeLinecap: "round",
        strokeWidth: 1.2
      }
    },
    {
      tag: "path",
      attrs: {
        stroke: "currentColor",
        d: "M8.91626 6.1311H9.27424",
        strokeLinecap: "round",
        strokeWidth: 1.2
      }
    },
    {
      tag: "path",
      attrs: {
        stroke: "currentColor",
        d: "M3.90454 6.1311H4.26252",
        strokeLinecap: "round",
        strokeWidth: 1.2
      }
    }
  ]
}, Ee = K(function(e, t) {
  return G(z, Object.assign({}, e, {
    id: "reply-to-comment-icon",
    ref: t,
    icon: Et
  }));
});
Ee.displayName = "ReplyToCommentIcon";
const Mt = {
  tag: "svg",
  attrs: {
    xmlns: "http://www.w3.org/2000/svg",
    fill: "none",
    viewBox: "0 0 17 17",
    width: "1em",
    height: "1em"
  },
  children: [{
    tag: "path",
    attrs: {
      fill: "currentColor",
      d: "M8.6106 15.4036C12.4766 15.4036 15.6106 12.2696 15.6106 8.40356C15.6106 4.53757 12.4766 1.40356 8.6106 1.40356C4.7446 1.40356 1.6106 4.53757 1.6106 8.40356C1.6106 12.2696 4.7446 15.4036 8.6106 15.4036ZM12.3351 6.82773C12.5694 6.59342 12.5694 6.21352 12.3351 5.9792C12.1007 5.74489 11.7208 5.74489 11.4865 5.9792L7.91079 9.55494L6.33506 7.9792C6.10074 7.74489 5.72084 7.74489 5.48653 7.9792C5.25221 8.21352 5.25221 8.59342 5.48653 8.82773L7.48653 10.8277C7.72084 11.062 8.10074 11.062 8.33506 10.8277L12.3351 6.82773Z",
      fillRule: "evenodd",
      clipRule: "evenodd"
    }
  }]
}, Me = K(function(e, t) {
  return G(z, Object.assign({}, e, {
    id: "resolved-icon",
    ref: t,
    icon: Mt
  }));
});
Me.displayName = "ResolvedIcon";
const Rt = {
  tag: "svg",
  attrs: {
    xmlns: "http://www.w3.org/2000/svg",
    fill: "none",
    viewBox: "0 0 17 17",
    width: "1em",
    height: "1em"
  },
  children: [{
    tag: "circle",
    attrs: {
      cx: 8.73,
      cy: 8.4,
      r: 6.4,
      stroke: "currentColor",
      strokeWidth: 1.2
    }
  }, {
    tag: "path",
    attrs: {
      stroke: "currentColor",
      d: "M6.02637 8.40356L8.02637 10.4036L12.0264 6.40356",
      strokeLinecap: "round",
      strokeLinejoin: "round",
      strokeWidth: 1.2
    }
  }]
}, Re = K(function(e, t) {
  return G(z, Object.assign({}, e, {
    id: "solve-icon",
    ref: t,
    icon: Rt
  }));
});
Re.displayName = "SolveIcon";
function be(n) {
  return {
    id: "d",
    body: n,
    documentStyle: {}
  };
}
const Oe = K((n, e) => {
  var R;
  const { comment: t, onSave: r, id: i, onCancel: o, autoFocus: c, unitId: x, type: h } = n, N = S(te), k = S(he), [V, U] = P(!1), b = S(mt), u = ee(null), v = h === de.UNIVER_DOC ? Ke : x, [T, y] = P(() => {
    var d, m, f;
    return ue.transform.getPlainText((f = (m = (d = u.current) == null ? void 0 : d.getDocumentData().body) == null ? void 0 : m.dataStream) != null ? f : "");
  });
  ae(() => {
    var m, f, w, p;
    y(ue.transform.getPlainText((w = (f = (m = u.current) == null ? void 0 : m.getDocumentData().body) == null ? void 0 : f.dataStream) != null ? w : ""));
    const d = (p = u.current) == null ? void 0 : p.selectionChange$.subscribe(() => {
      var s, _, A;
      y(ue.transform.getPlainText((A = (_ = (s = u.current) == null ? void 0 : s.getDocumentData().body) == null ? void 0 : _.dataStream) != null ? A : ""));
    });
    return () => d == null ? void 0 : d.unsubscribe();
  }, [(R = u.current) == null ? void 0 : R.selectionChange$]);
  const j = X(() => ({
    keyCodes: [{ keyCode: pe.ENTER }],
    handler: (d) => {
      d === pe.ENTER && N.executeCommand(
        ht.id
      );
    }
  }), [N]);
  vt(e, () => ({
    reply(d) {
      var f, w;
      if (!u.current)
        return;
      b.focus((f = u.current.getEditorId()) != null ? f : "");
      const m = be(d);
      (w = u.current) == null || w.setDocumentData(m, [{
        startOffset: m.body.dataStream.length - 2,
        endOffset: m.body.dataStream.length - 2,
        collapsed: !0
      }]);
    }
  }));
  const C = () => {
    if (u.current) {
      const d = ze.deepClone(u.current.getDocumentData().body);
      U(!1), r == null || r({
        ...t,
        text: d
      }), u.current.replaceText(""), setTimeout(() => {
        var m, f;
        (m = u.current) == null || m.setSelectionRanges([]), (f = u.current) == null || f.blur();
      }, 10);
    }
  };
  return /* @__PURE__ */ I("div", { onClick: (d) => d.preventDefault(), children: [
    /* @__PURE__ */ l(
      ft,
      {
        className: "univer-w-full",
        editorRef: u,
        autoFocus: c,
        keyboardEventConfig: j,
        placeholder: k.t("threadCommentUI.editor.placeholder"),
        initialValue: (t == null ? void 0 : t.text) && be(t.text),
        onFocusChange: (d) => d && U(d),
        isSingle: !1,
        maxHeight: 64,
        onClickOutside: () => {
          setTimeout(() => {
            b.focus(v);
          }, 30);
        }
      }
    ),
    V ? /* @__PURE__ */ I("div", { className: "univer-mt-3 univer-flex univer-flex-row univer-justify-end univer-gap-2", children: [
      /* @__PURE__ */ l(
        le,
        {
          onClick: () => {
            var d;
            o == null || o(), U(!1), (d = u.current) == null || d.replaceText("", !0), N.executeCommand(F.id);
          },
          children: k.t("threadCommentUI.editor.cancel")
        }
      ),
      /* @__PURE__ */ l(
        le,
        {
          variant: "primary",
          disabled: !T,
          onClick: C,
          children: k.t(i ? "threadCommentUI.editor.save" : "threadCommentUI.editor.reply")
        }
      )
    ] }) : null
  ] });
}), Ot = (n) => {
  const { dataStream: e, customRanges: t } = n, r = e.endsWith(`\r
`) ? e.length - 2 : e.length, i = [];
  let o = 0;
  return t == null || t.forEach((c) => {
    o < c.startIndex && i.push({
      type: "text",
      content: e.slice(o, c.startIndex)
    }), i.push({
      type: "mention",
      content: {
        label: e.slice(c.startIndex, c.endIndex + 1),
        id: c.rangeId
      }
    }), o = c.endIndex + 1;
  }), i.push({
    type: "text",
    content: e.slice(o, r)
  }), i;
}, Lt = (n) => {
  const { paragraphs: e = [] } = n;
  let t = 0;
  return e.map((r) => {
    const i = qe(n, t, r.startIndex);
    return t = r.startIndex + 1, Ot(i);
  });
}, Ht = (n) => {
  let e = "";
  const t = [];
  return n.forEach((r) => {
    switch (r.type) {
      case "text":
        e += r.content;
        break;
      case "mention": {
        const i = e.length;
        e += r.content.label;
        const o = e.length - 1;
        t.push({
          rangeId: r.content.id,
          rangeType: Ye.MENTION,
          startIndex: i,
          endIndex: o,
          properties: {},
          wholeEntity: !0
        });
        break;
      }
    }
  }), e += `\r
`, {
    textRuns: [],
    paragraphs: [
      {
        startIndex: e.length - 2,
        paragraphStyle: {}
      }
    ],
    sectionBreaks: [
      {
        startIndex: e.length - 1
      }
    ],
    dataStream: e,
    customRanges: t
  };
}, Le = "__mock__", Vt = (n) => {
  const { item: e, unitId: t, subUnitId: r, editing: i, onEditingChange: o, onReply: c, resolved: x, isRoot: h, onClose: N, onDeleteComment: k, type: V } = n, U = S(te), b = S(he), u = S(fe), v = u.getUser(e.personId), T = Z(u.currentUser$), y = (T == null ? void 0 : T.userID) === e.personId, j = e.id === Le, [C, R] = P(!1), d = Qe(Xe), m = d == null ? void 0 : d.avatarFallback, f = () => {
    (k == null ? void 0 : k(e)) !== !1 && (U.executeCommand(
      h ? Ne.id : lt.id,
      {
        unitId: t,
        subUnitId: r,
        commentId: e.id
      }
    ), h && (N == null || N()));
  };
  return /* @__PURE__ */ I("div", { className: "univer-relative univer-mb-3 univer-pl-[30px]", onMouseLeave: () => R(!1), onMouseEnter: () => R(!0), children: [
    /* @__PURE__ */ l(
      "div",
      {
        className: "univer-absolute univer-left-0 univer-top-0 univer-h-6 univer-w-6 univer-rounded-full univer-bg-cover univer-bg-center univer-bg-no-repeat",
        style: {
          backgroundImage: `url(${(v == null ? void 0 : v.avatar) || m})`
        }
      }
    ),
    v ? /* @__PURE__ */ I("div", { className: "univer-mb-1 univer-flex univer-h-6 univer-items-center univer-justify-between", children: [
      /* @__PURE__ */ l("div", { className: "univer-text-sm univer-font-medium univer-leading-5", children: (v == null ? void 0 : v.name) || " " }),
      /* @__PURE__ */ I("div", { children: [
        j || x ? null : C && v ? /* @__PURE__ */ l(
          "div",
          {
            className: "univer-ml-1 univer-inline-flex univer-h-6 univer-w-6 univer-cursor-pointer univer-items-center univer-justify-center univer-rounded-sm univer-text-base hover:univer-bg-gray-50",
            onClick: () => c(v),
            children: /* @__PURE__ */ l(Ee, {})
          }
        ) : null,
        y && !j && !x ? /* @__PURE__ */ l(
          ut,
          {
            overlay: /* @__PURE__ */ l("div", { className: "univer-rounded-lg", children: /* @__PURE__ */ I(
              "ul",
              {
                className: "univer-m-0 univer-box-border univer-grid univer-list-none univer-p-1.5 univer-text-sm [&_a]:univer-block [&_a]:univer-cursor-pointer [&_a]:univer-rounded [&_a]:univer-px-2 [&_a]:univer-py-1.5 [&_a]:univer-transition-colors",
                children: [
                  /* @__PURE__ */ l("li", { children: /* @__PURE__ */ l(
                    "a",
                    {
                      className: "hover:univer-bg-gray-200",
                      onClick: () => o == null ? void 0 : o(!0),
                      children: b.t("threadCommentUI.item.edit")
                    }
                  ) }),
                  /* @__PURE__ */ l("li", { children: /* @__PURE__ */ l(
                    "a",
                    {
                      className: "hover:univer-bg-gray-200",
                      onClick: f,
                      children: b.t("threadCommentUI.item.delete")
                    }
                  ) })
                ]
              }
            ) }),
            children: /* @__PURE__ */ l(
              "div",
              {
                className: "univer-ml-1 univer-inline-flex univer-h-6 univer-w-6 univer-cursor-pointer univer-items-center univer-justify-center univer-rounded-sm univer-text-base hover:univer-bg-gray-50",
                children: /* @__PURE__ */ l(De, {})
              }
            )
          }
        ) : null
      ] })
    ] }) : null,
    /* @__PURE__ */ l(
      "time",
      {
        className: "univer-mb-1 univer-text-xs/normal univer-text-gray-600 dark:!univer-text-gray-200",
        children: e.dT
      }
    ),
    i ? /* @__PURE__ */ l(
      Oe,
      {
        type: V,
        id: e.id,
        comment: e,
        onCancel: () => o == null ? void 0 : o(!1),
        autoFocus: !0,
        unitId: t,
        subUnitId: r,
        onSave: ({ text: w, attachments: p }) => {
          o == null || o(!1), U.executeCommand(
            st.id,
            {
              unitId: t,
              subUnitId: r,
              payload: {
                commentId: e.id,
                text: w,
                attachments: p
              }
            }
          );
        }
      }
    ) : /* @__PURE__ */ l(
      "div",
      {
        className: "univer-text-sm univer-text-gray-900 dark:!univer-text-white",
        children: Lt(e.text).map((w, p) => /* @__PURE__ */ l("div", { className: "univer-break-words", children: w.map((s, _) => {
          switch (s.type) {
            case "mention":
              return /* @__PURE__ */ I("a", { className: "univer-text-primary-600", children: [
                s.content.label,
                " "
              ] }, _);
            default:
              return s.content;
          }
        }) }, p))
      }
    )
  ] });
}, jt = (n) => {
  var se, Q, a;
  const {
    id: e,
    unitId: t,
    subUnitId: r,
    refStr: i,
    showEdit: o = !0,
    onClick: c,
    showHighlight: x,
    onClose: h,
    getSubUnitName: N,
    prefix: k,
    autoFocus: V,
    onMouseEnter: U,
    onMouseLeave: b,
    onAddComment: u,
    onDeleteComment: v,
    onResolve: T,
    type: y,
    style: j,
    full: C
  } = n, R = S(Se), [d, m] = P(!1), [f, w] = P(""), p = X(() => R.commentUpdate$.pipe(tt(16)), [R]);
  Z(p);
  const s = e ? R.getCommentWithChildren(t, r, e) : null, _ = S(te), A = S(fe), D = s == null ? void 0 : s.root.resolved, E = Z(A.currentUser$), L = ee(null), q = [
    ...s ? [s.root] : (
      // mock empty comment
      [{
        id: Le,
        text: {
          dataStream: `
\r`
        },
        personId: (se = E == null ? void 0 : E.userID) != null ? se : "",
        ref: i != null ? i : "",
        dT: "",
        unitId: t,
        subUnitId: r,
        threadId: ""
      }]
    ),
    ...(Q = s == null ? void 0 : s.children) != null ? Q : []
  ], O = ee(null), ne = (g) => {
    g.stopPropagation(), D ? _.executeCommand(F.id, {
      unitId: t,
      subUnitId: r,
      commentId: e
    }) : _.executeCommand(F.id), _.executeCommand(ot.id, {
      unitId: t,
      subUnitId: r,
      commentId: e,
      resolved: !D
    }), T == null || T(!D);
  }, J = (g) => {
    g.stopPropagation(), _.executeCommand(F.id), !(s != null && s.root && (v == null ? void 0 : v(s.root)) === !1) && (_.executeCommand(
      Ne.id,
      {
        unitId: t,
        subUnitId: r,
        commentId: e
      }
    ), h == null || h());
  };
  ae(() => b == null ? void 0 : b(), []);
  const re = N((a = s == null ? void 0 : s.root.subUnitId) != null ? a : r), ie = o && !f && !D, oe = `${i || (s == null ? void 0 : s.root.ref) || ""}${re ? " · " : ""}${re}`;
  return /* @__PURE__ */ I(
    "div",
    {
      id: `${k}-${t}-${r}-${e}`,
      className: ve("univer-relative univer-box-border univer-rounded-md univer-bg-white univer-p-4 dark:!univer-bg-gray-900 dark:!univer-text-white", dt, {
        "univer-w-[278px]": !C,
        "univer-w-full": C,
        "univer-shadow": !D && (x || d || k === "cell")
      }),
      style: j,
      onClick: c,
      onMouseEnter: () => {
        U == null || U(), m(!0);
      },
      onMouseLeave: () => {
        b == null || b(), m(!1);
      },
      children: [
        !D && x && /* @__PURE__ */ l(
          "div",
          {
            className: "univer-absolute univer-left-0 univer-right-0 univer-top-0 univer-h-1.5 univer-rounded-t-md univer-bg-yellow-400"
          }
        ),
        /* @__PURE__ */ I(
          "div",
          {
            className: "univer-mb-4 univer-flex univer-flex-row univer-items-center univer-justify-between univer-text-sm univer-leading-5",
            children: [
              /* @__PURE__ */ I("div", { className: "univer-flex univer-flex-1 univer-flex-row univer-items-center univer-overflow-hidden", children: [
                /* @__PURE__ */ l(
                  "div",
                  {
                    className: "univer-mr-2 univer-h-3.5 univer-w-[3px] univer-flex-shrink-0 univer-flex-grow-0 univer-rounded-sm univer-bg-yellow-500"
                  }
                ),
                /* @__PURE__ */ l(at, { showIfEllipsis: !0, title: oe, children: /* @__PURE__ */ l(
                  "span",
                  {
                    className: "univer-flex-1 univer-truncate",
                    children: oe
                  }
                ) })
              ] }),
              !!s && /* @__PURE__ */ I("div", { className: "univer-flex univer-flex-shrink-0 univer-flex-grow-0 univer-flex-row", children: [
                /* @__PURE__ */ l(
                  "div",
                  {
                    className: ve("univer-ml-1 univer-inline-flex univer-h-6 univer-w-6 univer-cursor-pointer univer-items-center univer-justify-center univer-rounded-[3px] univer-text-base hover:univer-bg-gray-50 dark:hover:!univer-bg-gray-800", {
                      "univer-text-green-500": D
                    }),
                    onClick: ne,
                    children: D ? /* @__PURE__ */ l(Me, {}) : /* @__PURE__ */ l(Re, {})
                  }
                ),
                (E == null ? void 0 : E.userID) === s.root.personId ? /* @__PURE__ */ l(
                  "div",
                  {
                    className: "univer-ml-1 univer-inline-flex univer-h-6 univer-w-6 univer-cursor-pointer univer-items-center univer-justify-center univer-rounded-[3px] univer-text-base hover:univer-bg-gray-50 dark:hover:!univer-bg-gray-800",
                    onClick: J,
                    children: /* @__PURE__ */ l(Ue, {})
                  }
                ) : null
              ] })
            ]
          }
        ),
        /* @__PURE__ */ l(
          "div",
          {
            ref: O,
            className: ve("univer-max-h-80 univer-overflow-y-auto univer-overflow-x-hidden", ct),
            children: q.map(
              (g) => /* @__PURE__ */ l(
                Vt,
                {
                  unitId: t,
                  subUnitId: r,
                  item: g,
                  isRoot: g.id === (s == null ? void 0 : s.root.id),
                  editing: f === g.id,
                  resolved: s == null ? void 0 : s.root.resolved,
                  type: y,
                  onClose: h,
                  onEditingChange: (M) => {
                    w(M ? g.id : "");
                  },
                  onReply: (M) => {
                    M && requestAnimationFrame(() => {
                      var H;
                      (H = L.current) == null || H.reply(Ht([
                        {
                          type: "mention",
                          content: {
                            id: M.userID,
                            label: `@${M.name}`
                          }
                        },
                        {
                          type: "text",
                          content: " "
                        }
                      ]));
                    });
                  },
                  onAddComment: u,
                  onDeleteComment: v
                },
                g.id
              )
            )
          }
        ),
        ie && /* @__PURE__ */ l("div", { children: /* @__PURE__ */ l(
          Oe,
          {
            ref: L,
            type: y,
            unitId: t,
            subUnitId: r,
            onSave: async ({ text: g, attachments: M }) => {
              const H = {
                text: g,
                attachments: M,
                dT: rt(),
                id: Je(),
                ref: i,
                personId: E == null ? void 0 : E.userID,
                parentId: s == null ? void 0 : s.root.id,
                unitId: t,
                subUnitId: r,
                threadId: s == null ? void 0 : s.root.threadId
              };
              (u == null ? void 0 : u(H)) !== !1 && (await _.executeCommand(
                it.id,
                {
                  unitId: t,
                  subUnitId: r,
                  comment: H
                }
              ), O.current && (O.current.scrollTop = O.current.scrollHeight));
            },
            autoFocus: V || !s,
            onCancel: () => {
              s || h == null || h();
            }
          },
          `${V}`
        ) })
      ]
    }
  );
}, Yt = (n) => {
  const {
    unitId: e,
    subUnitId$: t,
    type: r,
    onAdd: i,
    getSubUnitName: o,
    onResolve: c,
    sortComments: x,
    onItemLeave: h,
    onItemEnter: N,
    disableAdd: k,
    tempComment: V,
    onAddComment: U,
    onDeleteComment: b,
    showComments: u
  } = n, [v, T] = P("all"), [y, j] = P("all"), C = S(he), R = S(fe), d = S(Se), [m, f] = P(() => d.getUnit(e)), w = S(Y), p = Z(w.activeCommentId$), s = Z(d.commentUpdate$), _ = S(te), A = Z(t), D = ee(!0), E = "panel", L = Z(R.currentUser$), q = X(() => {
    var H;
    const a = v === "all" ? m : (H = m.filter(($) => $.subUnitId === A)) != null ? H : [], g = x != null ? x : (($) => $), M = a.map(($) => {
      var B;
      return { ...$.root, children: (B = $.children) != null ? B : [], users: $.relativeUsers };
    });
    if (u) {
      const $ = /* @__PURE__ */ new Map();
      return M.forEach((B) => {
        $.set(B.id, B);
      }), [...u, ""].map((B) => $.get(B)).filter(Boolean);
    } else
      return g(M);
  }, [u, v, m, x, A]), O = X(() => [
    ...q.filter((a) => !a.resolved),
    ...q.filter((a) => a.resolved)
  ], [q]), ne = X(() => y === "resolved" ? O.filter((a) => a.resolved) : y === "unsolved" ? O.filter((a) => !a.resolved) : y === "concern_me" && L != null && L.userID ? O.filter((a) => a == null ? void 0 : a.users.has(L.userID)) : O, [O, L == null ? void 0 : L.userID, y]), J = V ? [V, ...ne] : ne, re = J.filter((a) => !a.resolved), ie = J.filter((a) => a.resolved), oe = y !== "all" || v !== "all", se = () => {
    j("all"), T("all");
  };
  ae(() => {
    e && f(
      d.getUnit(e)
    );
  }, [e, d, s]), ae(() => {
    var $;
    if (!p)
      return;
    if (!D.current) {
      D.current = !0;
      return;
    }
    const { unitId: a, subUnitId: g, commentId: M } = p, H = `${E}-${a}-${g}-${M}`;
    ($ = document.getElementById(H)) == null || $.scrollIntoView({ block: "center" });
  }, [p]);
  const Q = (a) => /* @__PURE__ */ l(
    jt,
    {
      prefix: E,
      getSubUnitName: o,
      id: a.id,
      unitId: a.unitId,
      subUnitId: a.subUnitId,
      refStr: a.ref,
      type: r,
      showEdit: (p == null ? void 0 : p.commentId) === a.id,
      showHighlight: (p == null ? void 0 : p.commentId) === a.id,
      onClick: () => {
        D.current = !1, a.resolved ? _.executeCommand(F.id) : _.executeCommand(
          F.id,
          {
            unitId: a.unitId,
            subUnitId: a.subUnitId,
            commentId: a.id,
            temp: !1
          }
        );
      },
      onMouseEnter: () => N == null ? void 0 : N(a),
      onMouseLeave: () => h == null ? void 0 : h(a),
      onAddComment: U,
      onDeleteComment: b,
      onResolve: (g) => c == null ? void 0 : c(a.id, g)
    },
    a.id
  );
  return /* @__PURE__ */ I("div", { className: "univer-flex univer-min-h-full univer-flex-col univer-pb-3", children: [
    /* @__PURE__ */ I("div", { className: "univer-mt-3 univer-flex univer-flex-row univer-justify-between", children: [
      r === de.UNIVER_SHEET ? /* @__PURE__ */ l(
        ge,
        {
          borderless: !0,
          value: v,
          options: [
            {
              value: "current",
              label: C.t("threadCommentUI.filter.sheet.current")
            },
            {
              value: "all",
              label: C.t("threadCommentUI.filter.sheet.all")
            }
          ],
          onChange: T
        }
      ) : null,
      /* @__PURE__ */ l(
        ge,
        {
          borderless: !0,
          value: y,
          options: [
            {
              value: "all",
              label: C.t("threadCommentUI.filter.status.all")
            },
            {
              value: "resolved",
              label: C.t("threadCommentUI.filter.status.resolved")
            },
            {
              value: "unsolved",
              label: C.t("threadCommentUI.filter.status.unsolved")
            },
            {
              value: "concern_me",
              label: C.t("threadCommentUI.filter.status.concernMe")
            }
          ],
          onChange: j
        }
      )
    ] }),
    J.length === 0 ? /* @__PURE__ */ I(
      "div",
      {
        className: "univer-flex univer-flex-1 univer-flex-col univer-items-center univer-justify-center univer-text-sm univer-text-gray-600 dark:!univer-text-gray-200",
        children: [
          C.t("threadCommentUI.panel.empty"),
          oe ? /* @__PURE__ */ l("div", { className: "univer-mt-2 univer-flex univer-flex-row", children: /* @__PURE__ */ l(le, { onClick: se, children: C.t("threadCommentUI.panel.reset") }) }) : k ? null : /* @__PURE__ */ l("div", { className: "univer-mt-2 univer-flex univer-flex-row", children: /* @__PURE__ */ I(le, { onClick: i, children: [
            /* @__PURE__ */ l(Te, { className: "univer-mr-1.5" }),
            C.t("threadCommentUI.panel.addComment")
          ] }) })
        ]
      }
    ) : /* @__PURE__ */ I("div", { className: "univer-mt-3 univer-flex univer-flex-col univer-gap-3", children: [
      re.map(Q),
      ie.length > 0 && /* @__PURE__ */ l("div", { className: "univer-text-xs", children: C.t("threadCommentUI.panel.solved") }),
      ie.map(Q)
    ] })
  ] });
};
export {
  F as SetActiveCommentOperation,
  gt as THREAD_COMMENT_PANEL,
  Yt as ThreadCommentPanel,
  Y as ThreadCommentPanelService,
  jt as ThreadCommentTree,
  xt as ToggleSheetCommentPanelOperation,
  ce as UniverThreadCommentUIPlugin
};
