// @pdfme/pdf-lib/dist/cjs/src/core/acroform/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PDFAcroText = exports.PDFAcroTerminal = exports.PDFAcroSignature = exports.PDFAcroRadioButton = exports.PDFAcroPushButton = exports.PDFAcroNonTerminal = exports.PDFAcroListBox = exports.PDFAcroForm = exports.PDFAcroField = exports.PDFAcroComboBox = exports.PDFAcroChoice = exports.PDFAcroCheckBox = exports.PDFAcroButton = void 0;
var PDFAcroButton_1 = require("./PDFAcroButton");
Object.defineProperty(exports, "PDFAcroButton", { enumerable: true, get: function () { return __importDefault(PDFAcroButton_1).default; } });
var PDFAcroCheckBox_1 = require("./PDFAcroCheckBox");
Object.defineProperty(exports, "PDFAcroCheckBox", { enumerable: true, get: function () { return __importDefault(PDFAcroCheckBox_1).default; } });
var PDFAcroChoice_1 = require("./PDFAcroChoice");
Object.defineProperty(exports, "PDFAcroChoice", { enumerable: true, get: function () { return __importDefault(PDFAcroChoice_1).default; } });
var PDFAcroComboBox_1 = require("./PDFAcroComboBox");
Object.defineProperty(exports, "PDFAcroComboBox", { enumerable: true, get: function () { return __importDefault(PDFAcroComboBox_1).default; } });
var PDFAcroField_1 = require("./PDFAcroField");
Object.defineProperty(exports, "PDFAcroField", { enumerable: true, get: function () { return __importDefault(PDFAcroField_1).default; } });
var PDFAcroForm_1 = require("./PDFAcroForm");
Object.defineProperty(exports, "PDFAcroForm", { enumerable: true, get: function () { return __importDefault(PDFAcroForm_1).default; } });
var PDFAcroListBox_1 = require("./PDFAcroListBox");
Object.defineProperty(exports, "PDFAcroListBox", { enumerable: true, get: function () { return __importDefault(PDFAcroListBox_1).default; } });
var PDFAcroNonTerminal_1 = require("./PDFAcroNonTerminal");
Object.defineProperty(exports, "PDFAcroNonTerminal", { enumerable: true, get: function () { return __importDefault(PDFAcroNonTerminal_1).default; } });
var PDFAcroPushButton_1 = require("./PDFAcroPushButton");
Object.defineProperty(exports, "PDFAcroPushButton", { enumerable: true, get: function () { return __importDefault(PDFAcroPushButton_1).default; } });
var PDFAcroRadioButton_1 = require("./PDFAcroRadioButton");
Object.defineProperty(exports, "PDFAcroRadioButton", { enumerable: true, get: function () { return __importDefault(PDFAcroRadioButton_1).default; } });
var PDFAcroSignature_1 = require("./PDFAcroSignature");
Object.defineProperty(exports, "PDFAcroSignature", { enumerable: true, get: function () { return __importDefault(PDFAcroSignature_1).default; } });
var PDFAcroTerminal_1 = require("./PDFAcroTerminal");
Object.defineProperty(exports, "PDFAcroTerminal", { enumerable: true, get: function () { return __importDefault(PDFAcroTerminal_1).default; } });
var PDFAcroText_1 = require("./PDFAcroText");
Object.defineProperty(exports, "PDFAcroText", { enumerable: true, get: function () { return __importDefault(PDFAcroText_1).default; } });
__exportStar(require("./flags"), exports);
__exportStar(require("./utils"), exports);
